

<!DOCTYPE html>
<html lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>pypath: A Python module for molecular signaling prior knowledge processing &mdash; pypath 0.10.6 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="#" class="icon icon-home" alt="Documentation Home"> pypath
          

          
          </a>

          
            
            
              <div class="version">
                0.10.6
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#"><em>pypath:</em> A Python module for molecular signaling prior knowledge processing</a><ul>
<li><a class="reference internal" href="#webservice">Webservice</a><ul>
<li><a class="reference internal" href="#query-types">Query types</a></li>
<li><a class="reference internal" href="#interaction-datasets">Interaction datasets</a></li>
<li><a class="reference internal" href="#mouse-and-rat">Mouse and rat</a></li>
<li><a class="reference internal" href="#examples">Examples</a><ul>
<li><a class="reference internal" href="#molecular-interaction-network">Molecular interaction network</a></li>
<li><a class="reference internal" href="#enzyme-substrate-interactions">Enzyme-substrate interactions</a></li>
<li><a class="reference internal" href="#molecular-complexes">Molecular complexes</a></li>
<li><a class="reference internal" href="#annotations">Annotations</a></li>
<li><a class="reference internal" href="#roles-in-inter-cellular-communication">Roles in inter-cellular communication</a></li>
<li><a class="reference internal" href="#exploring-possible-parameters">Exploring possible parameters</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#can-i-use-omnipath-in-r">Can I use OmniPath in R?</a></li>
<li><a class="reference internal" href="#installation">Installation</a><ul>
<li><a class="reference internal" href="#linux">Linux</a></li>
<li><a class="reference internal" href="#igraph-c-library-cairo-and-pycairo">igraph C library, cairo and pycairo</a></li>
<li><a class="reference internal" href="#directly-from-git">Directly from git</a></li>
<li><a class="reference internal" href="#with-pip">With pip</a></li>
<li><a class="reference internal" href="#build-source-distribution">Build source distribution</a></li>
<li><a class="reference internal" href="#mac-os-x">Mac OS X</a><ul>
<li><a class="reference internal" href="#troubleshooting">Troubleshooting</a></li>
</ul>
</li>
<li><a class="reference internal" href="#microsoft-windows">Microsoft Windows</a><ul>
<li><a class="reference internal" href="#with-anaconda">With Anaconda</a></li>
<li><a class="reference internal" href="#with-other-python-distributions">With other Python distributions</a></li>
<li><a class="reference internal" href="#known-issues">Known issues</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#release-history">Release History</a><ul>
<li><a class="reference internal" href="#id3">0.1.0</a></li>
<li><a class="reference internal" href="#id4">0.2.0</a></li>
<li><a class="reference internal" href="#id5">0.3.0</a></li>
<li><a class="reference internal" href="#id6">0.4.0</a></li>
<li><a class="reference internal" href="#id7">0.5.0</a></li>
<li><a class="reference internal" href="#id8">0.5.32</a></li>
<li><a class="reference internal" href="#id9">0.6.31</a></li>
<li><a class="reference internal" href="#id10">0.7.0</a></li>
<li><a class="reference internal" href="#id11">0.7.74</a></li>
<li><a class="reference internal" href="#id12">0.7.93</a></li>
<li><a class="reference internal" href="#id13">0.7.110</a></li>
<li><a class="reference internal" href="#id14">0.8</a></li>
<li><a class="reference internal" href="#id15">0.9</a></li>
<li><a class="reference internal" href="#id16">0.10.0</a></li>
<li><a class="reference internal" href="#upcoming">Upcoming</a></li>
</ul>
</li>
<li><a class="reference internal" href="#features">Features</a><ul>
<li><a class="reference internal" href="#id-conversion">ID conversion</a></li>
<li><a class="reference internal" href="#pathways">Pathways</a></li>
<li><a class="reference internal" href="#structural-features">Structural features</a></li>
<li><a class="reference internal" href="#sequences">Sequences</a></li>
<li><a class="reference internal" href="#tissue-expression">Tissue expression</a></li>
<li><a class="reference internal" href="#functional-annotations">Functional annotations</a></li>
<li><a class="reference internal" href="#drug-compounds">Drug compounds</a></li>
<li><a class="reference internal" href="#technical">Technical</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">pypath</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="#" class="icon icon-home"></a> &raquo;</li>
        
      <li><em>pypath:</em> A Python module for molecular signaling prior knowledge processing</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="pypath-a-python-module-for-molecular-signaling-prior-knowledge-processing">
<h1><em>pypath:</em> A Python module for molecular signaling prior knowledge processing<a class="headerlink" href="#pypath-a-python-module-for-molecular-signaling-prior-knowledge-processing" title="Permalink to this headline">¶</a></h1>
<blockquote>
<div><p><strong>Important:</strong> New module structure and new network API</p>
<p>Around the end of December we added a new network API to <code class="docutils literal notranslate"><span class="pre">pypath</span></code> which
is not based on <code class="docutils literal notranslate"><span class="pre">igraph</span></code> any more and provides a modular and versatile
access interface to the network data (since version <code class="docutils literal notranslate"><span class="pre">0.9</span></code>). In January
we reorganized the submodules in <code class="docutils literal notranslate"><span class="pre">pypath</span></code> in order to create a clear
structure (since version <code class="docutils literal notranslate"><span class="pre">0.10</span></code>). These are important milestones
towards version <code class="docutils literal notranslate"><span class="pre">1.0</span></code> and we hope they will make <code class="docutils literal notranslate"><span class="pre">pypath</span></code> more
convenient to use for everyone. By 18 February we merged these changes
to the master branch however the <em>pypath guide</em> is still to be updated.
Apologies for this inconvenience and please don’t hesitate to ask
questions by opening an issue on github. The old <code class="docutils literal notranslate"><span class="pre">igraph</span></code> based network
class is still available in the <code class="docutils literal notranslate"><span class="pre">pypath.legacy</span></code> module.</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Py2/3</dt>
<dd class="field-odd"><p>Although we still keep the compatibility with Python 2, we don’t
test <code class="docutils literal notranslate"><span class="pre">pypath</span></code> in this environment and very few people uses it
already. We highly recommend to use <code class="docutils literal notranslate"><span class="pre">pypath</span></code> in Python 3.6+.</p>
</dd>
<dt class="field-even">documentation</dt>
<dd class="field-even"><p><a class="reference external" href="http://saezlab.github.io/pypath">http://saezlab.github.io/pypath</a></p>
</dd>
<dt class="field-odd">issues</dt>
<dd class="field-odd"><p><a class="reference external" href="https://github.com/saezlab/pypath/issues">https://github.com/saezlab/pypath/issues</a></p>
</dd>
<dt class="field-even">contact</dt>
<dd class="field-even"><p><a class="reference external" href="mailto:omnipathdb&#37;&#52;&#48;gmail&#46;com">omnipathdb<span>&#64;</span>gmail<span>&#46;</span>com</a></p>
</dd>
<dt class="field-odd">developers</dt>
<dd class="field-odd"><p><code class="docutils literal notranslate"><span class="pre">pypath</span></code> is developed in the Saez Lab (<a class="reference external" href="http://saezlab.org">http://saezlab.org</a>) by
Olga Ivanova, Nicolàs Palacio and Dénes Türei; the R package and the
Cytoscape app are developed and maintained by Francesco Ceccarelli, Attila
Gábor, Alberto Valdeolivas and Nicolàs Palacio.</p>
</dd>
</dl>
<p><strong>pypath</strong> is a Python module for processing molecular biology data resources,
combining them into databases and providing a versatile interface in Python
as well as exporting the data for access through other platforms such as
the R (the OmnipathR R/Bioconductor package), web service (at
<a class="reference external" href="http://omnipathdb.org">http://omnipathdb.org</a>), Cytoscape (the OmniPath Cytoscape app) and BEL
(Biological Expression Language).</p>
<p><strong>pypath</strong> provides access to more than 100 resources! It builds 5 major
combined databases and within these we can distinguish different datasets.
The 5 major databases are interactions (molecular interaction network or
pathways), enzyme-substrate relationships, protein complexes, molecular
annotations (functional roles, localizations, and more) and inter-cellular
communication roles.</p>
<p><strong>pypath</strong> consists of a number of submodules and each of them again contains
a number of submodules. Overall <strong>pypath</strong> consists of around 100 modules.
The most important higher level submodules:</p>
<ul class="simple">
<li><p><em>pypath.core:</em> contains the database classes e.g. network, complex,
annotations, etc</p></li>
<li><p><em>pypath.inputs:</em> contains the resource specific methods which directly
downlad and preprocess data from the original sources</p></li>
<li><p><em>pypath.omnipath:</em> higher level applications, e.g. a database manager, a
web server</p></li>
<li><p><em>pypath.utils:</em> stand alone useful utilities, e.g. identifier translator,
Gene Ontology processor, BioPax processor, etc</p></li>
</ul>
<div class="section" id="webservice">
<h2>Webservice<a class="headerlink" href="#webservice" title="Permalink to this headline">¶</a></h2>
<p><strong>New webservice</strong> from 14 June 2018: the queries slightly changed, have been
largely extended. See the examples below.</p>
<p>The webservice implements a very simple REST style API, you can make requests
by the HTTP protocol (browser, wget, curl or whatever). After defining the
query type and optionally a set of molecular entities (proteins) you can
add further GET parameters encoded in the URL.</p>
<div class="section" id="query-types">
<h3>Query types<a class="headerlink" href="#query-types" title="Permalink to this headline">¶</a></h3>
<p>The webservice currently recognizes 7 types of queries: <code class="docutils literal notranslate"><span class="pre">interactions</span></code>,
<code class="docutils literal notranslate"><span class="pre">enz_sub</span></code>, <code class="docutils literal notranslate"><span class="pre">annotations</span></code>, <code class="docutils literal notranslate"><span class="pre">complexes</span></code>, <code class="docutils literal notranslate"><span class="pre">intercell</span></code>, <code class="docutils literal notranslate"><span class="pre">queries</span></code> and
<code class="docutils literal notranslate"><span class="pre">info</span></code>.
The query types <code class="docutils literal notranslate"><span class="pre">resources</span></code>, <code class="docutils literal notranslate"><span class="pre">network</span></code> and <code class="docutils literal notranslate"><span class="pre">about</span></code> have not been
implemented yet in the new webservice.</p>
</div>
<div class="section" id="interaction-datasets">
<h3>Interaction datasets<a class="headerlink" href="#interaction-datasets" title="Permalink to this headline">¶</a></h3>
<p>The instance of the <code class="docutils literal notranslate"><span class="pre">pypath</span></code> webserver running at the domain
<a class="reference external" href="http://omnipathdb.org/">http://omnipathdb.org/</a>, serves not only the OmniPath data but also other
datasets. Each of them has a short name what you can use in the queries
(e.g. <code class="docutils literal notranslate"><span class="pre">&amp;datasets=omnipath,pathwayextra</span></code>).</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">omnipath</span></code>: the OmniPath data as defined in the paper, an arbitrary
optimum between coverage and quality</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pathwayextra</span></code>: activity flow interactions without literature reference</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">kinaseextra</span></code>: enzyme-substrate interactions without literature reference</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ligrecextra</span></code>: ligand-receptor interactions without literature reference</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">tfregulons</span></code>: transcription factor (TF)-target interactions from DoRothEA</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">mirnatarget</span></code>: miRNA-mRNA and TF-miRNA interactions</p></li>
</ul>
<p>TF-target interactions from TF Regulons, a large collection additional
enzyme-substrate interactions, and literature curated miRNA-mRNA interacions
combined from 4 databases.</p>
</div>
<div class="section" id="mouse-and-rat">
<h3>Mouse and rat<a class="headerlink" href="#mouse-and-rat" title="Permalink to this headline">¶</a></h3>
<p>Except the miRNA interactions all interactions are available for human, mouse
and rat. The rodent data has been translated from human using the NCBI
Homologene database. Many human proteins do not have known homolog in rodents
hence rodent datasets are smaller than their human counterparts. Note, if you
work with mouse omics data you might do better to translate your dataset to
human (for example using the <code class="docutils literal notranslate"><span class="pre">pypath.homology</span></code> module) and use human
interaction data.</p>
</div>
<div class="section" id="examples">
<h3>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h3>
<p>A request without any parameter provides the main webpage:</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org">http://omnipathdb.org</a></p>
</div></blockquote>
<p>The <code class="docutils literal notranslate"><span class="pre">info</span></code> returns a HTML page with comprehensive information about the
resources. The list here should be and will be updated as currently OmniPath
includes much more databases:</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org/info">http://omnipathdb.org/info</a></p>
</div></blockquote>
<div class="section" id="molecular-interaction-network">
<h4>Molecular interaction network<a class="headerlink" href="#molecular-interaction-network" title="Permalink to this headline">¶</a></h4>
<p>The <code class="docutils literal notranslate"><span class="pre">interactions</span></code> query accepts some parameters and returns interactions in
tabular format. This example returns all interactions of EGFR (P00533), with
sources and references listed.</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org/interactions/?partners=P00533&amp;fields=sources,references">http://omnipathdb.org/interactions/?partners=P00533&amp;fields=sources,references</a></p>
</div></blockquote>
<p>By default only the OmniPath dataset used, to include any other dataset you
have to set additional parameters. For example to query the transcriptional regulators of EGFR:</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org/interactions/?targets=EGFR&amp;types=TF">http://omnipathdb.org/interactions/?targets=EGFR&amp;types=TF</a></p>
</div></blockquote>
<p>The TF Regulons database assigns confidence levels to the interactions. You
might want to select only the highest confidence, <em>A</em> category:</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org/interactions/?targets=EGFR&amp;types=TF&amp;tfregulons_levels=A">http://omnipathdb.org/interactions/?targets=EGFR&amp;types=TF&amp;tfregulons_levels=A</a></p>
</div></blockquote>
<p>Show the transcriptional targets of Smad2 homology translated to rat including
the confidence levels from TF Regulons:</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org/interactions/?genesymbols=1&amp;fields=type,ncbi_tax_id,tfregulons_level&amp;organisms=10116&amp;sources=Smad2&amp;types=TF">http://omnipathdb.org/interactions/?genesymbols=1&amp;fields=type,ncbi_tax_id,tfregulons_level&amp;organisms=10116&amp;sources=Smad2&amp;types=TF</a></p>
</div></blockquote>
<p>Query interactions from PhosphoNetworks which is part of the <em>kinaseextra</em>
dataset:</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org/interactions/?genesymbols=1&amp;fields=sources&amp;databases=PhosphoNetworks&amp;datasets=kinaseextra">http://omnipathdb.org/interactions/?genesymbols=1&amp;fields=sources&amp;databases=PhosphoNetworks&amp;datasets=kinaseextra</a></p>
</div></blockquote>
<p>Get the interactions from Signor, SPIKE and SignaLink3:</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org/interactions/?genesymbols=1&amp;fields=sources,references&amp;databases=Signor,SPIKE,SignaLink3">http://omnipathdb.org/interactions/?genesymbols=1&amp;fields=sources,references&amp;databases=Signor,SPIKE,SignaLink3</a></p>
</div></blockquote>
<p>All interactions of MAP1LC3B:</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org/interactions/?genesymbols=1&amp;partners=MAP1LC3B">http://omnipathdb.org/interactions/?genesymbols=1&amp;partners=MAP1LC3B</a></p>
</div></blockquote>
<p>By default <code class="docutils literal notranslate"><span class="pre">partners</span></code> queries the interaction where either the source or the
arget is among the partners. If you set the <code class="docutils literal notranslate"><span class="pre">source_target</span></code> parameter to
<code class="docutils literal notranslate"><span class="pre">AND</span></code> both the source and the target must be in the queried set:</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org/interactions/?genesymbols=1&amp;fields=sources,references&amp;sources=ATG3,ATG7,ATG4B,SQSTM1&amp;targets=MAP1LC3B,MAP1LC3A,MAP1LC3C,Q9H0R8,GABARAP,GABARAPL2&amp;source_target=AND">http://omnipathdb.org/interactions/?genesymbols=1&amp;fields=sources,references&amp;sources=ATG3,ATG7,ATG4B,SQSTM1&amp;targets=MAP1LC3B,MAP1LC3A,MAP1LC3C,Q9H0R8,GABARAP,GABARAPL2&amp;source_target=AND</a></p>
</div></blockquote>
<p>As you see above you can use UniProt IDs and Gene Symbols in the queries and
also mix them. Get the miRNA regulating NOTCH1:</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org/interactions/?genesymbols=1&amp;fields=sources,references&amp;datasets=mirnatarget&amp;targets=NOTCH1">http://omnipathdb.org/interactions/?genesymbols=1&amp;fields=sources,references&amp;datasets=mirnatarget&amp;targets=NOTCH1</a></p>
</div></blockquote>
<p>Note: with the exception of mandatory fields and genesymbols, the columns
appear exactly in the order you provided in your query.</p>
</div>
<div class="section" id="enzyme-substrate-interactions">
<h4>Enzyme-substrate interactions<a class="headerlink" href="#enzyme-substrate-interactions" title="Permalink to this headline">¶</a></h4>
<p>Another query type available is <code class="docutils literal notranslate"><span class="pre">ptms</span></code> which provides enzyme-substrate
interactions. It is very similar to the <code class="docutils literal notranslate"><span class="pre">interactions</span></code>:</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org/enz_sub?genesymbols=1&amp;fields=sources,references,isoforms&amp;enzymes=FYN">http://omnipathdb.org/enz_sub?genesymbols=1&amp;fields=sources,references,isoforms&amp;enzymes=FYN</a></p>
</div></blockquote>
<p>Is there any ubiquitination reaction?</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org/ens_sub?genesymbols=1&amp;fields=sources,references&amp;types=ubiquitination">http://omnipathdb.org/ens_sub?genesymbols=1&amp;fields=sources,references&amp;types=ubiquitination</a></p>
</div></blockquote>
<p>And acetylation in mouse?</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org/ptms?genesymbols=1&amp;fields=sources,references&amp;types=acetylation&amp;organisms=10090">http://omnipathdb.org/ptms?genesymbols=1&amp;fields=sources,references&amp;types=acetylation&amp;organisms=10090</a></p>
</div></blockquote>
<p>Rat interactions, both directly from rat and homology translated from human,
from the PhosphoSite database:</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org/enz_sub?genesymbols=1&amp;fields=sources,references&amp;organisms=10116&amp;databases=PhosphoSite,PhosphoSite_noref">http://omnipathdb.org/enz_sub?genesymbols=1&amp;fields=sources,references&amp;organisms=10116&amp;databases=PhosphoSite,PhosphoSite_noref</a></p>
</div></blockquote>
</div>
<div class="section" id="molecular-complexes">
<h4>Molecular complexes<a class="headerlink" href="#molecular-complexes" title="Permalink to this headline">¶</a></h4>
<p>The <code class="docutils literal notranslate"><span class="pre">complexes</span></code> query provides a comprehensive database of more than 22,000
protein complexes. For example, to query all complexes from CORUM and PDB
containing MTOR (P42345):</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org/complexes?proteins=P42345&amp;databases=CORUM,PDB">http://omnipathdb.org/complexes?proteins=P42345&amp;databases=CORUM,PDB</a></p>
</div></blockquote>
</div>
<div class="section" id="annotations">
<h4>Annotations<a class="headerlink" href="#annotations" title="Permalink to this headline">¶</a></h4>
<p>The <code class="docutils literal notranslate"><span class="pre">annotations</span></code> query provides a large variety of data about proteins,
complexes and in the future other kinds of molecules. For example an
annotation can tell if a protein is a kinase, or if it is expressed in the
hearth muscle. These data come from dozens of databases and each kind of
annotation record contains different fields. Because of this here we have
a <code class="docutils literal notranslate"><span class="pre">record_id</span></code> field which is unique within the records of each database.
Each row contains one key value pair and you need to use the <code class="docutils literal notranslate"><span class="pre">record_id</span></code>
to connect the related key-value pairs. You can easily do this with <code class="docutils literal notranslate"><span class="pre">tidyr</span></code>
and <code class="docutils literal notranslate"><span class="pre">dplyr</span></code> in R or <code class="docutils literal notranslate"><span class="pre">pandas</span></code> in Python. An example to query the pathway
annotations from SignaLink:</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org/annotations?databases=SignaLink3">http://omnipathdb.org/annotations?databases=SignaLink3</a></p>
</div></blockquote>
<p>Or the tissue expression of BMP7 from Human Protein Atlas:</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org/annotations?databases=HPA_tissue&amp;proteins=BMP7">http://omnipathdb.org/annotations?databases=HPA_tissue&amp;proteins=BMP7</a></p>
</div></blockquote>
</div>
<div class="section" id="roles-in-inter-cellular-communication">
<h4>Roles in inter-cellular communication<a class="headerlink" href="#roles-in-inter-cellular-communication" title="Permalink to this headline">¶</a></h4>
<p>Another query type is <code class="docutils literal notranslate"><span class="pre">intercell</span></code> providing information on the roles in
inter-cellular signaling. E.g. if a protein is a ligand, a receptor, an
extracellular matrix (ECM) component, etc. This query type is very similar
to <code class="docutils literal notranslate"><span class="pre">annotations</span></code> but here the data does not come from original sources but
combined from several databases by us. However we refer also to the original
databases whenever the <code class="docutils literal notranslate"><span class="pre">class_type</span></code> is <code class="docutils literal notranslate"><span class="pre">sub</span></code> (subclass). E.g. the main
class <code class="docutils literal notranslate"><span class="pre">ligand</span></code> is a combination of <code class="docutils literal notranslate"><span class="pre">Ramilowski</span> <span class="pre">2015</span></code>, <code class="docutils literal notranslate"><span class="pre">CellPhoneDB</span></code>,
<code class="docutils literal notranslate"><span class="pre">HPMR</span></code> and many other databases, hence besides the <code class="docutils literal notranslate"><span class="pre">ligand</span></code> category you
will find sub-categories like <code class="docutils literal notranslate"><span class="pre">ligand_ramilowski</span></code>, <code class="docutils literal notranslate"><span class="pre">ligand_cellphonedb</span></code>
and so on. An example how to get all intercell annotations for 4 selected
proteins:</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org/intercell?proteins=EGFR,ULK1,ATG4A,BMP8B">http://omnipathdb.org/intercell?proteins=EGFR,ULK1,ATG4A,BMP8B</a></p>
</div></blockquote>
<p>Or all the main classes for one protein:</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org/intercell?levels=main&amp;proteins=P00533">http://omnipathdb.org/intercell?levels=main&amp;proteins=P00533</a></p>
</div></blockquote>
<p>Or a list of all ECM proteins:</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org/intercell?categories=ecm">http://omnipathdb.org/intercell?categories=ecm</a></p>
</div></blockquote>
</div>
<div class="section" id="exploring-possible-parameters">
<h4>Exploring possible parameters<a class="headerlink" href="#exploring-possible-parameters" title="Permalink to this headline">¶</a></h4>
<p>Sometimes the names and values of the query parameters are not intuitive,
even though in many cases the server accepts multiple alternatives. To see
the possible parameters with all possible values you can use the <code class="docutils literal notranslate"><span class="pre">queries</span></code>
query type. The server checks the parameter names and values exactly against
these rules and if any of them don’t match you will get an error message
instead of reply. To see the parameters for the <code class="docutils literal notranslate"><span class="pre">interactions</span></code> query:</p>
<blockquote>
<div><p><a class="reference external" href="http://omnipathdb.org/queries/interactions">http://omnipathdb.org/queries/interactions</a></p>
</div></blockquote>
</div>
</div>
</div>
<div class="section" id="can-i-use-omnipath-in-r">
<h2>Can I use OmniPath in R?<a class="headerlink" href="#can-i-use-omnipath-in-r" title="Permalink to this headline">¶</a></h2>
<p>You can download the data from the webservice and load into R. Thanks to
our colleague Attila Gabor we have a dedicated package for this:</p>
<blockquote>
<div><p><a class="reference external" href="https://github.com/saezlab/OmnipathR">https://github.com/saezlab/OmnipathR</a></p>
</div></blockquote>
</div>
<div class="section" id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<div class="section" id="linux">
<h3>Linux<a class="headerlink" href="#linux" title="Permalink to this headline">¶</a></h3>
<p>In almost any up-to-date Linux distribution the dependencies of <strong>pypath</strong> are
built-in, or provided by the distributors. You can simply install <strong>pypath</strong>
by <strong>pip</strong> (see below).
If any non mandatory dependency is still missing, you can install them the
usual way by <em>pip</em> or your package manager.</p>
</div>
<div class="section" id="igraph-c-library-cairo-and-pycairo">
<h3>igraph C library, cairo and pycairo<a class="headerlink" href="#igraph-c-library-cairo-and-pycairo" title="Permalink to this headline">¶</a></h3>
<p>For the legacy network class or the <code class="docutils literal notranslate"><span class="pre">igraph</span></code> conversion from the current
network class <em>python-igraph</em> must be installed.
<em>python(2)-igraph</em> is a Python interface to use the igraph C library. The
C library must be installed. The same goes for <em>cairo</em>, <em>py(2)cairo</em> and
<em>graphviz</em>.</p>
</div>
<div class="section" id="directly-from-git">
<h3>Directly from git<a class="headerlink" href="#directly-from-git" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip install git+https://github.com/saezlab/pypath.git
</pre></div>
</div>
</div>
<div class="section" id="with-pip">
<h3>With pip<a class="headerlink" href="#with-pip" title="Permalink to this headline">¶</a></h3>
<p>Download the package from /dist, and install with pip:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip install pypath-x.y.z.tar.gz
</pre></div>
</div>
</div>
<div class="section" id="build-source-distribution">
<h3>Build source distribution<a class="headerlink" href="#build-source-distribution" title="Permalink to this headline">¶</a></h3>
<p>Clone the git repo, and run setup.py:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python setup.py sdist
</pre></div>
</div>
</div>
<div class="section" id="mac-os-x">
<h3>Mac OS X<a class="headerlink" href="#mac-os-x" title="Permalink to this headline">¶</a></h3>
<p>Recently the installation on Mac should not be more complicated than on Linux:
you can simply install by <strong>pip</strong> (see above).</p>
<p>When <code class="docutils literal notranslate"><span class="pre">igraph</span></code> was a mandatory dependency and it didn’t provide wheels
the OS X installation was not straightforward primarily because cairo needs to
be compiled from source. If you want igraph and cairo we provide two scripts
<a class="reference external" href="src/scripts">here</a>: the <strong>mac-install-brew.sh</strong> installs everything with HomeBrew and
<strong>mac-install-conda.sh</strong> installs from Anaconda distribution. With these
scripts, installation of igraph, cairo and graphviz goes smoothly most of the
time and options are available to omit the last two. To know more, see
the description in the script header. There is a third script
<strong>mac-install-source.sh</strong> which compiles everything from source and presumes
only Python 2.7 and Xcode installed. We do not recommend this as it is time
consuming and troubleshooting requires expertise.</p>
<div class="section" id="troubleshooting">
<h4>Troubleshooting<a class="headerlink" href="#troubleshooting" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">no</span> <span class="pre">module</span> <span class="pre">named</span> <span class="pre">...</span></code> when you try to load a module in Python. Did
the installation of the module run without error? Try to run again the specific
part from the mac install shell script to see if any error comes up. Is the
path where the module has been installed in your <code class="docutils literal notranslate"><span class="pre">$PYTHONPATH</span></code>? Try <code class="docutils literal notranslate"><span class="pre">echo</span>
<span class="pre">$PYTHONPATH</span></code> to see the current paths. Add your local install directories if
those are not there, e.g.
<code class="docutils literal notranslate"><span class="pre">export</span> <span class="pre">PYTHONPATH=&quot;/Users/me/local/python2.7/site-packages:$PYTHONPATH&quot;</span></code>.
If it works afterwards, don’t forget to append these export path statements to
your <code class="docutils literal notranslate"><span class="pre">~/.bash_profile</span></code>, so these will be set every time you launch a new
shell.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pkgconfig</span></code> not found. Check if the <code class="docutils literal notranslate"><span class="pre">$PKG_CONFIG_PATH</span></code> variable is
set correctly, and pointing on a directory where pkgconfig really can be
found.</p></li>
<li><p>Error while trying to install py(2)cairo by pip. py(2)cairo could not be
installed by pip, but only by waf. Please set the <code class="docutils literal notranslate"><span class="pre">$PKG_CONFIG_PATH</span></code> before.
See <strong>mac-install-source.sh</strong> on how to install with waf.</p></li>
<li><p>Error at pygraphviz build: <code class="docutils literal notranslate"><span class="pre">graphviz/cgraph.h</span> <span class="pre">file</span> <span class="pre">not</span> <span class="pre">found</span></code>. This is
because the directory of graphviz detected wrong by pkgconfig. See
<strong>mac-install-source.sh</strong> how to set include dirs and library dirs by
<code class="docutils literal notranslate"><span class="pre">--global-option</span></code> parameters.</p></li>
<li><p>Can not install bioservices, because installation of jurko-suds fails. Ok,
this fails because pip is not able to install the recent version of
setuptools, because a very old version present in the system path. The
development version of jurko-suds does not require setuptools, so you can
install it directly from git as it is done in <strong>mac-install-source.sh</strong>.</p></li>
<li><p>In <strong>Anaconda</strong>, <em>pypath</em> can be imported, but the modules and classes are
missing. Apparently Anaconda has some built-in stuff called <em>pypath</em>. This
has nothing to do with this module. Please be aware that Anaconda installs a
completely separated Python distribution, and does not detect modules in the
main Python installation. You need to install all modules within Anaconda’s
directory. <strong>mac-install-conda.sh</strong> does exactly this. If you still
experience issues, please contact us.</p></li>
</ul>
</div>
</div>
<div class="section" id="microsoft-windows">
<h3>Microsoft Windows<a class="headerlink" href="#microsoft-windows" title="Permalink to this headline">¶</a></h3>
<p>Not many people have used <em>pypath</em> on Microsoft computers so far. Please share
your experiences and contact us if you encounter any issue. We appreciate
your feedback, and it would be nice to have better support for other computer
systems.</p>
<div class="section" id="with-anaconda">
<h4>With Anaconda<a class="headerlink" href="#with-anaconda" title="Permalink to this headline">¶</a></h4>
<p>The same workflow like you see in <code class="docutils literal notranslate"><span class="pre">mac-install-conda.sh</span></code> should work for
Anaconda on Windows. The only problem you certainly will encounter is that not
all the channels have packages for all platforms. If certain channel provides
no package for Windows, or for your Python version, you just need to find an
other one. For this, do a search:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>anaconda search -t conda &lt;package name&gt;
</pre></div>
</div>
<p>For example, if you search for <em>pycairo</em>, you will find out that <em>vgauther</em>
provides it for osx-64, but only for Python 3.4, while <em>richlewis</em> provides
also for Python 3.5. And for win-64 platform, there is the channel of
<em>KristanAmstrong</em>. Go along all the commands in <code class="docutils literal notranslate"><span class="pre">mac-install-conda.sh</span></code>, and
modify the channel if necessary, until all packages install successfully.</p>
</div>
<div class="section" id="with-other-python-distributions">
<h4>With other Python distributions<a class="headerlink" href="#with-other-python-distributions" title="Permalink to this headline">¶</a></h4>
<p>Here the basic principles are the same as everywhere: first try to install all
external dependencies, after <em>pip</em> install should work. On Windows certain
packages can not be installed by compiled from source by <em>pip</em>, instead the
easiest to install them precompiled. These are in our case <em>fisher, lxml,
numpy (mkl version), pycairo, igraph, pygraphviz, scipy and statsmodels</em>. The
precompiled packages are available <a class="reference external" href="http://www.lfd.uci.edu/~gohlke/pythonlibs/">here</a>.
We tested the setup with Python 3.4.3 and Python 2.7.11. The former should just
work fine, while with the latter we have issues to be resolved.</p>
</div>
<div class="section" id="known-issues">
<h4>Known issues<a class="headerlink" href="#known-issues" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p><em>“No module fabric available.”</em> – or <em>pysftp</em> missing: this is not
important, only certain data download methods rely on these modules, but
likely you won’t call those at all.</p></li>
<li><p>Progress indicator floods terminal: sorry about that, will be fixed soon.</p></li>
<li><p>Encoding related exceptions in Python2: these might occur at some points in
the module, please send the traceback if you encounter one, and we will fix
as soon as possible.</p></li>
<li><p>For Mac OS X (v &gt;= 10.11 El Capitan) import of pypath fails with error:
“libcurl link-time ssl backend (openssl) is different from compile-time ssl
backend (none/other)”. To fix it, you may need to reinstall pycurl library
using special flags. More information and steps can be found
<a class="reference external" href="https://cscheng.info/2018/01/26/installing-pycurl-on-macos-high-sierra.html">here</a>.</p></li>
</ul>
<p><em>Special thanks to Jorge Ferreira for testing pypath on Windows!</em></p>
</div>
</div>
</div>
<div class="section" id="release-history">
<h2>Release History<a class="headerlink" href="#release-history" title="Permalink to this headline">¶</a></h2>
<p>Main improvements in the past releases:</p>
<div class="section" id="id3">
<h3>0.1.0<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>First release of PyPath, for initial testing.</p></li>
</ul>
</div>
<div class="section" id="id4">
<h3>0.2.0<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Lots of small improvements in almost every module</p></li>
<li><p>Networks can be read from local files, remote files, lists or provided by any function</p></li>
<li><p>Almost all redistributed data have been removed, every source downloaded from the original provider.</p></li>
</ul>
</div>
<div class="section" id="id5">
<h3>0.3.0<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>First version with partial Python 3 support.</p></li>
</ul>
</div>
<div class="section" id="id6">
<h3>0.4.0<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><strong>pyreact</strong> module with <strong>BioPaxReader</strong> and <strong>PyReact</strong> classes added</p></li>
<li><p>Process description databases, BioPax and PathwayCommons SIF conversion rules are supported</p></li>
<li><p>Format definitions for 6 process description databases included.</p></li>
</ul>
</div>
<div class="section" id="id7">
<h3>0.5.0<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Many classes have been added to the <strong>plot</strong> module</p></li>
<li><p>All figures and tables in the manuscript can be generated automatically</p></li>
<li><p>This is supported by a new module, <strong>analysis</strong>, which implements a generic workflow in its <strong>Workflow</strong> class.</p></li>
</ul>
</div>
<div class="section" id="id8">
<h3>0.5.32<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><cite>chembl</cite>, <cite>unichem</cite>, <cite>mysql</cite> and <cite>mysql_connect</cite> modules made Python3 compatible</p></li>
</ul>
</div>
<div class="section" id="id9">
<h3>0.6.31<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Orthology translation of network</p></li>
<li><p>Homologene UniProt dict to translate between different organisms UniProt-to-UniProt</p></li>
<li><p>Orthology translation of PTMs</p></li>
<li><p>Better processing of PhosphoSite regulatory sites</p></li>
</ul>
</div>
<div class="section" id="id10">
<h3>0.7.0<a class="headerlink" href="#id10" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>TF-target, miRNA-mRNA and TF-miRNA interactions from many databases</p></li>
</ul>
</div>
<div class="section" id="id11">
<h3>0.7.74<a class="headerlink" href="#id11" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>New web server based on <cite>pandas</cite> data frames</p></li>
<li><p>New module <cite>export</cite> for generating data frames of interactions or enzyme-substrate interactions</p></li>
<li><p>New module <cite>websrvtab</cite> for exporting data frames for the web server</p></li>
<li><p>TF-target interactions from DoRothEA</p></li>
</ul>
</div>
<div class="section" id="id12">
<h3>0.7.93<a class="headerlink" href="#id12" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>New <cite>dataio</cite> methods for Gene Ontology</p></li>
</ul>
</div>
<div class="section" id="id13">
<h3>0.7.110<a class="headerlink" href="#id13" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Many new docstrings</p></li>
</ul>
</div>
<div class="section" id="id14">
<h3>0.8<a class="headerlink" href="#id14" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>New module <cite>complex</cite>: a comprehensive database of complexes</p></li>
<li><p>New module <cite>annot</cite>: database of protein annotations (function, location)</p></li>
<li><p>New module <cite>intercell</cite>: special methods for data integration focusing on intercellular communication</p></li>
<li><p>New module <cite>bel</cite>: BEL integration</p></li>
<li><p>Module <cite>go</cite> and all the connected <cite>dataio</cite> methods have been rewritten offering a workaround for
data access despite GO’s terrible web services and providing much more versatile query methods</p></li>
<li><p>Removed MySQL support (e.g. loading mapping tables from MySQL)</p></li>
<li><p>Modules <cite>mapping</cite>, <cite>reflists</cite>, <cite>complex</cite>, <cite>ptm</cite>, <cite>annot</cite>, <cite>go</cite> became services:
these modules build databases and provide query methods, sometimes they even automatically
delete data to free memory</p></li>
<li><p>New interaction category in <cite>data_formats</cite>: <cite>ligand_receptor</cite></p></li>
<li><p>Improved logging and control over verbosity</p></li>
<li><p>Better control over parameters by the <cite>settings</cite> module</p></li>
<li><p>Many methods in <cite>dataio</cite> have been improved or fixed, docs and code style largely improved</p></li>
<li><p>Started to add tests especially for methods in <cite>dataio</cite></p></li>
</ul>
</div>
<div class="section" id="id15">
<h3>0.9<a class="headerlink" href="#id15" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>The network database is not dependent any more on <cite>python-igraph</cite> hence it
has been removed from the mandatory dependencies</p></li>
<li><p>New API for the network, interactions, evidences, molecular entities</p></li>
</ul>
</div>
<div class="section" id="id16">
<h3>0.10.0<a class="headerlink" href="#id16" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>New module structure: modules grouped into <cite>core</cite>, <cite>inputs</cite>, <cite>internals</cite>,
<cite>legacy</cite>, <cite>omnipath</cite>, <cite>resources</cite>, <cite>share</cite> and <cite>utils</cite> submodules.</p></li>
</ul>
</div>
<div class="section" id="upcoming">
<h3>Upcoming<a class="headerlink" href="#upcoming" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>New, more flexible network reader class</p></li>
<li><p>Full support for multi-species molecular interaction networks
(e.g. pathogene-host)</p></li>
<li><p>Better support for not protein only molecular interaction networks
(metabolites, drug compounds, RNA)</p></li>
</ul>
</div>
</div>
<div class="section" id="features">
<h2>Features<a class="headerlink" href="#features" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><p><em>Warning:</em>
The sections below are outdated, will be updated soon</p>
</div></blockquote>
<p>In the beginning the primary aim of <strong>pypath</strong> was to build networks from
multiple sources using an igraph object as the fundament of the integrated
data structure. From version 0.7 and 0.8 this design principle started to
change. Today <strong>pypath</strong> builds a number of different databases each having
<strong>pandas.DataFrame</strong> as a final format. Each of these integrates a specific
kind of data from various databases (e.g. protein complexes, interactions,
enzyme-PTM relationships, etc). <strong>pypath</strong> has many submodules with standalone
functionality which can be used in other modules and scripts. For example
the ID conversion module <strong>pypath.mapping</strong>.</p>
<p>Submodules perform various features, e.g. graph visualization, working with
rug compound data, searching drug targets and compounds in <strong>ChEMBL</strong>.</p>
<div class="section" id="id-conversion">
<h3>ID conversion<a class="headerlink" href="#id-conversion" title="Permalink to this headline">¶</a></h3>
<p>The ID conversion module <code class="docutils literal notranslate"><span class="pre">utils.mapping</span></code> can be used independently. It has
the feature to translate secondary UniProt IDs to primaries, and Trembl IDs to
SwissProt, using primary Gene Symbols to find the connections. This module
automatically loads and stores the necessary conversion tables. Many tables
are predefined, such as all the IDs in <strong>UniProt mapping service,</strong> while
users are able to load any table from <strong>file</strong> or <strong>MySQL,</strong> using the classes
provided in the module <code class="docutils literal notranslate"><span class="pre">input_formats</span></code>.</p>
</div>
<div class="section" id="pathways">
<h3>Pathways<a class="headerlink" href="#pathways" title="Permalink to this headline">¶</a></h3>
<p><strong>pypath</strong> includes data and predefined format descriptions for more than 25
high quality, literature curated databases. The input formats are defined in
the <code class="docutils literal notranslate"><span class="pre">data_formats</span></code> module. For some resources data downloaded on the fly,
where it is not possible, data is redistributed with the module. Descriptions
and comprehensive information about the resources is available in the
<code class="docutils literal notranslate"><span class="pre">descriptions</span></code> module.</p>
</div>
<div class="section" id="structural-features">
<h3>Structural features<a class="headerlink" href="#structural-features" title="Permalink to this headline">¶</a></h3>
<p>One of the modules called <code class="docutils literal notranslate"><span class="pre">intera</span></code> provides many classes for representing
structures and mechanisms behind protein interactions. These are <code class="docutils literal notranslate"><span class="pre">Residue</span></code>
(optionally mutated), <code class="docutils literal notranslate"><span class="pre">Motif</span></code>, <code class="docutils literal notranslate"><span class="pre">Ptm</span></code>, <code class="docutils literal notranslate"><span class="pre">Domain</span></code>, <code class="docutils literal notranslate"><span class="pre">DomainMotif</span></code>,
<code class="docutils literal notranslate"><span class="pre">DomainDomain</span></code> and <code class="docutils literal notranslate"><span class="pre">Interface</span></code>. All these classes have <code class="docutils literal notranslate"><span class="pre">__eq__()</span></code>
methods to test equality between instances, and also <code class="docutils literal notranslate"><span class="pre">__contains__()</span></code>
methods to look up easily if a residue is within a short motif or protein
domain, or is the target residue of a PTM.</p>
</div>
<div class="section" id="sequences">
<h3>Sequences<a class="headerlink" href="#sequences" title="Permalink to this headline">¶</a></h3>
<p>The module <code class="docutils literal notranslate"><span class="pre">seq</span></code> contains a simple class for quick lookup any residue or
segment in <strong>UniProt</strong> protein sequences while being aware of isoforms.</p>
</div>
<div class="section" id="tissue-expression">
<h3>Tissue expression<a class="headerlink" href="#tissue-expression" title="Permalink to this headline">¶</a></h3>
<p>For three protein expression databases there are functions and modules for
downloading and combining the expression data with the network. These are the
Human Protein Atlas, the ProteomicsDB and GIANT. The <code class="docutils literal notranslate"><span class="pre">giant</span></code> and
<code class="docutils literal notranslate"><span class="pre">proteomicsdb</span></code> modules can be used also as stand alone Python clients for
these resources.</p>
</div>
<div class="section" id="functional-annotations">
<h3>Functional annotations<a class="headerlink" href="#functional-annotations" title="Permalink to this headline">¶</a></h3>
<p><strong>GSEA</strong> and <strong>Gene Ontology</strong> are two approaches for annotating genes and
gene products, and enrichment analysis technics aims to use these annotations
to highlight the biological functions a given set of genes is related to. Here
the <code class="docutils literal notranslate"><span class="pre">enrich</span></code> module gives abstract classes to calculate enrichment
statistics, while the <code class="docutils literal notranslate"><span class="pre">go</span></code> and the <code class="docutils literal notranslate"><span class="pre">gsea</span></code> modules give access to GO and
GSEA data, and make it easy to count enrichment statistics for sets of genes.</p>
</div>
<div class="section" id="drug-compounds">
<h3>Drug compounds<a class="headerlink" href="#drug-compounds" title="Permalink to this headline">¶</a></h3>
<p><strong>UniChem</strong> submodule provides an interface to effectively query the UniChem
service, use connectivity search with custom settings, and translate SMILEs to
ChEMBL IDs with ChEMBL web service.</p>
<p><strong>ChEMBL</strong> submodule queries directly your own ChEMBL MySQL instance, has the
features to search targets and compounds from custom assay types and
relationship types, to get activity values, binding domains, and action types.
You need to download the ChEMBL MySQL dump, and load into your own server.</p>
</div>
<div class="section" id="technical">
<h3>Technical<a class="headerlink" href="#technical" title="Permalink to this headline">¶</a></h3>
<p>The module <code class="docutils literal notranslate"><span class="pre">pypath.curl</span></code> provides a very flexible <strong>download manager</strong>
built on top of <code class="docutils literal notranslate"><span class="pre">pycurl</span></code>. The classes <code class="docutils literal notranslate"><span class="pre">pypath.curl.Curl()</span></code> and
<code class="docutils literal notranslate"><span class="pre">pypath.curl.FileOpener</span></code> accept numerous arguments to deal in a smart
way with local <strong>cache</strong>, authentication, redirects, uncompression, character
encodings, FTP and HTTP transactions, and many other stuff. Cache can grow to
several GBs, and takes place in <code class="docutils literal notranslate"><span class="pre">~/.pypath/cache</span></code> by default. If you
experience issues using <code class="docutils literal notranslate"><span class="pre">pypath</span></code> these are most often related to failed
downloads which often result nonsense cache contents. To debug such issues
you can see the cache file names and cache usage in the log, and you can use
the context managers in <code class="docutils literal notranslate"><span class="pre">pypath.curl</span></code> to show, delete or bypass the cache
for some particular method calls (<code class="docutils literal notranslate"><span class="pre">pypath.curl.cache_print_on()</span></code>,
<code class="docutils literal notranslate"><span class="pre">pypath.curl.cache_delete_on()</span></code> and <code class="docutils literal notranslate"><span class="pre">pypath.curl.cache_off()</span></code>.
You can always set up an alternative cache directory for the entire session
using the <code class="docutils literal notranslate"><span class="pre">pypath.settings</span></code> module.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">pypath.session</span></code> and <code class="docutils literal notranslate"><span class="pre">pypath.log</span></code> modules take care of setting up
session level parameters and logging. Each session has a random 5 character
identifier e.g. <code class="docutils literal notranslate"><span class="pre">y5jzx</span></code>. The default log file in this case is
<code class="docutils literal notranslate"><span class="pre">pypath_log/pypath-y5jzx.log</span></code>. The log messages are flushed every 2 seconds
by default. You can always change these things using the <code class="docutils literal notranslate"><span class="pre">settings</span></code> module.
In this module you can get and set the values of various parameters using
the <code class="docutils literal notranslate"><span class="pre">pypath.settings.setup()</span></code> and the <code class="docutils literal notranslate"><span class="pre">pypath.settings.get()</span></code> methods.</p>
<p>A simple <strong>webservice</strong> comes with this module: the <code class="docutils literal notranslate"><span class="pre">server</span></code> module based on
<code class="docutils literal notranslate"><span class="pre">twisted.web.server</span></code> opens a custom port and serves plain text tables over
HTTP with REST style querying.</p>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Dénes Türei, Nicolàs Palacio, Olga Ivanova. 
European Molecular Biology Laboratory, Heidelberg Germany
European Bioinformatics Institute, Hinxton UK
University Hospital RWTH, Aachen Germany
University Hospital Heidelberg Germany

    </p>
  </div>
    
    
      Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>