
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>&lt;no title&gt; &#8212; pypath 0.7.97 documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="pypath: A Python module for molecular interaction networks" href="index.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="index.html" title="pypath: A Python module for molecular interaction networks"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">pypath 0.7.97 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-pypath.main"></span><dl class="class">
<dt id="pypath.main.PyPath">
<em class="property">class </em><code class="descclassname">pypath.main.</code><code class="descname">PyPath</code><span class="sig-paren">(</span><em>ncbi_tax_id=9606</em>, <em>default_name_type={'drug': 'chembl'</em>, <em>'lncrna': 'lncrna-genesymbol'</em>, <em>'mirna': 'mirbase'</em>, <em>'protein': 'uniprot'}</em>, <em>copy=None</em>, <em>mysql=(None</em>, <em>'mapping')</em>, <em>chembl_mysql=(None</em>, <em>'chembl')</em>, <em>name='unnamed'</em>, <em>outdir='results'</em>, <em>loglevel='INFO'</em>, <em>loops=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath" title="Permalink to this definition">Â¶</a></dt>
<dd><ul class="simple">
<li><dl class="first docutils">
<dt>Attributes:</dt>
<dd><ul class="first last">
<li><em>acsn_effects</em> []:</li>
<li><em>add_genesets</em> []:</li>
<li><em>add_grouped_eattr</em> []:</li>
<li><em>add_grouped_set_eattr</em> []:</li>
<li><em>add_list_eattr</em> []:</li>
<li><em>add_set_eattr</em> []:</li>
<li><em>add_update_edge</em> []:</li>
<li><em>add_update_vertex</em> []:</li>
<li><em>affected_by</em> []:</li>
<li><em>affects</em> []:</li>
<li><em>all_between</em> []:</li>
<li><em>all_neighbours</em> []:</li>
<li><em>apply_list</em> []:</li>
<li><em>apply_negative</em> []:</li>
<li><em>attach_network</em> []:</li>
<li><em>basic_stats</em> []:</li>
<li><em>basic_stats_intergroup</em> []:</li>
<li><em>cancer_drivers_list</em> []:</li>
<li><em>cancer_gene_census_list</em> []:</li>
<li><em>clean_graph</em> []:</li>
<li><em>collapse_by_name</em> []:</li>
<li><em>combine_attr</em> []:</li>
<li><em>communities</em> []:</li>
<li><em>complex_comembership_network</em> []:</li>
<li><em>complexes</em> []:</li>
<li><em>complexes_in_network</em> []:</li>
<li><em>compounds_from_chembl</em> []:</li>
<li><em>consistency</em> []:</li>
<li><em>copy</em> []:</li>
<li><em>copy_edges</em> []:</li>
<li><em>count_sol</em> []:</li>
<li><em>coverage</em> []:</li>
<li><em>curation_effort</em> []:</li>
<li><em>curation_stats</em> []:</li>
<li><em>curation_tab</em> []:</li>
<li><em>curators_work</em> []:</li>
<li><em>databases_similarity</em> []:</li>
<li><em>default_name_type</em> []:</li>
<li><em>degree_dist</em> []:</li>
<li><em>degree_dists</em> []:</li>
<li><em>delete_by_source</em> []:</li>
<li><em>delete_by_taxon</em> []:</li>
<li><em>delete_unknown</em> []:</li>
<li><em>delete_unmapped</em> []:</li>
<li><em>dgenesymbol</em> []:</li>
<li><em>dgenesymbols</em> []:</li>
<li><em>dgs</em> []:</li>
<li><em>dgss</em> []:</li>
<li><em>disease_genes_list</em> []:</li>
<li><em>dneighbors</em> []:</li>
<li><em>dp</em> []:</li>
<li><em>dproteins</em> []:</li>
<li><em>dps</em> []:</li>
<li><em>druggability_list</em> []:</li>
<li><em>duniprot</em> []:</li>
<li><em>duniprots</em> []:</li>
<li><em>dup</em> []:</li>
<li><em>dups</em> []:</li>
<li><em>dv</em> []:</li>
<li><em>dvs</em> []:</li>
<li><em>edge_exists</em> []:</li>
<li><em>edge_loc</em> []:</li>
<li><em>edge_names</em> []:</li>
<li><em>edges_3d</em> []:</li>
<li><em>edges_expression</em> []:</li>
<li><em>edges_in_comlexes</em> []:</li>
<li><em>edges_ptms</em> []:</li>
<li><em>edgeseq_inverse</em> []:</li>
<li><em>export_dot</em> []:</li>
<li><em>export_edgelist</em> []:</li>
<li><em>export_graphml</em> []:</li>
<li><em>export_ptms_tab</em> []:</li>
<li><em>export_sif</em> []:</li>
<li><em>export_struct_tab</em> []:</li>
<li><em>export_tab</em> []:</li>
<li><em>filters</em> []:</li>
<li><em>find_all_paths</em> []:</li>
<li><em>find_all_paths2</em> []:</li>
<li><em>find_complex</em> []:</li>
<li><em>first_neighbours</em> []:</li>
<li><em>fisher_enrichment</em> []:</li>
<li><em>geneset_enrichment</em> []:</li>
<li><em>genesymbol</em> []:</li>
<li><em>genesymbol_labels</em> []:</li>
<li><em>genesymbols</em> []:</li>
<li><em>get_attrs</em> []:</li>
<li><em>get_directed</em> []:</li>
<li><em>get_dirs_signs</em> []:</li>
<li><em>get_edge</em> []:</li>
<li><em>get_edges</em> []:</li>
<li><em>get_function</em> []:</li>
<li><em>get_giant</em> []:</li>
<li><em>get_max</em> []:</li>
<li><em>get_network</em> []:</li>
<li><em>get_node</em> []:</li>
<li><em>get_node_d</em> []:</li>
<li><em>get_node_pair</em> []:</li>
<li><em>get_nodes</em> []:</li>
<li><em>get_nodes_d</em> []:</li>
<li><em>get_pathways</em> []:</li>
<li><em>get_proteomicsdb</em> []:</li>
<li><em>get_sub</em> []:</li>
<li><em>get_taxon</em> []:</li>
<li><em>go_annotate</em> []:</li>
<li><em>go_dict</em> []:</li>
<li><em>go_enrichment</em> []:</li>
<li><em>gs</em> []:</li>
<li><em>gs_affected_by</em> []:</li>
<li><em>gs_affects</em> []:</li>
<li><em>gs_edge</em> []:</li>
<li><em>gs_in_directed</em> []:</li>
<li><em>gs_in_undirected</em> []:</li>
<li><em>gs_inhibited_by</em> []:</li>
<li><em>gs_inhibits</em> []:</li>
<li><em>gs_neighborhood</em> []:</li>
<li><em>gs_neighbors</em> []:</li>
<li><em>gs_stimulated_by</em> []:</li>
<li><em>gs_stimulates</em> []:</li>
<li><em>gss</em> []:</li>
<li><em>guide2pharma</em> []:</li>
<li><em>having_attr</em> []:</li>
<li><em>having_eattr</em> []:</li>
<li><em>having_ptm</em> []:</li>
<li><em>having_vattr</em> []:</li>
<li><em>homology_translation</em> []:</li>
<li><em>htp_stats</em> []:</li>
<li><em>in_complex</em> []:</li>
<li><em>in_directed</em> []:</li>
<li><em>in_undirected</em> []:</li>
<li><em>info</em> []:</li>
<li><em>init_complex_attr</em> []:</li>
<li><em>init_edge_attr</em> []:</li>
<li><em>init_gsea</em> []:</li>
<li><em>init_network</em> []:</li>
<li><em>init_vertex_attr</em> []:</li>
<li><em>intergroup_shortest_paths</em> []:</li>
<li><em>intogen_cancer_drivers_list</em> []:</li>
<li><em>jaccard_edges</em> []:</li>
<li><em>jaccard_meta</em> []:</li>
<li><em>kegg_directions</em> []:</li>
<li><em>kegg_pathways</em> []:</li>
<li><em>kinase_stats</em> []:</li>
<li><em>kinases_list</em> []:</li>
<li><em>label_by_go</em> []:</li>
<li><em>laudanna_directions</em> []:</li>
<li><em>laudanna_effects</em> []:</li>
<li><em>licence</em> []:</li>
<li><em>list_resources</em> []:</li>
<li><em>load_3dcomplexes</em> []:</li>
<li><em>load_3did_ddi</em> []:</li>
<li><em>load_3did_ddi2</em> []:</li>
<li><em>load_3did_dmi</em> []:</li>
<li><em>load_3did_interfaces</em> []:</li>
<li><em>load_all_pathways</em> []:</li>
<li><em>load_compleat</em> []:</li>
<li><em>load_complexportal</em> []:</li>
<li><em>load_comppi</em> []:</li>
<li><em>load_corum</em> []:</li>
<li><em>load_dbptm</em> []:</li>
<li><em>load_ddi</em> []:</li>
<li><em>load_ddis</em> []:</li>
<li><em>load_depod_dmi</em> []:</li>
<li><em>load_disgenet</em> []:</li>
<li><em>load_dmi</em> []:</li>
<li><em>load_dmis</em> []:</li>
<li><em>load_domino_dmi</em> []:</li>
<li><em>load_elm</em> []:</li>
<li><em>load_expression</em> []:</li>
<li><em>load_go</em> []:</li>
<li><em>load_havugimana</em> []:</li>
<li><em>load_hpa</em> []:</li>
<li><em>load_hprd_ptms</em> []:</li>
<li><em>load_ielm</em> []:</li>
<li><em>load_interfaces</em> []:</li>
<li><em>load_li2012_ptms</em> []:</li>
<li><em>load_ligand_receptor_network</em> []:</li>
<li><em>load_list</em> []:</li>
<li><em>load_lmpid</em> []:</li>
<li><em>load_mappings</em> []:</li>
<li><em>load_mimp_dmi</em> []:</li>
<li><em>load_mutations</em> []:</li>
<li><em>load_negatives</em> []:</li>
<li><em>load_old_omnipath</em> []:</li>
<li><em>load_omnipath</em> []:</li>
<li><em>load_pathways</em> []:</li>
<li><em>load_pdb</em> []:</li>
<li><em>load_pepcyber</em> []:</li>
<li><em>load_pfam</em> []:</li>
<li><em>load_pfam2</em> []:</li>
<li><em>load_pfam3</em> []:</li>
<li><em>load_phospho_dmi</em> []:</li>
<li><em>load_phosphoelm</em> []:</li>
<li><em>load_pisa</em> []:</li>
<li><em>load_pnetworks_dmi</em> []:</li>
<li><em>load_psite_phos</em> []:</li>
<li><em>load_psite_reg</em> []:</li>
<li><em>load_ptms</em> []:</li>
<li><em>load_ptms2</em> []:</li>
<li><em>load_reflist</em> []:</li>
<li><em>load_reflists</em> []:</li>
<li><em>load_resource</em> []:</li>
<li><em>load_resources</em> []:</li>
<li><em>load_signor_ptms</em> []:</li>
<li><em>load_tfregulons</em> []:</li>
<li><em>lookup_cache</em> []:</li>
<li><em>loop_edges</em> []:</li>
<li><em>map_edge</em> []:</li>
<li><em>map_item</em> []:</li>
<li><em>map_list</em> []:</li>
<li><em>mean_reference_per_interaction</em> []:</li>
<li><em>merge_lists</em> []:</li>
<li><em>merge_nodes</em> []:</li>
<li><em>mimp_directions</em> []:</li>
<li><em>mutated_edges</em> []:</li>
<li><em>names2vids</em> []:</li>
<li><em>negative_report</em> []:</li>
<li><em>neighborhood</em> []:</li>
<li><em>neighbors</em> []:</li>
<li><em>neighbourhood_network</em> []:</li>
<li><em>network_filter</em> []:</li>
<li><em>network_stats</em> []:</li>
<li><em>new_edges</em> []:</li>
<li><em>new_nodes</em> []:</li>
<li><em>node_exists</em> []:</li>
<li><em>numof_directed_edges</em> []:</li>
<li><em>numof_reference_interaction_pairs</em> []:</li>
<li><em>numof_references</em> []:</li>
<li><em>numof_undirected_edges</em> []:</li>
<li><em>orthology_translation</em> []:</li>
<li><em>p</em> []:</li>
<li><em>pathway_attributes</em> []:</li>
<li><em>pathway_members</em> []:</li>
<li><em>pathway_names</em> []:</li>
<li><em>pathway_similarity</em> []:</li>
<li><em>pathways_table</em> []:</li>
<li><em>pfam_regions</em> []:</li>
<li><em>phosphonetworks_directions</em> []:</li>
<li><em>phosphopoint_directions</em> []:</li>
<li><em>phosphorylation_directions</em> []:</li>
<li><em>phosphorylation_signs</em> []:</li>
<li><em>phosphosite_directions</em> []:</li>
<li><em>prdb_tissue_expr</em> []:</li>
<li><em>process_direction</em> []:</li>
<li><em>process_directions</em> []:</li>
<li><em>process_dmi</em> []:</li>
<li><em>process_sign</em> []:</li>
<li><em>protein</em> []:</li>
<li><em>protein_edge</em> []:</li>
<li><em>proteins</em> []:</li>
<li><em>proteome_list</em> []:</li>
<li><em>ps</em> []:</li>
<li><em>random_walk_with_return</em> []:</li>
<li><em>random_walk_with_return2</em> []:</li>
<li><em>read_data_file</em> []:</li>
<li><em>read_from_cache</em> []:</li>
<li><em>read_list_file</em> []:</li>
<li><em>receptors_list</em> []:</li>
<li><em>reference_edge_ratio</em> []:</li>
<li><em>reference_hist</em> []:</li>
<li><em>reload</em> []:</li>
<li><em>remove_htp</em> []:</li>
<li><em>remove_undirected</em> []:</li>
<li><em>run_batch</em> []:</li>
<li><em>save_network</em> []:</li>
<li><em>save_session</em> []:</li>
<li><em>search_attr_and</em> []:</li>
<li><em>search_attr_or</em> []:</li>
<li><em>second_neighbours</em> []:</li>
<li><em>select_by_go</em> []:</li>
<li><em>separate</em> []:</li>
<li><em>separate_by_category</em> []:</li>
<li><em>sequences</em> []:</li>
<li><em>set_boolean_vattr</em> []:</li>
<li><em>set_categories</em> []:</li>
<li><em>set_chembl_mysql</em> []:</li>
<li><em>set_disease_genes</em> []:</li>
<li><em>set_druggability</em> []:</li>
<li><em>set_drugtargets</em> []:</li>
<li><em>set_kinases</em> []:</li>
<li><em>set_receptors</em> []:</li>
<li><em>set_signaling_proteins</em> []:</li>
<li><em>set_tfs</em> []:</li>
<li><em>set_transcription_factors</em> []:</li>
<li><em>shortest_path_dist</em> []:</li>
<li><em>signaling_proteins_list</em> []:</li>
<li><em>signor_pathways</em> []:</li>
<li><em>similarity_groups</em> []:</li>
<li><em>small_plot</em> []:</li>
<li><em>sorensen_pathways</em> []:</li>
<li><em>source_diagram</em> []:</li>
<li><em>source_network</em> []:</li>
<li><em>source_similarity</em> []:</li>
<li><em>source_stats</em> []:</li>
<li><em>sources_hist</em> []:</li>
<li><em>sources_overlap</em> []:</li>
<li><em>sources_venn_data</em> []:</li>
<li><em>straight_between</em> []:</li>
<li><em>string_effects</em> []:</li>
<li><em>sum_in_complex</em> []:</li>
<li><em>table_latex</em> []:</li>
<li><em>tfs_list</em> []:</li>
<li><em>third_source_directions</em> []:</li>
<li><em>tissue_network</em> []:</li>
<li><em>transcription_factors</em> []:</li>
<li><em>translate_refsdir</em> []:</li>
<li><em>uniprot</em> []:</li>
<li><em>uniprots</em> []:</li>
<li><em>uniq_node_list</em> []:</li>
<li><em>uniq_ptm</em> []:</li>
<li><em>uniq_ptms</em> []:</li>
<li><em>up</em> []:</li>
<li><em>up_affected_by</em> []:</li>
<li><em>up_affects</em> []:</li>
<li><em>up_edge</em> []:</li>
<li><em>up_in_directed</em> []:</li>
<li><em>up_in_undirected</em> []:</li>
<li><em>up_inhibited_by</em> []:</li>
<li><em>up_inhibits</em> []:</li>
<li><em>up_neighborhood</em> []:</li>
<li><em>up_neighbors</em> []:</li>
<li><em>up_stimulated_by</em> []:</li>
<li><em>up_stimulates</em> []:</li>
<li><em>update_adjlist</em> []:</li>
<li><em>update_attrs</em> []:</li>
<li><em>update_cats</em> []:</li>
<li><em>update_db_dict</em> []:</li>
<li><em>update_pathway_types</em> []:</li>
<li><em>update_pathways</em> []:</li>
<li><em>update_sources</em> []:</li>
<li><em>update_vertex_sources</em> []:</li>
<li><em>update_vindex</em> []:</li>
<li><em>update_vname</em> []:</li>
<li><em>ups</em> []:</li>
<li><em>v</em> []:</li>
<li><em>vertex_pathways</em> []:</li>
<li><em>vs</em> []:</li>
<li><em>vsgs</em> []:</li>
<li><em>vsup</em> []:</li>
<li><em>wang_effects</em> []:</li>
<li><em>write_table</em> []:</li>
</ul>
</dd>
</dl>
</li>
</ul>
<dl class="method">
<dt id="pypath.main.PyPath.add_update_vertex">
<code class="descname">add_update_vertex</code><span class="sig-paren">(</span><em>defAttrs</em>, <em>originalName</em>, <em>originalNameType</em>, <em>extraAttrs={}</em>, <em>add=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.add_update_vertex"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.add_update_vertex" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Updates the attributes of one node in the network.
Optionally it creates a new node and sets the attributes,
but it is not efficient as igraph needs to reindex vertices
after this operation, so better to create new nodes and
edges in batch.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.all_between">
<code class="descname">all_between</code><span class="sig-paren">(</span><em>nameA</em>, <em>nameB</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.all_between"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.all_between" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns all edges between two given vertex names. Similar to
straight_between(), but checks both directions, and returns
list of edge ids in [undirected, straight, reversed] format,
for both nameA -&gt; nameB and nameB -&gt; nameA edges.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.apply_list">
<code class="descname">apply_list</code><span class="sig-paren">(</span><em>name</em>, <em>node_or_edge='node'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.apply_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.apply_list" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Creates vertex or edge attribute based on a list.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.attach_network">
<code class="descname">attach_network</code><span class="sig-paren">(</span><em>edgeList=False</em>, <em>regulator=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.attach_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.attach_network" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Adds edges to the network from edgeList obtained from file or
other input method.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.basic_stats">
<code class="descname">basic_stats</code><span class="sig-paren">(</span><em>latex=False, caption='', latex_hdr=True, fontsize=8, font='HelveticaNeueLTStd-LtCn', fname=None, header_format='%s', row_order=None, by_category=True, use_cats=['p', 'm', 'i', 'r'], urls=True, annots=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.basic_stats"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.basic_stats" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns basic numbers about the network resources, e.g. edge and
node counts.</p>
<dl class="docutils">
<dt>latex</dt>
<dd>Return table in a LaTeX document. This can be compiled by
PDFLaTeX:
latex stats.tex</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.cancer_gene_census_list">
<code class="descname">cancer_gene_census_list</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.cancer_gene_census_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.cancer_gene_census_list" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Loads the list of cancer driver proteins from the COSMIC Cancer
Gene Census.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.clean_graph">
<code class="descname">clean_graph</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.clean_graph"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.clean_graph" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Removes multiple edges, unknown molecules and those from wrong taxon.
Multiple edges will be combined by <cite>combine_attr()</cite> method.
Loops will be deleted unless the <cite>loops</cite> attribute set to <cite>True</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.collapse_by_name">
<code class="descname">collapse_by_name</code><span class="sig-paren">(</span><em>graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.collapse_by_name"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.collapse_by_name" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Collapses nodes with the same name with copying and merging
all edges and attributes.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.combine_attr">
<code class="descname">combine_attr</code><span class="sig-paren">(</span><em>lst</em>, <em>num_method=&lt;built-in function max&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.combine_attr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.combine_attr" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Combines multiple attributes into one. This method attempts
to find out which is the best way to combine attributes.</p>
<blockquote>
<div><ul class="simple">
<li>if there is only one value or one of them is None, then returns
the one available</li>
<li>lists: concatenates unique values of lists</li>
<li>numbers: returns the greater by default
or calls <cite>num_method()</cite> if given.</li>
<li>sets: returns the union</li>
<li>dicts: calls <cite>common.merge_dicts()</cite></li>
<li>Direction: calls their special <cite>merge()</cite> method</li>
</ul>
</div></blockquote>
<p>Works on more than 2 attributes recursively.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>lst</strong> (<em>list</em>) â List of one or two attribute values.</li>
<li><strong>num_method</strong> (<em>callable</em>) â Method to merge numeric attributes.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.copy_edges">
<code class="descname">copy_edges</code><span class="sig-paren">(</span><em>sources</em>, <em>target</em>, <em>move=False</em>, <em>graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.copy_edges"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.copy_edges" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Copies edges of one node to another,
keeping attributes and directions.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sources</strong> (<em>list</em>) â Vertex IDs to copy from.</li>
<li><strong>target</strong> (<em>int</em>) â Vertex ID to copy for.</li>
<li><strong>move</strong> (<em>bool</em>) â Whether perform copy or move, i.e. remove or keep
the source edges.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.count_sol">
<code class="descname">count_sol</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.count_sol"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.count_sol" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Counts nodes with zero degree.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.curation_effort">
<code class="descname">curation_effort</code><span class="sig-paren">(</span><em>sum_by_source=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.curation_effort"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.curation_effort" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns the total number of reference-interactions pairs.</p>
<dl class="docutils">
<dt>&#64;sum_by_source <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd>If True, counts the refrence-interaction pairs by
sources, and returns the sum of these values.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.delete_by_taxon">
<code class="descname">delete_by_taxon</code><span class="sig-paren">(</span><em>tax</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.delete_by_taxon"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.delete_by_taxon" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Removes the proteins of all organisms which are not listed.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>tax</strong> (<em>list</em>) â List of NCBI Taxonomy IDs of the organisms.
E.g. [7227, 9606]</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.delete_unknown">
<code class="descname">delete_unknown</code><span class="sig-paren">(</span><em>tax</em>, <em>typ='protein'</em>, <em>defaultNameType=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.delete_unknown"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.delete_unknown" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Removes those proteins which are not in the list of all default
IDs of the organisms. By default, it means to remove all protein
nodes not having a human SwissProt ID.</p>
<dl class="docutils">
<dt>&#64;tax <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd>List of NCBI Taxonomy IDs of the organisms of interest.
E.g. [7227, 9606]</dd>
<dt>&#64;typ <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>Molecule type. E.g. âproteinâ or âmirnaâ</dd>
<dt>&#64;defaultNameType <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>The default name type of the given molecular species.
For proteins itâs âuniprotâ by default.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.dgenesymbol">
<code class="descname">dgenesymbol</code><span class="sig-paren">(</span><em>genesymbol</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.dgenesymbol"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.dgenesymbol" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the GeneSymbol
can be found in the default directed network,
otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;genesymbol <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>GeneSymbol.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.dgs">
<code class="descname">dgs</code><span class="sig-paren">(</span><em>genesymbol</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.dgs" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the GeneSymbol
can be found in the default directed network,
otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;genesymbol <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>GeneSymbol.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.disease_genes_list">
<code class="descname">disease_genes_list</code><span class="sig-paren">(</span><em>dataset='curated'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.disease_genes_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.disease_genes_list" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Loads the list of all disease related genes from DisGeNet.
This resource is human only.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.dp">
<code class="descname">dp</code><span class="sig-paren">(</span><em>identifier</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.dp" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Same as <code class="docutils literal notranslate"><span class="pre">PyPath.get_node</span></code>, just for the directed graph.
Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the identifier
is a valid vertex index in the default directed graph,
or a UniProt ID or GeneSymbol which can be found in the
default directed network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;identifier <span class="classifier-delimiter">:</span> <span class="classifier">int, str</span></dt>
<dd>Vertex index (int) or GeneSymbol (str) or UniProt ID (str).</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.druggability_list">
<code class="descname">druggability_list</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.druggability_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.druggability_list" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Loads the list of druggable proteins from DgiDB.
This resource is human only.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.duniprot">
<code class="descname">duniprot</code><span class="sig-paren">(</span><em>uniprot</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.duniprot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.duniprot" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Same as <code class="docutils literal notranslate"><span class="pre">PyPath.uniprot(),</span> <span class="pre">just</span> <span class="pre">for</span> <span class="pre">directed</span> <span class="pre">graph.</span>
<span class="pre">Returns</span> <span class="pre">``igraph.Vertex()</span></code> object if the UniProt
can be found in the default directed network,
otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;uniprot <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>UniProt ID.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.duniprots">
<code class="descname">duniprots</code><span class="sig-paren">(</span><em>uniprots</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.duniprots"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.duniprots" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns list of <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object
for a list of UniProt IDs omitting those
could not be found in the default
directed graph.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.dup">
<code class="descname">dup</code><span class="sig-paren">(</span><em>uniprot</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.dup" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Same as <code class="docutils literal notranslate"><span class="pre">PyPath.uniprot(),</span> <span class="pre">just</span> <span class="pre">for</span> <span class="pre">directed</span> <span class="pre">graph.</span>
<span class="pre">Returns</span> <span class="pre">``igraph.Vertex()</span></code> object if the UniProt
can be found in the default directed network,
otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;uniprot <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>UniProt ID.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.dups">
<code class="descname">dups</code><span class="sig-paren">(</span><em>uniprots</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.dups" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns list of <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object
for a list of UniProt IDs omitting those
could not be found in the default
directed graph.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.dv">
<code class="descname">dv</code><span class="sig-paren">(</span><em>identifier</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.dv" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Same as <code class="docutils literal notranslate"><span class="pre">PyPath.get_node</span></code>, just for the directed graph.
Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the identifier
is a valid vertex index in the default directed graph,
or a UniProt ID or GeneSymbol which can be found in the
default directed network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;identifier <span class="classifier-delimiter">:</span> <span class="classifier">int, str</span></dt>
<dd>Vertex index (int) or GeneSymbol (str) or UniProt ID (str).</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.edge_exists">
<code class="descname">edge_exists</code><span class="sig-paren">(</span><em>nameA</em>, <em>nameB</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.edge_exists"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.edge_exists" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns a tuple of vertice indices if edge doesnât exists,
otherwise edge id. Not sensitive to direction.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.edges_expression">
<code class="descname">edges_expression</code><span class="sig-paren">(</span><em>func=&lt;function &lt;lambda&gt;&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.edges_expression"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.edges_expression" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Executes function <cite>func</cite> for each pairs of connected proteins in the
network, for every expression dataset. By default, <cite>func</cite> simply
gives the product the (normalized) expression values.</p>
<dl class="docutils">
<dt>func <span class="classifier-delimiter">:</span> <span class="classifier">callable</span></dt>
<dd>Function to handle 2 vectors (pandas.Series() objects), should
return one vector of the same length.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.edges_in_comlexes">
<code class="descname">edges_in_comlexes</code><span class="sig-paren">(</span><em>csources=['corum'], graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.edges_in_comlexes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.edges_in_comlexes" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Creates edge attributes <code class="docutils literal notranslate"><span class="pre">complexes</span></code> and <code class="docutils literal notranslate"><span class="pre">in_complex</span></code>.
These are both dicts where the keys are complex resources.
The values in <code class="docutils literal notranslate"><span class="pre">complexes</span></code> are the list of complex names
both the source and the target vertices belong to.
The values <code class="docutils literal notranslate"><span class="pre">in_complex</span></code> are boolean values whether there
is at least one complex in the given resources both the
source and the target vertex of the edge belong to.</p>
<dl class="docutils">
<dt>&#64;csources <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd>List of complex resources. Should be already loaded.</dd>
<dt>&#64;graph <span class="classifier-delimiter">:</span> <span class="classifier">igraph.Graph()</span></dt>
<dd>The graph object to do the calculations on.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.export_dot">
<code class="descname">export_dot</code><span class="sig-paren">(</span><em>nodes=None</em>, <em>edges=None</em>, <em>directed=True</em>, <em>labels='genesymbol'</em>, <em>edges_filter=&lt;function &lt;lambda&gt;&gt;</em>, <em>nodes_filter=&lt;function &lt;lambda&gt;&gt;</em>, <em>edge_sources=None</em>, <em>dir_sources=None</em>, <em>graph=None</em>, <em>return_object=False</em>, <em>save_dot=None</em>, <em>save_graphics=None</em>, <em>prog='neato'</em>, <em>format=None</em>, <em>hide=False</em>, <em>font=None</em>, <em>auto_edges=False</em>, <em>hide_nodes=[]</em>, <em>defaults={}</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.export_dot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.export_dot" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Builds a pygraphviz.AGraph() object with filtering the edges
and vertices along arbitrary criteria.
Returns the Agraph object if requesred, or exports the dot
file, or saves the graphics.</p>
<p>&#64;nodes : list
List of vertex ids to be included.
&#64;edges : list
List of edge ids to be included.
&#64;directed : bool
Create a directed or undirected graph.
&#64;labels : str
Name type to be used as id/label in the dot format.
&#64;edges_filter : function
Function to filter edges, accepting igraph.Edge as argument.
&#64;nodes_filter : function
Function to filter vertices, accepting igraph.Vertex as argument.
&#64;edge_sources : list
Sources to be included.
&#64;dir_sources : list
Direction and effect sources to be included.
&#64;graph : igraph.Graph
The graph object to export.
&#64;return_object : bool
Whether to return the pygraphviz.AGraph object.
&#64;save_dot : str
Filename to export the dot file to.
&#64;save_graphics : str
Filename to export the graphics, the extension defines the format.
&#64;prog : str
The graphviz layout algorithm to use.
&#64;format : str
The graphics format passed to pygraphviz.AGrapg().draw().
&#64;hide : bool
Hide filtered edges instead of omit them.
&#64;hide nodes : list
Nodes to hide. List of vertex ids.
&#64;auto_edges : str
Automatic, built-in style for edges.
âDIRECTIONSâ or âRESOURCE_CATEGORIESâ are supported.
&#64;font : str
Font to use for labels.
For using more than one fonts refer to graphviz attributes with constant values
or define callbacks or mapping dictionaries.
&#64;defaults : dict
Default values for graphviz attributes, labeled with the entity, e.g.
<cite>{âedge_penwidthâ: 0.2}</cite>.
&#64;**kwargs : constant, callable or dict
Graphviz attributes, labeled by the target entity. E.g. <cite>edge_penwidth</cite>,
âvertex_shape` or <cite>graph_label</cite>.
If the value is constant, this value will be used.
If the value is dict, and has <cite>_name</cite> as key, for every instance of the
given entity, the value of the attribute defined by <cite>_name</cite> will be looked
up in the dict, and the corresponding value will be given to the graphviz
attribute. If the key <cite>_name</cite> is missing from the dict, igraph vertex and
edge indices will be looked up among the keys.
If the value is callable, it will be called with the current instance of
the entity and the returned value will be used for the graphviz attribute.
E.g. <cite>edge_arrowhead(edge)</cite> or <cite>vertex_fillcolor(vertex)</cite>
Example:</p>
<blockquote>
<div><p>import pypath
from pypath import data_formats
net = pypath.PyPath()
net.init_network(pfile = âcache/default.pickleâ)
#net.init_network({âarnâ: data_formats.omnipath[âarnâ]})
tgf = [v.index for v in net.graph.vs if âTGFâ in v[âslk_pathwaysâ]]
dot = net.export_dot(nodes = tgf, save_graphics = âtgf_slk.pdfâ, prog = âdotâ,</p>
<blockquote>
<div>main_title = âTGF-beta pathwayâ, return_object = True,
label_font = âHelveticaNeueLTStd Med Cnâ,
edge_sources = [âSignaLink3â],
dir_sources = [âSignaLink3â], hide = True)</div></blockquote>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.export_edgelist">
<code class="descname">export_edgelist</code><span class="sig-paren">(</span><em>fname, graph=None, names=['name'], edge_attributes=[], sep='\t'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.export_edgelist"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.export_edgelist" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Write edge list to text file with attributes</p>
<p>&#64;param fname: the name of the file or a stream to read from.
&#64;param graph: the igraph object containing the network
&#64;param names: list with the vertex attribute names to be printed</p>
<blockquote>
<div>for source and target vertices</div></blockquote>
<dl class="docutils">
<dt>&#64;param edge_attributes: list with the edge attribute names</dt>
<dd>to be printed</dd>
</dl>
<p>&#64;param sep: string used to separate columns</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.export_tab">
<code class="descname">export_tab</code><span class="sig-paren">(</span><em>outfile=None</em>, <em>extra_node_attrs={}</em>, <em>extra_edge_attrs={}</em>, <em>unique_pairs=True</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.export_tab"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.export_tab" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Exports the network in a tabular format.</p>
<p>By default UniProt IDs, Gene Symbols, source databases, literature
references, directionality and sign information and interaction type
are included.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>outfile</strong> (<em>str</em>) â Name of the output file. If <cite>None</cite> a file name
ânetrowk-&lt;session id&gt;.tabâ is used.</li>
<li><strong>extra_node_attrs</strong> (<em>dict</em>) â Additional node attributes to be included in the exported table.
Keys are column ames used in the header while values are names
of vertex attributes. In the header <cite>_A</cite> and <cite>_B</cite> suffixes will
be appended to the column names so the values can be assigned to
A and B side interaction partners.</li>
<li><strong>extra_edge_attrs</strong> (<em>dict</em>) â Additional edge attributes to be included in the exported table.
Keys are column ames used in the header while values are names
of edge attributes.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.find_all_paths">
<code class="descname">find_all_paths</code><span class="sig-paren">(</span><em>start</em>, <em>end</em>, <em>mode='OUT'</em>, <em>maxlen=2</em>, <em>graph=None</em>, <em>silent=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.find_all_paths"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.find_all_paths" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Finds all paths up to length <cite>maxlen</cite> between groups of
vertices. This function is needed only becaues igraph`s
get_all_shortest_paths() finds only the shortest, not any
path up to a defined length.</p>
<dl class="docutils">
<dt>&#64;start <span class="classifier-delimiter">:</span> <span class="classifier">int or list</span></dt>
<dd>Indices of the starting node(s) of the paths.</dd>
<dt>&#64;end <span class="classifier-delimiter">:</span> <span class="classifier">int or list</span></dt>
<dd>Indices of the target node(s) of the paths.</dd>
<dt>&#64;mode <span class="classifier-delimiter">:</span> <span class="classifier">âINâ, âOUTâ, âALLâ</span></dt>
<dd>Passed to igraph.Graph.neighbors()</dd>
<dt>&#64;maxlen <span class="classifier-delimiter">:</span> <span class="classifier">int</span></dt>
<dd>Maximum length of paths in steps, i.e. if maxlen = 3, then
the longest path may consist of 3 edges and 4 nodes.</dd>
<dt>&#64;graph <span class="classifier-delimiter">:</span> <span class="classifier">igraph.Graph object</span></dt>
<dd>The graph you want to find paths in. self.graph by default.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.find_complex">
<code class="descname">find_complex</code><span class="sig-paren">(</span><em>search</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.find_complex"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.find_complex" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Finds complexes by their non standard names.
E.g. to find DNA polymerases you can use the search
term <cite>DNA pol</cite> which will be tested against complex names
in CORUM.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.genesymbol">
<code class="descname">genesymbol</code><span class="sig-paren">(</span><em>genesymbol</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.genesymbol"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.genesymbol" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the GeneSymbol
can be found in the default undirected network,
otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;genesymbol <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>GeneSymbol.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.genesymbol_labels">
<code class="descname">genesymbol_labels</code><span class="sig-paren">(</span><em>graph=None</em>, <em>remap_all=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.genesymbol_labels"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.genesymbol_labels" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Creats vertex attribute <code class="docutils literal notranslate"><span class="pre">label</span></code> and fills up with Gene Symbols
of all proteins where the Gene Symbol can be looked up based on
the default name of the protein vertex.
If the attribute <code class="docutils literal notranslate"><span class="pre">label</span></code> had been already initialized,
updates this attribute or recreates if <code class="docutils literal notranslate"><span class="pre">remap_all</span></code>
is <code class="docutils literal notranslate"><span class="pre">True</span></code>.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.get_directed">
<code class="descname">get_directed</code><span class="sig-paren">(</span><em>graph=False</em>, <em>conv_edges=False</em>, <em>mutual=False</em>, <em>ret=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.get_directed"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.get_directed" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Converts <code class="docutils literal notranslate"><span class="pre">graph</span></code> undirected <code class="docutils literal notranslate"><span class="pre">igraph.Graph</span></code> object to a directed one.
By default it converts the graph in <code class="docutils literal notranslate"><span class="pre">PyPath.graph</span></code> and places the directed
instance in <code class="docutils literal notranslate"><span class="pre">PyPath.dgraph</span></code>.</p>
<dl class="docutils">
<dt>&#64;graph <span class="classifier-delimiter">:</span> <span class="classifier">igraph.Graph</span></dt>
<dd>Undirected graph object.</dd>
<dt>&#64;conv_edges <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd>Whether to convert undirected edges (those without explicit
direction information) to an arbitrary direction edge or
a pair of opposite edges.
Otherwise those will be deleted. Default is <code class="docutils literal notranslate"><span class="pre">False</span></code>.</dd>
<dt>&#64;mutual <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd>If <code class="docutils literal notranslate"><span class="pre">conv_edges</span></code> is <code class="docutils literal notranslate"><span class="pre">True</span></code>, whether to convert the
undirected edges to a single, arbitrary directed edge,
or a pair of opposite directed edges. Default is <code class="docutils literal notranslate"><span class="pre">False</span></code>.</dd>
<dt>&#64;ret <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd>Return the directed graph instance, or return <code class="docutils literal notranslate"><span class="pre">None</span></code>.
Default is <code class="docutils literal notranslate"><span class="pre">False</span></code> (returns <code class="docutils literal notranslate"><span class="pre">None</span></code>).</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.get_edge">
<code class="descname">get_edge</code><span class="sig-paren">(</span><em>source</em>, <em>target</em>, <em>directed=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.get_edge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.get_edge" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Edge</span></code> object if an edge exist between
the 2 proteins, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>source</strong> (<em>int</em><em>,</em><em>str</em>) â Vertex index or UniProt ID or GeneSymbol</li>
<li><strong>target</strong> (<em>int</em><em>,</em><em>str</em>) â Vertex index or UniProt ID or GeneSymbol</li>
<li><strong>directed</strong> (<em>bool</em>) â To be passed to igraph.Graph.get_eid()</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.get_edges">
<code class="descname">get_edges</code><span class="sig-paren">(</span><em>sources</em>, <em>targets</em>, <em>directed=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.get_edges"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.get_edges" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns a generator with all edges between source and target vertices.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sources</strong> (<em>iterable</em>) â Source vertex IDs, names or labels.</li>
<li><strong>targets</strong> (<em>iterable</em>) â Target vertec IDs, names or labels.</li>
<li><strong>directed</strong> (<em>bool</em>) â Passed to <cite>igraph.get_eid()</cite>.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.get_giant">
<code class="descname">get_giant</code><span class="sig-paren">(</span><em>replace=False</em>, <em>graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.get_giant"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.get_giant" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns the giant component of the graph, or
replaces the igraph object with only the giant
component.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.get_node">
<code class="descname">get_node</code><span class="sig-paren">(</span><em>identifier</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.get_node"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.get_node" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the identifier
is a valid vertex index in the default undirected graph,
or a UniProt ID or GeneSymbol which can be found in the
default undirected network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;identifier <span class="classifier-delimiter">:</span> <span class="classifier">int, str</span></dt>
<dd>Vertex index (int) or GeneSymbol (str) or UniProt ID (str).</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.get_node_d">
<code class="descname">get_node_d</code><span class="sig-paren">(</span><em>identifier</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.get_node_d"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.get_node_d" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Same as <code class="docutils literal notranslate"><span class="pre">PyPath.get_node</span></code>, just for the directed graph.
Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the identifier
is a valid vertex index in the default directed graph,
or a UniProt ID or GeneSymbol which can be found in the
default directed network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;identifier <span class="classifier-delimiter">:</span> <span class="classifier">int, str</span></dt>
<dd>Vertex index (int) or GeneSymbol (str) or UniProt ID (str).</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.go_annotate">
<code class="descname">go_annotate</code><span class="sig-paren">(</span><em>aspects=('C'</em>, <em>'F'</em>, <em>'P')</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.go_annotate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.go_annotate" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Annotates protein nodes with GO terms. In the <code class="docutils literal notranslate"><span class="pre">go</span></code> vertex
attribute each node is annotated by a dict of sets where keys are
one letter codes of GO aspects and values are sets of GO accessions.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.gs">
<code class="descname">gs</code><span class="sig-paren">(</span><em>genesymbol</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.gs" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the GeneSymbol
can be found in the default undirected network,
otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;genesymbol <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>GeneSymbol.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.gs_edge">
<code class="descname">gs_edge</code><span class="sig-paren">(</span><em>source</em>, <em>target</em>, <em>directed=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.gs_edge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.gs_edge" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Edge</span></code> object if an edge exist between
the 2 proteins, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;source <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>GeneSymbol</dd>
<dt>&#64;target <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>GeneSymbol</dd>
<dt>&#64;directed <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd>To be passed to igraph.Graph.get_eid()</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.homology_translation">
<code class="descname">homology_translation</code><span class="sig-paren">(</span><em>target</em>, <em>source=None</em>, <em>only_swissprot=True</em>, <em>graph=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.homology_translation" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Translates the current object to another organism by orthology.
Proteins without known ortholog will be deleted.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>target</strong> (<em>int</em>) â NCBI Taxonomy ID of the target organism.
E.g. 10090 for mouse.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.init_edge_attr">
<code class="descname">init_edge_attr</code><span class="sig-paren">(</span><em>attr</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.init_edge_attr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.init_edge_attr" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Fills edge attribute with its default values, creates
lists if in <cite>edgeAttrs</cite> the attribute is registered as list.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.init_network">
<code class="descname">init_network</code><span class="sig-paren">(</span><em>lst={'arn': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8ce60&gt;</em>, <em>'biogrid': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01fc8&gt;</em>, <em>'ca1': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8cd40&gt;</em>, <em>'ccmap': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01320&gt;</em>, <em>'cellphonedb': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb1d128&gt;</em>, <em>'dbptm': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01758&gt;</em>, <em>'death': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01050&gt;</em>, <em>'depod': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb015a8&gt;</em>, <em>'dip': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb013b0&gt;</em>, <em>'domino': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb016c8&gt;</em>, <em>'elm': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01680&gt;</em>, <em>'guide2pharma': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8cc68&gt;</em>, <em>'hpmr': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb1d098&gt;</em>, <em>'hprd': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb10170&gt;</em>, <em>'hprd_p': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb017a0&gt;</em>, <em>'innatedb': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01440&gt;</em>, <em>'intact': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01f80&gt;</em>, <em>'lmpid': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb015f0&gt;</em>, <em>'macrophage': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8cfc8&gt;</em>, <em>'matrixdb': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01518&gt;</em>, <em>'mppi': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01368&gt;</em>, <em>'nrf2': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8cf38&gt;</em>, <em>'pdz': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01098&gt;</em>, <em>'phelm': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01638&gt;</em>, <em>'psite': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01560&gt;</em>, <em>'ramilowski2015': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb1d1b8&gt;</em>, <em>'signalink3': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8cc20&gt;</em>, <em>'signor': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01290&gt;</em>, <em>'spike': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8cbd8&gt;</em>, <em>'trip': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8cb00&gt;}</em>, <em>exclude=[]</em>, <em>cache_files={}</em>, <em>pfile=False</em>, <em>save=False</em>, <em>reread=False</em>, <em>redownload=False</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.init_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.init_network" title="Permalink to this definition">Â¶</a></dt>
<dd><p>This is a lazy way to start the module, load data
and build the high confidence, literature curated
part of the signaling network.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.init_vertex_attr">
<code class="descname">init_vertex_attr</code><span class="sig-paren">(</span><em>attr</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.init_vertex_attr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.init_vertex_attr" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Fills vertex attribute with its default values, creates
lists if in <cite>vertexAttrs</cite> the attribute is registered as list.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.intogen_cancer_drivers_list">
<code class="descname">intogen_cancer_drivers_list</code><span class="sig-paren">(</span><em>intogen_file</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.intogen_cancer_drivers_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.intogen_cancer_drivers_list" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Loads the list of cancer driver proteins from IntOGen data.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.kinases_list">
<code class="descname">kinases_list</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.kinases_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.kinases_list" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Loads the list of all known kinases in the proteome from kinase.com.
This resource is human only.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.label_by_go">
<code class="descname">label_by_go</code><span class="sig-paren">(</span><em>label</em>, <em>go_terms</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.label_by_go"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.label_by_go" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Assigns a boolean vertex attribute to nodes which tells whether
the node is annotated by all or any (see <code class="docutils literal notranslate"><span class="pre">method</span></code> parameter of
<code class="docutils literal notranslate"><span class="pre">select_by_go</span></code>) the GO terms.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_compleat">
<code class="descname">load_compleat</code><span class="sig-paren">(</span><em>graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_compleat"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_compleat" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Loads complexes from Compleat. Loads data into vertex attribute
<cite>graph.vs[âcomplexesâ][âcompleatâ]</cite>.
This resource is human only.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_complexportal">
<code class="descname">load_complexportal</code><span class="sig-paren">(</span><em>graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_complexportal"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_complexportal" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Loads complexes from ComplexPortal. Loads data into vertex attribute
<cite>graph.vs[âcomplexesâ][âcomplexportalâ]</cite>.
This resource is human only.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_corum">
<code class="descname">load_corum</code><span class="sig-paren">(</span><em>graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_corum"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_corum" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Loads complexes from CORUM database. Loads data into vertex attribute
<cite>graph.vs[âcomplexesâ][âcorumâ]</cite>.
This resource is human only.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_ddi">
<code class="descname">load_ddi</code><span class="sig-paren">(</span><em>ddi</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_ddi"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_ddi" title="Permalink to this definition">Â¶</a></dt>
<dd><p>ddi is either a list of intera.DomainDomain objects,
or a function resulting this list</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_disgenet">
<code class="descname">load_disgenet</code><span class="sig-paren">(</span><em>dataset='curated'</em>, <em>score=0.0</em>, <em>umls=False</em>, <em>full_data=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_disgenet"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_disgenet" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Assigns DisGeNet disease-gene associations to the proteins
in the network. Disease annotations will be added to the <cite>dis</cite>
vertex attribute.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>score</strong> (<em>float</em>) â Confidence score from DisGeNet. Only associations
above the score provided will be considered.</li>
<li><strong>ulms</strong> (<em>bool</em>) â By default we assign a list of disease names to
each protein. To use Unified Medical Language System IDs instead
set this to <cite>True</cite>.</li>
<li><strong>full_data</strong> (<em>bool</em>) â By default we load only disease names. Set this
to <cite>True</cite> if you wish to load additional annotations like number
of PubMed IDs, number of SNPs and original sources.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_dmi">
<code class="descname">load_dmi</code><span class="sig-paren">(</span><em>dmi</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_dmi"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_dmi" title="Permalink to this definition">Â¶</a></dt>
<dd><p>dmi is either a list of intera.DomainMotif objects,
or a function resulting this list</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_expression">
<code class="descname">load_expression</code><span class="sig-paren">(</span><em>array=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_expression"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_expression" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Expression data can be loaded into vertex attributes,
or into a pandas DataFrame â the latter offers faster
ways to process and use these huge matrices.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_go">
<code class="descname">load_go</code><span class="sig-paren">(</span><em>aspects=('C'</em>, <em>'F'</em>, <em>'P')</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.load_go" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Annotates protein nodes with GO terms. In the <code class="docutils literal notranslate"><span class="pre">go</span></code> vertex
attribute each node is annotated by a dict of sets where keys are
one letter codes of GO aspects and values are sets of GO accessions.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_havugimana">
<code class="descname">load_havugimana</code><span class="sig-paren">(</span><em>graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_havugimana"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_havugimana" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Loads complexes from Havugimana 2012. Loads data into vertex attribute
<cite>graph.vs[âcomplexesâ][âhavugimanaâ]</cite>.
This resource is human only.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_hpa">
<code class="descname">load_hpa</code><span class="sig-paren">(</span><em>normal=True</em>, <em>pathology=True</em>, <em>cancer=True</em>, <em>summarize_pathology=True</em>, <em>tissues=None</em>, <em>quality=set(['Supported'</em>, <em>'Approved'])</em>, <em>levels={'High': 3</em>, <em>'Low': 1</em>, <em>'Medium': 2</em>, <em>'Not detected': 0}</em>, <em>graph=None</em>, <em>na_value=0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_hpa"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_hpa" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Loads Human Protein Atlas data into vertex attributes.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_ligand_receptor_network">
<code class="descname">load_ligand_receptor_network</code><span class="sig-paren">(</span><em>lig_rec_resources=True</em>, <em>inference_from_go=True</em>, <em>sources={'arn': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8ce60&gt;</em>, <em>'ca1': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8cd40&gt;</em>, <em>'cellphonedb': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb1d128&gt;</em>, <em>'death': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01050&gt;</em>, <em>'guide2pharma': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8cc68&gt;</em>, <em>'hpmr': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb1d098&gt;</em>, <em>'macrophage': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8cfc8&gt;</em>, <em>'nrf2': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8cf38&gt;</em>, <em>'pdz': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01098&gt;</em>, <em>'ramilowski2015': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb1d1b8&gt;</em>, <em>'signalink3': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8cc20&gt;</em>, <em>'signor': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01290&gt;</em>, <em>'spike': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8cbd8&gt;</em>, <em>'trip': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8cb00&gt;}</em>, <em>keep_undirected=False</em>, <em>keep_rec_rec=False</em>, <em>keep_lig_lig=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_ligand_receptor_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_ligand_receptor_network" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Initializes a ligand-receptor network.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_mutations">
<code class="descname">load_mutations</code><span class="sig-paren">(</span><em>attributes=None</em>, <em>gdsc_datadir=None</em>, <em>mutation_file=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_mutations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_mutations" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Mutations are listed in vertex attributes. Mutation() objects
offers methods to identify residues and look up in Ptm(), Motif()
and Domain() objects, to check if those residues are
modified, or are in some short motif or domain.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_old_omnipath">
<code class="descname">load_old_omnipath</code><span class="sig-paren">(</span><em>kinase_substrate_extra=False</em>, <em>remove_htp=False</em>, <em>htp_threshold=1</em>, <em>keep_directed=False</em>, <em>min_refs_undirected=2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_old_omnipath"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_old_omnipath" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Loads the OmniPath network as it was before August 2016.
Furthermore it gives some more options.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_omnipath">
<code class="descname">load_omnipath</code><span class="sig-paren">(</span><em>kinase_substrate_extra=False</em>, <em>remove_htp=True</em>, <em>htp_threshold=1</em>, <em>keep_directed=True</em>, <em>min_refs_undirected=2</em>, <em>old_omnipath_resources=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_omnipath"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_omnipath" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Loads the OmniPath network.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_pathways">
<code class="descname">load_pathways</code><span class="sig-paren">(</span><em>source</em>, <em>graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_pathways"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_pathways" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Generic method to load pathway annotations from a resource.
We donât recommend calling this method but either specific
methods for a single source e.g. <cite>kegg_pathways()</cite>
or <cite>sirnor_pathways()</cite> or call <cite>load_all_pathways()</cite> to
load all resources.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>source</strong> (<em>str</em>) â Name of the source, this need to match a method in the dict
in <cite>get_pathways()</cite> method and the edge and vertex attributes
with pathway annotations will be called â&lt;source&gt;_pathwaysâ.</li>
<li><strong>graph</strong> (<em>igraph.Graph</em>) â A graph, by default the default the <cite>graph</cite> attribute of the
current instance.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_ptms2">
<code class="descname">load_ptms2</code><span class="sig-paren">(</span><em>input_methods=None, map_by_homology_from=[9606], homology_only_swissprot=True, ptm_homology_strict=False, nonhuman_direct_lookup=True, inputargs={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_ptms2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_ptms2" title="Permalink to this definition">Â¶</a></dt>
<dd><p>This is a new method which will replace <cite>load_ptms</cite>.
It uses <cite>pypath.ptm.PtmAggregator</cite>, a newly introduced
module for combining enzyme-substrate data from multiple
resources using homology translation on users demand.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>input_methods</strong> (<em>list</em>) â Resources to collect enzyme-substrate
interactions from. E.g. <cite>[âSignorâ, âphosphoELMâ]</cite>. By default
it contains Signor, PhosphoSitePlus, HPRD, phosphoELM, dbPTM,
PhosphoNetworks, Li2012 and MIMP.</li>
<li><strong>map_by_homology_from</strong> (<em>list</em>) â List of NCBI Taxonomy IDs of
source taxons used for homology translation of enzyme-substrate
interactions. If you have a human network and you add here
<cite>[10090, 10116]</cite> then mouse and rat interactions from the source
databases will be translated to human.</li>
<li><strong>homology_only_swissprot</strong> (<em>bool</em>) â <cite>True</cite> by default which means
only SwissProt IDs are accepted at homology translateion, Trembl
IDs will be dropped.</li>
<li><strong>ptm_homology_strict</strong> (<em>bool</em>) â For homology translation use
PhosphoSiteâs PTM homology table. This guarantees that only
truely homologous sites will be included. Otherwise we only
check if at the same numeric offset in the homologous sequence
the appropriate residue can be find.</li>
<li><strong>nonhuman_direct_lookup</strong> (<em>bool</em>) â Fetch also directly nonhuman
data from the resources whereever itâs available. PhosphoSite
contains mouse enzyme-substrate interactions and it is possible
to extract these directly beside translating the human ones
to mouse.</li>
<li><strong>inputargs</strong> (<em>dict</em>) â Additional arguments passed to <cite>PtmProcessor</cite>.
A <cite>dict</cite> can be supplied for each resource, e.g.
<cite>{âSignorâ: {â¦}, âPhosphoSiteâ: {â¦}, â¦}</cite>.
Those not used by <cite>PtmProcessor</cite> are forwarded to the
<cite>pypath.dataio</cite> methods.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_resources">
<code class="descname">load_resources</code><span class="sig-paren">(</span><em>lst={'arn': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8ce60&gt;</em>, <em>'biogrid': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01fc8&gt;</em>, <em>'ca1': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8cd40&gt;</em>, <em>'ccmap': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01320&gt;</em>, <em>'cellphonedb': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb1d128&gt;</em>, <em>'dbptm': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01758&gt;</em>, <em>'death': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01050&gt;</em>, <em>'depod': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb015a8&gt;</em>, <em>'dip': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb013b0&gt;</em>, <em>'domino': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb016c8&gt;</em>, <em>'elm': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01680&gt;</em>, <em>'guide2pharma': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8cc68&gt;</em>, <em>'hpmr': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb1d098&gt;</em>, <em>'hprd': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb10170&gt;</em>, <em>'hprd_p': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb017a0&gt;</em>, <em>'innatedb': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01440&gt;</em>, <em>'intact': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01f80&gt;</em>, <em>'lmpid': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb015f0&gt;</em>, <em>'macrophage': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8cfc8&gt;</em>, <em>'matrixdb': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01518&gt;</em>, <em>'mppi': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01368&gt;</em>, <em>'nrf2': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8cf38&gt;</em>, <em>'pdz': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01098&gt;</em>, <em>'phelm': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01638&gt;</em>, <em>'psite': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01560&gt;</em>, <em>'ramilowski2015': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb1d1b8&gt;</em>, <em>'signalink3': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8cc20&gt;</em>, <em>'signor': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcb01290&gt;</em>, <em>'spike': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8cbd8&gt;</em>, <em>'trip': &lt;pypath.input_formats.ReadSettings instance at 0x7faddcd8cb00&gt;}</em>, <em>exclude=[]</em>, <em>cache_files={}</em>, <em>reread=False</em>, <em>redownload=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_resources"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_resources" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Loads multiple resources, and cleans up after.
Looks up ID types, and loads all ID conversion
tables from UniProt if necessary. This is much
faster than loading the ID conversion and the
resources one by one.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_tfregulons">
<code class="descname">load_tfregulons</code><span class="sig-paren">(</span><em>levels=set(['A'</em>, <em>'B'])</em>, <em>only_curated=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_tfregulons"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_tfregulons" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Adds TF-target interactions from TF regulons to the network.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>levels</strong> (<em>set</em>) â Confidence levels to be used.</li>
<li><strong>only_curated</strong> (<em>bool</em>) â Retrieve only literature curated interactions.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>TF regulons is a comprehensive resource of TF-target interactions
combining multiple lines of evidences: literature curated databases,
ChIP-Seq data, PWM based prediction using HOCOMOCO and JASPAR matrices
and prediction from GTEx expression data by ARACNe.</p>
<p>For details see <a class="reference external" href="https://github.com/saezlab/DoRothEA">https://github.com/saezlab/DoRothEA</a>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pypath</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pa</span> <span class="o">=</span> <span class="n">pypath</span><span class="o">.</span><span class="n">PyPath</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pa</span><span class="o">.</span><span class="n">load_tfregulons</span><span class="p">(</span><span class="n">levels</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;A&#39;</span><span class="p">})</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.map_edge">
<code class="descname">map_edge</code><span class="sig-paren">(</span><em>edge</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.map_edge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.map_edge" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Translates molecule names in dict representing an edge.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.map_item">
<code class="descname">map_item</code><span class="sig-paren">(</span><em>item</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.map_item"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.map_item" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Translates the name in item representing a molecule.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.map_list">
<code class="descname">map_list</code><span class="sig-paren">(</span><em>lst</em>, <em>singleList=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.map_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.map_list" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Only a wrapper for map_edge()</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.merge_lists">
<code class="descname">merge_lists</code><span class="sig-paren">(</span><em>nameA</em>, <em>nameB</em>, <em>name=None</em>, <em>and_or='and'</em>, <em>delete=False</em>, <em>func='max'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.merge_lists"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.merge_lists" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Merges two lists in <cite>lists</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.merge_nodes">
<code class="descname">merge_nodes</code><span class="sig-paren">(</span><em>nodes</em>, <em>primary=None</em>, <em>graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.merge_nodes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.merge_nodes" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Merges all attributes and all edges of selected nodes
and assigns them to the primary node
(by default the one with lowest ID).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>nodes</strong> (<em>list</em>) â List of edge IDs.</li>
<li><strong>primary</strong> (<em>int</em>) â ID of the primary edge;
if None the lowest ID selected.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.mutated_edges">
<code class="descname">mutated_edges</code><span class="sig-paren">(</span><em>sample</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.mutated_edges"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.mutated_edges" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Compares the mutated residues and the modified residues in PTMs.
Interactions are marked as mutated if the target residue in the
underlying PTM is mutated.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.network_filter">
<code class="descname">network_filter</code><span class="sig-paren">(</span><em>p=2.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.network_filter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.network_filter" title="Permalink to this definition">Â¶</a></dt>
<dd><p>This function aims to cut the number of edges in the network,
without loosing nodes, to make the network less connected,
less hairball-like, more usable for analysis.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.network_stats">
<code class="descname">network_stats</code><span class="sig-paren">(</span><em>outfile=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.network_stats"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.network_stats" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Calculates basic statistics for the whole network
and each of sources. Writes the results in a tab file.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.orthology_translation">
<code class="descname">orthology_translation</code><span class="sig-paren">(</span><em>target</em>, <em>source=None</em>, <em>only_swissprot=True</em>, <em>graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.orthology_translation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.orthology_translation" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Translates the current object to another organism by orthology.
Proteins without known ortholog will be deleted.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>target</strong> (<em>int</em>) â NCBI Taxonomy ID of the target organism.
E.g. 10090 for mouse.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.p">
<code class="descname">p</code><span class="sig-paren">(</span><em>identifier</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.p" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the identifier
is a valid vertex index in the default undirected graph,
or a UniProt ID or GeneSymbol which can be found in the
default undirected network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;identifier <span class="classifier-delimiter">:</span> <span class="classifier">int, str</span></dt>
<dd>Vertex index (int) or GeneSymbol (str) or UniProt ID (str).</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.pathway_members">
<code class="descname">pathway_members</code><span class="sig-paren">(</span><em>pathway</em>, <em>source</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.pathway_members"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.pathway_members" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns an iterator with the members of a single pathway.
Apart from the pathway name you need to supply its source
database too.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.pathway_names">
<code class="descname">pathway_names</code><span class="sig-paren">(</span><em>source</em>, <em>graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.pathway_names"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.pathway_names" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns the names of all pathways having at least one member
in the current graph.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.process_dmi">
<code class="descname">process_dmi</code><span class="sig-paren">(</span><em>source</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.process_dmi"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.process_dmi" title="Permalink to this definition">Â¶</a></dt>
<dd><p>This is an universal function
for loading domain-motif objects
like load_phospho_dmi() for PTMs.
TODO this will replace load_elm, load_ielm, etc</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.protein">
<code class="descname">protein</code><span class="sig-paren">(</span><em>identifier</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.protein" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Same as <code class="docutils literal notranslate"><span class="pre">PyPath.get_node</span></code>, just for the directed graph.
Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the identifier
is a valid vertex index in the default directed graph,
or a UniProt ID or GeneSymbol which can be found in the
default directed network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;identifier <span class="classifier-delimiter">:</span> <span class="classifier">int, str</span></dt>
<dd>Vertex index (int) or GeneSymbol (str) or UniProt ID (str).</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.protein_edge">
<code class="descname">protein_edge</code><span class="sig-paren">(</span><em>source</em>, <em>target</em>, <em>directed=True</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.protein_edge" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Edge</span></code> object if an edge exist between
the 2 proteins, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>source</strong> (<em>int</em><em>,</em><em>str</em>) â Vertex index or UniProt ID or GeneSymbol</li>
<li><strong>target</strong> (<em>int</em><em>,</em><em>str</em>) â Vertex index or UniProt ID or GeneSymbol</li>
<li><strong>directed</strong> (<em>bool</em>) â To be passed to igraph.Graph.get_eid()</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.proteome_list">
<code class="descname">proteome_list</code><span class="sig-paren">(</span><em>swissprot=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.proteome_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.proteome_list" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Loads the whole proteome as a list.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.random_walk_with_return">
<code class="descname">random_walk_with_return</code><span class="sig-paren">(</span><em>q</em>, <em>graph=None</em>, <em>c=0.5</em>, <em>niter=1000</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.random_walk_with_return"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.random_walk_with_return" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Random walk with return (RWR) starting from one or more query nodes.
Returns affinity (probability) vector of all nodes in the graph.</p>
<blockquote>
<div><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name" colspan="2">param int,list q:</th></tr>
<tr class="field-odd field"><td>&#160;</td><td class="field-body">Vertex IDs of query nodes.</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">param igraph.Graph graph:</th></tr>
<tr class="field-even field"><td>&#160;</td><td class="field-body">An <cite>igraph.Graph</cite> object.</td>
</tr>
<tr class="field-odd field"><th class="field-name">param float c:</th><td class="field-body">Probability of restart.</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">param int niter:</th></tr>
<tr class="field-even field"><td>&#160;</td><td class="field-body">Number of iterations.</td>
</tr>
</tbody>
</table>
</div></blockquote>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">igraph</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pypath</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pa</span> <span class="o">=</span> <span class="n">pypath</span><span class="o">.</span><span class="n">PyPath</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pa</span><span class="o">.</span><span class="n">init_network</span><span class="p">({</span>
<span class="go">        &#39;signor&#39;: pypath.data_formats.pathway[&#39;signor&#39;]</span>
<span class="go">    })</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">q</span> <span class="o">=</span> <span class="p">[</span>
<span class="go">        pa.gs(&#39;EGFR&#39;).index,</span>
<span class="go">        pa.gs(&#39;ATG4B&#39;).index</span>
<span class="go">    ]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rwr</span> <span class="o">=</span> <span class="n">pa</span><span class="o">.</span><span class="n">random_walk_with_return</span><span class="p">(</span><span class="n">q</span> <span class="o">=</span> <span class="n">q</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">palette</span> <span class="o">=</span> <span class="n">igraph</span><span class="o">.</span><span class="n">RainbowPalette</span><span class="p">(</span><span class="n">n</span> <span class="o">=</span> <span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">colors</span>  <span class="o">=</span> <span class="p">[</span><span class="n">palette</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">i</span><span class="p">)))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">rwr</span> <span class="o">/</span> <span class="nb">max</span><span class="p">(</span><span class="n">rwr</span><span class="p">)</span> <span class="o">*</span> <span class="mi">99</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">igraph</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="p">,</span> <span class="n">vertex_color</span> <span class="o">=</span> <span class="n">colors</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.random_walk_with_return2">
<code class="descname">random_walk_with_return2</code><span class="sig-paren">(</span><em>q</em>, <em>c=0.5</em>, <em>niter=1000</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.random_walk_with_return2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.random_walk_with_return2" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Literally does random walks.
Only for testing of the other method, to be deleted later.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.read_data_file">
<code class="descname">read_data_file</code><span class="sig-paren">(</span><em>settings</em>, <em>keep_raw=False</em>, <em>cache_files={}</em>, <em>reread=False</em>, <em>redownload=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.read_data_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.read_data_file" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Interaction data with node and edge attributes can be read
from simple text based files. This function works not only
with files, but with lists as well. Any other function can
be written to download a preprocess data, and then give it
to this function to finally attach to the network.</p>
<dl class="docutils">
<dt>&#64;settings <span class="classifier-delimiter">:</span> <span class="classifier">ReadSettings instance</span></dt>
<dd>The detailed definition of the input format. Instead of
the file name you can give a function name, which will
be executed, and the returned data will be used.</dd>
<dt>&#64;keep_raw <span class="classifier-delimiter">:</span> <span class="classifier">boolean</span></dt>
<dd>To keep the raw data read by this function, in order for
debugging purposes, or further use.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.receptors_list">
<code class="descname">receptors_list</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.receptors_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.receptors_list" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Loads the Human Plasma Membrane Receptome as a list.
This resource is human only.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.save_session">
<code class="descname">save_session</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.save_session"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.save_session" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Save current state into pickle dump.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.select_by_go">
<code class="descname">select_by_go</code><span class="sig-paren">(</span><em>go_terms</em>, <em>go_desc=None</em>, <em>aspects=('C'</em>, <em>'F'</em>, <em>'P')</em>, <em>method='ANY'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.select_by_go"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.select_by_go" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Selects the nodes annotated by certain GO terms.</p>
<p>Returns set of vertex IDs.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>method</strong> (<em>str</em>) â If <cite>ANY</cite> nodes annotated with any of the terms returned.
If <cite>ALL</cite> nodes annotated with all the terms returned.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.separate">
<code class="descname">separate</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.separate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.separate" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Separates networks from different sources.
Returns dict of igraph objects.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.separate_by_category">
<code class="descname">separate_by_category</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.separate_by_category"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.separate_by_category" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Separate networks based on categories.
Returns dict of igraph objects.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.set_chembl_mysql">
<code class="descname">set_chembl_mysql</code><span class="sig-paren">(</span><em>title</em>, <em>config_file=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.set_chembl_mysql"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.set_chembl_mysql" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Sets the ChEMBL MySQL config according to
<cite>title</cite> section in <cite>config_file</cite> ini style config.</p>
<blockquote>
<div><p>title (str): section title in ini file
config_file (str, NoneType): config file name;</p>
<blockquote>
<div>if None, the <cite>mysql_config/defaults.mysql</cite>
will be used</div></blockquote>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.set_disease_genes">
<code class="descname">set_disease_genes</code><span class="sig-paren">(</span><em>dataset='curated'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.set_disease_genes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.set_disease_genes" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Creates a vertex attribute named <cite>dis</cite> with boolean values <em>True</em>
if the protein encoded by a disease related gene according to
DisGeNet.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>dataset</strong> (<em>str</em>) â Which dataset to use from DisGeNet. Default is <cite>curated</cite>.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.set_druggability">
<code class="descname">set_druggability</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.set_druggability"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.set_druggability" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Creates a vertex attribute <cite>dgb</cite> with value <em>True</em> if
the protein is druggable, otherwise <em>False</em>.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.set_drugtargets">
<code class="descname">set_drugtargets</code><span class="sig-paren">(</span><em>pchembl=5.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.set_drugtargets"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.set_drugtargets" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Creates a vertex attribute <cite>dtg</cite> with value <em>True</em> if
the protein has at least one compound binding with
affinity higher than <cite>pchembl</cite>, otherwise <em>False</em>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>pchembl</strong> (<em>float</em>) â Pchembl threshold.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.set_kinases">
<code class="descname">set_kinases</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.set_kinases"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.set_kinases" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Creates a vertex attribute <cite>kin</cite> with value <em>True</em> if
the protein is a kinase, otherwise <em>False</em>.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.set_receptors">
<code class="descname">set_receptors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.set_receptors"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.set_receptors" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Creates a vertex attribute <cite>rec</cite> with value <em>True</em> if
the protein is a receptor, otherwise <em>False</em>.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.set_signaling_proteins">
<code class="descname">set_signaling_proteins</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.set_signaling_proteins"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.set_signaling_proteins" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Creates a vertex attribute <cite>kin</cite> with value <em>True</em> if
the protein is a kinase, otherwise <em>False</em>.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.set_transcription_factors">
<code class="descname">set_transcription_factors</code><span class="sig-paren">(</span><em>classes=['a', 'b', 'other']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.set_transcription_factors"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.set_transcription_factors" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Creates a vertex attribute <cite>tf</cite> with value <em>True</em> if
the protein is a transcription factor, otherwise <em>False</em>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>classes</strong> (<em>list</em>) â Classes to use from TF Census. Default is <cite>[âaâ, âbâ, âotherâ]</cite>.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.shortest_path_dist">
<code class="descname">shortest_path_dist</code><span class="sig-paren">(</span><em>graph=None</em>, <em>subset=None</em>, <em>outfile=None</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.shortest_path_dist"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.shortest_path_dist" title="Permalink to this definition">Â¶</a></dt>
<dd><p>subset is a tuple of two lists if you wish to look for
paths between elements of two groups, or a list if you
wish to look for shortest paths within this group</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.signaling_proteins_list">
<code class="descname">signaling_proteins_list</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.signaling_proteins_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.signaling_proteins_list" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Compiles a list of signaling proteins (as opposed to other proteins
like metabolic enzymes, matrix proteins), by looking up a few simple
keywords in short description of GO terms.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.small_plot">
<code class="descname">small_plot</code><span class="sig-paren">(</span><em>graph</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.small_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.small_plot" title="Permalink to this definition">Â¶</a></dt>
<dd><p>This method is deprecated, do not use it.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.source_network">
<code class="descname">source_network</code><span class="sig-paren">(</span><em>font='HelveticaNeueLTStd'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.source_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.source_network" title="Permalink to this definition">Â¶</a></dt>
<dd><p>For EMBL branding, use Helvetica Neue Linotype Standard light</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.straight_between">
<code class="descname">straight_between</code><span class="sig-paren">(</span><em>nameA</em>, <em>nameB</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.straight_between"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.straight_between" title="Permalink to this definition">Â¶</a></dt>
<dd><p>This does actually the same as get_edge(), but by names
instead of vertex ids.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.sum_in_complex">
<code class="descname">sum_in_complex</code><span class="sig-paren">(</span><em>csources=['corum'], graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.sum_in_complex"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.sum_in_complex" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns the total number of edges in the network falling
between two members of the same complex.
Returns as a dict by complex resources.
Calls :py:func:pypath.pypath.Pypath.edges_in_comlexes()
to do the calculations.</p>
<dl class="docutils">
<dt>&#64;csources <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd>List of complex resources. Should be already loaded.</dd>
<dt>&#64;graph <span class="classifier-delimiter">:</span> <span class="classifier">igraph.Graph()</span></dt>
<dd>The graph object to do the calculations on.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.tfs_list">
<code class="descname">tfs_list</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.tfs_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.tfs_list" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Loads the list of all known transcription factors from TF census
(Vaquerizas 2009). This resource is human only.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.third_source_directions">
<code class="descname">third_source_directions</code><span class="sig-paren">(</span><em>graph=None</em>, <em>use_string_effects=False</em>, <em>use_laudanna_data=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.third_source_directions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.third_source_directions" title="Permalink to this definition">Â¶</a></dt>
<dd><p>This method calls a series of methods to get
additional direction &amp; effect information
from sources having no literature curated references,
but giving sufficient evidence about the directionality
for interactions already supported by literature
evidences from other sources.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.tissue_network">
<code class="descname">tissue_network</code><span class="sig-paren">(</span><em>tissue</em>, <em>graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.tissue_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.tissue_network" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns a network which includes the proteins expressed in
certain tissue according to ProteomicsDB.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>tissue</strong> (<em>str</em>) â Tissue name as used in ProteomicsDB.</li>
<li><strong>graph</strong> (<em>igraph.Graph</em>) â A graph object, by default the <cite>graph</cite> attribute of
the current instance.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.uniprot">
<code class="descname">uniprot</code><span class="sig-paren">(</span><em>uniprot</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.uniprot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.uniprot" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the UniProt
can be found in the default undirected network,
otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;uniprot <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>UniProt ID.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.uniprots">
<code class="descname">uniprots</code><span class="sig-paren">(</span><em>uniprots</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.uniprots"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.uniprots" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns list of <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object
for a list of UniProt IDs omitting those
could not be found in the default
undirected graph.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.up">
<code class="descname">up</code><span class="sig-paren">(</span><em>uniprot</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.up" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the UniProt
can be found in the default undirected network,
otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;uniprot <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>UniProt ID.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.up_edge">
<code class="descname">up_edge</code><span class="sig-paren">(</span><em>source</em>, <em>target</em>, <em>directed=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.up_edge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.up_edge" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Edge</span></code> object if an edge exist between
the 2 proteins, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;source <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>UniProt ID</dd>
<dt>&#64;target <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>UniProt ID</dd>
<dt>&#64;directed <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd>To be passed to igraph.Graph.get_eid()</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.update_adjlist">
<code class="descname">update_adjlist</code><span class="sig-paren">(</span><em>graph=None</em>, <em>mode='ALL'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.update_adjlist"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.update_adjlist" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Creates an adjacency list in a dict of sets format.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.update_cats">
<code class="descname">update_cats</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.update_cats"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.update_cats" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Makes sure that the <cite>has_cats</cite> attribute is an up to date
set of all categories in the current network.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.update_sources">
<code class="descname">update_sources</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.update_sources"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.update_sources" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Makes sure that the <cite>sources</cite> attribute is an up to date
list of all sources in the current network.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.update_vindex">
<code class="descname">update_vindex</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.update_vindex"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.update_vindex" title="Permalink to this definition">Â¶</a></dt>
<dd><p>This is deprecated.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.update_vname">
<code class="descname">update_vname</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.update_vname"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.update_vname" title="Permalink to this definition">Â¶</a></dt>
<dd><p>For fast lookup of node names and indexes, these are
hold in a list and a dict as well. However, every time
new nodes are added, these should be updated. This
function is automatically called after all operations
affecting node indices.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.ups">
<code class="descname">ups</code><span class="sig-paren">(</span><em>uniprots</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.ups" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns list of <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object
for a list of UniProt IDs omitting those
could not be found in the default
undirected graph.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.v">
<code class="descname">v</code><span class="sig-paren">(</span><em>identifier</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.v" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">igraph.Vertex()</span></code> object if the identifier
is a valid vertex index in the default undirected graph,
or a UniProt ID or GeneSymbol which can be found in the
default undirected network, otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;identifier <span class="classifier-delimiter">:</span> <span class="classifier">int, str</span></dt>
<dd>Vertex index (int) or GeneSymbol (str) or UniProt ID (str).</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.vertex_pathways">
<code class="descname">vertex_pathways</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.vertex_pathways"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.vertex_pathways" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Some resources assignes interactions some others
proteins to pathways.
This function converts pathway annotations from
edge attributes to vertex attributes.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.main.Direction">
<em class="property">class </em><code class="descclassname">pypath.main.</code><code class="descname">Direction</code><span class="sig-paren">(</span><em>nameA</em>, <em>nameB</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Object storing directionality information of an edge. Also includes
information about the reverse direction, mode of regulation and
sources of that information.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Attributes:</dt>
<dd><ul class="first last">
<li><em>dirs</em> [dict]: Dictionary containing the presence of
directionality of the given edge. Keys are <em>straight</em>,
<em>reverse</em> and <code class="docutils literal notranslate"><span class="pre">'undirected'</span></code> and their values denote the
presence/absence [bool].</li>
<li><em>negative</em> [dict]: Dictionary contianing the presence/absence
[bool] of negative interactions for both <em>straight</em> and
<em>reverse</em> directions.</li>
<li><em>negative_sources</em> [dict]: Contains the resource names [str]
supporting a negative interaction on <em>straight</em> and <em>reverse</em>
directions.</li>
<li><em>nodes</em> [list]: Contains the node names [str] sorted
alphabetically (<em>nameA</em>, <em>nameB</em>).</li>
<li><em>positive</em> [dict]:  Dictionary contianing the presence/absence
[bool] of positive interactions for both <em>straight</em> and
<em>reverse</em> directions.</li>
<li><em>positive_sources</em> [dict]: Contains the resource names [str]
supporting a positive interaction on <em>straight</em> and <em>reverse</em>
directions.</li>
<li><em>reverse</em> [tuple]: Contains the node names [str] in the original
order e.g. (source, target).</li>
<li><em>sources</em> [dict]: Contains the resource names [str] of a given
edge for each directionality (<em>straight</em>, <em>reverse</em> and
<code class="docutils literal notranslate"><span class="pre">'undirected'</span></code>). Values are sets containing the names of
those resources supporting such directionality.</li>
<li><em>straight</em> [tuple]: Contains the node names [str] in reverse
order e.g. (target, source).</li>
</ul>
</dd>
</dl>
</li>
</ul>
<dl class="method">
<dt id="pypath.main.Direction.check_nodes">
<code class="descname">check_nodes</code><span class="sig-paren">(</span><em>nodes</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.check_nodes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.check_nodes" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Checks if <em>nodes</em> is contained in the edge.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Arguments:</dt>
<dd><ul class="first last">
<li><em>nodes</em> [list]: Or [tuple], contains the names of the
edges to be checked.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[bool]: True if all elements in <em>nodes</em> are contained in
the objectâs node list.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.check_param">
<code class="descname">check_param</code><span class="sig-paren">(</span><em>di</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.check_param"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.check_param" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Checks if <em>di</em> is <code class="docutils literal notranslate"><span class="pre">'undirected'</span></code> or contains the nodes of
the current edge. Used internally to check that <em>di</em> is a valid
key for the object attributes declared on dictionaries.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Arguments:</dt>
<dd><ul class="first last">
<li><em>di</em> [tuple]: Or [str], key to be tested for validity.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[bool]: <code class="docutils literal notranslate"><span class="pre">True</span></code> if <em>di</em> is <code class="docutils literal notranslate"><span class="pre">'undirected'</span></code> or a tuple
of node names contained in the edge, <code class="docutils literal notranslate"><span class="pre">False</span></code> otherwise.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.consensus_edges">
<code class="descname">consensus_edges</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.consensus_edges"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.consensus_edges" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Infers the consensus edge(s) according to the number of
supporting sources. This includes direction and sign attributes.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[list]: Contains the consensus edge(s) along with the
consensus sign. If there is no major directionality, both
are returned. The structure is as follows:
<code class="docutils literal notranslate"><span class="pre">['&lt;source&gt;',</span> <span class="pre">'&lt;target&gt;',</span> <span class="pre">'&lt;(un)directed&gt;',</span> <span class="pre">'&lt;sign&gt;']</span></code></li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.get_dir">
<code class="descname">get_dir</code><span class="sig-paren">(</span><em>direction</em>, <em>sources=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.get_dir"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.get_dir" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns the state (or <em>sources</em> if specified) of the given
<em>direction</em>.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Arguments:</dt>
<dd><ul class="first last">
<li><em>direction</em> [tuple]: Or [str] (if <code class="docutils literal notranslate"><span class="pre">'undirected'</span></code>). Pair
of nodes from which direction information is to be
retrieved.</li>
<li><em>sources</em> [bool]: Optional, <code class="docutils literal notranslate"><span class="pre">'False'</span></code> by default.
Specifies if the <em>sources</em> information of the given
direction is to be retrieved instead.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[bool] or [set] (if <code class="docutils literal notranslate"><span class="pre">sources=True</span></code>). Presence/absence of
the requested direction (or the list of sources if
specified). Returns <code class="docutils literal notranslate"><span class="pre">None</span></code> if <em>direction</em> is not valid.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.get_dirs">
<code class="descname">get_dirs</code><span class="sig-paren">(</span><em>src</em>, <em>tgt</em>, <em>sources=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.get_dirs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.get_dirs" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns all directions with boolean values or list of sources.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Arguments:</dt>
<dd><ul class="first last">
<li><em>src</em> [str]: Source node.</li>
<li><em>tgt</em> [str]: Target node.</li>
<li><em>sources</em> [bool]: Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default.
Specifies whether to return the <em>sources</em> attribute
instead of <em>dirs</em>.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[list]: Contains the <em>dirs</em> (or <em>sources</em> if specified) of
the given edge.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.get_sign">
<code class="descname">get_sign</code><span class="sig-paren">(</span><em>direction</em>, <em>sign=None</em>, <em>sources=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.get_sign"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.get_sign" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Retrieves the sign information of the edge in the given
diretion. If specified in <em>sign</em>, only that signâs information
will be retrieved. If specified in <em>sources</em>, the sources of
that information will be retrieved instead.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Arguments:</dt>
<dd><ul class="first last">
<li><em>direction</em> [tuple]: Contains the pair of nodes specifying
the directionality of the edge from which th information
is to be retrieved.</li>
<li><em>sign</em> [str]: Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. Denotes
whether to retrieve the <code class="docutils literal notranslate"><span class="pre">'positive'</span></code> or <code class="docutils literal notranslate"><span class="pre">'negative'</span></code>
specific information.</li>
<li><em>sources</em> [bool]: Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default.
Specifies whether to return the sources instead of sign.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[list]: If <code class="docutils literal notranslate"><span class="pre">sign=None</span></code> containing [bool] values denoting
the presence of positive and negative sign on that
direction, if <code class="docutils literal notranslate"><span class="pre">sources=True</span></code> the [set] of sources for
each of them will be returned instead. If <em>sign</em> is
specified, returns [bool] or [set] (if <code class="docutils literal notranslate"><span class="pre">sources=True</span></code>)
of that specific direction and sign.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.has_sign">
<code class="descname">has_sign</code><span class="sig-paren">(</span><em>direction=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.has_sign"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.has_sign" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Checks whether the edge (or for a specific <em>direction</em>) has
any signed information (about positive/negative interactions).</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Arguments:</dt>
<dd><ul class="first last">
<li><em>direction</em> [tuple]: Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. If
specified, only the information of that direction is
checked for sign.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[bool]: <code class="docutils literal notranslate"><span class="pre">True</span></code> if there exist any information on the
sign of the interaction, <code class="docutils literal notranslate"><span class="pre">False</span></code> otherwise.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.is_directed">
<code class="descname">is_directed</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.is_directed"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.is_directed" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Checks if edge has any directionality information.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[bool]: Returns <code class="docutils literal notranslate"><span class="pre">True```if</span> <span class="pre">any</span> <span class="pre">of</span> <span class="pre">the</span> <span class="pre">*dirs*</span> <span class="pre">attribute</span>
<span class="pre">values</span> <span class="pre">is</span> <span class="pre">``True</span></code> (except <code class="docutils literal notranslate"><span class="pre">'undirected'</span></code>), <code class="docutils literal notranslate"><span class="pre">False</span></code>
otherwise.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.is_inhibition">
<code class="descname">is_inhibition</code><span class="sig-paren">(</span><em>direction=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.is_inhibition"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.is_inhibition" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Checks if any (or for a specific <em>direction</em>) interaction is
inhibition (negative interaction).</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Arguments:</dt>
<dd><ul class="first last">
<li><em>direction</em> [tuple]: Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. If
specified, checks the <em>negative</em> attribute of that
specific directionality. If not specified, checks both.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[bool]: <code class="docutils literal notranslate"><span class="pre">True</span></code> if any interaction (or the specified
<em>direction</em>) is inhibitory (negative).</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.is_stimulation">
<code class="descname">is_stimulation</code><span class="sig-paren">(</span><em>direction=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.is_stimulation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.is_stimulation" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Checks if any (or for a specific <em>direction</em>) interaction is
activation (positive interaction).</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Arguments:</dt>
<dd><ul class="first last">
<li><em>direction</em> [tuple]: Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. If
specified, checks the <em>positive</em> attribute of that
specific directionality. If not specified, checks both.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[bool]: <code class="docutils literal notranslate"><span class="pre">True</span></code> if any interaction (or the specified
<em>direction</em>) is activatory (positive).</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.majority_dir">
<code class="descname">majority_dir</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.majority_dir"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.majority_dir" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Infers which is the major directionality of the edge by number
of supporting sources.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[tuple]: Contains the pair of nodes denoting the consensus
directionality. If the number of sources on both
directions is equal, <code class="docutils literal notranslate"><span class="pre">None</span></code> is returned. If there is no
directionality information, <code class="docutils literal notranslate"><span class="pre">'undirected'`</span></code> will be
returned.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.majority_sign">
<code class="descname">majority_sign</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.majority_sign"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.majority_sign" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Infers which is the major sign (activation/inhibition) of the
edge by number of supporting sources on both directions.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[dict]: Keys are the node tuples on both directions
(straight/reverse) and values can be either <code class="docutils literal notranslate"><span class="pre">None</span></code> if
that direction has no sign information or a list of
two [bool] elements corresponding to majority of positive
and majority of negative support. In case both elements of
the list are <code class="docutils literal notranslate"><span class="pre">True</span></code>, this means the number of supporting
sources for both signs in that direction is equal.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.merge">
<code class="descname">merge</code><span class="sig-paren">(</span><em>other</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.merge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.merge" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Merges current edge with another (if and only if they are the
same class and contain the same nodes). Updates the attributes
<em>dirs</em>, <em>sources</em>, <em>positive</em>, <em>negative</em>, <em>positive_sources</em>
and <em>negative_sources</em>.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Arguments:</dt>
<dd><ul class="first last">
<li><em>other</em> [pypath.main.Direction]: The new edge object to be
merged with the current one.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.negative_reverse">
<code class="descname">negative_reverse</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.negative_reverse"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.negative_reverse" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Checks if the reverse directionality is a negative interaction.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Returns</dt>
<dd><ul class="first last">
<li>[bool]: <code class="docutils literal notranslate"><span class="pre">True</span></code> if there is supporting information on the
reverse direction of the edge as inhibition. <code class="docutils literal notranslate"><span class="pre">False</span></code>
otherwise.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.negative_sources_reverse">
<code class="descname">negative_sources_reverse</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.negative_sources_reverse"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.negative_sources_reverse" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Retrieves the list of sources for the reverse direction and
negative sign.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[set]: Contains the names of the sources supporting the
reverse directionality of the edge with a negative sign.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.negative_sources_straight">
<code class="descname">negative_sources_straight</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.negative_sources_straight"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.negative_sources_straight" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Retrieves the list of sources for the straight direction and
negative sign.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[set]: Contains the names of the sources supporting the
straight directionality of the edge with a negative sign.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.negative_straight">
<code class="descname">negative_straight</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.negative_straight"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.negative_straight" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Checks if the straight directionality is a negative interaction.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Returns</dt>
<dd><ul class="first last">
<li>[bool]: <code class="docutils literal notranslate"><span class="pre">True</span></code> if there is supporting information on the
straight direction of the edge as inhibition. <code class="docutils literal notranslate"><span class="pre">False</span></code>
otherwise.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.positive_reverse">
<code class="descname">positive_reverse</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.positive_reverse"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.positive_reverse" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Checks if the reverse directionality is a positive interaction.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Returns</dt>
<dd><ul class="first last">
<li>[bool]: <code class="docutils literal notranslate"><span class="pre">True</span></code> if there is supporting information on the
reverse direction of the edge as activation. <code class="docutils literal notranslate"><span class="pre">False</span></code>
otherwise.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.positive_sources_reverse">
<code class="descname">positive_sources_reverse</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.positive_sources_reverse"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.positive_sources_reverse" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Retrieves the list of sources for the reverse direction and
positive sign.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[set]: Contains the names of the sources supporting the
reverse directionality of the edge with a positive sign.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.positive_sources_straight">
<code class="descname">positive_sources_straight</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.positive_sources_straight"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.positive_sources_straight" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Retrieves the list of sources for the straight direction and
positive sign.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[set]: Contains the names of the sources supporting the
straight directionality of the edge with a positive sign.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.positive_straight">
<code class="descname">positive_straight</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.positive_straight"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.positive_straight" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Checks if the straight directionality is a positive interaction.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Returns</dt>
<dd><ul class="first last">
<li>[bool]: <code class="docutils literal notranslate"><span class="pre">True</span></code> if there is supporting information on the
straight direction of the edge as activation. <code class="docutils literal notranslate"><span class="pre">False</span></code>
otherwise.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.reload">
<code class="descname">reload</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.reload"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.reload" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Reloads the object from the module level.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.set_dir">
<code class="descname">set_dir</code><span class="sig-paren">(</span><em>direction</em>, <em>source</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.set_dir"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.set_dir" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Adds directionality information with the corresponding data
source named. Modifies self attributes <em>dirs</em> and <em>sources</em>.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Arguments:</dt>
<dd><ul class="first last">
<li><em>direction</em> [tuple]: Or [str], the directionality key for
which the value on <em>dirs</em> has to be set <code class="docutils literal notranslate"><span class="pre">True</span></code>.</li>
<li><em>source</em> [set]: Contains the name(s) of the source(s) from
which such information was obtained.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.set_sign">
<code class="descname">set_sign</code><span class="sig-paren">(</span><em>direction</em>, <em>sign</em>, <em>source</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.set_sign"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.set_sign" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Sets sign and source information on a given direction of the
edge. Modifies the attributes <em>positive</em> and <em>positive_sources</em>
or <em>negative</em> and <em>negative_sources</em> depending on the sign.
Direction is also updated accordingly, which also modifies the
attributes <em>dirs</em>, <em>sources</em>,</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Arguments:</dt>
<dd><ul class="first last">
<li><em>direction</em> [tuple]: Pair of edge nodes specifying the
direction from which the information is to be set/updated.</li>
<li><em>sign</em> [str]: Specifies the type of interaction. If
<code class="docutils literal notranslate"><span class="pre">'positive'</span></code>, is considered activation, otherwise, is
assumed to be negative (inhibition).</li>
<li><em>source</em> [set]: Contains the name(s) of the source(s) from
which the information was obtained.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.sources_reverse">
<code class="descname">sources_reverse</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.sources_reverse"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.sources_reverse" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Retrieves the list of sources for the reverse direction.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[set]: Contains the names of the sources supporting the
reverse directionality of the edge.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.sources_straight">
<code class="descname">sources_straight</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.sources_straight"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.sources_straight" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Retrieves the list of sources for the straight direction.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[set]: Contains the names of the sources supporting the
straight directionality of the edge.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.sources_undirected">
<code class="descname">sources_undirected</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.sources_undirected"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.sources_undirected" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Retrieves the list of sources without directed information.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[set]: Contains the names of the sources supporting the
edge presence but without specific directionality
information.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.src">
<code class="descname">src</code><span class="sig-paren">(</span><em>undirected=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.src"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.src" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns the name(s) of the source node(s) for each existing
direction on the interaction.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Arguments:</dt>
<dd><ul class="first last">
<li><em>undirected</em> [bool]: Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[list]: Contains the name(s) for the source node(s). This
means if the interaction is bidirectional, the list will
contain both identifiers of the edge. If the interaction
is undirected, an empty list will be returned.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.src_by_source">
<code class="descname">src_by_source</code><span class="sig-paren">(</span><em>source</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.src_by_source"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.src_by_source" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns the name(s) of the source node(s) for each existing
direction on the interaction for a specific <em>source</em>.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Arguments:</dt>
<dd><ul class="first last">
<li><em>source</em> [str]: Name of the source according to which the
information is to be retrieved.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[list]: Contains the name(s) for the source node(s)
according to the specified <em>source</em>. This means if the
interaction is bidirectional, the list will contain both
identifiers of the edge. If the specified <em>source</em> is not
found or invalid, an empty list will be returned.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.tgt">
<code class="descname">tgt</code><span class="sig-paren">(</span><em>undirected=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.tgt"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.tgt" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns the name(s) of the target node(s) for each existing
direction on the interaction.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Arguments:</dt>
<dd><ul class="first last">
<li><em>undirected</em> [bool]: Optional, <code class="docutils literal notranslate"><span class="pre">False</span></code> by default.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[list]: Contains the name(s) for the target node(s). This
means if the interaction is bidirectional, the list will
contain both identifiers of the edge. If the interaction
is undirected, an empty list will be returned.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.tgt_by_source">
<code class="descname">tgt_by_source</code><span class="sig-paren">(</span><em>source</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.tgt_by_source"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.tgt_by_source" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns the name(s) of the target node(s) for each existing
direction on the interaction for a specific <em>source</em>.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Arguments:</dt>
<dd><ul class="first last">
<li><em>source</em> [str]: Name of the source according to which the
information is to be retrieved.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[list]: Contains the name(s) for the target node(s)
according to the specified <em>source</em>. This means if the
interaction is bidirectional, the list will contain both
identifiers of the edge. If the specified <em>source</em> is not
found or invalid, an empty list will be returned.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.translate">
<code class="descname">translate</code><span class="sig-paren">(</span><em>ids</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.translate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.translate" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Translates the node names/identifiers according to the
dictionary <em>ids</em>.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Arguments:</dt>
<dd><ul class="first last">
<li><em>ids</em> [dict]: Dictionary containing (at least) the current
names of the nodes as keys and their translation as
values.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[pypath.main.Direction]: The copy of current edge object
with translated node names.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.unset_dir">
<code class="descname">unset_dir</code><span class="sig-paren">(</span><em>direction</em>, <em>source=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.unset_dir"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.unset_dir" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Removes directionality and/or source information of the
specified <em>direction</em>. Modifies attribute <em>dirs</em> and <em>sources</em>
(or <em>sources</em> only if <code class="docutils literal notranslate"><span class="pre">source=True</span></code>).</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Arguments:</dt>
<dd><ul class="first last">
<li><em>direction</em> [tuple]: Or [str] (if <code class="docutils literal notranslate"><span class="pre">'undirected'</span></code>) the
pair of nodes specifying the directionality from which the
information is to be removed.</li>
<li><em>source</em> [set]: Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. If
specified, determines which source(s) is(are) to be
removed from <em>sources</em> attribute in the specified
<em>direction</em>.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.unset_sign">
<code class="descname">unset_sign</code><span class="sig-paren">(</span><em>direction</em>, <em>sign</em>, <em>source=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.unset_sign"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.unset_sign" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Removes sign and/or source information of the specified
<em>direction</em> and <em>sign</em>. Modifies attribute <em>positive</em> and
<em>positive_sources</em> or <em>negative</em> and <em>negative_sources</em> (or
<em>positive_sources</em>/<em>negative_sources</em> only if <code class="docutils literal notranslate"><span class="pre">source=True</span></code>).</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Arguments:</dt>
<dd><ul class="first last">
<li><em>direction</em> [tuple]: The pair of nodes specifying the
directionality from which the information is to be
removed.</li>
<li><em>sign</em> [str]: Sign from which the information is to be
removed. Must be either <code class="docutils literal notranslate"><span class="pre">'positive'</span></code> or <code class="docutils literal notranslate"><span class="pre">'negative'</span></code>.</li>
<li><em>source</em> [set]: Optional, <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. If
specified, determines which source(s) is(are) to be
removed from the sources in the specified <em>direction</em> and
<em>sign</em>.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.which_dirs">
<code class="descname">which_dirs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.which_dirs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.which_dirs" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns the pair(s) of nodes for which there is information
about their directionality.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Returns:</dt>
<dd><ul class="first last">
<li>[list]: List of tuples containing the nodes for which
theit attribute <em>dirs</em> is <code class="docutils literal notranslate"><span class="pre">True</span></code>.</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd></dl>

</dd></dl>



          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter"><em>pypath:</em> A Python module for molecular interaction networks</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/main.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="index.html" title="pypath: A Python module for molecular interaction networks"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">pypath 0.7.97 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2018, DÃ©nes TÃ¼rei.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.7.5.
    </div>
  </body>
</html>