%% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\ifdefined\pdfpxdimen
   \let\sphinxpxdimen\pdfpxdimen\else\newdimen\sphinxpxdimen
\fi \sphinxpxdimen=.75bp\relax

\PassOptionsToPackage{warn}{textcomp}
\usepackage[utf8]{inputenc}
\ifdefined\DeclareUnicodeCharacter
 \ifdefined\DeclareUnicodeCharacterAsOptional
  \DeclareUnicodeCharacter{"00A0}{\nobreakspace}
  \DeclareUnicodeCharacter{"2500}{\sphinxunichar{2500}}
  \DeclareUnicodeCharacter{"2502}{\sphinxunichar{2502}}
  \DeclareUnicodeCharacter{"2514}{\sphinxunichar{2514}}
  \DeclareUnicodeCharacter{"251C}{\sphinxunichar{251C}}
  \DeclareUnicodeCharacter{"2572}{\textbackslash}
 \else
  \DeclareUnicodeCharacter{00A0}{\nobreakspace}
  \DeclareUnicodeCharacter{2500}{\sphinxunichar{2500}}
  \DeclareUnicodeCharacter{2502}{\sphinxunichar{2502}}
  \DeclareUnicodeCharacter{2514}{\sphinxunichar{2514}}
  \DeclareUnicodeCharacter{251C}{\sphinxunichar{251C}}
  \DeclareUnicodeCharacter{2572}{\textbackslash}
 \fi
\fi
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{amsmath,amssymb,amstext}
\usepackage{babel}
\usepackage{times}
\usepackage[Bjarne]{fncychap}
\usepackage{sphinx}

\usepackage{geometry}

% Include hyperref last.
\usepackage{hyperref}
% Fix anchor placement for figures with captions.
\usepackage{hypcap}% it must be loaded after hyperref.
% Set up styles of URL: it should be placed after hyperref.
\urlstyle{same}
\addto\captionsenglish{\renewcommand{\contentsname}{Contents:}}

\addto\captionsenglish{\renewcommand{\figurename}{Fig.}}
\addto\captionsenglish{\renewcommand{\tablename}{Table}}
\addto\captionsenglish{\renewcommand{\literalblockname}{Listing}}

\addto\captionsenglish{\renewcommand{\literalblockcontinuedname}{continued from previous page}}
\addto\captionsenglish{\renewcommand{\literalblockcontinuesname}{continues on next page}}

\addto\extrasenglish{\def\pageautorefname{page}}

\setcounter{tocdepth}{1}



\title{pypath Documentation}
\date{Oct 22, 2018}
\release{0.7.97}
\author{Dénes Türei}
\newcommand{\sphinxlogo}{\vbox{}}
\renewcommand{\releasename}{Release}
\makeindex

\begin{document}

\maketitle
\sphinxtableofcontents
\phantomsection\label{\detokenize{index::doc}}

\begin{quote}\begin{description}
\item[{note}] \leavevmode
\sphinxcode{\sphinxupquote{pypath}} supports both Python 2.7 and Python 3.6+. In the beginning,
pypath has been developed only for Python 2.7. Then the code have been
adjusted to Py3 however we can not guarantee no incompatibilities
remained. If you find any method does not work please submit an issue on
github. For few years I develop and test \sphinxcode{\sphinxupquote{pypath}} in Python 3. Therefore
this is the better supported Python variant.

\item[{contributions}] \leavevmode
\sphinxhref{mailto:turei.denes@gmail.com}{turei.denes@gmail.com}

\item[{documentation}] \leavevmode
\sphinxurl{http://pypath.omnipathdb.org/}

\item[{issues}] \leavevmode
\sphinxurl{https://github.com/saezlab/pypath/issues}

\end{description}\end{quote}

\sphinxstylestrong{pypath} is a Python package built around igraph to work with molecular
network representations e.g. protein, miRNA and drug compound interaction
networks.


\chapter{Webservice}
\label{\detokenize{index:webservice}}
\sphinxstylestrong{New webservice} from 14 June 2018: the queries slightly changed, have been
largely extended. See the examples below.

One instance of the pypath webservice runs at the domain
\sphinxurl{http://omnipathdb.org/}, serving not only the OmniPath data but other datasets:
TF-target interactions from TF Regulons, a large collection additional
enzyme-substrate interactions, and literature curated miRNA-mRNA interacions
combined from 4 databases. The webservice implements a very simple REST style
API, you can make requests by HTTP protocol (browser, wget, curl or whatever).

The webservice currently recognizes 3 types of queries: \sphinxcode{\sphinxupquote{interactions}},
\sphinxcode{\sphinxupquote{ptms}} and \sphinxcode{\sphinxupquote{info}}. The query types \sphinxcode{\sphinxupquote{resources}}, \sphinxcode{\sphinxupquote{network}} and
\sphinxcode{\sphinxupquote{about}} have not been implemented yet in the new webservice.


\section{Mouse and rat}
\label{\detokenize{index:mouse-and-rat}}
Except the miRNA interactions all interactions are available for human, mouse
and rat. The rodent data has been translated from human using the NCBI
Homologene database. Many human proteins have no known homolog in rodents
hence rodent datasets are smaller than their human counterparts. Note, if you
work with mouse omics data you might do better to translate your dataset to
human (for example using the \sphinxcode{\sphinxupquote{pypath.homology}} module) and use human
interaction data.


\section{Examples}
\label{\detokenize{index:examples}}
A request without any parameter, gives some basic numbers about the actual
loaded dataset:
\begin{quote}

\sphinxurl{http://omnipathdb.org}
\end{quote}

The \sphinxcode{\sphinxupquote{info}} returns a HTML page with comprehensive information about the
resources:
\begin{quote}

\sphinxurl{http://omnipathdb.org/info}
\end{quote}

The \sphinxcode{\sphinxupquote{interactions}} query accepts some parameters and returns interactions in
tabular format. This example returns all interactions of EGFR (P00533), with
sources and references listed.
\begin{quote}

\sphinxurl{http://omnipathdb.org/interactions/?partners=P00533\&fields=sources,references}
\end{quote}

By default only the OmniPath dataset used, to query the TF Regulons or add the
extra enzyme-substrate interactions you need to set additional parameters. For
example to query the transcriptional regulators of EGFR:
\begin{quote}

\sphinxurl{http://omnipathdb.org/interactions/?targets=EGFR\&types=TF}
\end{quote}

The TF Regulons database assigns confidence levels to the interactions. You
might want to select only the highest confidence, \sphinxstyleemphasis{A} category:
\begin{quote}

\sphinxurl{http://omnipathdb.org/interactions/?targets=EGFR\&types=TF\&tfregulons\_levels=A}
\end{quote}

Show the transcriptional targets of Smad2 homology translated to rat including
the confidence levels from TF Regulons:
\begin{quote}

\sphinxurl{http://omnipathdb.org/interactions/?genesymbols=1\&fields=type,ncbi\_tax\_id,tfregulons\_level\&organisms=10116\&sources=Smad2\&types=TF}
\end{quote}

Query interactions from PhosphoNetworks which is part of the \sphinxstyleemphasis{kinaseextra}
dataset:
\begin{quote}

\sphinxurl{http://omnipathdb.org/interactions/?genesymbols=1\&fields=sources\&databases=PhosphoNetworks\&datasets=kinaseextra}
\end{quote}

Get the interactions from Signor, SPIKE and SignaLink3:
\begin{quote}

\sphinxurl{http://omnipathdb.org/interactions/?genesymbols=1\&fields=sources,references\&databases=Signor,SPIKE,SignaLink3}
\end{quote}

All interactions of MAP1LC3B:
\begin{quote}

\sphinxurl{http://omnipathdb.org/interactions/?genesymbols=1\&partners=MAP1LC3B}
\end{quote}

By default \sphinxcode{\sphinxupquote{partners}} queries the interaction where either the source or the
arget is among the partners. If you set the \sphinxcode{\sphinxupquote{source\_target}} parameter to
\sphinxcode{\sphinxupquote{AND}} both the source and the target must be in the queried set:
\begin{quote}

\sphinxurl{http://omnipathdb.org/interactions/?genesymbols=1\&fields=sources,references\&sources=ATG3,ATG7,ATG4B,SQSTM1\&targets=MAP1LC3B,MAP1LC3A,MAP1LC3C,Q9H0R8,GABARAP,GABARAPL2\&source\_target=AND}
\end{quote}

As you see above you can use UniProt IDs and Gene Symbols in the queries and
also mix them. Get the miRNA regulating NOTCH1:
\begin{quote}

\sphinxurl{http://omnipathdb.org/interactions/?genesymbols=1\&fields=sources,references\&datasets=mirnatarget\&targets=NOTCH1}
\end{quote}

Note: with the exception of mandatory fields and genesymbols, the columns
appear exactly in the order you provided in your query.

Another query type available is \sphinxcode{\sphinxupquote{ptms}} which provides enzyme-substrate
interactions. It is very similar to the \sphinxcode{\sphinxupquote{interactions}}:
\begin{quote}

\sphinxurl{http://omnipathdb.org/ptms?genesymbols=1\&fields=sources,references,isoforms\&enzymes=FYN}
\end{quote}

Is there any ubiquitination reaction?
\begin{quote}

\sphinxurl{http://omnipathdb.org/ptms?genesymbols=1\&fields=sources,references\&types=ubiquitination}
\end{quote}

And acetylation in mouse?
\begin{quote}

\sphinxurl{http://omnipathdb.org/ptms?genesymbols=1\&fields=sources,references\&types=acetylation\&organisms=10090}
\end{quote}

Rat interactions, both directly from rat and homology translated from human,
from the PhosphoSite database:
\begin{quote}

\sphinxurl{http://omnipathdb.org/ptms?genesymbols=1\&fields=sources,references\&organisms=10116\&databases=PhosphoSite,PhosphoSite\_noref}
\end{quote}


\chapter{Can I use OmniPath in R?}
\label{\detokenize{index:can-i-use-omnipath-in-r}}
You can download the data from the webservice and load into R. Look here for
an example:
\begin{quote}

\sphinxurl{https://github.com/saezlab/pypath/tree/master/r\_import}
\end{quote}


\chapter{Installation}
\label{\detokenize{index:installation}}

\section{Linux}
\label{\detokenize{index:linux}}
In almost any up-to-date Linux distribution the dependencies of \sphinxstylestrong{pypath} are
built-in, or provided by the distributors. You only need to install a couple
of things in your package manager (cairo, py(2)cairo, igraph,
python(2)-igraph, graphviz, pygraphviz), and after install \sphinxstylestrong{pypath} by \sphinxstyleemphasis{pip}
(see below). If any module still missing, you can install them the usual way
by \sphinxstyleemphasis{pip} or your package manager.


\section{igraph C library, cairo and pycairo}
\label{\detokenize{index:igraph-c-library-cairo-and-pycairo}}
\sphinxstyleemphasis{python(2)-igraph} is a Python interface to use the igraph C library. The
C library must be installed. The same goes for \sphinxstyleemphasis{cairo}, \sphinxstyleemphasis{py(2)cairo} and
\sphinxstyleemphasis{graphviz}.


\section{Directly from git}
\label{\detokenize{index:directly-from-git}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{pip} \PYG{n}{install} \PYG{n}{git}\PYG{o}{+}\PYG{n}{https}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{n}{github}\PYG{o}{.}\PYG{n}{com}\PYG{o}{/}\PYG{n}{saezlab}\PYG{o}{/}\PYG{n}{pypath}\PYG{o}{.}\PYG{n}{git}
\end{sphinxVerbatim}


\section{With pip}
\label{\detokenize{index:with-pip}}
Download the package from /dist, and install with pip:

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{pip} \PYG{n}{install} \PYG{n}{pypath}\PYG{o}{\PYGZhy{}}\PYG{n}{x}\PYG{o}{.}\PYG{n}{y}\PYG{o}{.}\PYG{n}{z}\PYG{o}{.}\PYG{n}{tar}\PYG{o}{.}\PYG{n}{gz}
\end{sphinxVerbatim}


\section{Build source distribution}
\label{\detokenize{index:build-source-distribution}}
Clone the git repo, and run setup.py:

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{python} \PYG{n}{setup}\PYG{o}{.}\PYG{n}{py} \PYG{n}{sdist}
\end{sphinxVerbatim}


\section{Mac OS X}
\label{\detokenize{index:mac-os-x}}
On OS X installation is not straightforward primarily because cairo needs to
be compiled from source. We provide 2 scripts here: the
\sphinxstylestrong{mac-install-brew.sh} installs everything with HomeBrew, and
\sphinxstylestrong{mac-install-conda.sh} installs from Anaconda distribution. With these
scripts installation of igraph, cairo and graphviz goes smoothly most of the
time, and options are available for omitting the 2 latter. To know more see
the description in the script header. There is a third script
\sphinxstylestrong{mac-install-source.sh} which compiles everything from source and presumes
only Python 2.7 and Xcode installed. We do not recommend this as it is time
consuming and troubleshooting requires expertise.


\subsection{Troubleshooting}
\label{\detokenize{index:troubleshooting}}\begin{itemize}
\item {} 
\sphinxcode{\sphinxupquote{no module named ...}} when you try to load a module in Python. Did
theinstallation of the module run without error? Try to run again the specific
part from the mac install shell script to see if any error comes up. Is the
path where the module has been installed in your \sphinxcode{\sphinxupquote{\$PYTHONPATH}}? Try \sphinxcode{\sphinxupquote{echo
\$PYTHONPATH}} to see the current paths. Add your local install directories if
those are not there, e.g.
\sphinxcode{\sphinxupquote{export PYTHONPATH="/Users/me/local/python2.7/site-packages:\$PYTHONPATH"}}.
If it works afterwards, don’t forget to append these export path statements to
your \sphinxcode{\sphinxupquote{\textasciitilde{}/.bash\_profile}}, so these will be set every time you launch a new
shell.

\item {} 
\sphinxcode{\sphinxupquote{pkgconfig}} not found. Check if the \sphinxcode{\sphinxupquote{\$PKG\_CONFIG\_PATH}} variable is
set correctly, and pointing on a directory where pkgconfig really can be
found.

\item {} 
Error while trying to install py(2)cairo by pip. py(2)cairo could not be
installed by pip, but only by waf. Please set the \sphinxcode{\sphinxupquote{\$PKG\_CONFIG\_PATH}} before.
See \sphinxstylestrong{mac-install-source.sh} on how to install with waf.

\item {} 
Error at pygraphviz build: \sphinxcode{\sphinxupquote{graphviz/cgraph.h file not found}}. This is
because the directory of graphviz detected wrong by pkgconfig. See
\sphinxstylestrong{mac-install-source.sh} how to set include dirs and library dirs by
\sphinxcode{\sphinxupquote{-{-}global-option}} parameters.

\item {} 
Can not install bioservices, because installation of jurko-suds fails. Ok,
this fails because pip is not able to install the recent version of
setuptools, because a very old version present in the system path. The
development version of jurko-suds does not require setuptools, so you can
install it directly from git as it is done in \sphinxstylestrong{mac-install-source.sh}.

\item {} 
In \sphinxstylestrong{Anaconda}, \sphinxstyleemphasis{pypath} can be imported, but the modules and classes are
missing. Apparently Anaconda has some built-in stuff called \sphinxstyleemphasis{pypath}. This
has nothing to do with this module. Please be aware that Anaconda installs a
completely separated Python distribution, and does not detect modules in the
main Python installation. You need to install all modules within Anaconda’s
directory. \sphinxstylestrong{mac-install-conda.sh} does exactly this. If you still
experience issues, please contact us.

\end{itemize}


\section{Microsoft Windows}
\label{\detokenize{index:microsoft-windows}}
Not many people have used \sphinxstyleemphasis{pypath} on Microsoft computers so far. Please share
your experiences and contact us if you encounter any issue. We appreciate
your feedback, and it would be nice to have better support for other computer
systems.


\subsection{With Anaconda}
\label{\detokenize{index:with-anaconda}}
The same workflow like you see in \sphinxcode{\sphinxupquote{mac-install-conda.sh}} should work for
Anaconda on Windows. The only problem you certainly will encounter is that not
all the channels have packages for all platforms. If certain channel provides
no package for Windows, or for your Python version, you just need to find an
other one. For this, do a search:

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{anaconda} \PYG{n}{search} \PYG{o}{\PYGZhy{}}\PYG{n}{t} \PYG{n}{conda} \PYG{o}{\PYGZlt{}}\PYG{n}{package} \PYG{n}{name}\PYG{o}{\PYGZgt{}}
\end{sphinxVerbatim}

For example, if you search for \sphinxstyleemphasis{pycairo}, you will find out that \sphinxstyleemphasis{vgauther}
provides it for osx-64, but only for Python 3.4, while \sphinxstyleemphasis{richlewis} provides
also for Python 3.5. And for win-64 platform, there is the channel of
\sphinxstyleemphasis{KristanAmstrong}. Go along all the commands in \sphinxcode{\sphinxupquote{mac-install-conda.sh}}, and
modify the channel if necessary, until all packages install successfully.


\subsection{With other Python distributions}
\label{\detokenize{index:with-other-python-distributions}}
Here the basic principles are the same as everywhere: first try to install all
external dependencies, after \sphinxstyleemphasis{pip} install should work. On Windows certain
packages can not be installed by compiled from source by \sphinxstyleemphasis{pip}, instead the
easiest to install them precompiled. These are in our case \sphinxstyleemphasis{fisher, lxml,
numpy (mkl version), pycairo, igraph, pygraphviz, scipy and statsmodels}. The
precompiled packages are available here:
\sphinxurl{http://www.lfd.uci.edu/~gohlke/pythonlibs/}. We tested the setup with Python
3.4.3 and Python 2.7.11. The former should just work fine, while with the
latter we have issues to be resolved.


\subsection{Known issues}
\label{\detokenize{index:known-issues}}\begin{itemize}
\item {} 
\sphinxstyleemphasis{“No module fabric available.”} \textendash{} or \sphinxstyleemphasis{pysftp} missing: this is not
important, only certain data download methods rely on these modules, but
likely you won’t call those at all.

\item {} 
Progress indicator floods terminal: sorry about that, will be fixed soon.

\item {} 
Encoding related exceptions in Python2: these might occur at some points in
the module, please send the traceback if you encounter one, and we will fix
as soon as possible.

\end{itemize}

\sphinxstyleemphasis{Special thanks to Jorge Ferreira for testing pypath on Windows!}


\chapter{Release History}
\label{\detokenize{index:release-history}}
Main improvements in the past releases:


\section{0.1.0:}
\label{\detokenize{index:id1}}\begin{itemize}
\item {} 
First release of pypath, for initial testing.

\end{itemize}


\section{0.2.0:}
\label{\detokenize{index:id2}}\begin{itemize}
\item {} 
Lots of small improvements in almost every module

\item {} 
Networks can be read from local files, remote files, lists or provided by
any function

\item {} 
Almost all redistributed data have been removed, every source downloaded
from the original provider.

\end{itemize}


\section{0.3.0:}
\label{\detokenize{index:id3}}\begin{itemize}
\item {} 
First version whith partial Python 3 support.

\end{itemize}


\section{0.4.0:}
\label{\detokenize{index:id4}}\begin{itemize}
\item {} 
\sphinxstylestrong{pyreact} module with \sphinxstylestrong{BioPaxReader} and \sphinxstylestrong{PyReact} classes added

\item {} 
Process description databases, BioPax and PathwayCommons SIF conversion
rules are supported

\item {} 
Format definitions for 6 process description databases included.

\end{itemize}


\section{0.5.0:}
\label{\detokenize{index:id5}}\begin{itemize}
\item {} 
Many classes have been added to the \sphinxstylestrong{plot} module

\item {} 
All figures and tables in the manuscript can be generated automatically

\item {} 
This is supported by a new module, \sphinxstylestrong{analysis}, which implements a generic

\end{itemize}

workflow in its \sphinxstylestrong{Workflow} class.


\section{0.7.74:}
\label{\detokenize{index:id6}}\begin{itemize}
\item {} 
\sphinxstylestrong{homology} module: finds the homologs of proteins using the NCBI

\end{itemize}

Homologene database and the homologs of PTM sites using UniProt sequences
and PhosphoSitePlus homology table
* \sphinxstylestrong{ptm} module: fully integrated way of processing enzyme-substrate
interactions from many databases and their translation by homology to other
species
* \sphinxstylestrong{export} module: creates \sphinxcode{\sphinxupquote{pandas.DataFrame}} or exports the network into
tabular file
* New webservice
* TF Regulons database included and provides much more comprehensive
transcriptional regulation resources, including literature curated, in silico
predicted, ChIP-Seq and expression pattern based approaches
* Many network resources added, including miRNA-mRNA and TF-miRNA interactions


\section{Upcoming:}
\label{\detokenize{index:upcoming}}\begin{itemize}
\item {} 
New, more flexible network reader class

\item {} 
Full support for multi-species molecular interaction networks

\end{itemize}

(e.g. pathogene-host)
* Better support for not protein only molecular interaction networks
(metabolites, drug compounds, RNA)
* ChEMBL webservice interface, interface for PubChem and eventually
forDrugBank
* Silent mode: a way to suppress messages and progress bars


\chapter{Features}
\label{\detokenize{index:features}}
The primary aim of \sphinxstylestrong{pypath} is to build up networks from multiple sources on
one igraph object. \sphinxstylestrong{pypath} handles ambiguous ID conversion, reads custom
edge and node attributes from text files and \sphinxstylestrong{MySQL}.

Submodules perform various features, e.g. graph visualization, working with
rug compound data, searching drug targets and compounds in \sphinxstylestrong{ChEMBL}.


\section{ID conversion}
\label{\detokenize{index:id-conversion}}
The ID conversion module \sphinxcode{\sphinxupquote{mapping}} can be used independently. It has the
feature to translate secondary UniProt IDs to primaries, and Trembl IDs to
SwissProt, using primary Gene Symbols to find the connections. This module
automatically loads and stores the necessary conversion tables. Many tables
are predefined, such as all the IDs in \sphinxstylestrong{UniProt mapping service,} while
users are able to load any table from \sphinxstylestrong{file} or \sphinxstylestrong{MySQL,} using the classes
provided in the module \sphinxcode{\sphinxupquote{input\_formats}}.


\section{Pathways}
\label{\detokenize{index:pathways}}
\sphinxstylestrong{pypath} includes data and predefined format descriptions for more than 25
high quality, literature curated databases. The inut formats are defined in
the \sphinxcode{\sphinxupquote{data\_formats}} module. For some resources data downloaded on the fly,
where it is not possible, data is redistributed with the module. Descriptions
and comprehensive information about the resources is available in the
\sphinxcode{\sphinxupquote{descriptions}} module.


\section{Structural features}
\label{\detokenize{index:structural-features}}
One of the modules called \sphinxcode{\sphinxupquote{intera}} provides many classes for representing
structures and mechanisms behind protein interactions. These are \sphinxcode{\sphinxupquote{Residue}}
(optionally mutated), \sphinxcode{\sphinxupquote{Motif}}, \sphinxcode{\sphinxupquote{Ptm}}, \sphinxcode{\sphinxupquote{Domain}}, \sphinxcode{\sphinxupquote{DomainMotif}},
\sphinxcode{\sphinxupquote{DomainDomain}} and \sphinxcode{\sphinxupquote{Interface}}. All these classes have \sphinxcode{\sphinxupquote{\_\_eq\_\_()}}
methods to test equality between instances, and also \sphinxcode{\sphinxupquote{\_\_contains\_\_()}}
methods to look up easily if a residue is within a short motif or protein
domain, or is the target residue of a PTM.


\section{Sequences}
\label{\detokenize{index:sequences}}
The module \sphinxcode{\sphinxupquote{seq}} contains a simple class for quick lookup any residue or
segment in \sphinxstylestrong{UniProt} protein sequences while being aware of isoforms.


\section{Tissue expression}
\label{\detokenize{index:tissue-expression}}
For 3 protein expression databases there are functions and modules for
downloading and combining the expression data with the network. These are the
Human Protein Atlas, the ProteomicsDB and GIANT. The \sphinxcode{\sphinxupquote{giant}} and
\sphinxcode{\sphinxupquote{proteomicsdb}} modules can be used also as stand alone Python clients for
these resources.


\section{Functional annotations}
\label{\detokenize{index:functional-annotations}}
\sphinxstylestrong{GSEA} and \sphinxstylestrong{Gene Ontology} are two approaches for annotating genes and
gene products, and enrichment analysis technics aims to use these annotations
to highlight the biological functions a given set of genes is related to. Here
the \sphinxcode{\sphinxupquote{enrich}} module gives abstract classes to calculate enrichment
statistics, while the \sphinxcode{\sphinxupquote{go}} and the \sphinxcode{\sphinxupquote{gsea}} modules give access to GO and
GSEA data, and make it easy to count enrichment statistics for sets of genes.


\section{Drug compounds}
\label{\detokenize{index:drug-compounds}}
\sphinxstylestrong{UniChem} submodule provides an interface to effectively query the UniChem
service, use connectivity search with custom settings, and translate SMILEs to
ChEMBL IDs with ChEMBL web service.

\sphinxstylestrong{ChEMBL} submodule queries directly your own ChEMBL MySQL instance, has the
features to search targets and compounds from custom assay types and
relationship types, to get activity values, binding domains, and action types.
You need to download the ChEMBL MySQL dump, and load into your own server.


\section{Technical}
\label{\detokenize{index:technical}}
\sphinxstylestrong{MySQL} submodule helps to manage MySQL connections and track queries. It is
able to run queries parallely to optimize CPU and memory usage on the server,
handling queues, and serve the result by server side or client side storage.
The \sphinxcode{\sphinxupquote{chembl}} and potentially the \sphinxcode{\sphinxupquote{mapping}} modules rely on this \sphinxcode{\sphinxupquote{mysql}}
module.

The most important function in module \sphinxcode{\sphinxupquote{dataio}} is a very flexible \sphinxstylestrong{download
manager} built around \sphinxcode{\sphinxupquote{curl}}. The function \sphinxcode{\sphinxupquote{dataio.curl()}} accepts
numerous arguments, tries to deal in a smart way with local \sphinxstylestrong{cache,}
authentication, redirects, uncompression, character encodings, FTP and HTTP
transactions, and many other stuff. Cache can grow to several GBs, and takes
place in \sphinxcode{\sphinxupquote{./cache}} by default. Please be aware of this, and use for example
symlinks in case of using multiple working directories.

A simple \sphinxstylestrong{webservice} comes with this module: the \sphinxcode{\sphinxupquote{server}} module based on
\sphinxcode{\sphinxupquote{twisted.web.server}} opens a custom port and serves plain text tables over
HTTP with REST style querying.



\renewcommand{\indexname}{Index}
\printindex
\end{document}