%% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\ifdefined\pdfpxdimen
   \let\sphinxpxdimen\pdfpxdimen\else\newdimen\sphinxpxdimen
\fi \sphinxpxdimen=.75bp\relax

\PassOptionsToPackage{warn}{textcomp}
\usepackage[utf8]{inputenc}
\ifdefined\DeclareUnicodeCharacter
 \ifdefined\DeclareUnicodeCharacterAsOptional
  \DeclareUnicodeCharacter{"00A0}{\nobreakspace}
  \DeclareUnicodeCharacter{"2500}{\sphinxunichar{2500}}
  \DeclareUnicodeCharacter{"2502}{\sphinxunichar{2502}}
  \DeclareUnicodeCharacter{"2514}{\sphinxunichar{2514}}
  \DeclareUnicodeCharacter{"251C}{\sphinxunichar{251C}}
  \DeclareUnicodeCharacter{"2572}{\textbackslash}
 \else
  \DeclareUnicodeCharacter{00A0}{\nobreakspace}
  \DeclareUnicodeCharacter{2500}{\sphinxunichar{2500}}
  \DeclareUnicodeCharacter{2502}{\sphinxunichar{2502}}
  \DeclareUnicodeCharacter{2514}{\sphinxunichar{2514}}
  \DeclareUnicodeCharacter{251C}{\sphinxunichar{251C}}
  \DeclareUnicodeCharacter{2572}{\textbackslash}
 \fi
\fi
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{amsmath,amssymb,amstext}
\usepackage{babel}
\usepackage{times}
\usepackage[Bjarne]{fncychap}
\usepackage{sphinx}

\usepackage{geometry}

% Include hyperref last.
\usepackage{hyperref}
% Fix anchor placement for figures with captions.
\usepackage{hypcap}% it must be loaded after hyperref.
% Set up styles of URL: it should be placed after hyperref.
\urlstyle{same}
\addto\captionsenglish{\renewcommand{\contentsname}{Contents:}}

\addto\captionsenglish{\renewcommand{\figurename}{Fig.}}
\addto\captionsenglish{\renewcommand{\tablename}{Table}}
\addto\captionsenglish{\renewcommand{\literalblockname}{Listing}}

\addto\captionsenglish{\renewcommand{\literalblockcontinuedname}{continued from previous page}}
\addto\captionsenglish{\renewcommand{\literalblockcontinuesname}{continues on next page}}

\addto\extrasenglish{\def\pageautorefname{page}}

\setcounter{tocdepth}{1}



\title{pypath Documentation}
\date{Oct 22, 2018}
\release{0.7.97}
\author{Dénes Türei}
\newcommand{\sphinxlogo}{\vbox{}}
\renewcommand{\releasename}{Release}
\makeindex

\begin{document}

\maketitle
\sphinxtableofcontents
\phantomsection\label{\detokenize{index::doc}}

\begin{itemize}
\item {} 
\DUrole{xref,std,std-ref}{genindex}

\item {} 
\DUrole{xref,std,std-ref}{modindex}

\item {} 
\DUrole{xref,std,std-ref}{search}

\end{itemize}
\phantomsection\label{\detokenize{main:module-pypath.main}}\index{pypath.main (module)}\index{PyPath (class in pypath.main)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{pypath.main.}}\sphinxbfcode{\sphinxupquote{PyPath}}}{\emph{ncbi\_tax\_id=9606}, \emph{default\_name\_type=\{'drug': 'chembl'}, \emph{'lncrna': 'lncrna-genesymbol'}, \emph{'mirna': 'mirbase'}, \emph{'protein': 'uniprot'\}}, \emph{copy=None}, \emph{mysql=(None}, \emph{'mapping')}, \emph{chembl\_mysql=(None}, \emph{'chembl')}, \emph{name='unnamed'}, \emph{outdir='results'}, \emph{loglevel='INFO'}, \emph{loops=False}}{}~\begin{itemize}
\item {} \begin{description}
\item[{Attributes:}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleemphasis{acsn\_effects} {[}{]}:

\item {} 
\sphinxstyleemphasis{add\_genesets} {[}{]}:

\item {} 
\sphinxstyleemphasis{add\_grouped\_eattr} {[}{]}:

\item {} 
\sphinxstyleemphasis{add\_grouped\_set\_eattr} {[}{]}:

\item {} 
\sphinxstyleemphasis{add\_list\_eattr} {[}{]}:

\item {} 
\sphinxstyleemphasis{add\_set\_eattr} {[}{]}:

\item {} 
\sphinxstyleemphasis{add\_update\_edge} {[}{]}:

\item {} 
\sphinxstyleemphasis{add\_update\_vertex} {[}{]}:

\item {} 
\sphinxstyleemphasis{affected\_by} {[}{]}:

\item {} 
\sphinxstyleemphasis{affects} {[}{]}:

\item {} 
\sphinxstyleemphasis{all\_between} {[}{]}:

\item {} 
\sphinxstyleemphasis{all\_neighbours} {[}{]}:

\item {} 
\sphinxstyleemphasis{apply\_list} {[}{]}:

\item {} 
\sphinxstyleemphasis{apply\_negative} {[}{]}:

\item {} 
\sphinxstyleemphasis{attach\_network} {[}{]}:

\item {} 
\sphinxstyleemphasis{basic\_stats} {[}{]}:

\item {} 
\sphinxstyleemphasis{basic\_stats\_intergroup} {[}{]}:

\item {} 
\sphinxstyleemphasis{cancer\_drivers\_list} {[}{]}:

\item {} 
\sphinxstyleemphasis{cancer\_gene\_census\_list} {[}{]}:

\item {} 
\sphinxstyleemphasis{clean\_graph} {[}{]}:

\item {} 
\sphinxstyleemphasis{collapse\_by\_name} {[}{]}:

\item {} 
\sphinxstyleemphasis{combine\_attr} {[}{]}:

\item {} 
\sphinxstyleemphasis{communities} {[}{]}:

\item {} 
\sphinxstyleemphasis{complex\_comembership\_network} {[}{]}:

\item {} 
\sphinxstyleemphasis{complexes} {[}{]}:

\item {} 
\sphinxstyleemphasis{complexes\_in\_network} {[}{]}:

\item {} 
\sphinxstyleemphasis{compounds\_from\_chembl} {[}{]}:

\item {} 
\sphinxstyleemphasis{consistency} {[}{]}:

\item {} 
\sphinxstyleemphasis{copy} {[}{]}:

\item {} 
\sphinxstyleemphasis{copy\_edges} {[}{]}:

\item {} 
\sphinxstyleemphasis{count\_sol} {[}{]}:

\item {} 
\sphinxstyleemphasis{coverage} {[}{]}:

\item {} 
\sphinxstyleemphasis{curation\_effort} {[}{]}:

\item {} 
\sphinxstyleemphasis{curation\_stats} {[}{]}:

\item {} 
\sphinxstyleemphasis{curation\_tab} {[}{]}:

\item {} 
\sphinxstyleemphasis{curators\_work} {[}{]}:

\item {} 
\sphinxstyleemphasis{databases\_similarity} {[}{]}:

\item {} 
\sphinxstyleemphasis{default\_name\_type} {[}{]}:

\item {} 
\sphinxstyleemphasis{degree\_dist} {[}{]}:

\item {} 
\sphinxstyleemphasis{degree\_dists} {[}{]}:

\item {} 
\sphinxstyleemphasis{delete\_by\_source} {[}{]}:

\item {} 
\sphinxstyleemphasis{delete\_by\_taxon} {[}{]}:

\item {} 
\sphinxstyleemphasis{delete\_unknown} {[}{]}:

\item {} 
\sphinxstyleemphasis{delete\_unmapped} {[}{]}:

\item {} 
\sphinxstyleemphasis{dgenesymbol} {[}{]}:

\item {} 
\sphinxstyleemphasis{dgenesymbols} {[}{]}:

\item {} 
\sphinxstyleemphasis{dgs} {[}{]}:

\item {} 
\sphinxstyleemphasis{dgss} {[}{]}:

\item {} 
\sphinxstyleemphasis{disease\_genes\_list} {[}{]}:

\item {} 
\sphinxstyleemphasis{dneighbors} {[}{]}:

\item {} 
\sphinxstyleemphasis{dp} {[}{]}:

\item {} 
\sphinxstyleemphasis{dproteins} {[}{]}:

\item {} 
\sphinxstyleemphasis{dps} {[}{]}:

\item {} 
\sphinxstyleemphasis{druggability\_list} {[}{]}:

\item {} 
\sphinxstyleemphasis{duniprot} {[}{]}:

\item {} 
\sphinxstyleemphasis{duniprots} {[}{]}:

\item {} 
\sphinxstyleemphasis{dup} {[}{]}:

\item {} 
\sphinxstyleemphasis{dups} {[}{]}:

\item {} 
\sphinxstyleemphasis{dv} {[}{]}:

\item {} 
\sphinxstyleemphasis{dvs} {[}{]}:

\item {} 
\sphinxstyleemphasis{edge\_exists} {[}{]}:

\item {} 
\sphinxstyleemphasis{edge\_loc} {[}{]}:

\item {} 
\sphinxstyleemphasis{edge\_names} {[}{]}:

\item {} 
\sphinxstyleemphasis{edges\_3d} {[}{]}:

\item {} 
\sphinxstyleemphasis{edges\_expression} {[}{]}:

\item {} 
\sphinxstyleemphasis{edges\_in\_comlexes} {[}{]}:

\item {} 
\sphinxstyleemphasis{edges\_ptms} {[}{]}:

\item {} 
\sphinxstyleemphasis{edgeseq\_inverse} {[}{]}:

\item {} 
\sphinxstyleemphasis{export\_dot} {[}{]}:

\item {} 
\sphinxstyleemphasis{export\_edgelist} {[}{]}:

\item {} 
\sphinxstyleemphasis{export\_graphml} {[}{]}:

\item {} 
\sphinxstyleemphasis{export\_ptms\_tab} {[}{]}:

\item {} 
\sphinxstyleemphasis{export\_sif} {[}{]}:

\item {} 
\sphinxstyleemphasis{export\_struct\_tab} {[}{]}:

\item {} 
\sphinxstyleemphasis{export\_tab} {[}{]}:

\item {} 
\sphinxstyleemphasis{filters} {[}{]}:

\item {} 
\sphinxstyleemphasis{find\_all\_paths} {[}{]}:

\item {} 
\sphinxstyleemphasis{find\_all\_paths2} {[}{]}:

\item {} 
\sphinxstyleemphasis{find\_complex} {[}{]}:

\item {} 
\sphinxstyleemphasis{first\_neighbours} {[}{]}:

\item {} 
\sphinxstyleemphasis{fisher\_enrichment} {[}{]}:

\item {} 
\sphinxstyleemphasis{geneset\_enrichment} {[}{]}:

\item {} 
\sphinxstyleemphasis{genesymbol} {[}{]}:

\item {} 
\sphinxstyleemphasis{genesymbol\_labels} {[}{]}:

\item {} 
\sphinxstyleemphasis{genesymbols} {[}{]}:

\item {} 
\sphinxstyleemphasis{get\_attrs} {[}{]}:

\item {} 
\sphinxstyleemphasis{get\_directed} {[}{]}:

\item {} 
\sphinxstyleemphasis{get\_dirs\_signs} {[}{]}:

\item {} 
\sphinxstyleemphasis{get\_edge} {[}{]}:

\item {} 
\sphinxstyleemphasis{get\_edges} {[}{]}:

\item {} 
\sphinxstyleemphasis{get\_function} {[}{]}:

\item {} 
\sphinxstyleemphasis{get\_giant} {[}{]}:

\item {} 
\sphinxstyleemphasis{get\_max} {[}{]}:

\item {} 
\sphinxstyleemphasis{get\_network} {[}{]}:

\item {} 
\sphinxstyleemphasis{get\_node} {[}{]}:

\item {} 
\sphinxstyleemphasis{get\_node\_d} {[}{]}:

\item {} 
\sphinxstyleemphasis{get\_node\_pair} {[}{]}:

\item {} 
\sphinxstyleemphasis{get\_nodes} {[}{]}:

\item {} 
\sphinxstyleemphasis{get\_nodes\_d} {[}{]}:

\item {} 
\sphinxstyleemphasis{get\_pathways} {[}{]}:

\item {} 
\sphinxstyleemphasis{get\_proteomicsdb} {[}{]}:

\item {} 
\sphinxstyleemphasis{get\_sub} {[}{]}:

\item {} 
\sphinxstyleemphasis{get\_taxon} {[}{]}:

\item {} 
\sphinxstyleemphasis{go\_annotate} {[}{]}:

\item {} 
\sphinxstyleemphasis{go\_dict} {[}{]}:

\item {} 
\sphinxstyleemphasis{go\_enrichment} {[}{]}:

\item {} 
\sphinxstyleemphasis{gs} {[}{]}:

\item {} 
\sphinxstyleemphasis{gs\_affected\_by} {[}{]}:

\item {} 
\sphinxstyleemphasis{gs\_affects} {[}{]}:

\item {} 
\sphinxstyleemphasis{gs\_edge} {[}{]}:

\item {} 
\sphinxstyleemphasis{gs\_in\_directed} {[}{]}:

\item {} 
\sphinxstyleemphasis{gs\_in\_undirected} {[}{]}:

\item {} 
\sphinxstyleemphasis{gs\_inhibited\_by} {[}{]}:

\item {} 
\sphinxstyleemphasis{gs\_inhibits} {[}{]}:

\item {} 
\sphinxstyleemphasis{gs\_neighborhood} {[}{]}:

\item {} 
\sphinxstyleemphasis{gs\_neighbors} {[}{]}:

\item {} 
\sphinxstyleemphasis{gs\_stimulated\_by} {[}{]}:

\item {} 
\sphinxstyleemphasis{gs\_stimulates} {[}{]}:

\item {} 
\sphinxstyleemphasis{gss} {[}{]}:

\item {} 
\sphinxstyleemphasis{guide2pharma} {[}{]}:

\item {} 
\sphinxstyleemphasis{having\_attr} {[}{]}:

\item {} 
\sphinxstyleemphasis{having\_eattr} {[}{]}:

\item {} 
\sphinxstyleemphasis{having\_ptm} {[}{]}:

\item {} 
\sphinxstyleemphasis{having\_vattr} {[}{]}:

\item {} 
\sphinxstyleemphasis{homology\_translation} {[}{]}:

\item {} 
\sphinxstyleemphasis{htp\_stats} {[}{]}:

\item {} 
\sphinxstyleemphasis{in\_complex} {[}{]}:

\item {} 
\sphinxstyleemphasis{in\_directed} {[}{]}:

\item {} 
\sphinxstyleemphasis{in\_undirected} {[}{]}:

\item {} 
\sphinxstyleemphasis{info} {[}{]}:

\item {} 
\sphinxstyleemphasis{init\_complex\_attr} {[}{]}:

\item {} 
\sphinxstyleemphasis{init\_edge\_attr} {[}{]}:

\item {} 
\sphinxstyleemphasis{init\_gsea} {[}{]}:

\item {} 
\sphinxstyleemphasis{init\_network} {[}{]}:

\item {} 
\sphinxstyleemphasis{init\_vertex\_attr} {[}{]}:

\item {} 
\sphinxstyleemphasis{intergroup\_shortest\_paths} {[}{]}:

\item {} 
\sphinxstyleemphasis{intogen\_cancer\_drivers\_list} {[}{]}:

\item {} 
\sphinxstyleemphasis{jaccard\_edges} {[}{]}:

\item {} 
\sphinxstyleemphasis{jaccard\_meta} {[}{]}:

\item {} 
\sphinxstyleemphasis{kegg\_directions} {[}{]}:

\item {} 
\sphinxstyleemphasis{kegg\_pathways} {[}{]}:

\item {} 
\sphinxstyleemphasis{kinase\_stats} {[}{]}:

\item {} 
\sphinxstyleemphasis{kinases\_list} {[}{]}:

\item {} 
\sphinxstyleemphasis{label\_by\_go} {[}{]}:

\item {} 
\sphinxstyleemphasis{laudanna\_directions} {[}{]}:

\item {} 
\sphinxstyleemphasis{laudanna\_effects} {[}{]}:

\item {} 
\sphinxstyleemphasis{licence} {[}{]}:

\item {} 
\sphinxstyleemphasis{list\_resources} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_3dcomplexes} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_3did\_ddi} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_3did\_ddi2} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_3did\_dmi} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_3did\_interfaces} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_all\_pathways} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_compleat} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_complexportal} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_comppi} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_corum} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_dbptm} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_ddi} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_ddis} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_depod\_dmi} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_disgenet} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_dmi} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_dmis} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_domino\_dmi} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_elm} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_expression} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_go} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_havugimana} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_hpa} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_hprd\_ptms} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_ielm} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_interfaces} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_li2012\_ptms} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_ligand\_receptor\_network} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_list} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_lmpid} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_mappings} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_mimp\_dmi} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_mutations} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_negatives} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_old\_omnipath} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_omnipath} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_pathways} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_pdb} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_pepcyber} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_pfam} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_pfam2} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_pfam3} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_phospho\_dmi} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_phosphoelm} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_pisa} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_pnetworks\_dmi} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_psite\_phos} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_psite\_reg} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_ptms} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_ptms2} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_reflist} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_reflists} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_resource} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_resources} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_signor\_ptms} {[}{]}:

\item {} 
\sphinxstyleemphasis{load\_tfregulons} {[}{]}:

\item {} 
\sphinxstyleemphasis{lookup\_cache} {[}{]}:

\item {} 
\sphinxstyleemphasis{loop\_edges} {[}{]}:

\item {} 
\sphinxstyleemphasis{map\_edge} {[}{]}:

\item {} 
\sphinxstyleemphasis{map\_item} {[}{]}:

\item {} 
\sphinxstyleemphasis{map\_list} {[}{]}:

\item {} 
\sphinxstyleemphasis{mean\_reference\_per\_interaction} {[}{]}:

\item {} 
\sphinxstyleemphasis{merge\_lists} {[}{]}:

\item {} 
\sphinxstyleemphasis{merge\_nodes} {[}{]}:

\item {} 
\sphinxstyleemphasis{mimp\_directions} {[}{]}:

\item {} 
\sphinxstyleemphasis{mutated\_edges} {[}{]}:

\item {} 
\sphinxstyleemphasis{names2vids} {[}{]}:

\item {} 
\sphinxstyleemphasis{negative\_report} {[}{]}:

\item {} 
\sphinxstyleemphasis{neighborhood} {[}{]}:

\item {} 
\sphinxstyleemphasis{neighbors} {[}{]}:

\item {} 
\sphinxstyleemphasis{neighbourhood\_network} {[}{]}:

\item {} 
\sphinxstyleemphasis{network\_filter} {[}{]}:

\item {} 
\sphinxstyleemphasis{network\_stats} {[}{]}:

\item {} 
\sphinxstyleemphasis{new\_edges} {[}{]}:

\item {} 
\sphinxstyleemphasis{new\_nodes} {[}{]}:

\item {} 
\sphinxstyleemphasis{node\_exists} {[}{]}:

\item {} 
\sphinxstyleemphasis{numof\_directed\_edges} {[}{]}:

\item {} 
\sphinxstyleemphasis{numof\_reference\_interaction\_pairs} {[}{]}:

\item {} 
\sphinxstyleemphasis{numof\_references} {[}{]}:

\item {} 
\sphinxstyleemphasis{numof\_undirected\_edges} {[}{]}:

\item {} 
\sphinxstyleemphasis{orthology\_translation} {[}{]}:

\item {} 
\sphinxstyleemphasis{p} {[}{]}:

\item {} 
\sphinxstyleemphasis{pathway\_attributes} {[}{]}:

\item {} 
\sphinxstyleemphasis{pathway\_members} {[}{]}:

\item {} 
\sphinxstyleemphasis{pathway\_names} {[}{]}:

\item {} 
\sphinxstyleemphasis{pathway\_similarity} {[}{]}:

\item {} 
\sphinxstyleemphasis{pathways\_table} {[}{]}:

\item {} 
\sphinxstyleemphasis{pfam\_regions} {[}{]}:

\item {} 
\sphinxstyleemphasis{phosphonetworks\_directions} {[}{]}:

\item {} 
\sphinxstyleemphasis{phosphopoint\_directions} {[}{]}:

\item {} 
\sphinxstyleemphasis{phosphorylation\_directions} {[}{]}:

\item {} 
\sphinxstyleemphasis{phosphorylation\_signs} {[}{]}:

\item {} 
\sphinxstyleemphasis{phosphosite\_directions} {[}{]}:

\item {} 
\sphinxstyleemphasis{prdb\_tissue\_expr} {[}{]}:

\item {} 
\sphinxstyleemphasis{process\_direction} {[}{]}:

\item {} 
\sphinxstyleemphasis{process\_directions} {[}{]}:

\item {} 
\sphinxstyleemphasis{process\_dmi} {[}{]}:

\item {} 
\sphinxstyleemphasis{process\_sign} {[}{]}:

\item {} 
\sphinxstyleemphasis{protein} {[}{]}:

\item {} 
\sphinxstyleemphasis{protein\_edge} {[}{]}:

\item {} 
\sphinxstyleemphasis{proteins} {[}{]}:

\item {} 
\sphinxstyleemphasis{proteome\_list} {[}{]}:

\item {} 
\sphinxstyleemphasis{ps} {[}{]}:

\item {} 
\sphinxstyleemphasis{random\_walk\_with\_return} {[}{]}:

\item {} 
\sphinxstyleemphasis{random\_walk\_with\_return2} {[}{]}:

\item {} 
\sphinxstyleemphasis{read\_data\_file} {[}{]}:

\item {} 
\sphinxstyleemphasis{read\_from\_cache} {[}{]}:

\item {} 
\sphinxstyleemphasis{read\_list\_file} {[}{]}:

\item {} 
\sphinxstyleemphasis{receptors\_list} {[}{]}:

\item {} 
\sphinxstyleemphasis{reference\_edge\_ratio} {[}{]}:

\item {} 
\sphinxstyleemphasis{reference\_hist} {[}{]}:

\item {} 
\sphinxstyleemphasis{reload} {[}{]}:

\item {} 
\sphinxstyleemphasis{remove\_htp} {[}{]}:

\item {} 
\sphinxstyleemphasis{remove\_undirected} {[}{]}:

\item {} 
\sphinxstyleemphasis{run\_batch} {[}{]}:

\item {} 
\sphinxstyleemphasis{save\_network} {[}{]}:

\item {} 
\sphinxstyleemphasis{save\_session} {[}{]}:

\item {} 
\sphinxstyleemphasis{search\_attr\_and} {[}{]}:

\item {} 
\sphinxstyleemphasis{search\_attr\_or} {[}{]}:

\item {} 
\sphinxstyleemphasis{second\_neighbours} {[}{]}:

\item {} 
\sphinxstyleemphasis{select\_by\_go} {[}{]}:

\item {} 
\sphinxstyleemphasis{separate} {[}{]}:

\item {} 
\sphinxstyleemphasis{separate\_by\_category} {[}{]}:

\item {} 
\sphinxstyleemphasis{sequences} {[}{]}:

\item {} 
\sphinxstyleemphasis{set\_boolean\_vattr} {[}{]}:

\item {} 
\sphinxstyleemphasis{set\_categories} {[}{]}:

\item {} 
\sphinxstyleemphasis{set\_chembl\_mysql} {[}{]}:

\item {} 
\sphinxstyleemphasis{set\_disease\_genes} {[}{]}:

\item {} 
\sphinxstyleemphasis{set\_druggability} {[}{]}:

\item {} 
\sphinxstyleemphasis{set\_drugtargets} {[}{]}:

\item {} 
\sphinxstyleemphasis{set\_kinases} {[}{]}:

\item {} 
\sphinxstyleemphasis{set\_receptors} {[}{]}:

\item {} 
\sphinxstyleemphasis{set\_signaling\_proteins} {[}{]}:

\item {} 
\sphinxstyleemphasis{set\_tfs} {[}{]}:

\item {} 
\sphinxstyleemphasis{set\_transcription\_factors} {[}{]}:

\item {} 
\sphinxstyleemphasis{shortest\_path\_dist} {[}{]}:

\item {} 
\sphinxstyleemphasis{signaling\_proteins\_list} {[}{]}:

\item {} 
\sphinxstyleemphasis{signor\_pathways} {[}{]}:

\item {} 
\sphinxstyleemphasis{similarity\_groups} {[}{]}:

\item {} 
\sphinxstyleemphasis{small\_plot} {[}{]}:

\item {} 
\sphinxstyleemphasis{sorensen\_pathways} {[}{]}:

\item {} 
\sphinxstyleemphasis{source\_diagram} {[}{]}:

\item {} 
\sphinxstyleemphasis{source\_network} {[}{]}:

\item {} 
\sphinxstyleemphasis{source\_similarity} {[}{]}:

\item {} 
\sphinxstyleemphasis{source\_stats} {[}{]}:

\item {} 
\sphinxstyleemphasis{sources\_hist} {[}{]}:

\item {} 
\sphinxstyleemphasis{sources\_overlap} {[}{]}:

\item {} 
\sphinxstyleemphasis{sources\_venn\_data} {[}{]}:

\item {} 
\sphinxstyleemphasis{straight\_between} {[}{]}:

\item {} 
\sphinxstyleemphasis{string\_effects} {[}{]}:

\item {} 
\sphinxstyleemphasis{sum\_in\_complex} {[}{]}:

\item {} 
\sphinxstyleemphasis{table\_latex} {[}{]}:

\item {} 
\sphinxstyleemphasis{tfs\_list} {[}{]}:

\item {} 
\sphinxstyleemphasis{third\_source\_directions} {[}{]}:

\item {} 
\sphinxstyleemphasis{tissue\_network} {[}{]}:

\item {} 
\sphinxstyleemphasis{transcription\_factors} {[}{]}:

\item {} 
\sphinxstyleemphasis{translate\_refsdir} {[}{]}:

\item {} 
\sphinxstyleemphasis{uniprot} {[}{]}:

\item {} 
\sphinxstyleemphasis{uniprots} {[}{]}:

\item {} 
\sphinxstyleemphasis{uniq\_node\_list} {[}{]}:

\item {} 
\sphinxstyleemphasis{uniq\_ptm} {[}{]}:

\item {} 
\sphinxstyleemphasis{uniq\_ptms} {[}{]}:

\item {} 
\sphinxstyleemphasis{up} {[}{]}:

\item {} 
\sphinxstyleemphasis{up\_affected\_by} {[}{]}:

\item {} 
\sphinxstyleemphasis{up\_affects} {[}{]}:

\item {} 
\sphinxstyleemphasis{up\_edge} {[}{]}:

\item {} 
\sphinxstyleemphasis{up\_in\_directed} {[}{]}:

\item {} 
\sphinxstyleemphasis{up\_in\_undirected} {[}{]}:

\item {} 
\sphinxstyleemphasis{up\_inhibited\_by} {[}{]}:

\item {} 
\sphinxstyleemphasis{up\_inhibits} {[}{]}:

\item {} 
\sphinxstyleemphasis{up\_neighborhood} {[}{]}:

\item {} 
\sphinxstyleemphasis{up\_neighbors} {[}{]}:

\item {} 
\sphinxstyleemphasis{up\_stimulated\_by} {[}{]}:

\item {} 
\sphinxstyleemphasis{up\_stimulates} {[}{]}:

\item {} 
\sphinxstyleemphasis{update\_adjlist} {[}{]}:

\item {} 
\sphinxstyleemphasis{update\_attrs} {[}{]}:

\item {} 
\sphinxstyleemphasis{update\_cats} {[}{]}:

\item {} 
\sphinxstyleemphasis{update\_db\_dict} {[}{]}:

\item {} 
\sphinxstyleemphasis{update\_pathway\_types} {[}{]}:

\item {} 
\sphinxstyleemphasis{update\_pathways} {[}{]}:

\item {} 
\sphinxstyleemphasis{update\_sources} {[}{]}:

\item {} 
\sphinxstyleemphasis{update\_vertex\_sources} {[}{]}:

\item {} 
\sphinxstyleemphasis{update\_vindex} {[}{]}:

\item {} 
\sphinxstyleemphasis{update\_vname} {[}{]}:

\item {} 
\sphinxstyleemphasis{ups} {[}{]}:

\item {} 
\sphinxstyleemphasis{v} {[}{]}:

\item {} 
\sphinxstyleemphasis{vertex\_pathways} {[}{]}:

\item {} 
\sphinxstyleemphasis{vs} {[}{]}:

\item {} 
\sphinxstyleemphasis{vsgs} {[}{]}:

\item {} 
\sphinxstyleemphasis{vsup} {[}{]}:

\item {} 
\sphinxstyleemphasis{wang\_effects} {[}{]}:

\item {} 
\sphinxstyleemphasis{write\_table} {[}{]}:

\end{itemize}

\end{description}

\end{itemize}
\index{add\_update\_vertex() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.add_update_vertex}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{add\_update\_vertex}}}{\emph{defAttrs}, \emph{originalName}, \emph{originalNameType}, \emph{extraAttrs=\{\}}, \emph{add=False}}{}
Updates the attributes of one node in the network.
Optionally it creates a new node and sets the attributes,
but it is not efficient as igraph needs to reindex vertices
after this operation, so better to create new nodes and
edges in batch.

\end{fulllineitems}

\index{all\_between() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.all_between}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{all\_between}}}{\emph{nameA}, \emph{nameB}}{}
Returns all edges between two given vertex names. Similar to
straight\_between(), but checks both directions, and returns
list of edge ids in {[}undirected, straight, reversed{]} format,
for both nameA -\textgreater{} nameB and nameB -\textgreater{} nameA edges.

\end{fulllineitems}

\index{apply\_list() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.apply_list}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{apply\_list}}}{\emph{name}, \emph{node\_or\_edge='node'}}{}
Creates vertex or edge attribute based on a list.

\end{fulllineitems}

\index{attach\_network() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.attach_network}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{attach\_network}}}{\emph{edgeList=False}, \emph{regulator=False}}{}
Adds edges to the network from edgeList obtained from file or
other input method.

\end{fulllineitems}

\index{basic\_stats() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.basic_stats}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{basic\_stats}}}{\emph{latex=False, caption='', latex\_hdr=True, fontsize=8, font='HelveticaNeueLTStd-LtCn', fname=None, header\_format='\%s', row\_order=None, by\_category=True, use\_cats={[}'p', 'm', 'i', 'r'{]}, urls=True, annots=False}}{}
Returns basic numbers about the network resources, e.g. edge and
node counts.
\begin{description}
\item[{latex}] \leavevmode
Return table in a LaTeX document. This can be compiled by
PDFLaTeX:
latex stats.tex

\end{description}

\end{fulllineitems}

\index{cancer\_gene\_census\_list() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.cancer_gene_census_list}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cancer\_gene\_census\_list}}}{}{}
Loads the list of cancer driver proteins from the COSMIC Cancer
Gene Census.

\end{fulllineitems}

\index{clean\_graph() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.clean_graph}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{clean\_graph}}}{}{}
Removes multiple edges, unknown molecules and those from wrong taxon.
Multiple edges will be combined by \sphinxtitleref{combine\_attr()} method.
Loops will be deleted unless the \sphinxtitleref{loops} attribute set to \sphinxtitleref{True}.

\end{fulllineitems}

\index{collapse\_by\_name() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.collapse_by_name}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{collapse\_by\_name}}}{\emph{graph=None}}{}
Collapses nodes with the same name with copying and merging
all edges and attributes.

\end{fulllineitems}

\index{combine\_attr() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.combine_attr}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{combine\_attr}}}{\emph{lst}, \emph{num\_method=\textless{}built-in function max\textgreater{}}}{}
Combines multiple attributes into one. This method attempts
to find out which is the best way to combine attributes.
\begin{itemize}
\item {} 
if there is only one value or one of them is None, then returns
the one available

\item {} 
lists: concatenates unique values of lists

\item {} 
numbers: returns the greater by default
or calls \sphinxtitleref{num\_method()} if given.

\item {} 
sets: returns the union

\item {} 
dicts: calls \sphinxtitleref{common.merge\_dicts()}

\item {} 
Direction: calls their special \sphinxtitleref{merge()} method

\end{itemize}

Works on more than 2 attributes recursively.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{lst}} (\sphinxstyleliteralemphasis{\sphinxupquote{list}}) \textendash{} List of one or two attribute values.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{num\_method}} (\sphinxstyleliteralemphasis{\sphinxupquote{callable}}) \textendash{} Method to merge numeric attributes.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{copy\_edges() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.copy_edges}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{copy\_edges}}}{\emph{sources}, \emph{target}, \emph{move=False}, \emph{graph=None}}{}
Copies edges of one node to another,
keeping attributes and directions.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{sources}} (\sphinxstyleliteralemphasis{\sphinxupquote{list}}) \textendash{} Vertex IDs to copy from.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{target}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} Vertex ID to copy for.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{move}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Whether perform copy or move, i.e. remove or keep
the source edges.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{count\_sol() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.count_sol}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{count\_sol}}}{}{}
Counts nodes with zero degree.

\end{fulllineitems}

\index{curation\_effort() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.curation_effort}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{curation\_effort}}}{\emph{sum\_by\_source=False}}{}
Returns the total number of reference-interactions pairs.
\begin{description}
\item[{@sum\_by\_source}] \leavevmode{[}bool{]}
If True, counts the refrence-interaction pairs by
sources, and returns the sum of these values.

\end{description}

\end{fulllineitems}

\index{delete\_by\_taxon() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.delete_by_taxon}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{delete\_by\_taxon}}}{\emph{tax}}{}
Removes the proteins of all organisms which are not listed.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{tax}} (\sphinxstyleliteralemphasis{\sphinxupquote{list}}) \textendash{} List of NCBI Taxonomy IDs of the organisms.
E.g. {[}7227, 9606{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{delete\_unknown() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.delete_unknown}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{delete\_unknown}}}{\emph{tax}, \emph{typ='protein'}, \emph{defaultNameType=None}}{}
Removes those proteins which are not in the list of all default
IDs of the organisms. By default, it means to remove all protein
nodes not having a human SwissProt ID.
\begin{description}
\item[{@tax}] \leavevmode{[}list{]}
List of NCBI Taxonomy IDs of the organisms of interest.
E.g. {[}7227, 9606{]}

\item[{@typ}] \leavevmode{[}str{]}
Molecule type. E.g. ‘protein’ or ‘mirna’

\item[{@defaultNameType}] \leavevmode{[}str{]}
The default name type of the given molecular species.
For proteins it’s ‘uniprot’ by default.

\end{description}

\end{fulllineitems}

\index{dgenesymbol() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.dgenesymbol}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{dgenesymbol}}}{\emph{genesymbol}}{}
Returns \sphinxcode{\sphinxupquote{igraph.Vertex()}} object if the GeneSymbol
can be found in the default directed network,
otherwise \sphinxcode{\sphinxupquote{None}}.
\begin{description}
\item[{@genesymbol}] \leavevmode{[}str{]}
GeneSymbol.

\end{description}

\end{fulllineitems}

\index{dgs() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.dgs}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{dgs}}}{\emph{genesymbol}}{}
Returns \sphinxcode{\sphinxupquote{igraph.Vertex()}} object if the GeneSymbol
can be found in the default directed network,
otherwise \sphinxcode{\sphinxupquote{None}}.
\begin{description}
\item[{@genesymbol}] \leavevmode{[}str{]}
GeneSymbol.

\end{description}

\end{fulllineitems}

\index{disease\_genes\_list() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.disease_genes_list}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{disease\_genes\_list}}}{\emph{dataset='curated'}}{}
Loads the list of all disease related genes from DisGeNet.
This resource is human only.

\end{fulllineitems}

\index{dp() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.dp}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{dp}}}{\emph{identifier}}{}
Same as \sphinxcode{\sphinxupquote{PyPath.get\_node}}, just for the directed graph.
Returns \sphinxcode{\sphinxupquote{igraph.Vertex()}} object if the identifier
is a valid vertex index in the default directed graph,
or a UniProt ID or GeneSymbol which can be found in the
default directed network, otherwise \sphinxcode{\sphinxupquote{None}}.
\begin{description}
\item[{@identifier}] \leavevmode{[}int, str{]}
Vertex index (int) or GeneSymbol (str) or UniProt ID (str).

\end{description}

\end{fulllineitems}

\index{druggability\_list() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.druggability_list}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{druggability\_list}}}{}{}
Loads the list of druggable proteins from DgiDB.
This resource is human only.

\end{fulllineitems}

\index{duniprot() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.duniprot}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{duniprot}}}{\emph{uniprot}}{}
Same as \sphinxcode{\sphinxupquote{PyPath.uniprot(), just for directed graph.
Returns {}`{}`igraph.Vertex()}} object if the UniProt
can be found in the default directed network,
otherwise \sphinxcode{\sphinxupquote{None}}.
\begin{description}
\item[{@uniprot}] \leavevmode{[}str{]}
UniProt ID.

\end{description}

\end{fulllineitems}

\index{duniprots() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.duniprots}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{duniprots}}}{\emph{uniprots}}{}
Returns list of \sphinxcode{\sphinxupquote{igraph.Vertex()}} object
for a list of UniProt IDs omitting those
could not be found in the default
directed graph.

\end{fulllineitems}

\index{dup() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.dup}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{dup}}}{\emph{uniprot}}{}
Same as \sphinxcode{\sphinxupquote{PyPath.uniprot(), just for directed graph.
Returns {}`{}`igraph.Vertex()}} object if the UniProt
can be found in the default directed network,
otherwise \sphinxcode{\sphinxupquote{None}}.
\begin{description}
\item[{@uniprot}] \leavevmode{[}str{]}
UniProt ID.

\end{description}

\end{fulllineitems}

\index{dups() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.dups}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{dups}}}{\emph{uniprots}}{}
Returns list of \sphinxcode{\sphinxupquote{igraph.Vertex()}} object
for a list of UniProt IDs omitting those
could not be found in the default
directed graph.

\end{fulllineitems}

\index{dv() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.dv}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{dv}}}{\emph{identifier}}{}
Same as \sphinxcode{\sphinxupquote{PyPath.get\_node}}, just for the directed graph.
Returns \sphinxcode{\sphinxupquote{igraph.Vertex()}} object if the identifier
is a valid vertex index in the default directed graph,
or a UniProt ID or GeneSymbol which can be found in the
default directed network, otherwise \sphinxcode{\sphinxupquote{None}}.
\begin{description}
\item[{@identifier}] \leavevmode{[}int, str{]}
Vertex index (int) or GeneSymbol (str) or UniProt ID (str).

\end{description}

\end{fulllineitems}

\index{edge\_exists() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.edge_exists}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{edge\_exists}}}{\emph{nameA}, \emph{nameB}}{}
Returns a tuple of vertice indices if edge doesn’t exists,
otherwise edge id. Not sensitive to direction.

\end{fulllineitems}

\index{edges\_expression() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.edges_expression}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{edges\_expression}}}{\emph{func=\textless{}function \textless{}lambda\textgreater{}\textgreater{}}}{}
Executes function \sphinxtitleref{func} for each pairs of connected proteins in the
network, for every expression dataset. By default, \sphinxtitleref{func} simply
gives the product the (normalized) expression values.
\begin{description}
\item[{func}] \leavevmode{[}callable{]}
Function to handle 2 vectors (pandas.Series() objects), should
return one vector of the same length.

\end{description}

\end{fulllineitems}

\index{edges\_in\_comlexes() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.edges_in_comlexes}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{edges\_in\_comlexes}}}{\emph{csources={[}'corum'{]}, graph=None}}{}
Creates edge attributes \sphinxcode{\sphinxupquote{complexes}} and \sphinxcode{\sphinxupquote{in\_complex}}.
These are both dicts where the keys are complex resources.
The values in \sphinxcode{\sphinxupquote{complexes}} are the list of complex names
both the source and the target vertices belong to.
The values \sphinxcode{\sphinxupquote{in\_complex}} are boolean values whether there
is at least one complex in the given resources both the
source and the target vertex of the edge belong to.
\begin{description}
\item[{@csources}] \leavevmode{[}list{]}
List of complex resources. Should be already loaded.

\item[{@graph}] \leavevmode{[}igraph.Graph(){]}
The graph object to do the calculations on.

\end{description}

\end{fulllineitems}

\index{export\_dot() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.export_dot}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{export\_dot}}}{\emph{nodes=None}, \emph{edges=None}, \emph{directed=True}, \emph{labels='genesymbol'}, \emph{edges\_filter=\textless{}function \textless{}lambda\textgreater{}\textgreater{}}, \emph{nodes\_filter=\textless{}function \textless{}lambda\textgreater{}\textgreater{}}, \emph{edge\_sources=None}, \emph{dir\_sources=None}, \emph{graph=None}, \emph{return\_object=False}, \emph{save\_dot=None}, \emph{save\_graphics=None}, \emph{prog='neato'}, \emph{format=None}, \emph{hide=False}, \emph{font=None}, \emph{auto\_edges=False}, \emph{hide\_nodes={[}{]}}, \emph{defaults=\{\}}, \emph{**kwargs}}{}
Builds a pygraphviz.AGraph() object with filtering the edges
and vertices along arbitrary criteria.
Returns the Agraph object if requesred, or exports the dot
file, or saves the graphics.

@nodes : list
List of vertex ids to be included.
@edges : list
List of edge ids to be included.
@directed : bool
Create a directed or undirected graph.
@labels : str
Name type to be used as id/label in the dot format.
@edges\_filter : function
Function to filter edges, accepting igraph.Edge as argument.
@nodes\_filter : function
Function to filter vertices, accepting igraph.Vertex as argument.
@edge\_sources : list
Sources to be included.
@dir\_sources : list
Direction and effect sources to be included.
@graph : igraph.Graph
The graph object to export.
@return\_object : bool
Whether to return the pygraphviz.AGraph object.
@save\_dot : str
Filename to export the dot file to.
@save\_graphics : str
Filename to export the graphics, the extension defines the format.
@prog : str
The graphviz layout algorithm to use.
@format : str
The graphics format passed to pygraphviz.AGrapg().draw().
@hide : bool
Hide filtered edges instead of omit them.
@hide nodes : list
Nodes to hide. List of vertex ids.
@auto\_edges : str
Automatic, built-in style for edges.
‘DIRECTIONS’ or ‘RESOURCE\_CATEGORIES’ are supported.
@font : str
Font to use for labels.
For using more than one fonts refer to graphviz attributes with constant values
or define callbacks or mapping dictionaries.
@defaults : dict
Default values for graphviz attributes, labeled with the entity, e.g.
\sphinxtitleref{\{‘edge\_penwidth’: 0.2\}}.
@**kwargs : constant, callable or dict
Graphviz attributes, labeled by the target entity. E.g. \sphinxtitleref{edge\_penwidth},
‘vertex\_shape{}` or \sphinxtitleref{graph\_label}.
If the value is constant, this value will be used.
If the value is dict, and has \sphinxtitleref{\_name} as key, for every instance of the
given entity, the value of the attribute defined by \sphinxtitleref{\_name} will be looked
up in the dict, and the corresponding value will be given to the graphviz
attribute. If the key \sphinxtitleref{\_name} is missing from the dict, igraph vertex and
edge indices will be looked up among the keys.
If the value is callable, it will be called with the current instance of
the entity and the returned value will be used for the graphviz attribute.
E.g. \sphinxtitleref{edge\_arrowhead(edge)} or \sphinxtitleref{vertex\_fillcolor(vertex)}
Example:
\begin{quote}

import pypath
from pypath import data\_formats
net = pypath.PyPath()
net.init\_network(pfile = ‘cache/default.pickle’)
\#net.init\_network(\{‘arn’: data\_formats.omnipath{[}‘arn’{]}\})
tgf = {[}v.index for v in net.graph.vs if ‘TGF’ in v{[}‘slk\_pathways’{]}{]}
dot = net.export\_dot(nodes = tgf, save\_graphics = ‘tgf\_slk.pdf’, prog = ‘dot’,
\begin{quote}

main\_title = ‘TGF-beta pathway’, return\_object = True,
label\_font = ‘HelveticaNeueLTStd Med Cn’,
edge\_sources = {[}‘SignaLink3’{]},
dir\_sources = {[}‘SignaLink3’{]}, hide = True)
\end{quote}
\end{quote}

\end{fulllineitems}

\index{export\_edgelist() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.export_edgelist}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{export\_edgelist}}}{\emph{fname, graph=None, names={[}'name'{]}, edge\_attributes={[}{]}, sep='\textbackslash{}t'}}{}
Write edge list to text file with attributes

@param fname: the name of the file or a stream to read from.
@param graph: the igraph object containing the network
@param names: list with the vertex attribute names to be printed
\begin{quote}

for source and target vertices
\end{quote}
\begin{description}
\item[{@param edge\_attributes: list with the edge attribute names}] \leavevmode
to be printed

\end{description}

@param sep: string used to separate columns

\end{fulllineitems}

\index{export\_tab() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.export_tab}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{export\_tab}}}{\emph{outfile=None}, \emph{extra\_node\_attrs=\{\}}, \emph{extra\_edge\_attrs=\{\}}, \emph{unique\_pairs=True}, \emph{**kwargs}}{}
Exports the network in a tabular format.

By default UniProt IDs, Gene Symbols, source databases, literature
references, directionality and sign information and interaction type
are included.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{outfile}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Name of the output file. If \sphinxtitleref{None} a file name
“netrowk-\textless{}session id\textgreater{}.tab” is used.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{extra\_node\_attrs}} (\sphinxstyleliteralemphasis{\sphinxupquote{dict}}) \textendash{} Additional node attributes to be included in the exported table.
Keys are column ames used in the header while values are names
of vertex attributes. In the header \sphinxtitleref{\_A} and \sphinxtitleref{\_B} suffixes will
be appended to the column names so the values can be assigned to
A and B side interaction partners.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{extra\_edge\_attrs}} (\sphinxstyleliteralemphasis{\sphinxupquote{dict}}) \textendash{} Additional edge attributes to be included in the exported table.
Keys are column ames used in the header while values are names
of edge attributes.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{find\_all\_paths() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.find_all_paths}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{find\_all\_paths}}}{\emph{start}, \emph{end}, \emph{mode='OUT'}, \emph{maxlen=2}, \emph{graph=None}, \emph{silent=False}}{}
Finds all paths up to length \sphinxtitleref{maxlen} between groups of
vertices. This function is needed only becaues igraph{}`s
get\_all\_shortest\_paths() finds only the shortest, not any
path up to a defined length.
\begin{description}
\item[{@start}] \leavevmode{[}int or list{]}
Indices of the starting node(s) of the paths.

\item[{@end}] \leavevmode{[}int or list{]}
Indices of the target node(s) of the paths.

\item[{@mode}] \leavevmode{[}‘IN’, ‘OUT’, ‘ALL’{]}
Passed to igraph.Graph.neighbors()

\item[{@maxlen}] \leavevmode{[}int{]}
Maximum length of paths in steps, i.e. if maxlen = 3, then
the longest path may consist of 3 edges and 4 nodes.

\item[{@graph}] \leavevmode{[}igraph.Graph object{]}
The graph you want to find paths in. self.graph by default.

\end{description}

\end{fulllineitems}

\index{find\_complex() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.find_complex}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{find\_complex}}}{\emph{search}}{}
Finds complexes by their non standard names.
E.g. to find DNA polymerases you can use the search
term \sphinxtitleref{DNA pol} which will be tested against complex names
in CORUM.

\end{fulllineitems}

\index{genesymbol() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.genesymbol}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{genesymbol}}}{\emph{genesymbol}}{}
Returns \sphinxcode{\sphinxupquote{igraph.Vertex()}} object if the GeneSymbol
can be found in the default undirected network,
otherwise \sphinxcode{\sphinxupquote{None}}.
\begin{description}
\item[{@genesymbol}] \leavevmode{[}str{]}
GeneSymbol.

\end{description}

\end{fulllineitems}

\index{genesymbol\_labels() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.genesymbol_labels}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{genesymbol\_labels}}}{\emph{graph=None}, \emph{remap\_all=False}}{}
Creats vertex attribute \sphinxcode{\sphinxupquote{label}} and fills up with Gene Symbols
of all proteins where the Gene Symbol can be looked up based on
the default name of the protein vertex.
If the attribute \sphinxcode{\sphinxupquote{label}} had been already initialized,
updates this attribute or recreates if \sphinxcode{\sphinxupquote{remap\_all}}
is \sphinxcode{\sphinxupquote{True}}.

\end{fulllineitems}

\index{get\_directed() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.get_directed}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_directed}}}{\emph{graph=False}, \emph{conv\_edges=False}, \emph{mutual=False}, \emph{ret=False}}{}
Converts \sphinxcode{\sphinxupquote{graph}} undirected \sphinxcode{\sphinxupquote{igraph.Graph}} object to a directed one.
By default it converts the graph in \sphinxcode{\sphinxupquote{PyPath.graph}} and places the directed
instance in \sphinxcode{\sphinxupquote{PyPath.dgraph}}.
\begin{description}
\item[{@graph}] \leavevmode{[}igraph.Graph{]}
Undirected graph object.

\item[{@conv\_edges}] \leavevmode{[}bool{]}
Whether to convert undirected edges (those without explicit
direction information) to an arbitrary direction edge or
a pair of opposite edges.
Otherwise those will be deleted. Default is \sphinxcode{\sphinxupquote{False}}.

\item[{@mutual}] \leavevmode{[}bool{]}
If \sphinxcode{\sphinxupquote{conv\_edges}} is \sphinxcode{\sphinxupquote{True}}, whether to convert the
undirected edges to a single, arbitrary directed edge,
or a pair of opposite directed edges. Default is \sphinxcode{\sphinxupquote{False}}.

\item[{@ret}] \leavevmode{[}bool{]}
Return the directed graph instance, or return \sphinxcode{\sphinxupquote{None}}.
Default is \sphinxcode{\sphinxupquote{False}} (returns \sphinxcode{\sphinxupquote{None}}).

\end{description}

\end{fulllineitems}

\index{get\_edge() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.get_edge}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_edge}}}{\emph{source}, \emph{target}, \emph{directed=True}}{}
Returns \sphinxcode{\sphinxupquote{igraph.Edge}} object if an edge exist between
the 2 proteins, otherwise \sphinxcode{\sphinxupquote{None}}.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{source}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{,}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Vertex index or UniProt ID or GeneSymbol

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{target}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{,}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Vertex index or UniProt ID or GeneSymbol

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{directed}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} To be passed to igraph.Graph.get\_eid()

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_edges() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.get_edges}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_edges}}}{\emph{sources}, \emph{targets}, \emph{directed=True}}{}
Returns a generator with all edges between source and target vertices.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{sources}} (\sphinxstyleliteralemphasis{\sphinxupquote{iterable}}) \textendash{} Source vertex IDs, names or labels.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{targets}} (\sphinxstyleliteralemphasis{\sphinxupquote{iterable}}) \textendash{} Target vertec IDs, names or labels.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{directed}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Passed to \sphinxtitleref{igraph.get\_eid()}.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_giant() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.get_giant}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_giant}}}{\emph{replace=False}, \emph{graph=None}}{}
Returns the giant component of the graph, or
replaces the igraph object with only the giant
component.

\end{fulllineitems}

\index{get\_node() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.get_node}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_node}}}{\emph{identifier}}{}
Returns \sphinxcode{\sphinxupquote{igraph.Vertex()}} object if the identifier
is a valid vertex index in the default undirected graph,
or a UniProt ID or GeneSymbol which can be found in the
default undirected network, otherwise \sphinxcode{\sphinxupquote{None}}.
\begin{description}
\item[{@identifier}] \leavevmode{[}int, str{]}
Vertex index (int) or GeneSymbol (str) or UniProt ID (str).

\end{description}

\end{fulllineitems}

\index{get\_node\_d() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.get_node_d}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_node\_d}}}{\emph{identifier}}{}
Same as \sphinxcode{\sphinxupquote{PyPath.get\_node}}, just for the directed graph.
Returns \sphinxcode{\sphinxupquote{igraph.Vertex()}} object if the identifier
is a valid vertex index in the default directed graph,
or a UniProt ID or GeneSymbol which can be found in the
default directed network, otherwise \sphinxcode{\sphinxupquote{None}}.
\begin{description}
\item[{@identifier}] \leavevmode{[}int, str{]}
Vertex index (int) or GeneSymbol (str) or UniProt ID (str).

\end{description}

\end{fulllineitems}

\index{go\_annotate() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.go_annotate}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{go\_annotate}}}{\emph{aspects=('C'}, \emph{'F'}, \emph{'P')}}{}
Annotates protein nodes with GO terms. In the \sphinxcode{\sphinxupquote{go}} vertex
attribute each node is annotated by a dict of sets where keys are
one letter codes of GO aspects and values are sets of GO accessions.

\end{fulllineitems}

\index{gs() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.gs}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{gs}}}{\emph{genesymbol}}{}
Returns \sphinxcode{\sphinxupquote{igraph.Vertex()}} object if the GeneSymbol
can be found in the default undirected network,
otherwise \sphinxcode{\sphinxupquote{None}}.
\begin{description}
\item[{@genesymbol}] \leavevmode{[}str{]}
GeneSymbol.

\end{description}

\end{fulllineitems}

\index{gs\_edge() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.gs_edge}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{gs\_edge}}}{\emph{source}, \emph{target}, \emph{directed=True}}{}
Returns \sphinxcode{\sphinxupquote{igraph.Edge}} object if an edge exist between
the 2 proteins, otherwise \sphinxcode{\sphinxupquote{None}}.
\begin{description}
\item[{@source}] \leavevmode{[}str{]}
GeneSymbol

\item[{@target}] \leavevmode{[}str{]}
GeneSymbol

\item[{@directed}] \leavevmode{[}bool{]}
To be passed to igraph.Graph.get\_eid()

\end{description}

\end{fulllineitems}

\index{homology\_translation() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.homology_translation}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{homology\_translation}}}{\emph{target}, \emph{source=None}, \emph{only\_swissprot=True}, \emph{graph=None}}{}
Translates the current object to another organism by orthology.
Proteins without known ortholog will be deleted.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{target}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} NCBI Taxonomy ID of the target organism.
E.g. 10090 for mouse.

\end{description}\end{quote}

\end{fulllineitems}

\index{init\_edge\_attr() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.init_edge_attr}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{init\_edge\_attr}}}{\emph{attr}}{}
Fills edge attribute with its default values, creates
lists if in \sphinxtitleref{edgeAttrs} the attribute is registered as list.

\end{fulllineitems}

\index{init\_network() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.init_network}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{init\_network}}}{\emph{lst=\{'arn': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517d88\textgreater{}}, \emph{'biogrid': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528cef0\textgreater{}}, \emph{'ca1': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517c68\textgreater{}}, \emph{'ccmap': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c248\textgreater{}}, \emph{'cellphonedb': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc52a7050\textgreater{}}, \emph{'dbptm': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c680\textgreater{}}, \emph{'death': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517f38\textgreater{}}, \emph{'depod': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c4d0\textgreater{}}, \emph{'dip': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c2d8\textgreater{}}, \emph{'domino': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c5f0\textgreater{}}, \emph{'elm': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c5a8\textgreater{}}, \emph{'guide2pharma': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517b90\textgreater{}}, \emph{'hpmr': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc52a2f80\textgreater{}}, \emph{'hprd': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc529b098\textgreater{}}, \emph{'hprd\_p': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c6c8\textgreater{}}, \emph{'innatedb': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c368\textgreater{}}, \emph{'intact': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528cea8\textgreater{}}, \emph{'lmpid': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c518\textgreater{}}, \emph{'macrophage': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517ef0\textgreater{}}, \emph{'matrixdb': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c440\textgreater{}}, \emph{'mppi': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c290\textgreater{}}, \emph{'nrf2': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517e60\textgreater{}}, \emph{'pdz': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517f80\textgreater{}}, \emph{'phelm': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c560\textgreater{}}, \emph{'psite': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c488\textgreater{}}, \emph{'ramilowski2015': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc52a70e0\textgreater{}}, \emph{'signalink3': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517b48\textgreater{}}, \emph{'signor': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c1b8\textgreater{}}, \emph{'spike': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517b00\textgreater{}}, \emph{'trip': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517a28\textgreater{}\}}, \emph{exclude={[}{]}}, \emph{cache\_files=\{\}}, \emph{pfile=False}, \emph{save=False}, \emph{reread=False}, \emph{redownload=False}, \emph{**kwargs}}{}
This is a lazy way to start the module, load data
and build the high confidence, literature curated
part of the signaling network.

\end{fulllineitems}

\index{init\_vertex\_attr() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.init_vertex_attr}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{init\_vertex\_attr}}}{\emph{attr}}{}
Fills vertex attribute with its default values, creates
lists if in \sphinxtitleref{vertexAttrs} the attribute is registered as list.

\end{fulllineitems}

\index{intogen\_cancer\_drivers\_list() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.intogen_cancer_drivers_list}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{intogen\_cancer\_drivers\_list}}}{\emph{intogen\_file}}{}
Loads the list of cancer driver proteins from IntOGen data.

\end{fulllineitems}

\index{kinases\_list() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.kinases_list}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{kinases\_list}}}{}{}
Loads the list of all known kinases in the proteome from kinase.com.
This resource is human only.

\end{fulllineitems}

\index{label\_by\_go() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.label_by_go}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{label\_by\_go}}}{\emph{label}, \emph{go\_terms}, \emph{**kwargs}}{}
Assigns a boolean vertex attribute to nodes which tells whether
the node is annotated by all or any (see \sphinxcode{\sphinxupquote{method}} parameter of
\sphinxcode{\sphinxupquote{select\_by\_go}}) the GO terms.

\end{fulllineitems}

\index{load\_compleat() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.load_compleat}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{load\_compleat}}}{\emph{graph=None}}{}
Loads complexes from Compleat. Loads data into vertex attribute
\sphinxtitleref{graph.vs{[}‘complexes’{]}{[}‘compleat’{]}}.
This resource is human only.

\end{fulllineitems}

\index{load\_complexportal() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.load_complexportal}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{load\_complexportal}}}{\emph{graph=None}}{}
Loads complexes from ComplexPortal. Loads data into vertex attribute
\sphinxtitleref{graph.vs{[}‘complexes’{]}{[}‘complexportal’{]}}.
This resource is human only.

\end{fulllineitems}

\index{load\_corum() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.load_corum}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{load\_corum}}}{\emph{graph=None}}{}
Loads complexes from CORUM database. Loads data into vertex attribute
\sphinxtitleref{graph.vs{[}‘complexes’{]}{[}‘corum’{]}}.
This resource is human only.

\end{fulllineitems}

\index{load\_ddi() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.load_ddi}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{load\_ddi}}}{\emph{ddi}}{}
ddi is either a list of intera.DomainDomain objects,
or a function resulting this list

\end{fulllineitems}

\index{load\_disgenet() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.load_disgenet}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{load\_disgenet}}}{\emph{dataset='curated'}, \emph{score=0.0}, \emph{umls=False}, \emph{full\_data=False}}{}
Assigns DisGeNet disease-gene associations to the proteins
in the network. Disease annotations will be added to the \sphinxtitleref{dis}
vertex attribute.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{score}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}) \textendash{} Confidence score from DisGeNet. Only associations
above the score provided will be considered.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ulms}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} By default we assign a list of disease names to
each protein. To use Unified Medical Language System IDs instead
set this to \sphinxtitleref{True}.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{full\_data}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} By default we load only disease names. Set this
to \sphinxtitleref{True} if you wish to load additional annotations like number
of PubMed IDs, number of SNPs and original sources.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{load\_dmi() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.load_dmi}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{load\_dmi}}}{\emph{dmi}}{}
dmi is either a list of intera.DomainMotif objects,
or a function resulting this list

\end{fulllineitems}

\index{load\_expression() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.load_expression}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{load\_expression}}}{\emph{array=False}}{}
Expression data can be loaded into vertex attributes,
or into a pandas DataFrame \textendash{} the latter offers faster
ways to process and use these huge matrices.

\end{fulllineitems}

\index{load\_go() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.load_go}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{load\_go}}}{\emph{aspects=('C'}, \emph{'F'}, \emph{'P')}}{}
Annotates protein nodes with GO terms. In the \sphinxcode{\sphinxupquote{go}} vertex
attribute each node is annotated by a dict of sets where keys are
one letter codes of GO aspects and values are sets of GO accessions.

\end{fulllineitems}

\index{load\_havugimana() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.load_havugimana}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{load\_havugimana}}}{\emph{graph=None}}{}
Loads complexes from Havugimana 2012. Loads data into vertex attribute
\sphinxtitleref{graph.vs{[}‘complexes’{]}{[}‘havugimana’{]}}.
This resource is human only.

\end{fulllineitems}

\index{load\_hpa() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.load_hpa}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{load\_hpa}}}{\emph{normal=True}, \emph{pathology=True}, \emph{cancer=True}, \emph{summarize\_pathology=True}, \emph{tissues=None}, \emph{quality=set({[}'Supported'}, \emph{'Approved'{]})}, \emph{levels=\{'High': 3}, \emph{'Low': 1}, \emph{'Medium': 2}, \emph{'Not detected': 0\}}, \emph{graph=None}, \emph{na\_value=0}}{}
Loads Human Protein Atlas data into vertex attributes.

\end{fulllineitems}

\index{load\_ligand\_receptor\_network() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.load_ligand_receptor_network}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{load\_ligand\_receptor\_network}}}{\emph{lig\_rec\_resources=True}, \emph{inference\_from\_go=True}, \emph{sources=\{'arn': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517d88\textgreater{}}, \emph{'ca1': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517c68\textgreater{}}, \emph{'cellphonedb': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc52a7050\textgreater{}}, \emph{'death': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517f38\textgreater{}}, \emph{'guide2pharma': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517b90\textgreater{}}, \emph{'hpmr': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc52a2f80\textgreater{}}, \emph{'macrophage': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517ef0\textgreater{}}, \emph{'nrf2': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517e60\textgreater{}}, \emph{'pdz': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517f80\textgreater{}}, \emph{'ramilowski2015': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc52a70e0\textgreater{}}, \emph{'signalink3': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517b48\textgreater{}}, \emph{'signor': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c1b8\textgreater{}}, \emph{'spike': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517b00\textgreater{}}, \emph{'trip': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517a28\textgreater{}\}}, \emph{keep\_undirected=False}, \emph{keep\_rec\_rec=False}, \emph{keep\_lig\_lig=False}}{}
Initializes a ligand-receptor network.

\end{fulllineitems}

\index{load\_mutations() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.load_mutations}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{load\_mutations}}}{\emph{attributes=None}, \emph{gdsc\_datadir=None}, \emph{mutation\_file=None}}{}
Mutations are listed in vertex attributes. Mutation() objects
offers methods to identify residues and look up in Ptm(), Motif()
and Domain() objects, to check if those residues are
modified, or are in some short motif or domain.

\end{fulllineitems}

\index{load\_old\_omnipath() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.load_old_omnipath}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{load\_old\_omnipath}}}{\emph{kinase\_substrate\_extra=False}, \emph{remove\_htp=False}, \emph{htp\_threshold=1}, \emph{keep\_directed=False}, \emph{min\_refs\_undirected=2}}{}
Loads the OmniPath network as it was before August 2016.
Furthermore it gives some more options.

\end{fulllineitems}

\index{load\_omnipath() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.load_omnipath}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{load\_omnipath}}}{\emph{kinase\_substrate\_extra=False}, \emph{remove\_htp=True}, \emph{htp\_threshold=1}, \emph{keep\_directed=True}, \emph{min\_refs\_undirected=2}, \emph{old\_omnipath\_resources=False}}{}
Loads the OmniPath network.

\end{fulllineitems}

\index{load\_pathways() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.load_pathways}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{load\_pathways}}}{\emph{source}, \emph{graph=None}}{}
Generic method to load pathway annotations from a resource.
We don’t recommend calling this method but either specific
methods for a single source e.g. \sphinxtitleref{kegg\_pathways()}
or \sphinxtitleref{sirnor\_pathways()} or call \sphinxtitleref{load\_all\_pathways()} to
load all resources.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{source}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Name of the source, this need to match a method in the dict
in \sphinxtitleref{get\_pathways()} method and the edge and vertex attributes
with pathway annotations will be called “\textless{}source\textgreater{}\_pathways”.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{graph}} (\sphinxstyleliteralemphasis{\sphinxupquote{igraph.Graph}}) \textendash{} A graph, by default the default the \sphinxtitleref{graph} attribute of the
current instance.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{load\_ptms2() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.load_ptms2}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{load\_ptms2}}}{\emph{input\_methods=None, map\_by\_homology\_from={[}9606{]}, homology\_only\_swissprot=True, ptm\_homology\_strict=False, nonhuman\_direct\_lookup=True, inputargs=\{\}}}{}
This is a new method which will replace \sphinxtitleref{load\_ptms}.
It uses \sphinxtitleref{pypath.ptm.PtmAggregator}, a newly introduced
module for combining enzyme-substrate data from multiple
resources using homology translation on users demand.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{input\_methods}} (\sphinxstyleliteralemphasis{\sphinxupquote{list}}) \textendash{} Resources to collect enzyme-substrate
interactions from. E.g. \sphinxtitleref{{[}‘Signor’, ‘phosphoELM’{]}}. By default
it contains Signor, PhosphoSitePlus, HPRD, phosphoELM, dbPTM,
PhosphoNetworks, Li2012 and MIMP.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{map\_by\_homology\_from}} (\sphinxstyleliteralemphasis{\sphinxupquote{list}}) \textendash{} List of NCBI Taxonomy IDs of
source taxons used for homology translation of enzyme-substrate
interactions. If you have a human network and you add here
\sphinxtitleref{{[}10090, 10116{]}} then mouse and rat interactions from the source
databases will be translated to human.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{homology\_only\_swissprot}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} \sphinxtitleref{True} by default which means
only SwissProt IDs are accepted at homology translateion, Trembl
IDs will be dropped.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ptm\_homology\_strict}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} For homology translation use
PhosphoSite’s PTM homology table. This guarantees that only
truely homologous sites will be included. Otherwise we only
check if at the same numeric offset in the homologous sequence
the appropriate residue can be find.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{nonhuman\_direct\_lookup}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Fetch also directly nonhuman
data from the resources whereever it’s available. PhosphoSite
contains mouse enzyme-substrate interactions and it is possible
to extract these directly beside translating the human ones
to mouse.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{inputargs}} (\sphinxstyleliteralemphasis{\sphinxupquote{dict}}) \textendash{} Additional arguments passed to \sphinxtitleref{PtmProcessor}.
A \sphinxtitleref{dict} can be supplied for each resource, e.g.
\sphinxtitleref{\{‘Signor’: \{…\}, ‘PhosphoSite’: \{…\}, …\}}.
Those not used by \sphinxtitleref{PtmProcessor} are forwarded to the
\sphinxtitleref{pypath.dataio} methods.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{load\_resources() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.load_resources}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{load\_resources}}}{\emph{lst=\{'arn': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517d88\textgreater{}}, \emph{'biogrid': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528cef0\textgreater{}}, \emph{'ca1': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517c68\textgreater{}}, \emph{'ccmap': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c248\textgreater{}}, \emph{'cellphonedb': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc52a7050\textgreater{}}, \emph{'dbptm': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c680\textgreater{}}, \emph{'death': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517f38\textgreater{}}, \emph{'depod': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c4d0\textgreater{}}, \emph{'dip': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c2d8\textgreater{}}, \emph{'domino': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c5f0\textgreater{}}, \emph{'elm': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c5a8\textgreater{}}, \emph{'guide2pharma': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517b90\textgreater{}}, \emph{'hpmr': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc52a2f80\textgreater{}}, \emph{'hprd': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc529b098\textgreater{}}, \emph{'hprd\_p': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c6c8\textgreater{}}, \emph{'innatedb': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c368\textgreater{}}, \emph{'intact': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528cea8\textgreater{}}, \emph{'lmpid': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c518\textgreater{}}, \emph{'macrophage': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517ef0\textgreater{}}, \emph{'matrixdb': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c440\textgreater{}}, \emph{'mppi': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c290\textgreater{}}, \emph{'nrf2': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517e60\textgreater{}}, \emph{'pdz': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517f80\textgreater{}}, \emph{'phelm': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c560\textgreater{}}, \emph{'psite': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c488\textgreater{}}, \emph{'ramilowski2015': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc52a70e0\textgreater{}}, \emph{'signalink3': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517b48\textgreater{}}, \emph{'signor': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc528c1b8\textgreater{}}, \emph{'spike': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517b00\textgreater{}}, \emph{'trip': \textless{}pypath.input\_formats.ReadSettings instance at 0x7f6bc5517a28\textgreater{}\}}, \emph{exclude={[}{]}}, \emph{cache\_files=\{\}}, \emph{reread=False}, \emph{redownload=False}}{}
Loads multiple resources, and cleans up after.
Looks up ID types, and loads all ID conversion
tables from UniProt if necessary. This is much
faster than loading the ID conversion and the
resources one by one.

\end{fulllineitems}

\index{load\_tfregulons() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.load_tfregulons}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{load\_tfregulons}}}{\emph{levels=set({[}'A'}, \emph{'B'{]})}, \emph{only\_curated=False}}{}
Adds TF-target interactions from TF regulons to the network.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{levels}} (\sphinxstyleliteralemphasis{\sphinxupquote{set}}) \textendash{} Confidence levels to be used.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{only\_curated}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} Retrieve only literature curated interactions.

\end{itemize}

\end{description}\end{quote}

TF regulons is a comprehensive resource of TF-target interactions
combining multiple lines of evidences: literature curated databases,
ChIP-Seq data, PWM based prediction using HOCOMOCO and JASPAR matrices
and prediction from GTEx expression data by ARACNe.

For details see \sphinxurl{https://github.com/saezlab/DoRothEA}.

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{k+kn}{import} \PYG{n+nn}{pypath}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{pa} \PYG{o}{=} \PYG{n}{pypath}\PYG{o}{.}\PYG{n}{PyPath}\PYG{p}{(}\PYG{p}{)}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{pa}\PYG{o}{.}\PYG{n}{load\PYGZus{}tfregulons}\PYG{p}{(}\PYG{n}{levels} \PYG{o}{=} \PYG{p}{\PYGZob{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{A}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{\PYGZcb{}}\PYG{p}{)}
\end{sphinxVerbatim}

\end{fulllineitems}

\index{map\_edge() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.map_edge}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{map\_edge}}}{\emph{edge}}{}
Translates molecule names in dict representing an edge.

\end{fulllineitems}

\index{map\_item() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.map_item}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{map\_item}}}{\emph{item}}{}
Translates the name in item representing a molecule.

\end{fulllineitems}

\index{map\_list() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.map_list}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{map\_list}}}{\emph{lst}, \emph{singleList=False}}{}
Only a wrapper for map\_edge()

\end{fulllineitems}

\index{merge\_lists() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.merge_lists}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{merge\_lists}}}{\emph{nameA}, \emph{nameB}, \emph{name=None}, \emph{and\_or='and'}, \emph{delete=False}, \emph{func='max'}}{}
Merges two lists in \sphinxtitleref{lists}.

\end{fulllineitems}

\index{merge\_nodes() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.merge_nodes}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{merge\_nodes}}}{\emph{nodes}, \emph{primary=None}, \emph{graph=None}}{}
Merges all attributes and all edges of selected nodes
and assigns them to the primary node
(by default the one with lowest ID).
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{nodes}} (\sphinxstyleliteralemphasis{\sphinxupquote{list}}) \textendash{} List of edge IDs.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{primary}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} ID of the primary edge;
if None the lowest ID selected.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{mutated\_edges() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.mutated_edges}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{mutated\_edges}}}{\emph{sample}}{}
Compares the mutated residues and the modified residues in PTMs.
Interactions are marked as mutated if the target residue in the
underlying PTM is mutated.

\end{fulllineitems}

\index{network\_filter() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.network_filter}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{network\_filter}}}{\emph{p=2.0}}{}
This function aims to cut the number of edges in the network,
without loosing nodes, to make the network less connected,
less hairball-like, more usable for analysis.

\end{fulllineitems}

\index{network\_stats() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.network_stats}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{network\_stats}}}{\emph{outfile=None}}{}
Calculates basic statistics for the whole network
and each of sources. Writes the results in a tab file.

\end{fulllineitems}

\index{orthology\_translation() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.orthology_translation}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{orthology\_translation}}}{\emph{target}, \emph{source=None}, \emph{only\_swissprot=True}, \emph{graph=None}}{}
Translates the current object to another organism by orthology.
Proteins without known ortholog will be deleted.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{target}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} NCBI Taxonomy ID of the target organism.
E.g. 10090 for mouse.

\end{description}\end{quote}

\end{fulllineitems}

\index{p() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.p}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{p}}}{\emph{identifier}}{}
Returns \sphinxcode{\sphinxupquote{igraph.Vertex()}} object if the identifier
is a valid vertex index in the default undirected graph,
or a UniProt ID or GeneSymbol which can be found in the
default undirected network, otherwise \sphinxcode{\sphinxupquote{None}}.
\begin{description}
\item[{@identifier}] \leavevmode{[}int, str{]}
Vertex index (int) or GeneSymbol (str) or UniProt ID (str).

\end{description}

\end{fulllineitems}

\index{pathway\_members() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.pathway_members}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{pathway\_members}}}{\emph{pathway}, \emph{source}}{}
Returns an iterator with the members of a single pathway.
Apart from the pathway name you need to supply its source
database too.

\end{fulllineitems}

\index{pathway\_names() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.pathway_names}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{pathway\_names}}}{\emph{source}, \emph{graph=None}}{}
Returns the names of all pathways having at least one member
in the current graph.

\end{fulllineitems}

\index{process\_dmi() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.process_dmi}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{process\_dmi}}}{\emph{source}, \emph{**kwargs}}{}
This is an universal function
for loading domain-motif objects
like load\_phospho\_dmi() for PTMs.
TODO this will replace load\_elm, load\_ielm, etc

\end{fulllineitems}

\index{protein() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.protein}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{protein}}}{\emph{identifier}}{}
Same as \sphinxcode{\sphinxupquote{PyPath.get\_node}}, just for the directed graph.
Returns \sphinxcode{\sphinxupquote{igraph.Vertex()}} object if the identifier
is a valid vertex index in the default directed graph,
or a UniProt ID or GeneSymbol which can be found in the
default directed network, otherwise \sphinxcode{\sphinxupquote{None}}.
\begin{description}
\item[{@identifier}] \leavevmode{[}int, str{]}
Vertex index (int) or GeneSymbol (str) or UniProt ID (str).

\end{description}

\end{fulllineitems}

\index{protein\_edge() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.protein_edge}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{protein\_edge}}}{\emph{source}, \emph{target}, \emph{directed=True}}{}
Returns \sphinxcode{\sphinxupquote{igraph.Edge}} object if an edge exist between
the 2 proteins, otherwise \sphinxcode{\sphinxupquote{None}}.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{source}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{,}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Vertex index or UniProt ID or GeneSymbol

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{target}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{,}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Vertex index or UniProt ID or GeneSymbol

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{directed}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} To be passed to igraph.Graph.get\_eid()

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{proteome\_list() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.proteome_list}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{proteome\_list}}}{\emph{swissprot=True}}{}
Loads the whole proteome as a list.

\end{fulllineitems}

\index{random\_walk\_with\_return() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.random_walk_with_return}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{random\_walk\_with\_return}}}{\emph{q}, \emph{graph=None}, \emph{c=0.5}, \emph{niter=1000}}{}
Random walk with return (RWR) starting from one or more query nodes.
Returns affinity (probability) vector of all nodes in the graph.
\begin{quote}
\begin{quote}\begin{description}
\item[{param int,list q}] \leavevmode
Vertex IDs of query nodes.

\item[{param igraph.Graph graph}] \leavevmode
An \sphinxtitleref{igraph.Graph} object.

\item[{param float c}] \leavevmode
Probability of restart.

\item[{param int niter}] \leavevmode
Number of iterations.

\end{description}\end{quote}
\end{quote}

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{k+kn}{import} \PYG{n+nn}{igraph}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{k+kn}{import} \PYG{n+nn}{pypath}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{pa} \PYG{o}{=} \PYG{n}{pypath}\PYG{o}{.}\PYG{n}{PyPath}\PYG{p}{(}\PYG{p}{)}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{pa}\PYG{o}{.}\PYG{n}{init\PYGZus{}network}\PYG{p}{(}\PYG{p}{\PYGZob{}}
\PYG{g+go}{        \PYGZsq{}signor\PYGZsq{}: pypath.data\PYGZus{}formats.pathway[\PYGZsq{}signor\PYGZsq{}]}
\PYG{g+go}{    \PYGZcb{})}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{q} \PYG{o}{=} \PYG{p}{[}
\PYG{g+go}{        pa.gs(\PYGZsq{}EGFR\PYGZsq{}).index,}
\PYG{g+go}{        pa.gs(\PYGZsq{}ATG4B\PYGZsq{}).index}
\PYG{g+go}{    ]}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{rwr} \PYG{o}{=} \PYG{n}{pa}\PYG{o}{.}\PYG{n}{random\PYGZus{}walk\PYGZus{}with\PYGZus{}return}\PYG{p}{(}\PYG{n}{q} \PYG{o}{=} \PYG{n}{q}\PYG{p}{)}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{palette} \PYG{o}{=} \PYG{n}{igraph}\PYG{o}{.}\PYG{n}{RainbowPalette}\PYG{p}{(}\PYG{n}{n} \PYG{o}{=} \PYG{l+m+mi}{100}\PYG{p}{)}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{colors}  \PYG{o}{=} \PYG{p}{[}\PYG{n}{palette}\PYG{o}{.}\PYG{n}{get}\PYG{p}{(}\PYG{n+nb}{int}\PYG{p}{(}\PYG{n+nb}{round}\PYG{p}{(}\PYG{n}{i}\PYG{p}{)}\PYG{p}{)}\PYG{p}{)} \PYG{k}{for} \PYG{n}{i} \PYG{o+ow}{in} \PYG{n}{rwr} \PYG{o}{/} \PYG{n+nb}{max}\PYG{p}{(}\PYG{n}{rwr}\PYG{p}{)} \PYG{o}{*} \PYG{l+m+mi}{99}\PYG{p}{]}
\PYG{g+gp}{\PYGZgt{}\PYGZgt{}\PYGZgt{} }\PYG{n}{igraph}\PYG{o}{.}\PYG{n}{plot}\PYG{p}{(}\PYG{n}{pa}\PYG{o}{.}\PYG{n}{graph}\PYG{p}{,} \PYG{n}{vertex\PYGZus{}color} \PYG{o}{=} \PYG{n}{colors}\PYG{p}{)}
\end{sphinxVerbatim}

\end{fulllineitems}

\index{random\_walk\_with\_return2() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.random_walk_with_return2}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{random\_walk\_with\_return2}}}{\emph{q}, \emph{c=0.5}, \emph{niter=1000}}{}
Literally does random walks.
Only for testing of the other method, to be deleted later.

\end{fulllineitems}

\index{read\_data\_file() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.read_data_file}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{read\_data\_file}}}{\emph{settings}, \emph{keep\_raw=False}, \emph{cache\_files=\{\}}, \emph{reread=False}, \emph{redownload=False}}{}
Interaction data with node and edge attributes can be read
from simple text based files. This function works not only
with files, but with lists as well. Any other function can
be written to download a preprocess data, and then give it
to this function to finally attach to the network.
\begin{description}
\item[{@settings}] \leavevmode{[}ReadSettings instance{]}
The detailed definition of the input format. Instead of
the file name you can give a function name, which will
be executed, and the returned data will be used.

\item[{@keep\_raw}] \leavevmode{[}boolean{]}
To keep the raw data read by this function, in order for
debugging purposes, or further use.

\end{description}

\end{fulllineitems}

\index{receptors\_list() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.receptors_list}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{receptors\_list}}}{}{}
Loads the Human Plasma Membrane Receptome as a list.
This resource is human only.

\end{fulllineitems}

\index{save\_session() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.save_session}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{save\_session}}}{}{}
Save current state into pickle dump.

\end{fulllineitems}

\index{select\_by\_go() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.select_by_go}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{select\_by\_go}}}{\emph{go\_terms}, \emph{go\_desc=None}, \emph{aspects=('C'}, \emph{'F'}, \emph{'P')}, \emph{method='ANY'}}{}
Selects the nodes annotated by certain GO terms.

Returns set of vertex IDs.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{method}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} If \sphinxtitleref{ANY} nodes annotated with any of the terms returned.
If \sphinxtitleref{ALL} nodes annotated with all the terms returned.

\end{description}\end{quote}

\end{fulllineitems}

\index{separate() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.separate}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{separate}}}{}{}
Separates networks from different sources.
Returns dict of igraph objects.

\end{fulllineitems}

\index{separate\_by\_category() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.separate_by_category}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{separate\_by\_category}}}{}{}
Separate networks based on categories.
Returns dict of igraph objects.

\end{fulllineitems}

\index{set\_chembl\_mysql() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.set_chembl_mysql}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{set\_chembl\_mysql}}}{\emph{title}, \emph{config\_file=None}}{}
Sets the ChEMBL MySQL config according to
\sphinxtitleref{title} section in \sphinxtitleref{config\_file} ini style config.
\begin{quote}

title (str): section title in ini file
config\_file (str, NoneType): config file name;
\begin{quote}

if None, the \sphinxtitleref{mysql\_config/defaults.mysql}
will be used
\end{quote}
\end{quote}

\end{fulllineitems}

\index{set\_disease\_genes() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.set_disease_genes}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{set\_disease\_genes}}}{\emph{dataset='curated'}}{}
Creates a vertex attribute named \sphinxtitleref{dis} with boolean values \sphinxstyleemphasis{True}
if the protein encoded by a disease related gene according to
DisGeNet.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{dataset}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Which dataset to use from DisGeNet. Default is \sphinxtitleref{curated}.

\end{description}\end{quote}

\end{fulllineitems}

\index{set\_druggability() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.set_druggability}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{set\_druggability}}}{}{}
Creates a vertex attribute \sphinxtitleref{dgb} with value \sphinxstyleemphasis{True} if
the protein is druggable, otherwise \sphinxstyleemphasis{False}.

\end{fulllineitems}

\index{set\_drugtargets() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.set_drugtargets}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{set\_drugtargets}}}{\emph{pchembl=5.0}}{}
Creates a vertex attribute \sphinxtitleref{dtg} with value \sphinxstyleemphasis{True} if
the protein has at least one compound binding with
affinity higher than \sphinxtitleref{pchembl}, otherwise \sphinxstyleemphasis{False}.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{pchembl}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}) \textendash{} Pchembl threshold.

\end{description}\end{quote}

\end{fulllineitems}

\index{set\_kinases() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.set_kinases}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{set\_kinases}}}{}{}
Creates a vertex attribute \sphinxtitleref{kin} with value \sphinxstyleemphasis{True} if
the protein is a kinase, otherwise \sphinxstyleemphasis{False}.

\end{fulllineitems}

\index{set\_receptors() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.set_receptors}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{set\_receptors}}}{}{}
Creates a vertex attribute \sphinxtitleref{rec} with value \sphinxstyleemphasis{True} if
the protein is a receptor, otherwise \sphinxstyleemphasis{False}.

\end{fulllineitems}

\index{set\_signaling\_proteins() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.set_signaling_proteins}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{set\_signaling\_proteins}}}{}{}
Creates a vertex attribute \sphinxtitleref{kin} with value \sphinxstyleemphasis{True} if
the protein is a kinase, otherwise \sphinxstyleemphasis{False}.

\end{fulllineitems}

\index{set\_transcription\_factors() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.set_transcription_factors}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{set\_transcription\_factors}}}{\emph{classes={[}'a', 'b', 'other'{]}}}{}
Creates a vertex attribute \sphinxtitleref{tf} with value \sphinxstyleemphasis{True} if
the protein is a transcription factor, otherwise \sphinxstyleemphasis{False}.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{classes}} (\sphinxstyleliteralemphasis{\sphinxupquote{list}}) \textendash{} Classes to use from TF Census. Default is \sphinxtitleref{{[}‘a’, ‘b’, ‘other’{]}}.

\end{description}\end{quote}

\end{fulllineitems}

\index{shortest\_path\_dist() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.shortest_path_dist}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{shortest\_path\_dist}}}{\emph{graph=None}, \emph{subset=None}, \emph{outfile=None}, \emph{**kwargs}}{}
subset is a tuple of two lists if you wish to look for
paths between elements of two groups, or a list if you
wish to look for shortest paths within this group

\end{fulllineitems}

\index{signaling\_proteins\_list() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.signaling_proteins_list}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{signaling\_proteins\_list}}}{}{}
Compiles a list of signaling proteins (as opposed to other proteins
like metabolic enzymes, matrix proteins), by looking up a few simple
keywords in short description of GO terms.

\end{fulllineitems}

\index{small\_plot() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.small_plot}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{small\_plot}}}{\emph{graph}, \emph{**kwargs}}{}
This method is deprecated, do not use it.

\end{fulllineitems}

\index{source\_network() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.source_network}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{source\_network}}}{\emph{font='HelveticaNeueLTStd'}}{}
For EMBL branding, use Helvetica Neue Linotype Standard light

\end{fulllineitems}

\index{straight\_between() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.straight_between}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{straight\_between}}}{\emph{nameA}, \emph{nameB}}{}
This does actually the same as get\_edge(), but by names
instead of vertex ids.

\end{fulllineitems}

\index{sum\_in\_complex() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.sum_in_complex}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sum\_in\_complex}}}{\emph{csources={[}'corum'{]}, graph=None}}{}
Returns the total number of edges in the network falling
between two members of the same complex.
Returns as a dict by complex resources.
Calls :py:func:pypath.pypath.Pypath.edges\_in\_comlexes()
to do the calculations.
\begin{description}
\item[{@csources}] \leavevmode{[}list{]}
List of complex resources. Should be already loaded.

\item[{@graph}] \leavevmode{[}igraph.Graph(){]}
The graph object to do the calculations on.

\end{description}

\end{fulllineitems}

\index{tfs\_list() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.tfs_list}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{tfs\_list}}}{}{}
Loads the list of all known transcription factors from TF census
(Vaquerizas 2009). This resource is human only.

\end{fulllineitems}

\index{third\_source\_directions() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.third_source_directions}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{third\_source\_directions}}}{\emph{graph=None}, \emph{use\_string\_effects=False}, \emph{use\_laudanna\_data=False}}{}
This method calls a series of methods to get
additional direction \& effect information
from sources having no literature curated references,
but giving sufficient evidence about the directionality
for interactions already supported by literature
evidences from other sources.

\end{fulllineitems}

\index{tissue\_network() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.tissue_network}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{tissue\_network}}}{\emph{tissue}, \emph{graph=None}}{}
Returns a network which includes the proteins expressed in
certain tissue according to ProteomicsDB.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{tissue}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Tissue name as used in ProteomicsDB.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{graph}} (\sphinxstyleliteralemphasis{\sphinxupquote{igraph.Graph}}) \textendash{} A graph object, by default the \sphinxtitleref{graph} attribute of
the current instance.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{uniprot() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.uniprot}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{uniprot}}}{\emph{uniprot}}{}
Returns \sphinxcode{\sphinxupquote{igraph.Vertex()}} object if the UniProt
can be found in the default undirected network,
otherwise \sphinxcode{\sphinxupquote{None}}.
\begin{description}
\item[{@uniprot}] \leavevmode{[}str{]}
UniProt ID.

\end{description}

\end{fulllineitems}

\index{uniprots() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.uniprots}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{uniprots}}}{\emph{uniprots}}{}
Returns list of \sphinxcode{\sphinxupquote{igraph.Vertex()}} object
for a list of UniProt IDs omitting those
could not be found in the default
undirected graph.

\end{fulllineitems}

\index{up() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.up}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{up}}}{\emph{uniprot}}{}
Returns \sphinxcode{\sphinxupquote{igraph.Vertex()}} object if the UniProt
can be found in the default undirected network,
otherwise \sphinxcode{\sphinxupquote{None}}.
\begin{description}
\item[{@uniprot}] \leavevmode{[}str{]}
UniProt ID.

\end{description}

\end{fulllineitems}

\index{up\_edge() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.up_edge}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{up\_edge}}}{\emph{source}, \emph{target}, \emph{directed=True}}{}
Returns \sphinxcode{\sphinxupquote{igraph.Edge}} object if an edge exist between
the 2 proteins, otherwise \sphinxcode{\sphinxupquote{None}}.
\begin{description}
\item[{@source}] \leavevmode{[}str{]}
UniProt ID

\item[{@target}] \leavevmode{[}str{]}
UniProt ID

\item[{@directed}] \leavevmode{[}bool{]}
To be passed to igraph.Graph.get\_eid()

\end{description}

\end{fulllineitems}

\index{update\_adjlist() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.update_adjlist}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{update\_adjlist}}}{\emph{graph=None}, \emph{mode='ALL'}}{}
Creates an adjacency list in a dict of sets format.

\end{fulllineitems}

\index{update\_cats() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.update_cats}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{update\_cats}}}{}{}
Makes sure that the \sphinxtitleref{has\_cats} attribute is an up to date
set of all categories in the current network.

\end{fulllineitems}

\index{update\_sources() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.update_sources}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{update\_sources}}}{}{}
Makes sure that the \sphinxtitleref{sources} attribute is an up to date
list of all sources in the current network.

\end{fulllineitems}

\index{update\_vindex() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.update_vindex}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{update\_vindex}}}{}{}
This is deprecated.

\end{fulllineitems}

\index{update\_vname() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.update_vname}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{update\_vname}}}{}{}
For fast lookup of node names and indexes, these are
hold in a list and a dict as well. However, every time
new nodes are added, these should be updated. This
function is automatically called after all operations
affecting node indices.

\end{fulllineitems}

\index{ups() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.ups}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{ups}}}{\emph{uniprots}}{}
Returns list of \sphinxcode{\sphinxupquote{igraph.Vertex()}} object
for a list of UniProt IDs omitting those
could not be found in the default
undirected graph.

\end{fulllineitems}

\index{v() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.v}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{v}}}{\emph{identifier}}{}
Returns \sphinxcode{\sphinxupquote{igraph.Vertex()}} object if the identifier
is a valid vertex index in the default undirected graph,
or a UniProt ID or GeneSymbol which can be found in the
default undirected network, otherwise \sphinxcode{\sphinxupquote{None}}.
\begin{description}
\item[{@identifier}] \leavevmode{[}int, str{]}
Vertex index (int) or GeneSymbol (str) or UniProt ID (str).

\end{description}

\end{fulllineitems}

\index{vertex\_pathways() (pypath.main.PyPath method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.PyPath.vertex_pathways}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{vertex\_pathways}}}{}{}
Some resources assignes interactions some others
proteins to pathways.
This function converts pathway annotations from
edge attributes to vertex attributes.

\end{fulllineitems}


\end{fulllineitems}

\index{Direction (class in pypath.main)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.Direction}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{pypath.main.}}\sphinxbfcode{\sphinxupquote{Direction}}}{\emph{nameA}, \emph{nameB}}{}
Object storing directionality information of an edge.
\begin{itemize}
\item {} \begin{description}
\item[{Arguments:}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleemphasis{nameA} {[}str{]}: Name of the source node.

\item {} 
\sphinxstyleemphasis{nameB} {[}str{]}: Name of the target node.

\end{itemize}

\end{description}

\item {} \begin{description}
\item[{Attributes:}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleemphasis{dirs} {[}dict{]}: Dictionary containing the presence of
directionality of the given edge. Keys are \sphinxcode{\sphinxupquote{straight}},
\sphinxcode{\sphinxupquote{reverse}} and \sphinxcode{\sphinxupquote{'undirected'}} and their values denote the
presence/absence {[}bool{]}.

\item {} 
\sphinxstyleemphasis{negative} {[}dict{]}: Dictionary contianing the presence/absence
{[}bool{]} of negative interactions for both \sphinxcode{\sphinxupquote{straight}} and
\sphinxcode{\sphinxupquote{reverse}} directions.

\item {} 
\sphinxstyleemphasis{negative\_sources} {[}dict{]}: Contains the resource names {[}str{]}
supporting a negative interaction on \sphinxcode{\sphinxupquote{straight}} and
\sphinxcode{\sphinxupquote{reverse}} directions.

\item {} 
\sphinxstyleemphasis{nodes} {[}list{]}: Contains the node names {[}str{]} sorted
alphabetically (\sphinxstyleemphasis{nameA}, \sphinxstyleemphasis{nameB}).

\item {} 
\sphinxstyleemphasis{positive} {[}dict{]}:  Dictionary contianing the presence/absence
{[}bool{]} of positive interactions for both \sphinxcode{\sphinxupquote{straight}} and
\sphinxcode{\sphinxupquote{reverse}} directions.

\item {} 
\sphinxstyleemphasis{positive\_sources} {[}dict{]}: Contains the resource names {[}str{]}
supporting a positive interaction on \sphinxcode{\sphinxupquote{straight}} and
\sphinxcode{\sphinxupquote{reverse}} directions.

\item {} 
\sphinxstyleemphasis{reverse} {[}tuple{]}: Contains the node names {[}str{]} in the original
order e.g. \sphinxcode{\sphinxupquote{(source, target)}}.

\item {} 
\sphinxstyleemphasis{sources} {[}dict{]}: Contains the resource names {[}str{]} of a given
edge for each directionality (\sphinxcode{\sphinxupquote{straight}}, \sphinxcode{\sphinxupquote{reverse}} and
\sphinxcode{\sphinxupquote{'undirected'}}). Values are sets containing the names of
those resources supporting such directionality.

\item {} 
\sphinxstyleemphasis{straight} {[}tuple{]}: Contains the node names {[}str{]} in reverse
order e.g. \sphinxcode{\sphinxupquote{(target, source)}}.

\end{itemize}

\end{description}

\end{itemize}
\index{consensus\_edges() (pypath.main.Direction method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.Direction.consensus_edges}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{consensus\_edges}}}{}{}
Returns list of edges based on majority consensus of directions and signs.

\end{fulllineitems}

\index{get\_dir() (pypath.main.Direction method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.Direction.get_dir}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_dir}}}{\emph{direction}, \emph{sources=False}}{}
Returns boolean or list of sources

\end{fulllineitems}

\index{get\_dirs() (pypath.main.Direction method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.Direction.get_dirs}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_dirs}}}{\emph{src}, \emph{tgt}, \emph{sources=False}}{}
Returns all directions with boolean values
or list of sources.

\end{fulllineitems}

\index{majority\_dir() (pypath.main.Direction method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.Direction.majority_dir}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{majority\_dir}}}{}{}
Returns directionality based on majority consensus.
Returns \sphinxtitleref{None} if the number of sources supporting the two opposite
directions are the same. Returns \sphinxtitleref{‘undirected’} if there is no directionality
information. Returns \sphinxtitleref{tuple} of IDs if one direction is supported by
more sources.

\end{fulllineitems}

\index{majority\_sign() (pypath.main.Direction method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.Direction.majority_sign}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{majority\_sign}}}{}{}
Returns signs based on majority consensus. Keys in the returned \sphinxtitleref{dict} are directions.
Values are \sphinxtitleref{None} if the direction lacks effect sign. Otherwise \sphinxtitleref{tuples} with their
first element \sphinxtitleref{True} if the number of sources supporting stimulation in the given
direction is greater or equal compared to those supporting inhibition. The second
value is the same for inhibition.

\end{fulllineitems}

\index{set\_dir() (pypath.main.Direction method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.Direction.set_dir}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{set\_dir}}}{\emph{direction}, \emph{source}}{}
Adds directionality information with
the corresponding data source named.

\end{fulllineitems}

\index{src() (pypath.main.Direction method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.Direction.src}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{src}}}{\emph{undirected=False}}{}
Returns the IDs of effector molecules in this directed
interaction. If the interaction is bidirectional, the
list will contain 2 IDs. If the interaction is undirec-
ted, an empty list will be returned.

\end{fulllineitems}

\index{tgt() (pypath.main.Direction method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.Direction.tgt}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{tgt}}}{\emph{undirected=False}}{}
Returns the IDs of the target molecules in the inter-
action. Same behaviour as \sphinxtitleref{Direction.src()}.

\end{fulllineitems}

\index{unset\_dir() (pypath.main.Direction method)}

\begin{fulllineitems}
\phantomsection\label{\detokenize{main:pypath.main.Direction.unset_dir}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{unset\_dir}}}{\emph{direction}, \emph{source=None}}{}
Removes directionality information,
or single source.

\end{fulllineitems}


\end{fulllineitems}

\begin{quote}\begin{description}
\item[{note}] \leavevmode
\sphinxcode{\sphinxupquote{pypath}} supports both Python 2.7 and Python 3.6+. In the beginning,
pypath has been developed only for Python 2.7. Then the code have been
adjusted to Py3 however we can not guarantee no incompatibilities
remained. If you find any method does not work please submit an issue on
github. For few years I develop and test \sphinxcode{\sphinxupquote{pypath}} in Python 3. Therefore
this is the better supported Python variant.

\item[{contributions}] \leavevmode
\sphinxhref{mailto:turei.denes@gmail.com}{turei.denes@gmail.com}

\item[{documentation}] \leavevmode
\sphinxurl{http://pypath.omnipathdb.org/}

\item[{issues}] \leavevmode
\sphinxurl{https://github.com/saezlab/pypath/issues}

\end{description}\end{quote}

\sphinxstylestrong{pypath} is a Python package built around igraph to work with molecular
network representations e.g. protein, miRNA and drug compound interaction
networks.


\chapter{Webservice}
\label{\detokenize{index:webservice}}\label{\detokenize{index::doc}}
\sphinxstylestrong{New webservice} from 14 June 2018: the queries slightly changed, have been
largely extended. See the examples below.

One instance of the pypath webservice runs at the domain
\sphinxurl{http://omnipathdb.org/}, serving not only the OmniPath data but other datasets:
TF-target interactions from TF Regulons, a large collection additional
enzyme-substrate interactions, and literature curated miRNA-mRNA interacions
combined from 4 databases. The webservice implements a very simple REST style
API, you can make requests by HTTP protocol (browser, wget, curl or whatever).

The webservice currently recognizes 3 types of queries: \sphinxcode{\sphinxupquote{interactions}},
\sphinxcode{\sphinxupquote{ptms}} and \sphinxcode{\sphinxupquote{info}}. The query types \sphinxcode{\sphinxupquote{resources}}, \sphinxcode{\sphinxupquote{network}} and
\sphinxcode{\sphinxupquote{about}} have not been implemented yet in the new webservice.


\section{Mouse and rat}
\label{\detokenize{index:mouse-and-rat}}
Except the miRNA interactions all interactions are available for human, mouse
and rat. The rodent data has been translated from human using the NCBI
Homologene database. Many human proteins have no known homolog in rodents
hence rodent datasets are smaller than their human counterparts. Note, if you
work with mouse omics data you might do better to translate your dataset to
human (for example using the \sphinxcode{\sphinxupquote{pypath.homology}} module) and use human
interaction data.


\section{Examples}
\label{\detokenize{index:examples}}
A request without any parameter, gives some basic numbers about the actual
loaded dataset:
\begin{quote}

\sphinxurl{http://omnipathdb.org}
\end{quote}

The \sphinxcode{\sphinxupquote{info}} returns a HTML page with comprehensive information about the
resources:
\begin{quote}

\sphinxurl{http://omnipathdb.org/info}
\end{quote}

The \sphinxcode{\sphinxupquote{interactions}} query accepts some parameters and returns interactions in
tabular format. This example returns all interactions of EGFR (P00533), with
sources and references listed.
\begin{quote}

\sphinxurl{http://omnipathdb.org/interactions/?partners=P00533\&fields=sources,references}
\end{quote}

By default only the OmniPath dataset used, to query the TF Regulons or add the
extra enzyme-substrate interactions you need to set additional parameters. For
example to query the transcriptional regulators of EGFR:
\begin{quote}

\sphinxurl{http://omnipathdb.org/interactions/?targets=EGFR\&types=TF}
\end{quote}

The TF Regulons database assigns confidence levels to the interactions. You
might want to select only the highest confidence, \sphinxstyleemphasis{A} category:
\begin{quote}

\sphinxurl{http://omnipathdb.org/interactions/?targets=EGFR\&types=TF\&tfregulons\_levels=A}
\end{quote}

Show the transcriptional targets of Smad2 homology translated to rat including
the confidence levels from TF Regulons:
\begin{quote}

\sphinxurl{http://omnipathdb.org/interactions/?genesymbols=1\&fields=type,ncbi\_tax\_id,tfregulons\_level\&organisms=10116\&sources=Smad2\&types=TF}
\end{quote}

Query interactions from PhosphoNetworks which is part of the \sphinxstyleemphasis{kinaseextra}
dataset:
\begin{quote}

\sphinxurl{http://omnipathdb.org/interactions/?genesymbols=1\&fields=sources\&databases=PhosphoNetworks\&datasets=kinaseextra}
\end{quote}

Get the interactions from Signor, SPIKE and SignaLink3:
\begin{quote}

\sphinxurl{http://omnipathdb.org/interactions/?genesymbols=1\&fields=sources,references\&databases=Signor,SPIKE,SignaLink3}
\end{quote}

All interactions of MAP1LC3B:
\begin{quote}

\sphinxurl{http://omnipathdb.org/interactions/?genesymbols=1\&partners=MAP1LC3B}
\end{quote}

By default \sphinxcode{\sphinxupquote{partners}} queries the interaction where either the source or the
arget is among the partners. If you set the \sphinxcode{\sphinxupquote{source\_target}} parameter to
\sphinxcode{\sphinxupquote{AND}} both the source and the target must be in the queried set:
\begin{quote}

\sphinxurl{http://omnipathdb.org/interactions/?genesymbols=1\&fields=sources,references\&sources=ATG3,ATG7,ATG4B,SQSTM1\&targets=MAP1LC3B,MAP1LC3A,MAP1LC3C,Q9H0R8,GABARAP,GABARAPL2\&source\_target=AND}
\end{quote}

As you see above you can use UniProt IDs and Gene Symbols in the queries and
also mix them. Get the miRNA regulating NOTCH1:
\begin{quote}

\sphinxurl{http://omnipathdb.org/interactions/?genesymbols=1\&fields=sources,references\&datasets=mirnatarget\&targets=NOTCH1}
\end{quote}

Note: with the exception of mandatory fields and genesymbols, the columns
appear exactly in the order you provided in your query.

Another query type available is \sphinxcode{\sphinxupquote{ptms}} which provides enzyme-substrate
interactions. It is very similar to the \sphinxcode{\sphinxupquote{interactions}}:
\begin{quote}

\sphinxurl{http://omnipathdb.org/ptms?genesymbols=1\&fields=sources,references,isoforms\&enzymes=FYN}
\end{quote}

Is there any ubiquitination reaction?
\begin{quote}

\sphinxurl{http://omnipathdb.org/ptms?genesymbols=1\&fields=sources,references\&types=ubiquitination}
\end{quote}

And acetylation in mouse?
\begin{quote}

\sphinxurl{http://omnipathdb.org/ptms?genesymbols=1\&fields=sources,references\&types=acetylation\&organisms=10090}
\end{quote}

Rat interactions, both directly from rat and homology translated from human,
from the PhosphoSite database:
\begin{quote}

\sphinxurl{http://omnipathdb.org/ptms?genesymbols=1\&fields=sources,references\&organisms=10116\&databases=PhosphoSite,PhosphoSite\_noref}
\end{quote}


\chapter{Can I use OmniPath in R?}
\label{\detokenize{index:can-i-use-omnipath-in-r}}
You can download the data from the webservice and load into R. Look here for
an example:
\begin{quote}

\sphinxurl{https://github.com/saezlab/pypath/tree/master/r\_import}
\end{quote}


\chapter{Installation}
\label{\detokenize{index:installation}}

\section{Linux}
\label{\detokenize{index:linux}}
In almost any up-to-date Linux distribution the dependencies of \sphinxstylestrong{pypath} are
built-in, or provided by the distributors. You only need to install a couple
of things in your package manager (cairo, py(2)cairo, igraph,
python(2)-igraph, graphviz, pygraphviz), and after install \sphinxstylestrong{pypath} by \sphinxstyleemphasis{pip}
(see below). If any module still missing, you can install them the usual way
by \sphinxstyleemphasis{pip} or your package manager.


\section{igraph C library, cairo and pycairo}
\label{\detokenize{index:igraph-c-library-cairo-and-pycairo}}
\sphinxstyleemphasis{python(2)-igraph} is a Python interface to use the igraph C library. The
C library must be installed. The same goes for \sphinxstyleemphasis{cairo}, \sphinxstyleemphasis{py(2)cairo} and
\sphinxstyleemphasis{graphviz}.


\section{Directly from git}
\label{\detokenize{index:directly-from-git}}
\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{pip} \PYG{n}{install} \PYG{n}{git}\PYG{o}{+}\PYG{n}{https}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{n}{github}\PYG{o}{.}\PYG{n}{com}\PYG{o}{/}\PYG{n}{saezlab}\PYG{o}{/}\PYG{n}{pypath}\PYG{o}{.}\PYG{n}{git}
\end{sphinxVerbatim}


\section{With pip}
\label{\detokenize{index:with-pip}}
Download the package from /dist, and install with pip:

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{pip} \PYG{n}{install} \PYG{n}{pypath}\PYG{o}{\PYGZhy{}}\PYG{n}{x}\PYG{o}{.}\PYG{n}{y}\PYG{o}{.}\PYG{n}{z}\PYG{o}{.}\PYG{n}{tar}\PYG{o}{.}\PYG{n}{gz}
\end{sphinxVerbatim}


\section{Build source distribution}
\label{\detokenize{index:build-source-distribution}}
Clone the git repo, and run setup.py:

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{python} \PYG{n}{setup}\PYG{o}{.}\PYG{n}{py} \PYG{n}{sdist}
\end{sphinxVerbatim}


\section{Mac OS X}
\label{\detokenize{index:mac-os-x}}
On OS X installation is not straightforward primarily because cairo needs to
be compiled from source. We provide 2 scripts here: the
\sphinxstylestrong{mac-install-brew.sh} installs everything with HomeBrew, and
\sphinxstylestrong{mac-install-conda.sh} installs from Anaconda distribution. With these
scripts installation of igraph, cairo and graphviz goes smoothly most of the
time, and options are available for omitting the 2 latter. To know more see
the description in the script header. There is a third script
\sphinxstylestrong{mac-install-source.sh} which compiles everything from source and presumes
only Python 2.7 and Xcode installed. We do not recommend this as it is time
consuming and troubleshooting requires expertise.


\subsection{Troubleshooting}
\label{\detokenize{index:troubleshooting}}\begin{itemize}
\item {} 
\sphinxcode{\sphinxupquote{no module named ...}} when you try to load a module in Python. Did
theinstallation of the module run without error? Try to run again the specific
part from the mac install shell script to see if any error comes up. Is the
path where the module has been installed in your \sphinxcode{\sphinxupquote{\$PYTHONPATH}}? Try \sphinxcode{\sphinxupquote{echo
\$PYTHONPATH}} to see the current paths. Add your local install directories if
those are not there, e.g.
\sphinxcode{\sphinxupquote{export PYTHONPATH="/Users/me/local/python2.7/site-packages:\$PYTHONPATH"}}.
If it works afterwards, don’t forget to append these export path statements to
your \sphinxcode{\sphinxupquote{\textasciitilde{}/.bash\_profile}}, so these will be set every time you launch a new
shell.

\item {} 
\sphinxcode{\sphinxupquote{pkgconfig}} not found. Check if the \sphinxcode{\sphinxupquote{\$PKG\_CONFIG\_PATH}} variable is
set correctly, and pointing on a directory where pkgconfig really can be
found.

\item {} 
Error while trying to install py(2)cairo by pip. py(2)cairo could not be
installed by pip, but only by waf. Please set the \sphinxcode{\sphinxupquote{\$PKG\_CONFIG\_PATH}} before.
See \sphinxstylestrong{mac-install-source.sh} on how to install with waf.

\item {} 
Error at pygraphviz build: \sphinxcode{\sphinxupquote{graphviz/cgraph.h file not found}}. This is
because the directory of graphviz detected wrong by pkgconfig. See
\sphinxstylestrong{mac-install-source.sh} how to set include dirs and library dirs by
\sphinxcode{\sphinxupquote{-{-}global-option}} parameters.

\item {} 
Can not install bioservices, because installation of jurko-suds fails. Ok,
this fails because pip is not able to install the recent version of
setuptools, because a very old version present in the system path. The
development version of jurko-suds does not require setuptools, so you can
install it directly from git as it is done in \sphinxstylestrong{mac-install-source.sh}.

\item {} 
In \sphinxstylestrong{Anaconda}, \sphinxstyleemphasis{pypath} can be imported, but the modules and classes are
missing. Apparently Anaconda has some built-in stuff called \sphinxstyleemphasis{pypath}. This
has nothing to do with this module. Please be aware that Anaconda installs a
completely separated Python distribution, and does not detect modules in the
main Python installation. You need to install all modules within Anaconda’s
directory. \sphinxstylestrong{mac-install-conda.sh} does exactly this. If you still
experience issues, please contact us.

\end{itemize}


\section{Microsoft Windows}
\label{\detokenize{index:microsoft-windows}}
Not many people have used \sphinxstyleemphasis{pypath} on Microsoft computers so far. Please share
your experiences and contact us if you encounter any issue. We appreciate
your feedback, and it would be nice to have better support for other computer
systems.


\subsection{With Anaconda}
\label{\detokenize{index:with-anaconda}}
The same workflow like you see in \sphinxcode{\sphinxupquote{mac-install-conda.sh}} should work for
Anaconda on Windows. The only problem you certainly will encounter is that not
all the channels have packages for all platforms. If certain channel provides
no package for Windows, or for your Python version, you just need to find an
other one. For this, do a search:

\fvset{hllines={, ,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{anaconda} \PYG{n}{search} \PYG{o}{\PYGZhy{}}\PYG{n}{t} \PYG{n}{conda} \PYG{o}{\PYGZlt{}}\PYG{n}{package} \PYG{n}{name}\PYG{o}{\PYGZgt{}}
\end{sphinxVerbatim}

For example, if you search for \sphinxstyleemphasis{pycairo}, you will find out that \sphinxstyleemphasis{vgauther}
provides it for osx-64, but only for Python 3.4, while \sphinxstyleemphasis{richlewis} provides
also for Python 3.5. And for win-64 platform, there is the channel of
\sphinxstyleemphasis{KristanAmstrong}. Go along all the commands in \sphinxcode{\sphinxupquote{mac-install-conda.sh}}, and
modify the channel if necessary, until all packages install successfully.


\subsection{With other Python distributions}
\label{\detokenize{index:with-other-python-distributions}}
Here the basic principles are the same as everywhere: first try to install all
external dependencies, after \sphinxstyleemphasis{pip} install should work. On Windows certain
packages can not be installed by compiled from source by \sphinxstyleemphasis{pip}, instead the
easiest to install them precompiled. These are in our case \sphinxstyleemphasis{fisher, lxml,
numpy (mkl version), pycairo, igraph, pygraphviz, scipy and statsmodels}. The
precompiled packages are available here:
\sphinxurl{http://www.lfd.uci.edu/~gohlke/pythonlibs/}. We tested the setup with Python
3.4.3 and Python 2.7.11. The former should just work fine, while with the
latter we have issues to be resolved.


\subsection{Known issues}
\label{\detokenize{index:known-issues}}\begin{itemize}
\item {} 
\sphinxstyleemphasis{“No module fabric available.”} \textendash{} or \sphinxstyleemphasis{pysftp} missing: this is not
important, only certain data download methods rely on these modules, but
likely you won’t call those at all.

\item {} 
Progress indicator floods terminal: sorry about that, will be fixed soon.

\item {} 
Encoding related exceptions in Python2: these might occur at some points in
the module, please send the traceback if you encounter one, and we will fix
as soon as possible.

\end{itemize}

\sphinxstyleemphasis{Special thanks to Jorge Ferreira for testing pypath on Windows!}


\chapter{Release History}
\label{\detokenize{index:release-history}}
Main improvements in the past releases:


\section{0.1.0:}
\label{\detokenize{index:id1}}\begin{itemize}
\item {} 
First release of pypath, for initial testing.

\end{itemize}


\section{0.2.0:}
\label{\detokenize{index:id2}}\begin{itemize}
\item {} 
Lots of small improvements in almost every module

\item {} 
Networks can be read from local files, remote files, lists or provided by
any function

\item {} 
Almost all redistributed data have been removed, every source downloaded
from the original provider.

\end{itemize}


\section{0.3.0:}
\label{\detokenize{index:id3}}\begin{itemize}
\item {} 
First version whith partial Python 3 support.

\end{itemize}


\section{0.4.0:}
\label{\detokenize{index:id4}}\begin{itemize}
\item {} 
\sphinxstylestrong{pyreact} module with \sphinxstylestrong{BioPaxReader} and \sphinxstylestrong{PyReact} classes added

\item {} 
Process description databases, BioPax and PathwayCommons SIF conversion
rules are supported

\item {} 
Format definitions for 6 process description databases included.

\end{itemize}


\section{0.5.0:}
\label{\detokenize{index:id5}}\begin{itemize}
\item {} 
Many classes have been added to the \sphinxstylestrong{plot} module

\item {} 
All figures and tables in the manuscript can be generated automatically

\item {} 
This is supported by a new module, \sphinxstylestrong{analysis}, which implements a generic

\end{itemize}

workflow in its \sphinxstylestrong{Workflow} class.


\section{0.7.74:}
\label{\detokenize{index:id6}}\begin{itemize}
\item {} 
\sphinxstylestrong{homology} module: finds the homologs of proteins using the NCBI

\end{itemize}

Homologene database and the homologs of PTM sites using UniProt sequences
and PhosphoSitePlus homology table
* \sphinxstylestrong{ptm} module: fully integrated way of processing enzyme-substrate
interactions from many databases and their translation by homology to other
species
* \sphinxstylestrong{export} module: creates \sphinxcode{\sphinxupquote{pandas.DataFrame}} or exports the network into
tabular file
* New webservice
* TF Regulons database included and provides much more comprehensive
transcriptional regulation resources, including literature curated, in silico
predicted, ChIP-Seq and expression pattern based approaches
* Many network resources added, including miRNA-mRNA and TF-miRNA interactions


\section{Upcoming:}
\label{\detokenize{index:upcoming}}\begin{itemize}
\item {} 
New, more flexible network reader class

\item {} 
Full support for multi-species molecular interaction networks

\end{itemize}

(e.g. pathogene-host)
* Better support for not protein only molecular interaction networks
(metabolites, drug compounds, RNA)
* ChEMBL webservice interface, interface for PubChem and eventually
forDrugBank
* Silent mode: a way to suppress messages and progress bars


\chapter{Features}
\label{\detokenize{index:features}}
The primary aim of \sphinxstylestrong{pypath} is to build up networks from multiple sources on
one igraph object. \sphinxstylestrong{pypath} handles ambiguous ID conversion, reads custom
edge and node attributes from text files and \sphinxstylestrong{MySQL}.

Submodules perform various features, e.g. graph visualization, working with
rug compound data, searching drug targets and compounds in \sphinxstylestrong{ChEMBL}.


\section{ID conversion}
\label{\detokenize{index:id-conversion}}
The ID conversion module \sphinxcode{\sphinxupquote{mapping}} can be used independently. It has the
feature to translate secondary UniProt IDs to primaries, and Trembl IDs to
SwissProt, using primary Gene Symbols to find the connections. This module
automatically loads and stores the necessary conversion tables. Many tables
are predefined, such as all the IDs in \sphinxstylestrong{UniProt mapping service,} while
users are able to load any table from \sphinxstylestrong{file} or \sphinxstylestrong{MySQL,} using the classes
provided in the module \sphinxcode{\sphinxupquote{input\_formats}}.


\section{Pathways}
\label{\detokenize{index:pathways}}
\sphinxstylestrong{pypath} includes data and predefined format descriptions for more than 25
high quality, literature curated databases. The inut formats are defined in
the \sphinxcode{\sphinxupquote{data\_formats}} module. For some resources data downloaded on the fly,
where it is not possible, data is redistributed with the module. Descriptions
and comprehensive information about the resources is available in the
\sphinxcode{\sphinxupquote{descriptions}} module.


\section{Structural features}
\label{\detokenize{index:structural-features}}
One of the modules called \sphinxcode{\sphinxupquote{intera}} provides many classes for representing
structures and mechanisms behind protein interactions. These are \sphinxcode{\sphinxupquote{Residue}}
(optionally mutated), \sphinxcode{\sphinxupquote{Motif}}, \sphinxcode{\sphinxupquote{Ptm}}, \sphinxcode{\sphinxupquote{Domain}}, \sphinxcode{\sphinxupquote{DomainMotif}},
\sphinxcode{\sphinxupquote{DomainDomain}} and \sphinxcode{\sphinxupquote{Interface}}. All these classes have \sphinxcode{\sphinxupquote{\_\_eq\_\_()}}
methods to test equality between instances, and also \sphinxcode{\sphinxupquote{\_\_contains\_\_()}}
methods to look up easily if a residue is within a short motif or protein
domain, or is the target residue of a PTM.


\section{Sequences}
\label{\detokenize{index:sequences}}
The module \sphinxcode{\sphinxupquote{seq}} contains a simple class for quick lookup any residue or
segment in \sphinxstylestrong{UniProt} protein sequences while being aware of isoforms.


\section{Tissue expression}
\label{\detokenize{index:tissue-expression}}
For 3 protein expression databases there are functions and modules for
downloading and combining the expression data with the network. These are the
Human Protein Atlas, the ProteomicsDB and GIANT. The \sphinxcode{\sphinxupquote{giant}} and
\sphinxcode{\sphinxupquote{proteomicsdb}} modules can be used also as stand alone Python clients for
these resources.


\section{Functional annotations}
\label{\detokenize{index:functional-annotations}}
\sphinxstylestrong{GSEA} and \sphinxstylestrong{Gene Ontology} are two approaches for annotating genes and
gene products, and enrichment analysis technics aims to use these annotations
to highlight the biological functions a given set of genes is related to. Here
the \sphinxcode{\sphinxupquote{enrich}} module gives abstract classes to calculate enrichment
statistics, while the \sphinxcode{\sphinxupquote{go}} and the \sphinxcode{\sphinxupquote{gsea}} modules give access to GO and
GSEA data, and make it easy to count enrichment statistics for sets of genes.


\section{Drug compounds}
\label{\detokenize{index:drug-compounds}}
\sphinxstylestrong{UniChem} submodule provides an interface to effectively query the UniChem
service, use connectivity search with custom settings, and translate SMILEs to
ChEMBL IDs with ChEMBL web service.

\sphinxstylestrong{ChEMBL} submodule queries directly your own ChEMBL MySQL instance, has the
features to search targets and compounds from custom assay types and
relationship types, to get activity values, binding domains, and action types.
You need to download the ChEMBL MySQL dump, and load into your own server.


\section{Technical}
\label{\detokenize{index:technical}}
\sphinxstylestrong{MySQL} submodule helps to manage MySQL connections and track queries. It is
able to run queries parallely to optimize CPU and memory usage on the server,
handling queues, and serve the result by server side or client side storage.
The \sphinxcode{\sphinxupquote{chembl}} and potentially the \sphinxcode{\sphinxupquote{mapping}} modules rely on this \sphinxcode{\sphinxupquote{mysql}}
module.

The most important function in module \sphinxcode{\sphinxupquote{dataio}} is a very flexible \sphinxstylestrong{download
manager} built around \sphinxcode{\sphinxupquote{curl}}. The function \sphinxcode{\sphinxupquote{dataio.curl()}} accepts
numerous arguments, tries to deal in a smart way with local \sphinxstylestrong{cache,}
authentication, redirects, uncompression, character encodings, FTP and HTTP
transactions, and many other stuff. Cache can grow to several GBs, and takes
place in \sphinxcode{\sphinxupquote{./cache}} by default. Please be aware of this, and use for example
symlinks in case of using multiple working directories.

A simple \sphinxstylestrong{webservice} comes with this module: the \sphinxcode{\sphinxupquote{server}} module based on
\sphinxcode{\sphinxupquote{twisted.web.server}} opens a custom port and serves plain text tables over
HTTP with REST style querying.


\renewcommand{\indexname}{Python Module Index}
\begin{sphinxtheindex}
\def\bigletter#1{{\Large\sffamily#1}\nopagebreak\vspace{1mm}}
\bigletter{p}
\item {\sphinxstyleindexentry{pypath.main}}\sphinxstyleindexpageref{main:\detokenize{module-pypath.main}}
\end{sphinxtheindex}

\renewcommand{\indexname}{Index}
\printindex
\end{document}