<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Welcome to pypath’s documentation! &mdash; pypath 0.1.1 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="pypath 0.1.1 documentation" href="#" />
    <link rel="next" title="Welcome to pypath’s documentation!" href="" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="#" title="Welcome to pypath’s documentation!"
             accesskey="N">next</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">pypath 0.1.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="welcome-to-pypath-s-documentation">
<h1>Welcome to pypath&#8217;s documentation!<a class="headerlink" href="#welcome-to-pypath-s-documentation" title="Permalink to this headline">¶</a></h1>
<p>Contents:</p>
<div class="toctree-wrapper compound">
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="">Welcome to pypath&#8217;s documentation!</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#examples">Examples</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#example-1-building-the-network">Example 1: building the network</a></li>
<li class="toctree-l2"><a class="reference internal" href="#example-2-using-the-mapper-class-for-translating-ids">Example 2: using the Mapper class for translating IDs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#module-pypath.pypath">Reference</a></li>
</ul>
</div>
</div>
<div class="section" id="examples">
<h1>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h1>
<div class="section" id="example-1-building-the-network">
<h2>Example 1: building the network<a class="headerlink" href="#example-1-building-the-network" title="Permalink to this headline">¶</a></h2>
<p>Import the module and create an instance. MySQL settings are stored by default in mysql_config/defaults.mysql, now you don&#8217;t need to modify it:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">pypath</span>
<span class="n">mysql_gelati</span> <span class="o">=</span> <span class="p">(</span><span class="bp">None</span><span class="p">,</span> <span class="s">&#39;mapping_gelati&#39;</span><span class="p">)</span>
<span class="n">mysql_chembl</span> <span class="o">=</span> <span class="p">(</span><span class="bp">None</span><span class="p">,</span> <span class="s">&#39;chembl_ebi&#39;</span><span class="p">)</span>
<span class="n">net</span> <span class="o">=</span> <span class="n">pypath</span><span class="o">.</span><span class="n">BioGraph</span><span class="p">(</span><span class="mi">9606</span><span class="p">,</span> <span class="n">mysql</span> <span class="o">=</span> <span class="n">mysql_gelati</span><span class="p">)</span>
</pre></div>
</div>
<p>The tables for ID conversions, and the network of the selected manually curated pathway databases can be initialized this way:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">net</span><span class="o">.</span><span class="n">init_network</span><span class="p">()</span>
</pre></div>
</div>
<p>Resources without references for each interaction are available separately. For example, to build a comprehensive phosphonetwork, you might want to load interactions from high-throughput data in PhosphoSite, kinase-substrate relationships from PhosphoNetworks and MIMP. This is the way to do it:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pypath.data_formats</span> <span class="kn">import</span> <span class="n">good</span>
<span class="n">net</span><span class="o">.</span><span class="n">load_resources</span><span class="p">(</span><span class="n">lst</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;mimp&#39;</span><span class="p">:</span> <span class="n">good</span><span class="p">[</span><span class="s">&#39;mimp&#39;</span><span class="p">]})</span>
<span class="n">net</span><span class="o">.</span><span class="n">load_resources</span><span class="p">(</span><span class="n">lst</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;pnetworks&#39;</span><span class="p">:</span> <span class="n">good</span><span class="p">[</span><span class="s">&#39;pnetworks&#39;</span><span class="p">]})</span>
<span class="n">net</span><span class="o">.</span><span class="n">load_resources</span><span class="p">(</span><span class="n">lst</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;psite_noref&#39;</span><span class="p">:</span> <span class="n">good</span><span class="p">[</span><span class="s">&#39;psite_noref&#39;</span><span class="p">]})</span>
</pre></div>
</div>
<p>In case you have your own interaction data, and you wish to merge this into the network, you first need to define its format, and then call the load_resource() function. In this simple example, the proteins are noted by Gene Symbols, their names are in columns #0 and #1, while from column #2 an edge attribute named &#8216;score&#8217; will be read. The file is tab separated, and the file name is &#8216;mylist.sif&#8217;. (This won&#8217;t work unless you really have a file like this.)</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">mylist</span> <span class="o">=</span> <span class="n">pypath</span><span class="o">.</span><span class="n">input_formats</span><span class="o">.</span><span class="n">ReadSettings</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="s">&quot;mylist&quot;</span><span class="p">,</span>
    <span class="n">separator</span> <span class="o">=</span> <span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">,</span> <span class="n">nameColA</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">nameColB</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
    <span class="n">nameTypeA</span> <span class="o">=</span> <span class="s">&quot;genesymbol&quot;</span><span class="p">,</span> <span class="n">nameTypeB</span> <span class="o">=</span> <span class="s">&quot;genesymbol&quot;</span><span class="p">,</span>
    <span class="n">isDirected</span> <span class="o">=</span> <span class="bp">False</span><span class="p">,</span> <span class="n">inFile</span> <span class="o">=</span> <span class="s">&#39;mylist.sif&#39;</span><span class="p">,</span>
    <span class="n">extraEdgeAttrs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s">&#39;score&#39;</span><span class="p">:</span> <span class="mi">2</span>
    <span class="p">})</span>
<span class="n">net</span><span class="o">.</span><span class="n">load_resource</span><span class="p">(</span><span class="n">mylist</span><span class="p">)</span>
</pre></div>
</div>
<p>Although proteins are identified by UniProt IDs, it is useful to have standard gene symbols as labels, to make things more human understandable.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">net</span><span class="o">.</span><span class="n">genesymbol_labels</span><span class="p">()</span>
</pre></div>
</div>
<p>Now the igraph object is in net.graph, vertex and edge attributes are accessible the usual way:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">net</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">vs</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&#39;name&#39;</span><span class="p">]</span>
<span class="n">net</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">vs</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&#39;label&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>Let&#8217;s see the default attributes of edges:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">net</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&#39;sources&#39;</span><span class="p">]</span>
<span class="n">net</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&#39;references&#39;</span><span class="p">]</span>
<span class="n">net</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&#39;dirs&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>The graph is undirected by default, and &#8216;Direction&#8217; objects can be found in &lt;edge&gt;[&#8216;dirs&#8217;] to describe the directions and signs of the edge:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="n">net</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">111</span><span class="p">][</span><span class="s">&#39;dirs&#39;</span><span class="p">]</span>
<span class="n">net</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">111</span><span class="p">]</span><span class="o">.</span><span class="n">is_directed</span><span class="p">()</span>
<span class="n">net</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">111</span><span class="p">][</span><span class="s">&#39;dirs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">is_inhibition</span><span class="p">()</span>
<span class="n">net</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">111</span><span class="p">][</span><span class="s">&#39;dirs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">which_dirs</span><span class="p">()</span>
<span class="n">net</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">111</span><span class="p">][</span><span class="s">&#39;dirs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">get_dir</span><span class="p">(</span><span class="n">net</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">111</span><span class="p">][</span><span class="s">&#39;dirs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">which_dirs</span><span class="p">()[</span><span class="mi">0</span><span class="p">])</span>
</pre></div>
</div>
<p>To convert the igraph object to a directed graph, use this function:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">net</span><span class="o">.</span><span class="n">get_directed</span><span class="p">()</span>
<span class="n">net</span><span class="o">.</span><span class="n">dgraph</span><span class="o">.</span><span class="n">ecount</span><span class="p">()</span>
<span class="n">net</span><span class="o">.</span><span class="n">dgraph</span><span class="o">.</span><span class="n">vcount</span><span class="p">()</span>
</pre></div>
</div>
<p>To replace the graph with the giant component of the directed graph:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">net</span><span class="o">.</span><span class="n">graph</span> <span class="o">=</span> <span class="n">net</span><span class="o">.</span><span class="n">get_giant</span><span class="p">(</span><span class="n">graph</span> <span class="o">=</span> <span class="n">net</span><span class="o">.</span><span class="n">dgraph</span><span class="p">)</span>
</pre></div>
</div>
<p>Get compounds for all the proteins in the network:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">net</span><span class="o">.</span><span class="n">compounds_from_chembl</span><span class="p">((</span><span class="bp">None</span><span class="p">,</span> <span class="s">&#39;chembl_ebi&#39;</span><span class="p">),</span> <span class="n">pchembl</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<p>This works only from EBI, as the ChEMBL MySQL instance running on an EBI virtual machine. To use it from home, start an SSH tunnel to EBI: ssh -o proxycommand=&#8221;ssh -p 2222 gate.ebi.ac.uk proxy %h&#8221; login.ebi.ac.uk -L 3306:172.22.68.151:3306
And while the tunnel is open, run the command above with mysql settings pointing to localhost:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">net</span><span class="o">.</span><span class="n">compounds_from_chembl</span><span class="p">((</span><span class="bp">None</span><span class="p">,</span> <span class="s">&#39;chembl&#39;</span><span class="p">),</span> <span class="n">pchembl</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<p>You can find the list of compounds and the requested data (now only the pchembl values) in vertex attributes:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">net</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">vs</span><span class="p">[</span><span class="mi">2</span><span class="p">][</span><span class="s">&#39;compounds_chembl&#39;</span><span class="p">]</span>
<span class="n">net</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">vs</span><span class="p">[</span><span class="mi">2</span><span class="p">][</span><span class="s">&#39;compounds_data&#39;</span><span class="p">]</span>
<span class="n">net</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">vs</span><span class="p">[</span><span class="mi">2</span><span class="p">][</span><span class="s">&#39;compounds_data&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s">&#39;pchembl&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>Loading PTMs. This function loads PTMs from all the available resources. It takes long time first, but once it saves the downloaded data under ./cache directory, it is much faster to run again.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">net</span><span class="o">.</span><span class="n">load_ptms</span><span class="p">()</span>
</pre></div>
</div>
<p>Individual PTM resources can be loaded the following way:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">net</span><span class="o">.</span><span class="n">load_phospho_dmi</span><span class="p">(</span><span class="n">source</span> <span class="o">=</span> <span class="s">&#39;phosphoELM&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>In the latter case, this function is needed to merge the identical PTMs loaded from multiple resources:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">net</span><span class="o">.</span><span class="n">uniq_ptms</span><span class="p">()</span>
</pre></div>
</div>
<p>PTMs are stored in objects. This example shows how to access the type of modification, and the number and name of the modified residue.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">net</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">70</span><span class="p">][</span><span class="s">&#39;ptm&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">ptm</span><span class="o">.</span><span class="n">typ</span>
<span class="n">net</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">70</span><span class="p">][</span><span class="s">&#39;ptm&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">ptm</span><span class="o">.</span><span class="n">protein</span>
<span class="n">net</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">70</span><span class="p">][</span><span class="s">&#39;ptm&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">ptm</span><span class="o">.</span><span class="n">residue</span><span class="o">.</span><span class="n">name</span>
<span class="n">net</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">70</span><span class="p">][</span><span class="s">&#39;ptm&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">ptm</span><span class="o">.</span><span class="n">residue</span><span class="o">.</span><span class="n">number</span>
</pre></div>
</div>
</div>
<div class="section" id="example-2-using-the-mapper-class-for-translating-ids">
<h2>Example 2: using the Mapper class for translating IDs<a class="headerlink" href="#example-2-using-the-mapper-class-for-translating-ids" title="Permalink to this headline">¶</a></h2>
<p>The mapping submodule of pypath can be used for ID conversion. Here is a basic example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pypath</span> <span class="kn">import</span> <span class="n">mapping</span>
<span class="n">m</span> <span class="o">=</span> <span class="n">mapping</span><span class="o">.</span><span class="n">Mapper</span><span class="p">(</span><span class="mi">9606</span><span class="p">,</span> <span class="n">mysql_conf</span> <span class="o">=</span> <span class="p">(</span><span class="bp">None</span><span class="p">,</span> <span class="s">&#39;mapping_gelati&#39;</span><span class="p">))</span>
<span class="n">m</span><span class="o">.</span><span class="n">load_mappings</span><span class="p">()</span>
<span class="n">result</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">gene_list</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;EGFR&#39;</span><span class="p">,</span> <span class="s">&#39;AKT1&#39;</span><span class="p">,</span> <span class="s">&#39;GABARAPL1&#39;</span><span class="p">,</span> <span class="s">&#39;TP53&#39;</span><span class="p">]</span>
<span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">gene_list</span><span class="p">:</span>
    <span class="n">result</span><span class="p">[</span><span class="n">g</span><span class="p">]</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">map_name</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="s">&#39;genesymbol&#39;</span><span class="p">,</span> <span class="s">&#39;uniprot&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="module-pypath.pypath">
<span id="reference"></span><h1>Reference<a class="headerlink" href="#module-pypath.pypath" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="pypath.pypath.BioGraph">
<em class="property">class </em><code class="descclassname">pypath.pypath.</code><code class="descname">BioGraph</code><span class="sig-paren">(</span><em>ncbi_tax_id=9606</em>, <em>default_name_type={'protein': 'uniprot'</em>, <em>'drug': 'chembl'</em>, <em>'mirna': 'mirbase'}</em>, <em>copy=None</em>, <em>mysql=(None</em>, <em>'mapping')</em>, <em>name='unnamed'</em>, <em>outdir='results'</em>, <em>loglevel='INFO'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.pypath.BioGraph.add_update_vertex">
<code class="descname">add_update_vertex</code><span class="sig-paren">(</span><em>defAttrs</em>, <em>originalName</em>, <em>originalNameType</em>, <em>extraAttrs={}</em>, <em>add=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.add_update_vertex"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.add_update_vertex" title="Permalink to this definition">¶</a></dt>
<dd><p>Updates the attributes of one node in the network.
Optionally it creates a new node and sets the attributes,
but it is not efficient as igraph needs to reindex vertices 
after this operation, so better to create new nodes and 
edges in batch.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.all_between">
<code class="descname">all_between</code><span class="sig-paren">(</span><em>nameA</em>, <em>nameB</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.all_between"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.all_between" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all edges between two given vertex names. Similar to 
straight_between(), but checks both directions, and returns 
list of edge ids in [undirected, straight, reveresed] format,
for both nameA -&gt; nameB and nameB -&gt; nameA edges.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.basic_stats">
<code class="descname">basic_stats</code><span class="sig-paren">(</span><em>latex=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.basic_stats"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.basic_stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns basic numbers about the network resources, e.g. edge and 
node counts.</p>
<dl class="docutils">
<dt>latex</dt>
<dd>Return table in a LaTeX document. This can be compiled by
PDFLaTeX: 
latex stats.tex</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.copy">
<code class="descname">copy</code><span class="sig-paren">(</span><em>other</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.copy"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.copy" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.count_sol">
<code class="descname">count_sol</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.count_sol"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.count_sol" title="Permalink to this definition">¶</a></dt>
<dd><p>Counts nodes with zero degree.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.delete_by_taxon">
<code class="descname">delete_by_taxon</code><span class="sig-paren">(</span><em>tax</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.delete_by_taxon"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.delete_by_taxon" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes the proteins of all organisms which are not listed.</p>
<dl class="docutils">
<dt>&#64;tax <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd>List of NCBI Taxonomy IDs of the organisms of interest.
E.g. [7227, 9606]</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.delete_unknown">
<code class="descname">delete_unknown</code><span class="sig-paren">(</span><em>tax</em>, <em>typ='protein'</em>, <em>defaultNameType=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.delete_unknown"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.delete_unknown" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes those proteins which are not in the list of all default
IDs of the organisms. By default, it means to remove all protein 
nodes not having a human SwissProt ID.</p>
<dl class="docutils">
<dt>&#64;tax <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd>List of NCBI Taxonomy IDs of the organisms of interest.
E.g. [7227, 9606]</dd>
<dt>&#64;typ <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>Molecule type. E.g. &#8216;protein&#8217; or &#8216;mirna&#8217;</dd>
<dt>&#64;defaultNameType <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>The default name type of the given molecular species.
For proteins it&#8217;s &#8216;uniprot&#8217; by default.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.edge_exists">
<code class="descname">edge_exists</code><span class="sig-paren">(</span><em>nameA</em>, <em>nameB</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.edge_exists"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.edge_exists" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a tuple of vertice indices if edge doesn&#8217;t exists, 
otherwise edge id. Not sensitive to direction.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.edges_expression">
<code class="descname">edges_expression</code><span class="sig-paren">(</span><em>func=&lt;function &lt;lambda&gt;&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.edges_expression"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.edges_expression" title="Permalink to this definition">¶</a></dt>
<dd><p>Executes function <cite>func</cite> for each pairs of connected proteins in the 
network, for every expression dataset. By default, <cite>func</cite> simply
gives the product the (normalized) expression values.</p>
<dl class="docutils">
<dt>func <span class="classifier-delimiter">:</span> <span class="classifier">callable</span></dt>
<dd>Function to handle 2 vectors (pandas.Series() objects), should
return one vector of the same length.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.find_all_paths">
<code class="descname">find_all_paths</code><span class="sig-paren">(</span><em>start</em>, <em>end</em>, <em>mode='OUT'</em>, <em>maxlen=None</em>, <em>graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.find_all_paths"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.find_all_paths" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds all paths up to length <cite>maxlen</cite> between groups of
vertices. This function is needed only becaues igraph`s
get_all_shortest_paths() finds only the shortest, not any
path up to a defined length.</p>
<dl class="docutils">
<dt>&#64;start <span class="classifier-delimiter">:</span> <span class="classifier">int or list</span></dt>
<dd>Indices of the starting node(s) of the paths.</dd>
<dt>&#64;end <span class="classifier-delimiter">:</span> <span class="classifier">int or list</span></dt>
<dd>Indices of the target node(s) of the paths.</dd>
<dt>&#64;mode <span class="classifier-delimiter">:</span> <span class="classifier">&#8216;IN&#8217;, &#8216;OUT&#8217;, &#8216;ALL&#8217;</span></dt>
<dd>Passed to igraph.Graph.neighbors()</dd>
<dt>&#64;maxlen <span class="classifier-delimiter">:</span> <span class="classifier">int</span></dt>
<dd>Maximum length of paths in steps, i.e. if maxlen = 3, then 
the longest path may consist of 3 edges and 4 nodes.</dd>
<dt>&#64;graph <span class="classifier-delimiter">:</span> <span class="classifier">igraph.Graph object</span></dt>
<dd>The graph you want to find paths in. self.graph by default.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.get_edge">
<code class="descname">get_edge</code><span class="sig-paren">(</span><em>nodes</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.get_edge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.get_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the edge id only if there is an edge from nodes[0] to nodes[1],
returns False if edge exists in opposite direction, or no edge exists 
between the two vertices, or any of the vertice ids doesn&#8217;t exist.
To find edges without regarding their direction, see edge_exists().</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.get_giant">
<code class="descname">get_giant</code><span class="sig-paren">(</span><em>replace=False</em>, <em>graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.get_giant"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.get_giant" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the giant component of the graph, or 
replaces the igraph object with only the giant 
component.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.init_network">
<code class="descname">init_network</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.init_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.init_network" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a lazy way to start the module, load data 
and build the high confidence, literature curated
part of the signaling network.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.load_ddi">
<code class="descname">load_ddi</code><span class="sig-paren">(</span><em>ddi</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.load_ddi"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.load_ddi" title="Permalink to this definition">¶</a></dt>
<dd><p>ddi is either a list of intera.DomainDomain objects, 
or a function resulting this list</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.load_dmi">
<code class="descname">load_dmi</code><span class="sig-paren">(</span><em>dmi</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.load_dmi"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.load_dmi" title="Permalink to this definition">¶</a></dt>
<dd><p>dmi is either a list of intera.DomainMotif objects, 
or a function resulting this list</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.load_expression">
<code class="descname">load_expression</code><span class="sig-paren">(</span><em>array=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.load_expression"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.load_expression" title="Permalink to this definition">¶</a></dt>
<dd><p>Expression data can be loaded into vertex attributes, 
or into a pandas DataFrame – the latter offers faster 
ways to process and use these huge matrices.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.load_mutations">
<code class="descname">load_mutations</code><span class="sig-paren">(</span><em>attributes=None</em>, <em>gdsc_datadir=None</em>, <em>mutation_file=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.load_mutations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.load_mutations" title="Permalink to this definition">¶</a></dt>
<dd><p>Mutations are listed in vertex attributes. Mutation() objects 
offers methods to identify residues and look up in Ptm(), Motif() 
and Domain() objects, to check if those residues are 
modified, or are in some short motif or domain.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.load_resources">
<code class="descname">load_resources</code><span class="sig-paren">(</span><em>lst={'netpath': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad170&gt;</em>, <em>'domino': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad758&gt;</em>, <em>'alz': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad4d0&gt;</em>, <em>'intact': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad5a8&gt;</em>, <em>'arn': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad320&gt;</em>, <em>'death': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad5f0&gt;</em>, <em>'elm': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad710&gt;</em>, <em>'ca1': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad200&gt;</em>, <em>'depod': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad518&gt;</em>, <em>'ccmap2': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f2409df80&gt;</em>, <em>'spike': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad050&gt;</em>, <em>'nci_pid': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad290&gt;</em>, <em>'dbptm': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad7a0&gt;</em>, <em>'phelm': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad6c8&gt;</em>, <em>'innatedb': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad488&gt;</em>, <em>'matrixdb': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad560&gt;</em>, <em>'mppi': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad098&gt;</em>, <em>'nrf2': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad3b0&gt;</em>, <em>'signor': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad638&gt;</em>, <em>'macrophage': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad440&gt;</em>, <em>'pdz': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad680&gt;</em>, <em>'dip': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad128&gt;</em>, <em>'slk': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f2409dab8&gt;</em>, <em>'psite': &lt;pypath.input_formats.ReadSettings instance at 0x7f0f240ad0e0&gt;}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.load_resources"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.load_resources" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads multiple resources, and cleans up after. 
Looks up ID types, and loads all ID conversion 
tables from UniProt if necessary. This is much
faster than loading the ID conversion and the 
resources one by one.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.map_list">
<code class="descname">map_list</code><span class="sig-paren">(</span><em>lst</em>, <em>singleList=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.map_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.map_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Only a wrapper for map_edge()</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.map_network">
<code class="descname">map_network</code><span class="sig-paren">(</span><em>mapping=('trembl'</em>, <em>'swissprot')</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.map_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.map_network" title="Permalink to this definition">¶</a></dt>
<dd><p>This function is not needed any more, will be removed soon.</p>
<p>In the whole network changes every uniprot id to primary.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.mutated_edges">
<code class="descname">mutated_edges</code><span class="sig-paren">(</span><em>sample</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.mutated_edges"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.mutated_edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Compares the mutated residues and the modified residues in PTMs.
Interactions are marked as mutated if the target residue in the 
underlying PTM is mutated.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.network_filter">
<code class="descname">network_filter</code><span class="sig-paren">(</span><em>p=2.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.network_filter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.network_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>This function aims to cut the number of edges in the network, 
without loosing nodes, to make the network less connected, 
less hairball-like, more usable for analysis.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.network_stats">
<code class="descname">network_stats</code><span class="sig-paren">(</span><em>outfile=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.network_stats"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.network_stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates basic statistics for the whole network 
and each of sources. Writes the results in a tab file.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.read_data_file">
<code class="descname">read_data_file</code><span class="sig-paren">(</span><em>settings</em>, <em>keep_raw=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.read_data_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.read_data_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Interaction data with node and edge attributes can be read 
from simple text based files. This function works not only
with files, but with lists as well. Any other function can 
be written to download a preprocess data, and then give it
to this function to finally attach to the network.</p>
<dl class="docutils">
<dt>&#64;settings <span class="classifier-delimiter">:</span> <span class="classifier">ReadSettings instance</span></dt>
<dd>The detailed definition of the input format. Instead of 
the file name you can give a function name, which will 
be executed, and the returned data will be used.</dd>
<dt>&#64;keep_raw <span class="classifier-delimiter">:</span> <span class="classifier">boolean</span></dt>
<dd>To keep the raw data read by this function, in order for
debugging purposes, or further use.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.shortest_path_dist">
<code class="descname">shortest_path_dist</code><span class="sig-paren">(</span><em>graph=None</em>, <em>subset=None</em>, <em>outfile=None</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.shortest_path_dist"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.shortest_path_dist" title="Permalink to this definition">¶</a></dt>
<dd><p>subset is a tuple of two lists if you wish to look for
paths between elements of two groups, or a list if you 
wish to look for shortest paths within this group</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.source_network">
<code class="descname">source_network</code><span class="sig-paren">(</span><em>font='HelveticaNeueLTStd'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.source_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.source_network" title="Permalink to this definition">¶</a></dt>
<dd><p>For EMBL branding, use Helvetica Neue Linotype Standard light</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.straight_between">
<code class="descname">straight_between</code><span class="sig-paren">(</span><em>nameA</em>, <em>nameB</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.straight_between"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.straight_between" title="Permalink to this definition">¶</a></dt>
<dd><p>This does actually the same as get_edge(), but by names
instead of vertex ids.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.update_vindex">
<code class="descname">update_vindex</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.update_vindex"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.update_vindex" title="Permalink to this definition">¶</a></dt>
<dd><p>See update_vname().</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.update_vname">
<code class="descname">update_vname</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.update_vname"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.update_vname" title="Permalink to this definition">¶</a></dt>
<dd><p>For fast lookup of node names and indexes, these are 
hold in a list and a dict as well. However, every time
new nodes are added, these should be updated. This 
function is automatically called after all operations
affecting node indices.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.BioGraph.vertex_pathways">
<code class="descname">vertex_pathways</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#BioGraph.vertex_pathways"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.BioGraph.vertex_pathways" title="Permalink to this definition">¶</a></dt>
<dd><p>Some resources assignes interactions some others 
proteins to pathways.
This function converts pathway annotations from 
edge attributes to vertex attributes.</p>
</dd></dl>

</dd></dl>

<span class="target" id="module-pypath.intera"></span><dl class="class">
<dt id="pypath.intera.Residue">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">Residue</code><span class="sig-paren">(</span><em>number</em>, <em>name</em>, <em>identifier</em>, <em>id_type='uniprot'</em>, <em>isoform=1</em>, <em>mutated=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Residue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Residue" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.intera.Mutation">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">Mutation</code><span class="sig-paren">(</span><em>original</em>, <em>mutated</em>, <em>sample</em>, <em>properties={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Mutation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Mutation" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.intera.Motif">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">Motif</code><span class="sig-paren">(</span><em>protein</em>, <em>start</em>, <em>end</em>, <em>id_type='uniprot'</em>, <em>regex=None</em>, <em>instance=None</em>, <em>isoform=1</em>, <em>motif_name=None</em>, <em>prob=None</em>, <em>elm=None</em>, <em>description=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Motif"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Motif" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.intera.Ptm">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">Ptm</code><span class="sig-paren">(</span><em>protein</em>, <em>id_type='uniprot'</em>, <em>typ='unknown'</em>, <em>motif=None</em>, <em>residue=None</em>, <em>source=None</em>, <em>isoform=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Ptm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Ptm" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.intera.Domain">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">Domain</code><span class="sig-paren">(</span><em>protein</em>, <em>id_type='uniprot'</em>, <em>domain=None</em>, <em>domain_id_type='pfam'</em>, <em>start=None</em>, <em>end=None</em>, <em>isoform=1</em>, <em>chains={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Domain"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Domain" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.intera.DomainMotif">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">DomainMotif</code><span class="sig-paren">(</span><em>domain</em>, <em>ptm</em>, <em>sources=None</em>, <em>refs=None</em>, <em>pdbs=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#DomainMotif"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.DomainMotif" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.intera.DomainDomain">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">DomainDomain</code><span class="sig-paren">(</span><em>domain_a</em>, <em>domain_b</em>, <em>pdbs=None</em>, <em>sources=None</em>, <em>refs=None</em>, <em>contact_residues=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#DomainDomain"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.DomainDomain" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.intera.Interface">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">Interface</code><span class="sig-paren">(</span><em>id_a</em>, <em>id_b</em>, <em>source</em>, <em>id_type='uniprot'</em>, <em>pdb=None</em>, <em>css=None</em>, <em>stab_en=None</em>, <em>solv_en=None</em>, <em>area=None</em>, <em>isoform_a=1</em>, <em>isoform_b=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Interface"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Interface" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.intera.Interface.add_residues">
<code class="descname">add_residues</code><span class="sig-paren">(</span><em>res_a</em>, <em>res_b</em>, <em>typ='undefined'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Interface.add_residues"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Interface.add_residues" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds one pair of residues of type <a href="#id1"><span class="problematic" id="id2">`</span></a>typ&#8217;, 
where <a href="#id3"><span class="problematic" id="id4">`</span></a>res_a&#8217; and <a href="#id5"><span class="problematic" id="id6">`</span></a>res_b&#8217; are tuples of 
residue number in sequence and residue type,
e.g. (124, &#8216;S&#8217;) &#8211; (means Serine #124)
<a href="#id7"><span class="problematic" id="id8">`</span></a>typ&#8217; can be undefined, hbonds, sbridges, ssbonds or covbonds</p>
</dd></dl>

<dl class="method">
<dt id="pypath.intera.Interface.bond_types">
<code class="descname">bond_types</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Interface.bond_types"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Interface.bond_types" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the bond types present in this interface</p>
</dd></dl>

<dl class="method">
<dt id="pypath.intera.Interface.get_bonds">
<code class="descname">get_bonds</code><span class="sig-paren">(</span><em>typ=None</em>, <em>mode=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Interface.get_bonds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Interface.get_bonds" title="Permalink to this definition">¶</a></dt>
<dd><p>Gives a generator to iterate throught bonds in 
this interface. If no type given, bonds of all types 
returned.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.intera.Interface.numof_residues">
<code class="descname">numof_residues</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Interface.numof_residues"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Interface.numof_residues" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of residue pairs by bound type</p>
</dd></dl>

</dd></dl>

<span class="target" id="module-pypath.mapping"></span><dl class="class">
<dt id="pypath.mapping.MappingTable">
<em class="property">class </em><code class="descclassname">pypath.mapping.</code><code class="descname">MappingTable</code><span class="sig-paren">(</span><em>one</em>, <em>two</em>, <em>typ</em>, <em>source</em>, <em>param</em>, <em>ncbi_tax_id</em>, <em>mysql=None</em>, <em>log=None</em>, <em>cache=True</em>, <em>cachedir='cache'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#MappingTable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.MappingTable" title="Permalink to this definition">¶</a></dt>
<dd><p>To initialize ID conversion tables for the first time
data is downloaded from UniProt and read to dictionaries.
It takes a couple of seconds. Data is saved to pickle 
dumps, this way after tables load much faster.</p>
<dl class="method">
<dt id="pypath.mapping.MappingTable.read_mapping_uniprot">
<code class="descname">read_mapping_uniprot</code><span class="sig-paren">(</span><em>param</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#MappingTable.read_mapping_uniprot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.MappingTable.read_mapping_uniprot" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads ID mappings directly from UniProt.
See the names of possible identifiers here:
<a class="reference external" href="http://www.uniprot.org/help/programmatic_access">http://www.uniprot.org/help/programmatic_access</a></p>
<p>&#64;param : UniprotMapping instance</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.mapping.Mapper">
<em class="property">class </em><code class="descclassname">pypath.mapping.</code><code class="descname">Mapper</code><span class="sig-paren">(</span><em>ncbi_tax_id=9606</em>, <em>mysql_conf=(None</em>, <em>'mapping')</em>, <em>log=None</em>, <em>cache=True</em>, <em>cachedir='cache'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#Mapper"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.Mapper" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.mapping.Mapper.load_mappings">
<code class="descname">load_mappings</code><span class="sig-paren">(</span><em>maps=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#Mapper.load_mappings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.Mapper.load_mappings" title="Permalink to this definition">¶</a></dt>
<dd><p>mapList is a list of mappings to load;
elements of mapList are dicts containing the 
id names, molecule type, and preferred source
e.g. (&#8220;one&#8221;: &#8220;uniprot&#8221;, &#8220;two&#8221;: &#8220;refseq&#8221;, &#8220;typ&#8221;: &#8220;protein&#8221;, 
&#8220;src&#8221;: &#8220;mysql&#8221;, &#8220;par&#8221;: &#8220;mysql_param/file_param&#8221;)
by default those are loaded from pickle files</p>
</dd></dl>

<dl class="method">
<dt id="pypath.mapping.Mapper.load_uniprot_mapping">
<code class="descname">load_uniprot_mapping</code><span class="sig-paren">(</span><em>filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#Mapper.load_uniprot_mapping"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.Mapper.load_uniprot_mapping" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a wrapper to load a ... mapping table.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.mapping.Mapper.map_name">
<code class="descname">map_name</code><span class="sig-paren">(</span><em>name</em>, <em>nameType</em>, <em>targetNameType</em>, <em>strict=False</em>, <em>silent=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#Mapper.map_name"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.Mapper.map_name" title="Permalink to this definition">¶</a></dt>
<dd><p>This function should be used to convert individual IDs.
It takes care about everything, you don&#8217;t need to think
on the details. How does it work: looks up dictionaries 
between the original and target ID type, if doesn&#8217;t 
find, attempts to load from the predefined inputs.
If the original name is genesymbol, first it looks up
among the preferred gene names from UniProt, if not 
found, it takes an attempt with the alternative gene
names. If the gene symbol still couldn&#8217;t be found, and 
strict = False, the last attempt only the first 5 chara-
cters of the gene symbol matched. If the target name 
type is uniprot, then it converts all the ACs to primary. 
Then, for the Trembl IDs it looks up the preferred gene 
names, and find Swissprot IDs with the same preferred 
gene name.</p>
<dl class="docutils">
<dt>&#64;name <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>The original name which shall be converted.</dd>
<dt>&#64;nameType <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd><p class="first">The type of the name. 
Available by default:</p>
<blockquote>
<div>genesymbol (gene name)
entrez (Entrez Gene ID [#])
refseqp (NCBI RefSeq Protein ID [NP_*|XP_*])
ensp (Ensembl protein ID [ENSP*])
enst (Ensembl transcript ID [ENST*])
ensg (Ensembl genomic DNA ID [ENSG*])
hgnc (HGNC ID [HGNC:#])
gi (GI number [#])
embl (DDBJ/EMBL/GeneBank CDS accession)
embl_id (DDBJ/EMBL/GeneBank accession)</div></blockquote>
<p class="last">To use other IDs, you need to define the input method
and load the table before calling Mapper.map_name().</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.mapping.Mapper.primary_uniprot">
<code class="descname">primary_uniprot</code><span class="sig-paren">(</span><em>lst</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#Mapper.primary_uniprot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.Mapper.primary_uniprot" title="Permalink to this definition">¶</a></dt>
<dd><p>For a list of UniProt IDs returns the list of primary ids.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.mapping.Mapper.trembl_swissprot">
<code class="descname">trembl_swissprot</code><span class="sig-paren">(</span><em>lst</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#Mapper.trembl_swissprot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.Mapper.trembl_swissprot" title="Permalink to this definition">¶</a></dt>
<dd><p>For a list of Trembl and Swissprot IDs, returns possibly
only Swissprot, mapping from Trembl to gene names, and 
then back to Swissprot.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.mapping.Mapper.which_table">
<code class="descname">which_table</code><span class="sig-paren">(</span><em>nameType</em>, <em>targetNameType</em>, <em>load=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#Mapper.which_table"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.Mapper.which_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the table which is suitable to convert an ID of 
nameType to targetNameType. If no such table have been loaded
yet, it attempts to load from UniProt.</p>
</dd></dl>

</dd></dl>

<span class="target" id="module-pypath.moduland"></span><span class="target" id="module-pypath.gdsc"></span><span class="target" id="module-pypath.dataio"></span><dl class="class">
<dt id="pypath.dataio.ResidueMapper">
<em class="property">class </em><code class="descclassname">pypath.dataio.</code><code class="descname">ResidueMapper</code><a class="reference internal" href="_modules/pypath/dataio.html#ResidueMapper"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.ResidueMapper" title="Permalink to this definition">¶</a></dt>
<dd><p>This class stores and serves the PDB &#8211;&gt; UniProt 
residue level mapping. Attempts to download the 
mapping, and stores it for further use. Converts 
PDB residue numbers to the corresponding UniProt ones.</p>
<dl class="method">
<dt id="pypath.dataio.ResidueMapper.clean">
<code class="descname">clean</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#ResidueMapper.clean"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.ResidueMapper.clean" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes cached mappings, freeing up memory.</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_complexportal">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_complexportal</code><span class="sig-paren">(</span><em>species=9606</em>, <em>zipped=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_complexportal"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_complexportal" title="Permalink to this definition">¶</a></dt>
<dd><p>Complex dataset from IntAct.
See more:
<a class="reference external" href="http://www.ebi.ac.uk/intact/complex/">http://www.ebi.ac.uk/intact/complex/</a>
<a class="reference external" href="http://nar.oxfordjournals.org/content/early/2014/10/13/nar.gku975.full.pdf">http://nar.oxfordjournals.org/content/early/2014/10/13/nar.gku975.full.pdf</a></p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_domino_ptms">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_domino_ptms</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_domino_ptms"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_domino_ptms" title="Permalink to this definition">¶</a></dt>
<dd><p>The table comes from dataio.get_domino(), having the following fields:
header = [&#8216;uniprot-A&#8217;, &#8216;uniprot-B&#8217;, &#8216;isoform-A&#8217;, &#8216;isoform-B&#8217;, #3
&#8216;exp. method&#8217;, &#8216;references&#8217;, &#8216;taxon-A&#8217;, &#8216;taxon-B&#8217;, #7
&#8216;role-A&#8217;, &#8216;role-B&#8217;, &#8216;binding-site-range-A&#8217;, &#8216;binding-site-range-B&#8217;, #11
&#8216;domains-A&#8217;, &#8216;domains-B&#8217;, &#8216;ptm-residue-A&#8217;, &#8216;ptm-residue-B&#8217;, #15
&#8216;ptm-type-mi-A&#8217;, &#8216;ptm-type-mi-B&#8217;, &#8216;ptm-type-A&#8217;, &#8216;ptm-type-B&#8217;, #19
&#8216;ptm-res-name-A&#8217;, &#8216;ptm-res-name-B&#8217;, &#8216;mutations-A&#8217;, &#8216;mutations-B&#8217;, #23
&#8216;mutation-effects-A&#8217;, &#8216;mutation-effects-B&#8217;, &#8216;domains-interpro-A&#8217;, #26 
&#8216;domains-interpro-B&#8217;, &#8216;negative&#8217;] #28</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_i3d">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_i3d</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_i3d"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_i3d" title="Permalink to this definition">¶</a></dt>
<dd><p>Interaction3D contains residue numbers in given chains in 
given PDB stuctures, so we need to add an offset to get the residue 
numbers valid for UniProt sequences. Offsets can be obtained from
Instruct, or from the Pfam PDB-chain-UniProt mapping table.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_instruct">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_instruct</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_instruct"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_instruct" title="Permalink to this definition">¶</a></dt>
<dd><p>Instruct contains residue numbers in UniProt sequences, it means
no further calculations of offsets in chains of PDB structures needed.
Chains are not given, only a set of PDB structures supporting the 
domain-domain // protein-protein interaction.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_instruct_offsets">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_instruct_offsets</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_instruct_offsets"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_instruct_offsets" title="Permalink to this definition">¶</a></dt>
<dd><p>These offsets should be understood as from UniProt to PDB.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_switches_elm">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_switches_elm</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_switches_elm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_switches_elm" title="Permalink to this definition">¶</a></dt>
<dd><p>switches.elm is a resource containing functional switches in molecular regulation,
in domain-motif level resolution, classified into categories according to their
mechanism.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.read_complexes_havugimana">
<code class="descclassname">pypath.dataio.</code><code class="descname">read_complexes_havugimana</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#read_complexes_havugimana"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.read_complexes_havugimana" title="Permalink to this definition">¶</a></dt>
<dd><p>Supplement Table S3/1 from Havugimana 2012
Cell. 150(5): 1068–1081.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.read_table">
<code class="descclassname">pypath.dataio.</code><code class="descname">read_table</code><span class="sig-paren">(</span><em>fileObject</em>, <em>cols</em>, <em>sep='\t'</em>, <em>sep2=None</em>, <em>rem=[]</em>, <em>hdr=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#read_table"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.read_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Generic function to read data tables.</p>
<dl class="docutils">
<dt>fileObject <span class="classifier-delimiter">:</span> <span class="classifier">file-like</span></dt>
<dd>Any file like object: file opened for read, or StringIO buffer</dd>
<dt>cols <span class="classifier-delimiter">:</span> <span class="classifier">dict</span></dt>
<dd>Dictionary of columns to read. Keys identifying fields are returned 
in the result. Values are column numbers.</dd>
<dt>sepLevel1 <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>Field separator of the file.</dd>
<dt>sepLevel2 <span class="classifier-delimiter">:</span> <span class="classifier">dict</span></dt>
<dd>Subfield separators and prefixes. 
E.g. {2: &#8216;,&#8217;, 3: &#8216;|&#8217;}</dd>
<dt>hdr <span class="classifier-delimiter">:</span> <span class="classifier">int</span></dt>
<dd>Number of header lines. If None, no headers assumed.</dd>
<dt>rem <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd>Strings to remove. For each line these elements will be replaced with &#8216;&#8217;.</dd>
</dl>
</dd></dl>

<span class="target" id="module-pypath.chembl"></span><span class="target" id="module-pypath.unichem"></span><span class="target" id="module-pypath.seq"></span><span class="target" id="module-pypath.mysql"></span><span class="target" id="module-pypath.mysql_connect"></span><span class="target" id="module-pypath.seq"></span></div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Welcome to pypath&#8217;s documentation!</a></li>
<li><a class="reference internal" href="#examples">Examples</a><ul>
<li><a class="reference internal" href="#example-1-building-the-network">Example 1: building the network</a></li>
<li><a class="reference internal" href="#example-2-using-the-mapper-class-for-translating-ids">Example 2: using the Mapper class for translating IDs</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-pypath.pypath">Reference</a></li>
</ul>

  <h4>Next topic</h4>
  <p class="topless"><a href=""
                        title="next chapter">Welcome to pypath&#8217;s documentation!</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="#" title="Welcome to pypath’s documentation!"
             >next</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">pypath 0.1.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Dénes Türei.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>