

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Welcome to pypath’s documentation! &mdash; pypath 0.1.72 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="pypath 0.1.72 documentation" href="#"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="#" class="icon icon-home"> pypath
          

          
          </a>

          
            
            
              <div class="version">
                0.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <!-- Local TOC -->
                <div class="local-toc"><ul>
<li><a class="reference internal" href="#">Welcome to pypath&#8217;s documentation!</a></li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
<li><a class="reference internal" href="#examples">Examples</a><ul>
<li><a class="reference internal" href="#example-1-building-the-network">Example 1: building the network</a></li>
<li><a class="reference internal" href="#example-2-using-the-mapper-class-for-translating-ids">Example 2: using the Mapper class for translating IDs</a></li>
</ul>
</li>
<li><a class="reference internal" href="#reference">Reference</a></li>
</ul>
</div>
            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="#">pypath</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          





<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="#">Docs</a> &raquo;</li>
      
    <li>Welcome to pypath&#8217;s documentation!</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/index.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="welcome-to-pypath-s-documentation">
<h1>Welcome to pypath&#8217;s documentation!<a class="headerlink" href="#welcome-to-pypath-s-documentation" title="Permalink to this headline">¶</a></h1>
<p>Contents:</p>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<blockquote>
<div><ul class="simple">
<li><a class="reference internal" href="genindex.html"><span>Index</span></a></li>
<li><a class="reference internal" href="py-modindex.html"><span>Module Index</span></a></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="examples">
<h1>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h1>
<div class="section" id="example-1-building-the-network">
<h2>Example 1: building the network<a class="headerlink" href="#example-1-building-the-network" title="Permalink to this headline">¶</a></h2>
<p>Import the module and create an instance:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">pypath</span>
<span class="n">pa</span> <span class="o">=</span> <span class="n">pypath</span><span class="o">.</span><span class="n">PyPath</span><span class="p">()</span>
</pre></div>
</div>
<p>The tables for ID conversions, and the network of the selected manually curated pathway databases (<a class="reference external" href="http://omnipathdb.org/">OmniPath</a>) can be initialized this way:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">pa</span><span class="o">.</span><span class="n">init_network</span><span class="p">()</span>
</pre></div>
</div>
<p>Resources without references for each interaction are available separately. For example, to build a comprehensive phosphonetwork, you might want to load interactions from high-throughput data in PhosphoSite, kinase-substrate relationships from PhosphoNetworks and MIMP. This is the way to do it:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pypath.data_formats</span> <span class="kn">import</span> <span class="n">ptm_misc</span>
<span class="n">pa</span><span class="o">.</span><span class="n">load_resources</span><span class="p">(</span><span class="n">lst</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;mimp&#39;</span><span class="p">:</span> <span class="n">ptm_misc</span><span class="p">[</span><span class="s1">&#39;mimp&#39;</span><span class="p">]})</span>
<span class="n">pa</span><span class="o">.</span><span class="n">load_resources</span><span class="p">(</span><span class="n">lst</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;pnetworks&#39;</span><span class="p">:</span> <span class="n">ptm_misc</span><span class="p">[</span><span class="s1">&#39;pnetworks&#39;</span><span class="p">]})</span>
<span class="n">pa</span><span class="o">.</span><span class="n">load_resources</span><span class="p">(</span><span class="n">lst</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;psite_noref&#39;</span><span class="p">:</span> <span class="n">ptm_misc</span><span class="p">[</span><span class="s1">&#39;psite_noref&#39;</span><span class="p">]})</span>
</pre></div>
</div>
<p>In case you have your own interaction data, and you wish to merge this into the network, you first need to define its format, and then call the load_resource() function. In this simple example, the proteins are noted by Gene Symbols, their names are in columns #0 and #1, while from column #2 an edge attribute named &#8216;score&#8217; will be read. The file is tab separated, and the file name is &#8216;mylist.sif&#8217;. (This won&#8217;t work unless you really have a file like this.)</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">mylist</span> <span class="o">=</span> <span class="n">pypath</span><span class="o">.</span><span class="n">input_formats</span><span class="o">.</span><span class="n">ReadSettings</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;mylist&quot;</span><span class="p">,</span>
    <span class="n">separator</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">nameColA</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">nameColB</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
    <span class="n">nameTypeA</span> <span class="o">=</span> <span class="s2">&quot;genesymbol&quot;</span><span class="p">,</span> <span class="n">nameTypeB</span> <span class="o">=</span> <span class="s2">&quot;genesymbol&quot;</span><span class="p">,</span>
    <span class="n">isDirected</span> <span class="o">=</span> <span class="bp">False</span><span class="p">,</span> <span class="n">inFile</span> <span class="o">=</span> <span class="s1">&#39;mylist.sif&#39;</span><span class="p">,</span>
    <span class="n">extraEdgeAttrs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;score&#39;</span><span class="p">:</span> <span class="mi">2</span>
    <span class="p">})</span>
<span class="n">pa</span><span class="o">.</span><span class="n">load_resource</span><span class="p">(</span><span class="n">mylist</span><span class="p">)</span>
</pre></div>
</div>
<p>Although proteins are identified by UniProt IDs, it is useful to have standard gene symbols as labels, to make things more human understandable.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">pa</span><span class="o">.</span><span class="n">genesymbol_labels</span><span class="p">()</span>
</pre></div>
</div>
<p>Now the igraph object is in pa.graph, vertex and edge attributes are accessible the usual way:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">vs</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;name&#39;</span><span class="p">]</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">vs</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;label&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>Let&#8217;s see the default attributes of edges:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;sources&#39;</span><span class="p">]</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;references&#39;</span><span class="p">]</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;dirs&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>The graph is undirected by default, and &#8216;Direction&#8217; objects can be found in &lt;edge&gt;[&#8216;dirs&#8217;] to describe the directions and signs of the edge:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">111</span><span class="p">][</span><span class="s1">&#39;dirs&#39;</span><span class="p">]</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">111</span><span class="p">]</span><span class="o">.</span><span class="n">is_directed</span><span class="p">()</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">111</span><span class="p">][</span><span class="s1">&#39;dirs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">is_inhibition</span><span class="p">()</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">111</span><span class="p">][</span><span class="s1">&#39;dirs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">which_dirs</span><span class="p">()</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">111</span><span class="p">][</span><span class="s1">&#39;dirs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">get_dir</span><span class="p">(</span><span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">111</span><span class="p">][</span><span class="s1">&#39;dirs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">which_dirs</span><span class="p">()[</span><span class="mi">0</span><span class="p">])</span>
</pre></div>
</div>
<p>To convert the igraph object to a directed graph, use this function:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">pa</span><span class="o">.</span><span class="n">get_directed</span><span class="p">()</span>
<span class="n">pa</span><span class="o">.</span><span class="n">dgraph</span><span class="o">.</span><span class="n">ecount</span><span class="p">()</span>
<span class="n">pa</span><span class="o">.</span><span class="n">dgraph</span><span class="o">.</span><span class="n">vcount</span><span class="p">()</span>
</pre></div>
</div>
<p>To replace the graph with the giant component of the directed graph:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">pa</span><span class="o">.</span><span class="n">graph</span> <span class="o">=</span> <span class="n">pa</span><span class="o">.</span><span class="n">get_giant</span><span class="p">(</span><span class="n">graph</span> <span class="o">=</span> <span class="n">pa</span><span class="o">.</span><span class="n">dgraph</span><span class="p">)</span>
</pre></div>
</div>
<p>Get compounds for all the proteins in the network:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">pa</span><span class="o">.</span><span class="n">compounds_from_chembl</span><span class="p">((</span><span class="bp">None</span><span class="p">,</span> <span class="s1">&#39;chembl_ebi&#39;</span><span class="p">),</span> <span class="n">pchembl</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<p>This works only if you have a MySQL server hosting an instance of ChEMBL:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">pa</span><span class="o">.</span><span class="n">compounds_from_chembl</span><span class="p">((</span><span class="bp">None</span><span class="p">,</span> <span class="s1">&#39;chembl&#39;</span><span class="p">),</span> <span class="n">pchembl</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<p>You can find the list of compounds and the requested data (now only the pchembl values) in vertex attributes:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">vs</span><span class="p">[</span><span class="mi">2</span><span class="p">][</span><span class="s1">&#39;compounds_chembl&#39;</span><span class="p">]</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">vs</span><span class="p">[</span><span class="mi">2</span><span class="p">][</span><span class="s1">&#39;compounds_data&#39;</span><span class="p">]</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">vs</span><span class="p">[</span><span class="mi">2</span><span class="p">][</span><span class="s1">&#39;compounds_data&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;pchembl&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>Loading PTMs. This function loads PTMs from all the available resources. It takes long time first, but once it saves the downloaded data under <cite>./cache</cite> directory, it is much faster to run again.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">pa</span><span class="o">.</span><span class="n">load_ptms</span><span class="p">()</span>
</pre></div>
</div>
<p>Individual PTM resources can be loaded the following way:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">pa</span><span class="o">.</span><span class="n">load_phospho_dmi</span><span class="p">(</span><span class="n">source</span> <span class="o">=</span> <span class="s1">&#39;phosphoELM&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>In the latter case, this function is needed to merge the identical PTMs loaded from multiple resources:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">pa</span><span class="o">.</span><span class="n">uniq_ptms</span><span class="p">()</span>
</pre></div>
</div>
<p>PTMs are stored in objects. This example shows how to access the type of modification, and the number and name of the modified residue.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">70</span><span class="p">][</span><span class="s1">&#39;ptm&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">ptm</span><span class="o">.</span><span class="n">typ</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">70</span><span class="p">][</span><span class="s1">&#39;ptm&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">ptm</span><span class="o">.</span><span class="n">protein</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">70</span><span class="p">][</span><span class="s1">&#39;ptm&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">ptm</span><span class="o">.</span><span class="n">residue</span><span class="o">.</span><span class="n">name</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">70</span><span class="p">][</span><span class="s1">&#39;ptm&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">ptm</span><span class="o">.</span><span class="n">residue</span><span class="o">.</span><span class="n">number</span>
</pre></div>
</div>
</div>
<div class="section" id="example-2-using-the-mapper-class-for-translating-ids">
<h2>Example 2: using the Mapper class for translating IDs<a class="headerlink" href="#example-2-using-the-mapper-class-for-translating-ids" title="Permalink to this headline">¶</a></h2>
<p>The mapping submodule of pypath can be used for ID conversion. Here is a basic example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pypath</span> <span class="kn">import</span> <span class="n">mapping</span>
<span class="n">m</span> <span class="o">=</span> <span class="n">mapping</span><span class="o">.</span><span class="n">Mapper</span><span class="p">(</span><span class="mi">9606</span><span class="p">,</span> <span class="n">mysql_conf</span> <span class="o">=</span> <span class="p">(</span><span class="bp">None</span><span class="p">,</span> <span class="s1">&#39;mapping_gelati&#39;</span><span class="p">))</span>
<span class="n">m</span><span class="o">.</span><span class="n">load_mappings</span><span class="p">()</span>
<span class="n">result</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">gene_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;EGFR&#39;</span><span class="p">,</span> <span class="s1">&#39;AKT1&#39;</span><span class="p">,</span> <span class="s1">&#39;GABARAPL1&#39;</span><span class="p">,</span> <span class="s1">&#39;TP53&#39;</span><span class="p">]</span>
<span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">gene_list</span><span class="p">:</span>
    <span class="n">result</span><span class="p">[</span><span class="n">g</span><span class="p">]</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">map_name</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="s1">&#39;genesymbol&#39;</span><span class="p">,</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="reference">
<h1>Reference<a class="headerlink" href="#reference" title="Permalink to this headline">¶</a></h1>
<span class="target" id="module-pypath.pypath"></span><dl class="class">
<dt id="pypath.pypath.PyPath">
<em class="property">class </em><code class="descclassname">pypath.pypath.</code><code class="descname">PyPath</code><span class="sig-paren">(</span><em>ncbi_tax_id=9606</em>, <em>default_name_type={'protein': 'uniprot'</em>, <em>'drug': 'chembl'</em>, <em>'mirna': 'mirbase'}</em>, <em>copy=None</em>, <em>mysql=(None</em>, <em>'mapping')</em>, <em>chembl_mysql=(None</em>, <em>'chembl')</em>, <em>name='unnamed'</em>, <em>outdir='results'</em>, <em>loglevel='INFO'</em>, <em>loops=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.pypath.PyPath.add_update_vertex">
<code class="descname">add_update_vertex</code><span class="sig-paren">(</span><em>defAttrs</em>, <em>originalName</em>, <em>originalNameType</em>, <em>extraAttrs={}</em>, <em>add=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.add_update_vertex"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.add_update_vertex" title="Permalink to this definition">¶</a></dt>
<dd><p>Updates the attributes of one node in the network.
Optionally it creates a new node and sets the attributes,
but it is not efficient as igraph needs to reindex vertices 
after this operation, so better to create new nodes and 
edges in batch.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.all_between">
<code class="descname">all_between</code><span class="sig-paren">(</span><em>nameA</em>, <em>nameB</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.all_between"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.all_between" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all edges between two given vertex names. Similar to 
straight_between(), but checks both directions, and returns 
list of edge ids in [undirected, straight, reveresed] format,
for both nameA -&gt; nameB and nameB -&gt; nameA edges.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.basic_stats">
<code class="descname">basic_stats</code><span class="sig-paren">(</span><em>latex=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.basic_stats"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.basic_stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns basic numbers about the network resources, e.g. edge and 
node counts.</p>
<dl class="docutils">
<dt>latex</dt>
<dd>Return table in a LaTeX document. This can be compiled by
PDFLaTeX: 
latex stats.tex</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.count_sol">
<code class="descname">count_sol</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.count_sol"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.count_sol" title="Permalink to this definition">¶</a></dt>
<dd><p>Counts nodes with zero degree.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.curation_effort">
<code class="descname">curation_effort</code><span class="sig-paren">(</span><em>sum_by_source=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.curation_effort"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.curation_effort" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the total number of reference-interactions pairs.</p>
<dl class="docutils">
<dt>&#64;sum_by_source <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd>If True, counts the refrence-interaction pairs by
sources, and returns the sum of these values.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.delete_by_taxon">
<code class="descname">delete_by_taxon</code><span class="sig-paren">(</span><em>tax</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.delete_by_taxon"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.delete_by_taxon" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes the proteins of all organisms which are not listed.</p>
<dl class="docutils">
<dt>&#64;tax <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd>List of NCBI Taxonomy IDs of the organisms of interest.
E.g. [7227, 9606]</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.delete_unknown">
<code class="descname">delete_unknown</code><span class="sig-paren">(</span><em>tax</em>, <em>typ='protein'</em>, <em>defaultNameType=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.delete_unknown"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.delete_unknown" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes those proteins which are not in the list of all default
IDs of the organisms. By default, it means to remove all protein 
nodes not having a human SwissProt ID.</p>
<dl class="docutils">
<dt>&#64;tax <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd>List of NCBI Taxonomy IDs of the organisms of interest.
E.g. [7227, 9606]</dd>
<dt>&#64;typ <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>Molecule type. E.g. &#8216;protein&#8217; or &#8216;mirna&#8217;</dd>
<dt>&#64;defaultNameType <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>The default name type of the given molecular species.
For proteins it&#8217;s &#8216;uniprot&#8217; by default.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.edge_exists">
<code class="descname">edge_exists</code><span class="sig-paren">(</span><em>nameA</em>, <em>nameB</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.edge_exists"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.edge_exists" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a tuple of vertice indices if edge doesn&#8217;t exists, 
otherwise edge id. Not sensitive to direction.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.edges_expression">
<code class="descname">edges_expression</code><span class="sig-paren">(</span><em>func=&lt;function &lt;lambda&gt;&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.edges_expression"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.edges_expression" title="Permalink to this definition">¶</a></dt>
<dd><p>Executes function <cite>func</cite> for each pairs of connected proteins in the 
network, for every expression dataset. By default, <cite>func</cite> simply
gives the product the (normalized) expression values.</p>
<dl class="docutils">
<dt>func <span class="classifier-delimiter">:</span> <span class="classifier">callable</span></dt>
<dd>Function to handle 2 vectors (pandas.Series() objects), should
return one vector of the same length.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.export_dot">
<code class="descname">export_dot</code><span class="sig-paren">(</span><em>nodes=None</em>, <em>edges=None</em>, <em>directed=True</em>, <em>labels='genesymbol'</em>, <em>edges_filter=&lt;function &lt;lambda&gt;&gt;</em>, <em>nodes_filter=&lt;function &lt;lambda&gt;&gt;</em>, <em>edge_sources=None</em>, <em>dir_sources=None</em>, <em>graph=None</em>, <em>return_object=False</em>, <em>save_dot=None</em>, <em>save_graphics=None</em>, <em>prog='neato'</em>, <em>format=None</em>, <em>hide=False</em>, <em>font=None</em>, <em>auto_edges=False</em>, <em>hide_nodes=[]</em>, <em>defaults={}</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.export_dot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.export_dot" title="Permalink to this definition">¶</a></dt>
<dd><p>Builds a pygraphviz.AGraph() object with filtering the edges 
and vertices along arbitrary criteria.
Returns the Agraph object if requesred, or exports the dot
file, or saves the graphics.</p>
<p>&#64;nodes : list
List of vertex ids to be included.
&#64;edges : list
List of edge ids to be included.
&#64;directed : bool
Create a directed or undirected graph.
&#64;labels : str
Name type to be used as id/label in the dot format.
&#64;edges_filter : function
Function to filter edges, accepting igraph.Edge as argument.
&#64;nodes_filter : function
Function to filter vertices, accepting igraph.Vertex as argument.
&#64;edge_sources : list
Sources to be included.
&#64;dir_sources : list
Direction and effect sources to be included.
&#64;graph : igraph.Graph
The graph object to export.
&#64;return_object : bool
Whether to return the pygraphviz.AGraph object.
&#64;save_dot : str
Filename to export the dot file to.
&#64;save_graphics : str
Filename to export the graphics, the extension defines the format.
&#64;prog : str
The graphviz layout algorithm to use.
&#64;format : str
The graphics format passed to pygraphviz.AGrapg().draw().
&#64;hide : bool
Hide filtered edges instead of omit them.
&#64;hide nodes : list
Nodes to hide. List of vertex ids.
&#64;auto_edges : str
Automatic, built-in style for edges. 
&#8216;DIRECTIONS&#8217; or &#8216;RESOURCE_CATEGORIES&#8217; are supported.
&#64;font : str
Font to use for labels. 
For using more than one fonts refer to graphviz attributes with constant values 
or define callbacks or mapping dictionaries.
&#64;defaults : dict
Default values for graphviz attributes, labeled with the entity, e.g.
<cite>{&#8216;edge_penwidth&#8217;: 0.2}</cite>.
&#64;**kwargs : constant, callable or dict
Graphviz attributes, labeled by the target entity. E.g. <cite>edge_penwidth</cite>,
&#8216;vertex_shape` or <cite>graph_label</cite>. 
If the value is constant, this value will be used.
If the value is dict, and has <cite>_name</cite> as key, for every instance of the
given entity, the value of the attribute defined by <cite>_name</cite> will be looked
up in the dict, and the corresponding value will be given to the graphviz
attribute. If the key <cite>_name</cite> is missing from the dict, igraph vertex and
edge indices will be looked up among the keys.
If the value is callable, it will be called with the current instance of
the entity and the returned value will be used for the graphviz attribute.
E.g. <cite>edge_arrowhead(edge)</cite> or <cite>vertex_fillcolor(vertex)</cite>
Example:</p>
<blockquote>
<div><p>import pypath
from pypath import data_formats
net = pypath.BioGraph()
net.init_network(pfile = &#8216;cache/default.pickle&#8217;)
#net.init_network({&#8216;arn&#8217;: data_formats.omnipath[&#8216;arn&#8217;]})
tgf = [v.index for v in net.graph.vs if &#8216;TGF&#8217; in v[&#8216;slk_pathways&#8217;]]
dot = net.export_dot(nodes = tgf, save_graphics = &#8216;tgf_slk.pdf&#8217;, prog = &#8216;dot&#8217;,</p>
<blockquote>
<div>main_title = &#8216;TGF-beta pathway&#8217;, return_object = True,
label_font = &#8216;HelveticaNeueLTStd Med Cn&#8217;, 
edge_sources = [&#8216;SignaLink3&#8217;], 
dir_sources = [&#8216;SignaLink3&#8217;], hide = True)</div></blockquote>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.find_all_paths">
<code class="descname">find_all_paths</code><span class="sig-paren">(</span><em>start</em>, <em>end</em>, <em>mode='OUT'</em>, <em>maxlen=None</em>, <em>graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.find_all_paths"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.find_all_paths" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds all paths up to length <cite>maxlen</cite> between groups of
vertices. This function is needed only becaues igraph`s
get_all_shortest_paths() finds only the shortest, not any
path up to a defined length.</p>
<dl class="docutils">
<dt>&#64;start <span class="classifier-delimiter">:</span> <span class="classifier">int or list</span></dt>
<dd>Indices of the starting node(s) of the paths.</dd>
<dt>&#64;end <span class="classifier-delimiter">:</span> <span class="classifier">int or list</span></dt>
<dd>Indices of the target node(s) of the paths.</dd>
<dt>&#64;mode <span class="classifier-delimiter">:</span> <span class="classifier">&#8216;IN&#8217;, &#8216;OUT&#8217;, &#8216;ALL&#8217;</span></dt>
<dd>Passed to igraph.Graph.neighbors()</dd>
<dt>&#64;maxlen <span class="classifier-delimiter">:</span> <span class="classifier">int</span></dt>
<dd>Maximum length of paths in steps, i.e. if maxlen = 3, then 
the longest path may consist of 3 edges and 4 nodes.</dd>
<dt>&#64;graph <span class="classifier-delimiter">:</span> <span class="classifier">igraph.Graph object</span></dt>
<dd>The graph you want to find paths in. self.graph by default.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.get_edge">
<code class="descname">get_edge</code><span class="sig-paren">(</span><em>nodes</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.get_edge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.get_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the edge id only if there is an edge from nodes[0] to nodes[1],
returns False if edge exists in opposite direction, or no edge exists 
between the two vertices, or any of the vertice ids doesn&#8217;t exist.
To find edges without regarding their direction, see edge_exists().</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.get_giant">
<code class="descname">get_giant</code><span class="sig-paren">(</span><em>replace=False</em>, <em>graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.get_giant"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.get_giant" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the giant component of the graph, or 
replaces the igraph object with only the giant 
component.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.init_network">
<code class="descname">init_network</code><span class="sig-paren">(</span><em>lst={'netpath': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e4d0&gt;</em>, <em>'domino': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e758&gt;</em>, <em>'alz': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e560&gt;</em>, <em>'trip': &lt;pypath.input_formats.ReadSettings instance at 0x7f29396fbb90&gt;</em>, <em>'hprd': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e830&gt;</em>, <em>'death': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e128&gt;</em>, <em>'elm': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e710&gt;</em>, <em>'ca1': &lt;pypath.input_formats.ReadSettings instance at 0x7f29396fbdd0&gt;</em>, <em>'depod': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e638&gt;</em>, <em>'ccmap': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e3b0&gt;</em>, <em>'arn': &lt;pypath.input_formats.ReadSettings instance at 0x7f29396fbef0&gt;</em>, <em>'phelm': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e6c8&gt;</em>, <em>'lmpid': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e680&gt;</em>, <em>'dbptm': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e7e8&gt;</em>, <em>'signalink3': &lt;pypath.input_formats.ReadSettings instance at 0x7f29396fbcf8&gt;</em>, <em>'spike': &lt;pypath.input_formats.ReadSettings instance at 0x7f29396fbcb0&gt;</em>, <em>'matrixdb': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e5a8&gt;</em>, <em>'innatedb': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e518&gt;</em>, <em>'mppi': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e440&gt;</em>, <em>'nrf2': &lt;pypath.input_formats.ReadSettings instance at 0x7f29396fbfc8&gt;</em>, <em>'signor': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e320&gt;</em>, <em>'macrophage': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e0e0&gt;</em>, <em>'pdz': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e170&gt;</em>, <em>'dip': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e488&gt;</em>, <em>'biogrid': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e368&gt;</em>, <em>'guide2pharma': &lt;pypath.input_formats.ReadSettings instance at 0x7f29396fbd40&gt;</em>, <em>'psite': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e5f0&gt;}</em>, <em>exclude=[]</em>, <em>cache_files={}</em>, <em>pfile=False</em>, <em>save=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.init_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.init_network" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a lazy way to start the module, load data 
and build the high confidence, literature curated
part of the signaling network.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.load_ddi">
<code class="descname">load_ddi</code><span class="sig-paren">(</span><em>ddi</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.load_ddi"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.load_ddi" title="Permalink to this definition">¶</a></dt>
<dd><p>ddi is either a list of intera.DomainDomain objects, 
or a function resulting this list</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.load_dmi">
<code class="descname">load_dmi</code><span class="sig-paren">(</span><em>dmi</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.load_dmi"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.load_dmi" title="Permalink to this definition">¶</a></dt>
<dd><p>dmi is either a list of intera.DomainMotif objects, 
or a function resulting this list</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.load_expression">
<code class="descname">load_expression</code><span class="sig-paren">(</span><em>array=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.load_expression"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.load_expression" title="Permalink to this definition">¶</a></dt>
<dd><p>Expression data can be loaded into vertex attributes, 
or into a pandas DataFrame – the latter offers faster 
ways to process and use these huge matrices.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.load_mutations">
<code class="descname">load_mutations</code><span class="sig-paren">(</span><em>attributes=None</em>, <em>gdsc_datadir=None</em>, <em>mutation_file=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.load_mutations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.load_mutations" title="Permalink to this definition">¶</a></dt>
<dd><p>Mutations are listed in vertex attributes. Mutation() objects 
offers methods to identify residues and look up in Ptm(), Motif() 
and Domain() objects, to check if those residues are 
modified, or are in some short motif or domain.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.load_resources">
<code class="descname">load_resources</code><span class="sig-paren">(</span><em>lst={'netpath': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e4d0&gt;</em>, <em>'domino': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e758&gt;</em>, <em>'alz': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e560&gt;</em>, <em>'trip': &lt;pypath.input_formats.ReadSettings instance at 0x7f29396fbb90&gt;</em>, <em>'hprd': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e830&gt;</em>, <em>'death': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e128&gt;</em>, <em>'elm': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e710&gt;</em>, <em>'ca1': &lt;pypath.input_formats.ReadSettings instance at 0x7f29396fbdd0&gt;</em>, <em>'depod': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e638&gt;</em>, <em>'ccmap': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e3b0&gt;</em>, <em>'arn': &lt;pypath.input_formats.ReadSettings instance at 0x7f29396fbef0&gt;</em>, <em>'phelm': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e6c8&gt;</em>, <em>'lmpid': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e680&gt;</em>, <em>'dbptm': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e7e8&gt;</em>, <em>'signalink3': &lt;pypath.input_formats.ReadSettings instance at 0x7f29396fbcf8&gt;</em>, <em>'spike': &lt;pypath.input_formats.ReadSettings instance at 0x7f29396fbcb0&gt;</em>, <em>'matrixdb': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e5a8&gt;</em>, <em>'innatedb': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e518&gt;</em>, <em>'mppi': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e440&gt;</em>, <em>'nrf2': &lt;pypath.input_formats.ReadSettings instance at 0x7f29396fbfc8&gt;</em>, <em>'signor': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e320&gt;</em>, <em>'macrophage': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e0e0&gt;</em>, <em>'pdz': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e170&gt;</em>, <em>'dip': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e488&gt;</em>, <em>'biogrid': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e368&gt;</em>, <em>'guide2pharma': &lt;pypath.input_formats.ReadSettings instance at 0x7f29396fbd40&gt;</em>, <em>'psite': &lt;pypath.input_formats.ReadSettings instance at 0x7f293970e5f0&gt;}</em>, <em>exclude=[]</em>, <em>cache_files={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.load_resources"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.load_resources" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads multiple resources, and cleans up after. 
Looks up ID types, and loads all ID conversion 
tables from UniProt if necessary. This is much
faster than loading the ID conversion and the 
resources one by one.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.map_list">
<code class="descname">map_list</code><span class="sig-paren">(</span><em>lst</em>, <em>singleList=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.map_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.map_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Only a wrapper for map_edge()</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.map_network">
<code class="descname">map_network</code><span class="sig-paren">(</span><em>mapping=('trembl'</em>, <em>'swissprot')</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.map_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.map_network" title="Permalink to this definition">¶</a></dt>
<dd><p>This function is not needed any more, will be removed soon.</p>
<p>In the whole network changes every uniprot id to primary.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.mutated_edges">
<code class="descname">mutated_edges</code><span class="sig-paren">(</span><em>sample</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.mutated_edges"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.mutated_edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Compares the mutated residues and the modified residues in PTMs.
Interactions are marked as mutated if the target residue in the 
underlying PTM is mutated.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.network_filter">
<code class="descname">network_filter</code><span class="sig-paren">(</span><em>p=2.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.network_filter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.network_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>This function aims to cut the number of edges in the network, 
without loosing nodes, to make the network less connected, 
less hairball-like, more usable for analysis.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.network_stats">
<code class="descname">network_stats</code><span class="sig-paren">(</span><em>outfile=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.network_stats"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.network_stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates basic statistics for the whole network 
and each of sources. Writes the results in a tab file.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.process_dmi">
<code class="descname">process_dmi</code><span class="sig-paren">(</span><em>source</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.process_dmi"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.process_dmi" title="Permalink to this definition">¶</a></dt>
<dd><p>This is an universal function 
for loading domain-motif objects
like load_phospho_dmi() for PTMs.
TODO this will replace load_elm, load_ielm, etc</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.read_data_file">
<code class="descname">read_data_file</code><span class="sig-paren">(</span><em>settings</em>, <em>keep_raw=False</em>, <em>cache_files={}</em>, <em>cache=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.read_data_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.read_data_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Interaction data with node and edge attributes can be read 
from simple text based files. This function works not only
with files, but with lists as well. Any other function can 
be written to download a preprocess data, and then give it
to this function to finally attach to the network.</p>
<dl class="docutils">
<dt>&#64;settings <span class="classifier-delimiter">:</span> <span class="classifier">ReadSettings instance</span></dt>
<dd>The detailed definition of the input format. Instead of 
the file name you can give a function name, which will 
be executed, and the returned data will be used.</dd>
<dt>&#64;keep_raw <span class="classifier-delimiter">:</span> <span class="classifier">boolean</span></dt>
<dd>To keep the raw data read by this function, in order for
debugging purposes, or further use.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.set_chembl_mysql">
<code class="descname">set_chembl_mysql</code><span class="sig-paren">(</span><em>title</em>, <em>config_file=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.set_chembl_mysql"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.set_chembl_mysql" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the ChEMBL MySQL config according to
<cite>title</cite> section in <cite>config_file</cite> ini style config.</p>
<blockquote>
<div><p>title (str): section title in ini file
config_file (str, NoneType): config file name;</p>
<blockquote>
<div>if None, the <cite>mysql_config/defaults.mysql</cite>
will be used</div></blockquote>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.set_drugtargets">
<code class="descname">set_drugtargets</code><span class="sig-paren">(</span><em>pchembl=5.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.set_drugtargets"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.set_drugtargets" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a vertex attribute <cite>dtg</cite> with value <em>True</em> if
the protein has at least one compound binding with 
affinity higher than <cite>pchembl</cite>, otherwise <em>False</em>.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.shortest_path_dist">
<code class="descname">shortest_path_dist</code><span class="sig-paren">(</span><em>graph=None</em>, <em>subset=None</em>, <em>outfile=None</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.shortest_path_dist"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.shortest_path_dist" title="Permalink to this definition">¶</a></dt>
<dd><p>subset is a tuple of two lists if you wish to look for
paths between elements of two groups, or a list if you 
wish to look for shortest paths within this group</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.source_network">
<code class="descname">source_network</code><span class="sig-paren">(</span><em>font='HelveticaNeueLTStd'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.source_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.source_network" title="Permalink to this definition">¶</a></dt>
<dd><p>For EMBL branding, use Helvetica Neue Linotype Standard light</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.straight_between">
<code class="descname">straight_between</code><span class="sig-paren">(</span><em>nameA</em>, <em>nameB</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.straight_between"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.straight_between" title="Permalink to this definition">¶</a></dt>
<dd><p>This does actually the same as get_edge(), but by names
instead of vertex ids.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.third_source_directions">
<code class="descname">third_source_directions</code><span class="sig-paren">(</span><em>graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.third_source_directions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.third_source_directions" title="Permalink to this definition">¶</a></dt>
<dd><p>This method calls a series of methods to get
additional direction &amp; effect information
from sources having no literature curated references,
but giving sufficient evidence about the directionality
for interactions already supported by literature 
evidences from other sources.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.update_vindex">
<code class="descname">update_vindex</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.update_vindex"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.update_vindex" title="Permalink to this definition">¶</a></dt>
<dd><p>See update_vname().</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.update_vname">
<code class="descname">update_vname</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.update_vname"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.update_vname" title="Permalink to this definition">¶</a></dt>
<dd><p>For fast lookup of node names and indexes, these are 
hold in a list and a dict as well. However, every time
new nodes are added, these should be updated. This 
function is automatically called after all operations
affecting node indices.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.PyPath.vertex_pathways">
<code class="descname">vertex_pathways</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#PyPath.vertex_pathways"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.PyPath.vertex_pathways" title="Permalink to this definition">¶</a></dt>
<dd><p>Some resources assignes interactions some others 
proteins to pathways.
This function converts pathway annotations from 
edge attributes to vertex attributes.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.pypath.Direction">
<em class="property">class </em><code class="descclassname">pypath.pypath.</code><code class="descname">Direction</code><span class="sig-paren">(</span><em>nameA</em>, <em>nameB</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#Direction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.Direction" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.pypath.Direction.get_dir">
<code class="descname">get_dir</code><span class="sig-paren">(</span><em>direction</em>, <em>sources=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#Direction.get_dir"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.Direction.get_dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns boolean or list of sources</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.Direction.get_dirs">
<code class="descname">get_dirs</code><span class="sig-paren">(</span><em>src</em>, <em>tgt</em>, <em>sources=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#Direction.get_dirs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.Direction.get_dirs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all directions with boolean values
or list of sources.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.Direction.set_dir">
<code class="descname">set_dir</code><span class="sig-paren">(</span><em>direction</em>, <em>source</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#Direction.set_dir"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.Direction.set_dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds directionality information with 
the corresponding data source named.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.Direction.src">
<code class="descname">src</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#Direction.src"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.Direction.src" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the IDs of effector molecules in this directed 
interaction. If the interaction is bidirectional, the 
list will contain 2 IDs. If the interaction is undirec-
ted, an empty list will be returned.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.Direction.tgt">
<code class="descname">tgt</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#Direction.tgt"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.Direction.tgt" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the IDs of the target moleculess in the inter-
action. Same behaviour as <cite>Direction.src()</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pypath.Direction.unset_dir">
<code class="descname">unset_dir</code><span class="sig-paren">(</span><em>direction</em>, <em>source=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pypath.html#Direction.unset_dir"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pypath.Direction.unset_dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes directionality information,
or single source.</p>
</dd></dl>

</dd></dl>

<span class="target" id="module-pypath.intera"></span><dl class="class">
<dt id="pypath.intera.Residue">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">Residue</code><span class="sig-paren">(</span><em>number</em>, <em>name</em>, <em>identifier</em>, <em>id_type='uniprot'</em>, <em>isoform=1</em>, <em>mutated=False</em>, <em>seq=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Residue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Residue" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.intera.Mutation">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">Mutation</code><span class="sig-paren">(</span><em>original</em>, <em>mutated</em>, <em>sample</em>, <em>properties={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Mutation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Mutation" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.intera.Motif">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">Motif</code><span class="sig-paren">(</span><em>protein</em>, <em>start</em>, <em>end</em>, <em>id_type='uniprot'</em>, <em>regex=None</em>, <em>instance=None</em>, <em>isoform=1</em>, <em>motif_name=None</em>, <em>prob=None</em>, <em>elm=None</em>, <em>description=None</em>, <em>seq=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Motif"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Motif" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.intera.Ptm">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">Ptm</code><span class="sig-paren">(</span><em>protein</em>, <em>id_type='uniprot'</em>, <em>typ='unknown'</em>, <em>motif=None</em>, <em>residue=None</em>, <em>source=None</em>, <em>isoform=1</em>, <em>seq=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Ptm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Ptm" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.intera.Domain">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">Domain</code><span class="sig-paren">(</span><em>protein</em>, <em>id_type='uniprot'</em>, <em>domain=None</em>, <em>domain_id_type='pfam'</em>, <em>start=None</em>, <em>end=None</em>, <em>isoform=1</em>, <em>chains={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Domain"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Domain" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.intera.DomainMotif">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">DomainMotif</code><span class="sig-paren">(</span><em>domain</em>, <em>ptm</em>, <em>sources=None</em>, <em>refs=None</em>, <em>pdbs=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#DomainMotif"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.DomainMotif" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.intera.DomainDomain">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">DomainDomain</code><span class="sig-paren">(</span><em>domain_a</em>, <em>domain_b</em>, <em>pdbs=None</em>, <em>sources=None</em>, <em>refs=None</em>, <em>contact_residues=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#DomainDomain"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.DomainDomain" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.intera.Interface">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">Interface</code><span class="sig-paren">(</span><em>id_a</em>, <em>id_b</em>, <em>source</em>, <em>id_type='uniprot'</em>, <em>pdb=None</em>, <em>css=None</em>, <em>stab_en=None</em>, <em>solv_en=None</em>, <em>area=None</em>, <em>isoform_a=1</em>, <em>isoform_b=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Interface"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Interface" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.intera.Interface.add_residues">
<code class="descname">add_residues</code><span class="sig-paren">(</span><em>res_a</em>, <em>res_b</em>, <em>typ='undefined'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Interface.add_residues"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Interface.add_residues" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds one pair of residues of type <cite>typ</cite>, 
where <cite>res_a</cite> and <cite>res_b</cite> are tuples of 
residue number in sequence and residue type,
e.g. (124, &#8216;S&#8217;) &#8211; (means Serine #124)
<cite>typ</cite> can be undefined, hbonds, sbridges, ssbonds or covbonds</p>
</dd></dl>

<dl class="method">
<dt id="pypath.intera.Interface.bond_types">
<code class="descname">bond_types</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Interface.bond_types"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Interface.bond_types" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the bond types present in this interface</p>
</dd></dl>

<dl class="method">
<dt id="pypath.intera.Interface.get_bonds">
<code class="descname">get_bonds</code><span class="sig-paren">(</span><em>typ=None</em>, <em>mode=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Interface.get_bonds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Interface.get_bonds" title="Permalink to this definition">¶</a></dt>
<dd><p>Gives a generator to iterate throught bonds in 
this interface. If no type given, bonds of all types 
returned.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.intera.Interface.numof_residues">
<code class="descname">numof_residues</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Interface.numof_residues"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Interface.numof_residues" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of residue pairs by bound type</p>
</dd></dl>

</dd></dl>

<span class="target" id="module-pypath.mapping"></span><dl class="class">
<dt id="pypath.mapping.MappingTable">
<em class="property">class </em><code class="descclassname">pypath.mapping.</code><code class="descname">MappingTable</code><span class="sig-paren">(</span><em>one</em>, <em>two</em>, <em>typ</em>, <em>source</em>, <em>param</em>, <em>ncbi_tax_id</em>, <em>mysql=None</em>, <em>log=None</em>, <em>cache=True</em>, <em>cachedir='cache'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#MappingTable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.MappingTable" title="Permalink to this definition">¶</a></dt>
<dd><p>To initialize ID conversion tables for the first time
data is downloaded from UniProt and read to dictionaries.
It takes a couple of seconds. Data is saved to pickle 
dumps, this way after tables load much faster.</p>
<dl class="method">
<dt id="pypath.mapping.MappingTable.read_mapping_uniprot">
<code class="descname">read_mapping_uniprot</code><span class="sig-paren">(</span><em>param</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#MappingTable.read_mapping_uniprot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.MappingTable.read_mapping_uniprot" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads ID mappings directly from UniProt.
See the names of possible identifiers here:
<a class="reference external" href="http://www.uniprot.org/help/programmatic_access">http://www.uniprot.org/help/programmatic_access</a></p>
<p>&#64;param : UniprotMapping instance</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.mapping.Mapper">
<em class="property">class </em><code class="descclassname">pypath.mapping.</code><code class="descname">Mapper</code><span class="sig-paren">(</span><em>ncbi_tax_id=9606</em>, <em>mysql_conf=(None</em>, <em>'mapping')</em>, <em>log=None</em>, <em>cache=True</em>, <em>cachedir='cache'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#Mapper"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.Mapper" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.mapping.Mapper.load_mappings">
<code class="descname">load_mappings</code><span class="sig-paren">(</span><em>maps=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#Mapper.load_mappings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.Mapper.load_mappings" title="Permalink to this definition">¶</a></dt>
<dd><p>mapList is a list of mappings to load;
elements of mapList are dicts containing the 
id names, molecule type, and preferred source
e.g. (&#8220;one&#8221;: &#8220;uniprot&#8221;, &#8220;two&#8221;: &#8220;refseq&#8221;, &#8220;typ&#8221;: &#8220;protein&#8221;, 
&#8220;src&#8221;: &#8220;mysql&#8221;, &#8220;par&#8221;: &#8220;mysql_param/file_param&#8221;)
by default those are loaded from pickle files</p>
</dd></dl>

<dl class="method">
<dt id="pypath.mapping.Mapper.load_uniprot_mapping">
<code class="descname">load_uniprot_mapping</code><span class="sig-paren">(</span><em>filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#Mapper.load_uniprot_mapping"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.Mapper.load_uniprot_mapping" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a wrapper to load a ... mapping table.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.mapping.Mapper.map_name">
<code class="descname">map_name</code><span class="sig-paren">(</span><em>name</em>, <em>nameType</em>, <em>targetNameType</em>, <em>strict=False</em>, <em>silent=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#Mapper.map_name"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.Mapper.map_name" title="Permalink to this definition">¶</a></dt>
<dd><p>This function should be used to convert individual IDs.
It takes care about everything, you don&#8217;t need to think
on the details. How does it work: looks up dictionaries 
between the original and target ID type, if doesn&#8217;t 
find, attempts to load from the predefined inputs.
If the original name is genesymbol, first it looks up
among the preferred gene names from UniProt, if not 
found, it takes an attempt with the alternative gene
names. If the gene symbol still couldn&#8217;t be found, and 
strict = False, the last attempt only the first 5 chara-
cters of the gene symbol matched. If the target name 
type is uniprot, then it converts all the ACs to primary. 
Then, for the Trembl IDs it looks up the preferred gene 
names, and find Swissprot IDs with the same preferred 
gene name.</p>
<dl class="docutils">
<dt>&#64;name <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>The original name which shall be converted.</dd>
<dt>&#64;nameType <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>The type of the name.
Available by default:
- genesymbol (gene name)
- entrez (Entrez Gene ID [#])
- refseqp (NCBI RefSeq Protein ID [NP_*|XP_*])
- ensp (Ensembl protein ID [ENSP*])
- enst (Ensembl transcript ID [ENST*])
- ensg (Ensembl genomic DNA ID [ENSG*])
- hgnc (HGNC ID [HGNC:#])
- gi (GI number [#])
- embl (DDBJ/EMBL/GeneBank CDS accession)
- embl_id (DDBJ/EMBL/GeneBank accession)
To use other IDs, you need to define the input method
and load the table before calling :py:func:Mapper.map_name().</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.mapping.Mapper.primary_uniprot">
<code class="descname">primary_uniprot</code><span class="sig-paren">(</span><em>lst</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#Mapper.primary_uniprot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.Mapper.primary_uniprot" title="Permalink to this definition">¶</a></dt>
<dd><p>For a list of UniProt IDs returns the list of primary ids.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.mapping.Mapper.trembl_swissprot">
<code class="descname">trembl_swissprot</code><span class="sig-paren">(</span><em>lst</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#Mapper.trembl_swissprot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.Mapper.trembl_swissprot" title="Permalink to this definition">¶</a></dt>
<dd><p>For a list of Trembl and Swissprot IDs, returns possibly
only Swissprot, mapping from Trembl to gene names, and 
then back to Swissprot.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.mapping.Mapper.which_table">
<code class="descname">which_table</code><span class="sig-paren">(</span><em>nameType</em>, <em>targetNameType</em>, <em>load=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#Mapper.which_table"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.Mapper.which_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the table which is suitable to convert an ID of 
nameType to targetNameType. If no such table have been loaded
yet, it attempts to load from UniProt.</p>
</dd></dl>

</dd></dl>

<span class="target" id="module-pypath.moduland"></span><span class="target" id="module-pypath.dataio"></span><dl class="class">
<dt id="pypath.dataio.ResidueMapper">
<em class="property">class </em><code class="descclassname">pypath.dataio.</code><code class="descname">ResidueMapper</code><a class="reference internal" href="_modules/pypath/dataio.html#ResidueMapper"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.ResidueMapper" title="Permalink to this definition">¶</a></dt>
<dd><p>This class stores and serves the PDB &#8211;&gt; UniProt 
residue level mapping. Attempts to download the 
mapping, and stores it for further use. Converts 
PDB residue numbers to the corresponding UniProt ones.</p>
<dl class="method">
<dt id="pypath.dataio.ResidueMapper.clean">
<code class="descname">clean</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#ResidueMapper.clean"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.ResidueMapper.clean" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes cached mappings, freeing up memory.</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="pypath.dataio.acsn_ppi">
<code class="descclassname">pypath.dataio.</code><code class="descname">acsn_ppi</code><span class="sig-paren">(</span><em>keep_in_complex_interactions=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#acsn_ppi"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.acsn_ppi" title="Permalink to this definition">¶</a></dt>
<dd><p>Processes ACSN data from local file.
Returns list of interactions.</p>
<dl class="docutils">
<dt>&#64;keep_in_complex_interactions <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd>Whether to include interactions from complex expansion.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.biogrid_interactions">
<code class="descclassname">pypath.dataio.</code><code class="descname">biogrid_interactions</code><span class="sig-paren">(</span><em>organism=9606</em>, <em>htp_limit=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#biogrid_interactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.biogrid_interactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes BioGRID interactions.
Keeps only the &#8220;low throughput&#8221; interactions.
Returns list of interactions.</p>
<dl class="docutils">
<dt>&#64;organism <span class="classifier-delimiter">:</span> <span class="classifier">int</span></dt>
<dd>NCBI Taxonomy ID of organism.</dd>
<dt>&#64;htp_limit <span class="classifier-delimiter">:</span> <span class="classifier">int</span></dt>
<dd>Exclude interactions only from references
cited at more than this number of interactions.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_acsn_effects">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_acsn_effects</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_acsn_effects"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_acsn_effects" title="Permalink to this definition">¶</a></dt>
<dd><p>Processes ACSN data, returns list of effects.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_ca1">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_ca1</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_ca1"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_ca1" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes the CA1 signaling network (Ma&#8217;ayan 2005).
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_ccmap">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_ccmap</code><span class="sig-paren">(</span><em>organism=9606</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_ccmap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_ccmap" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes CancerCellMap.
Returns list of interactions.</p>
<dl class="docutils">
<dt>&#64;organism <span class="classifier-delimiter">:</span> <span class="classifier">int</span></dt>
<dd>NCBI Taxonomy ID to match column #7 in nodes file.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_complexportal">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_complexportal</code><span class="sig-paren">(</span><em>species=9606</em>, <em>zipped=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_complexportal"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_complexportal" title="Permalink to this definition">¶</a></dt>
<dd><p>Complex dataset from IntAct.
See more:
<a class="reference external" href="http://www.ebi.ac.uk/intact/complex/">http://www.ebi.ac.uk/intact/complex/</a>
<a class="reference external" href="http://nar.oxfordjournals.org/content/early/2014/10/13/nar.gku975.full.pdf">http://nar.oxfordjournals.org/content/early/2014/10/13/nar.gku975.full.pdf</a></p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_dgidb">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_dgidb</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_dgidb"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_dgidb" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes the list of all human druggable proteins.
Returns a list of GeneSymbols.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_disgenet">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_disgenet</code><span class="sig-paren">(</span><em>dataset='curated'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_disgenet"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_disgenet" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes the list of all human disease related proteins
from DisGeNet.
Returns dict of dicts.</p>
<dl class="docutils">
<dt>&#64;dataset <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>Name of DisGeNet dataset to be obtained:
<cite>curated</cite>, <cite>literature</cite>, <cite>befree</cite> or <cite>all</cite>.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_domino_ptms">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_domino_ptms</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_domino_ptms"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_domino_ptms" title="Permalink to this definition">¶</a></dt>
<dd><p>The table comes from dataio.get_domino(), having the following fields:
header = [&#8216;uniprot-A&#8217;, &#8216;uniprot-B&#8217;, &#8216;isoform-A&#8217;, &#8216;isoform-B&#8217;, #3
&#8216;exp. method&#8217;, &#8216;references&#8217;, &#8216;taxon-A&#8217;, &#8216;taxon-B&#8217;, #7
&#8216;role-A&#8217;, &#8216;role-B&#8217;, &#8216;binding-site-range-A&#8217;, &#8216;binding-site-range-B&#8217;, #11
&#8216;domains-A&#8217;, &#8216;domains-B&#8217;, &#8216;ptm-residue-A&#8217;, &#8216;ptm-residue-B&#8217;, #15
&#8216;ptm-type-mi-A&#8217;, &#8216;ptm-type-mi-B&#8217;, &#8216;ptm-type-A&#8217;, &#8216;ptm-type-B&#8217;, #19
&#8216;ptm-res-name-A&#8217;, &#8216;ptm-res-name-B&#8217;, &#8216;mutations-A&#8217;, &#8216;mutations-B&#8217;, #23
&#8216;mutation-effects-A&#8217;, &#8216;mutation-effects-B&#8217;, &#8216;domains-interpro-A&#8217;, #26 
&#8216;domains-interpro-B&#8217;, &#8216;negative&#8217;] #28</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_elm_interactions">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_elm_interactions</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_elm_interactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_elm_interactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Downlods manually curated interactions from ELM.
This is the gold standard set of ELM.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_graphviz_attrs">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_graphviz_attrs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_graphviz_attrs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_graphviz_attrs" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads graphviz attribute list from graphviz.org.
Returns 3 dicts of dicts: graph_attrs, vertex_attrs and edge_attrs.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_guide2pharma">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_guide2pharma</code><span class="sig-paren">(</span><em>organism='human'</em>, <em>endogenous=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_guide2pharma"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_guide2pharma" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes Guide to Pharmacology data.
Returns list of dicts.</p>
<dl class="docutils">
<dt>&#64;organism <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>Name of the organism, e.g. <cite>human</cite>.</dd>
<dt>&#64;endogenous <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd>Whether to include only endogenous ligands interactions.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_hpmr">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_hpmr</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_hpmr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_hpmr" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes the list of all human receptors from
human receptor census (HPMR &#8211; Human Plasma Membrane Receptome).
Returns list of GeneSymbols.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_hprd">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_hprd</code><span class="sig-paren">(</span><em>in_vivo=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_hprd"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_hprd" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and preprocesses HPRD data.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_hprd_ptms">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_hprd_ptms</code><span class="sig-paren">(</span><em>in_vivo=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_hprd_ptms"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_hprd_ptms" title="Permalink to this definition">¶</a></dt>
<dd><p>Processes HPRD data and extracts PTMs.
Returns list of kinase-substrate interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_hsn">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_hsn</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_hsn"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_hsn" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes HumanSignalingNetwork version 6
(published 2014 Jan by Edwin Wang).
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_i3d">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_i3d</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_i3d"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_i3d" title="Permalink to this definition">¶</a></dt>
<dd><p>Interaction3D contains residue numbers in given chains in 
given PDB stuctures, so we need to add an offset to get the residue 
numbers valid for UniProt sequences. Offsets can be obtained from
Instruct, or from the Pfam PDB-chain-UniProt mapping table.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_instruct">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_instruct</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_instruct"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_instruct" title="Permalink to this definition">¶</a></dt>
<dd><p>Instruct contains residue numbers in UniProt sequences, it means
no further calculations of offsets in chains of PDB structures needed.
Chains are not given, only a set of PDB structures supporting the 
domain-domain // protein-protein interaction.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_instruct_offsets">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_instruct_offsets</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_instruct_offsets"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_instruct_offsets" title="Permalink to this definition">¶</a></dt>
<dd><p>These offsets should be understood as from UniProt to PDB.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_kinases">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_kinases</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_kinases"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_kinases" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes the list of all human kinases.
Returns a list of GeneSymbols.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_laudanna_directions">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_laudanna_directions</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_laudanna_directions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_laudanna_directions" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes the SignalingFlow edge attributes
from Laudanna Lab.
Returns list of directions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_laudanna_effects">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_laudanna_effects</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_laudanna_effects"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_laudanna_effects" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes the SignalingDirection edge attributes
from Laudanna Lab.
Returns list of effects.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_li2012">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_li2012</code><span class="sig-paren">(</span><em>filename=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_li2012"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_li2012" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads supplementary data of Li 2012 from local file.
Returns table (list of lists).</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_lit_bm_13">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_lit_bm_13</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_lit_bm_13"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_lit_bm_13" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes Lit-BM-13 dataset, the high confidence
literature curated interactions from CCSB.
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_phosphosite">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_phosphosite</code><span class="sig-paren">(</span><em>cache=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_phosphosite"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_phosphosite" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads curated and HTP data from Phosphosite,
from preprocessed cache file if available.
Processes BioPAX format.
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_phosphosite_curated">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_phosphosite_curated</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_phosphosite_curated"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_phosphosite_curated" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads literature curated PhosphoSite data,
from preprocessed cache file if available.
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_phosphosite_noref">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_phosphosite_noref</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_phosphosite_noref"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_phosphosite_noref" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads HTP PhosphoSite data,
from preprocessed cache file if available.
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_pmid">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_pmid</code><span class="sig-paren">(</span><em>idList</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_pmid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_pmid" title="Permalink to this definition">¶</a></dt>
<dd><p>For a list of doi or PMC IDs 
fetches the corresponding PMIDs.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_switches_elm">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_switches_elm</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_switches_elm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_switches_elm" title="Permalink to this definition">¶</a></dt>
<dd><p>switches.elm is a resource containing functional switches in molecular regulation,
in domain-motif level resolution, classified into categories according to their
mechanism.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_tfcensus">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_tfcensus</code><span class="sig-paren">(</span><em>classes=['a', 'b', 'other']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_tfcensus"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_tfcensus" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes list of all human transcripton factors.
Returns dict with lists of ENSGene IDs and HGNC Gene Names.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_wang_effects">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_wang_effects</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_wang_effects"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_wang_effects" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes Wang Lab HumanSignalingNetwork.
Returns list of effects.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.hprd_interactions">
<code class="descclassname">pypath.dataio.</code><code class="descname">hprd_interactions</code><span class="sig-paren">(</span><em>in_vivo=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#hprd_interactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.hprd_interactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Processes HPRD data and extracts interactions.
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.is_quoted">
<code class="descclassname">pypath.dataio.</code><code class="descname">is_quoted</code><span class="sig-paren">(</span><em>string</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#is_quoted"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.is_quoted" title="Permalink to this definition">¶</a></dt>
<dd><p>From <a class="reference external" href="http://stackoverflow.com/questions/1637762/test-if-string-is-url-encoded-in-php">http://stackoverflow.com/questions/1637762/test-if-string-is-url-encoded-in-php</a></p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.kegg_pathways">
<code class="descclassname">pypath.dataio.</code><code class="descname">kegg_pathways</code><span class="sig-paren">(</span><em>mapper=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#kegg_pathways"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.kegg_pathways" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes KEGG Pathways.
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.li2012_dmi">
<code class="descclassname">pypath.dataio.</code><code class="descname">li2012_dmi</code><span class="sig-paren">(</span><em>mapper=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#li2012_dmi"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.li2012_dmi" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts table read by <cite>pypath.dataio.get_li2012()</cite> to
list of <cite>pypath.intera.DomainMotif()</cite> objects.
Translates GeneSymbols to UniProt IDs.</p>
<dl class="docutils">
<dt>&#64;mapper <span class="classifier-delimiter">:</span> <span class="classifier">pypath.mapping.Mapper()</span></dt>
<dd>If not provided, a new <cite>Mapper()</cite> instance will be 
initialized, reserving more memory.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.li2012_interactions">
<code class="descclassname">pypath.dataio.</code><code class="descname">li2012_interactions</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#li2012_interactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.li2012_interactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts table read by <cite>pypath.dataio.get_li2012()</cite> to
list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.li2012_phospho">
<code class="descclassname">pypath.dataio.</code><code class="descname">li2012_phospho</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#li2012_phospho"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.li2012_phospho" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts table read by <cite>pypath.dataio.get_li2012()</cite> to
list of dicts of kinase-substrate interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.lmpid_dmi">
<code class="descclassname">pypath.dataio.</code><code class="descname">lmpid_dmi</code><span class="sig-paren">(</span><em>fname='LMPID_DATA_pubmed_ref.xml'</em>, <em>organism=9606</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#lmpid_dmi"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.lmpid_dmi" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts list of domain-motif interactions supplied by
<cite>pypath.dataio.load_lmpid()</cite> to list of 
<a href="#id1"><span class="problematic" id="id2">`</span></a>pypath.intera.DomainMotif() objects.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.lmpid_interactions">
<code class="descclassname">pypath.dataio.</code><code class="descname">lmpid_interactions</code><span class="sig-paren">(</span><em>fname='LMPID_DATA_pubmed_ref.xml'</em>, <em>organism=9606</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#lmpid_interactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.lmpid_interactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts list of domain-motif interactions supplied by
<cite>pypath.dataio.load_lmpid()</cite> to list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.load_lmpid">
<code class="descclassname">pypath.dataio.</code><code class="descname">load_lmpid</code><span class="sig-paren">(</span><em>fname='LMPID_DATA_pubmed_ref.xml'</em>, <em>organism=9606</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#load_lmpid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.load_lmpid" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads and processes LMPID data from local file
<cite>pypath.data/LMPID_DATA_pubmed_ref.xml</cite>.
The file was provided by LMPID authors and is now
redistributed with the module.
Returns list of domain-motif interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.load_macrophage">
<code class="descclassname">pypath.dataio.</code><code class="descname">load_macrophage</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#load_macrophage"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.load_macrophage" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads Macrophage from local file.
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.load_signor_ptms">
<code class="descclassname">pypath.dataio.</code><code class="descname">load_signor_ptms</code><span class="sig-paren">(</span><em>fname='signor_22052015.tab'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#load_signor_ptms"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.load_signor_ptms" title="Permalink to this definition">¶</a></dt>
<dd><p>This function is deprecated, you should not use it.
Loads and processes Signor PTMs from local file.
Returns dict of dicts.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.only_pmids">
<code class="descclassname">pypath.dataio.</code><code class="descname">only_pmids</code><span class="sig-paren">(</span><em>idList</em>, <em>strict=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#only_pmids"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.only_pmids" title="Permalink to this definition">¶</a></dt>
<dd><p>Return elements unchanged which compy to PubMed ID format,
and attempts to translate the DOIs and PMC IDs using NCBI
E-utils.
Returns list containing only PMIDs.</p>
<dl class="docutils">
<dt>&#64;idList <span class="classifier-delimiter">:</span> <span class="classifier">list, str</span></dt>
<dd>List of IDs or one single ID.</dd>
<dt>&#64;strict <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd>Whether keep in the list those IDs which are not PMIDs,
neither DOIs or PMC IDs or NIH manuscript IDs.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.open_pubmed">
<code class="descclassname">pypath.dataio.</code><code class="descname">open_pubmed</code><span class="sig-paren">(</span><em>pmid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#open_pubmed"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.open_pubmed" title="Permalink to this definition">¶</a></dt>
<dd><p>Opens PubMed record in web browser.</p>
<dl class="docutils">
<dt>&#64;pmid <span class="classifier-delimiter">:</span> <span class="classifier">str or int</span></dt>
<dd>PubMed ID</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.phosphosite_directions">
<code class="descclassname">pypath.dataio.</code><code class="descname">phosphosite_directions</code><span class="sig-paren">(</span><em>organism='human'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#phosphosite_directions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.phosphosite_directions" title="Permalink to this definition">¶</a></dt>
<dd><p>From curated and HTP PhosphoSite data generates a
list of directions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.reactions_biopax">
<code class="descclassname">pypath.dataio.</code><code class="descname">reactions_biopax</code><span class="sig-paren">(</span><em>biopax_file</em>, <em>organism=9606</em>, <em>protein_name_type='UniProt'</em>, <em>clean=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#reactions_biopax"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.reactions_biopax" title="Permalink to this definition">¶</a></dt>
<dd><p>Processes a BioPAX file and extracts binary interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.reactome_biopax">
<code class="descclassname">pypath.dataio.</code><code class="descname">reactome_biopax</code><span class="sig-paren">(</span><em>organism=9606</em>, <em>cache=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#reactome_biopax"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.reactome_biopax" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads Reactome human reactions in SBML format.
Returns File object.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.reactome_interactions">
<code class="descclassname">pypath.dataio.</code><code class="descname">reactome_interactions</code><span class="sig-paren">(</span><em>cacheFile=None</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#reactome_interactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.reactome_interactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes Reactome BioPAX.
Extracts binary interactions.
The applied criteria are very stringent, yields very few interactions.
Requires large free memory, approx. 2G.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.reactome_sbml">
<code class="descclassname">pypath.dataio.</code><code class="descname">reactome_sbml</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#reactome_sbml"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.reactome_sbml" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads Reactome human reactions in SBML format.
Returns gzip.GzipFile object.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.read_complexes_havugimana">
<code class="descclassname">pypath.dataio.</code><code class="descname">read_complexes_havugimana</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#read_complexes_havugimana"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.read_complexes_havugimana" title="Permalink to this definition">¶</a></dt>
<dd><p>Supplement Table S3/1 from Havugimana 2012
Cell. 150(5): 1068–1081.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.read_table">
<code class="descclassname">pypath.dataio.</code><code class="descname">read_table</code><span class="sig-paren">(</span><em>cols</em>, <em>fileObject=None</em>, <em>data=None</em>, <em>sep='\t'</em>, <em>sep2=None</em>, <em>rem=[]</em>, <em>hdr=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#read_table"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.read_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Generic function to read data tables.</p>
<dl class="docutils">
<dt>fileObject <span class="classifier-delimiter">:</span> <span class="classifier">file-like</span></dt>
<dd>Any file like object: file opened for read, or StringIO buffer</dd>
<dt>cols <span class="classifier-delimiter">:</span> <span class="classifier">dict</span></dt>
<dd>Dictionary of columns to read. Keys identifying fields are returned 
in the result. Values are column numbers.</dd>
<dt>sepLevel1 <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>Field separator of the file.</dd>
<dt>sepLevel2 <span class="classifier-delimiter">:</span> <span class="classifier">dict</span></dt>
<dd>Subfield separators and prefixes. 
E.g. {2: &#8216;,&#8217;, 3: &#8216;|&#8217;}</dd>
<dt>hdr <span class="classifier-delimiter">:</span> <span class="classifier">int</span></dt>
<dd>Number of header lines. If None, no headers assumed.</dd>
<dt>rem <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd>Strings to remove. For each line these elements will be replaced with &#8216;&#8217;.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.read_xls">
<code class="descclassname">pypath.dataio.</code><code class="descname">read_xls</code><span class="sig-paren">(</span><em>xls_file</em>, <em>sheet=''</em>, <em>csv_file=None</em>, <em>return_table=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#read_xls"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.read_xls" title="Permalink to this definition">¶</a></dt>
<dd><p>Generic function to read MS Excel XLS file, and convert one sheet
to CSV, or return as a list of lists</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.rolland_hi_ii_14">
<code class="descclassname">pypath.dataio.</code><code class="descname">rolland_hi_ii_14</code><span class="sig-paren">(</span><em>filename=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#rolland_hi_ii_14"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.rolland_hi_ii_14" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads the HI-II-14 unbiased interactome from the large scale screening
of from Rolland 2014.
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.signalink_interactions">
<code class="descclassname">pypath.dataio.</code><code class="descname">signalink_interactions</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#signalink_interactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.signalink_interactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads and processes SignaLink3 interactions from local file.
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.signor_interactions">
<code class="descclassname">pypath.dataio.</code><code class="descname">signor_interactions</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#signor_interactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.signor_interactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads the full dataset from Signor.
Return the file contents.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.signor_urls">
<code class="descclassname">pypath.dataio.</code><code class="descname">signor_urls</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#signor_urls"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.signor_urls" title="Permalink to this definition">¶</a></dt>
<dd><p>This function is deprecated.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.take_a_trip">
<code class="descclassname">pypath.dataio.</code><code class="descname">take_a_trip</code><span class="sig-paren">(</span><em>cachefile='trip.pickle'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#take_a_trip"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.take_a_trip" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads TRIP data from webpage and preprocesses it.
Saves preprocessed data into <cite>cachefile</cite> and next
time loads from this file.</p>
<dl class="docutils">
<dt>&#64;cachefile <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>Filename, located in <cite>./cache</cite>.
To disable cache, pass <cite>None</cite>.
To download again, remove file from <cite>./cache</cite>.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.trip_find_uniprot">
<code class="descclassname">pypath.dataio.</code><code class="descname">trip_find_uniprot</code><span class="sig-paren">(</span><em>soup</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#trip_find_uniprot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.trip_find_uniprot" title="Permalink to this definition">¶</a></dt>
<dd><p>Looks up a UniProt name in table downloaded from TRIP
webpage.</p>
<dl class="docutils">
<dt>&#64;soup <span class="classifier-delimiter">:</span> <span class="classifier">bs4.BeautifulSoup</span></dt>
<dd>The <cite>BeautifulSoup</cite> instance returned by <cite>pypath.dataio.trip_get_uniprot()</cite>.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.trip_get_uniprot">
<code class="descclassname">pypath.dataio.</code><code class="descname">trip_get_uniprot</code><span class="sig-paren">(</span><em>syn</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#trip_get_uniprot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.trip_get_uniprot" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads table from TRIP webpage and UniProt attempts to
look up the UniProt ID for one synonym.</p>
<dl class="docutils">
<dt>&#64;syn <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>The synonym as shown on TRIP webpage.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.trip_interactions">
<code class="descclassname">pypath.dataio.</code><code class="descname">trip_interactions</code><span class="sig-paren">(</span><em>exclude_methods=['Inference', 'Speculation'], predictions=False, species='Human', strict=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#trip_interactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.trip_interactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Obtains processed TRIP interactions by calling <cite>pypath.dataio.trip_process()</cite>
and returns list of interactions. All arguments are passed to 
<cite>trip_process()</cite>, see their definition there.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.trip_process">
<code class="descclassname">pypath.dataio.</code><code class="descname">trip_process</code><span class="sig-paren">(</span><em>exclude_methods=['Inference', 'Speculation'], predictions=False, species='Human', strict=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#trip_process"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.trip_process" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads TRIP data by calling <cite>pypath.dadio.take_a_trip()</cite> and
further provcesses it.
Returns dict of dict with TRIP data.</p>
<dl class="docutils">
<dt>&#64;exclude_methods <span class="classifier-delimiter">:</span> <span class="classifier">list</span></dt>
<dd>Interaction detection methods to be discarded.</dd>
<dt>&#64;predictions <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd>Whether to include predicted interactions.</dd>
<dt>&#64;species <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>Organism name, e.g. <cite>Human</cite>.</dd>
<dt>&#64;strict <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd>Whether include interactions with species not 
used as a bait or not specified.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.trip_process_table">
<code class="descclassname">pypath.dataio.</code><code class="descname">trip_process_table</code><span class="sig-paren">(</span><em>tab</em>, <em>result</em>, <em>intrs</em>, <em>trp_uniprot</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#trip_process_table"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.trip_process_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Processes one HTML table downloaded from TRIP webpage.</p>
<dl class="docutils">
<dt>&#64;tab <span class="classifier-delimiter">:</span> <span class="classifier">bs4.element.Tag()</span></dt>
<dd>One table of interactions from TRIP webpage.</dd>
<dt>&#64;result <span class="classifier-delimiter">:</span> <span class="classifier">dict</span></dt>
<dd>Dictionary the data should be filled in.</dd>
<dt>&#64;intrs <span class="classifier-delimiter">:</span> <span class="classifier">dict</span></dt>
<dd>Dictionary of already converted interactor IDs.
This serves as a cache so do not need to look up
the same ID twice.</dd>
<dt>&#64;trp_uniprot <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>UniProt ID of TRP domain containing protein.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.url_fix">
<code class="descclassname">pypath.dataio.</code><code class="descname">url_fix</code><span class="sig-paren">(</span><em>s</em>, <em>charset='utf-8'</em>, <em>force=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#url_fix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.url_fix" title="Permalink to this definition">¶</a></dt>
<dd><p>From <a class="reference external" href="http://stackoverflow.com/a/121017/854988">http://stackoverflow.com/a/121017/854988</a></p>
</dd></dl>

<span class="target" id="module-pypath.chembl"></span><span class="target" id="module-pypath.unichem"></span><span class="target" id="module-pypath.seq"></span><span class="target" id="module-pypath.mysql"></span><span class="target" id="module-pypath.mysql_connect"></span></div>


           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2016, Dénes Türei.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.1.72',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>