

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Welcome to pypath’s documentation! &mdash; pypath 0.5.1 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="pypath 0.5.1 documentation" href="#"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="#" class="icon icon-home"> pypath
          

          
          </a>

          
            
            
              <div class="version">
                0.5
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <!-- Local TOC -->
                <div class="local-toc"><ul>
<li><a class="reference internal" href="#">Welcome to pypath&#8217;s documentation!</a><ul>
<li><a class="reference internal" href="#source-code">Source code</a></li>
<li><a class="reference internal" href="#main-features">Main features</a></li>
</ul>
</li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
<li><a class="reference internal" href="#examples">Examples</a><ul>
<li><a class="reference internal" href="#example-1-building-the-network">Example 1: building the network</a></li>
<li><a class="reference internal" href="#example-2-using-the-mapper-class-for-translating-ids">Example 2: using the Mapper class for translating IDs</a></li>
<li><a class="reference internal" href="#example-3-pathways-annotations">Example 3: pathways annotations</a></li>
<li><a class="reference internal" href="#example-4-other-functional-annotations">Example 4: other functional annotations</a></li>
<li><a class="reference internal" href="#example-5-bypass-cache">Example 5: bypass cache</a></li>
<li><a class="reference internal" href="#example-6-saving-and-loading-a-session">Example 6: saving and loading a session:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#how-to-set-up-a-chembl-mysql-instance">How to set up a ChEMBL MySQL instance?</a></li>
<li><a class="reference internal" href="#how-to-access-omnipath-data-by-bioservices">How to access OmniPath data by bioservices</a></li>
<li><a class="reference internal" href="#reference">Reference</a></li>
</ul>
</div>
            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="#">pypath</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          

 



<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="#">Docs</a> &raquo;</li>
      
    <li>Welcome to pypath&#8217;s documentation!</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/index.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="welcome-to-pypath-s-documentation">
<h1>Welcome to pypath&#8217;s documentation!<a class="headerlink" href="#welcome-to-pypath-s-documentation" title="Permalink to this headline">¶</a></h1>
<p>pypath is a Python module for cellular signaling pathways analysis. It is accompanied and developed together with a high confidence, literature curated signaling network, <a class="reference external" href="http://omnipathdb.org/">OmniPath</a>.</p>
<div class="section" id="source-code">
<h2>Source code<a class="headerlink" href="#source-code" title="Permalink to this headline">¶</a></h2>
<p>pypath is free software, licensed under <a class="reference external" href="http://www.gnu.org/licenses/gpl.html">GPLv3</a>.
The code is available at <a class="reference external" href="https://github.com/saezlab/pypath">github</a>.
Recent package for PIP is available at <a class="reference external" href="http://pypath.omnipathdb.org/releases/latest/">http://pypath.omnipathdb.org/releases/latest/</a>, archives are available at <a class="reference external" href="http://pypath.omnipathdb.org/releases/archive/">http://pypath.omnipathdb.org/releases/archive/</a>.</p>
</div>
<div class="section" id="main-features">
<h2>Main features<a class="headerlink" href="#main-features" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Undirected or directed networks.</li>
<li>Easy, often seemless protein ID conversion.</li>
<li>Efficient handling of annotations, especially sources, literature references, directions, effect signs (stimulation/inhibition) and enzyme-substrate interactions.</li>
<li>Ready integration of dozens of bioinformatics resources, most of them being downloaded on the fly, directly from the original source.</li>
<li>Caching: after downloading data, a local copy saved, so faster and offline run is possible afterwards.</li>
<li>Based on <a class="reference external" href="http://igraph.org/">igraph</a>: plenty of graph methods are available with excellent computational performance.</li>
<li>Partial support for non-human species.</li>
<li>Partial support for other molecular species than proteins.</li>
</ul>
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<blockquote>
<div><ul class="simple">
<li><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></li>
<li><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="examples">
<h1>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h1>
<div class="section" id="example-1-building-the-network">
<h2>Example 1: building the network<a class="headerlink" href="#example-1-building-the-network" title="Permalink to this headline">¶</a></h2>
<p>Import the module and create an instance:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pypath</span>
<span class="n">pa</span> <span class="o">=</span> <span class="n">pypath</span><span class="o">.</span><span class="n">PyPath</span><span class="p">()</span>
</pre></div>
</div>
<p>The tables for ID conversions, and the network of the selected manually curated pathway databases (<a class="reference external" href="http://omnipathdb.org/">OmniPath</a>) can be initialized this way:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">pa</span><span class="o">.</span><span class="n">init_network</span><span class="p">()</span>
</pre></div>
</div>
<p>Resources without references for each interaction are available separately. For example, to build a comprehensive phosphonetwork, you might want to load interactions from high-throughput data in PhosphoSite, kinase-substrate relationships from PhosphoNetworks and MIMP. This is the way to do it:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pypath.data_formats</span> <span class="kn">import</span> <span class="n">ptm_misc</span>
<span class="n">pa</span><span class="o">.</span><span class="n">load_resources</span><span class="p">(</span><span class="n">lst</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;mimp&#39;</span><span class="p">:</span> <span class="n">ptm_misc</span><span class="p">[</span><span class="s1">&#39;mimp&#39;</span><span class="p">]})</span>
<span class="n">pa</span><span class="o">.</span><span class="n">load_resources</span><span class="p">(</span><span class="n">lst</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;pnetworks&#39;</span><span class="p">:</span> <span class="n">ptm_misc</span><span class="p">[</span><span class="s1">&#39;pnetworks&#39;</span><span class="p">]})</span>
<span class="n">pa</span><span class="o">.</span><span class="n">load_resources</span><span class="p">(</span><span class="n">lst</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;psite_noref&#39;</span><span class="p">:</span> <span class="n">ptm_misc</span><span class="p">[</span><span class="s1">&#39;psite_noref&#39;</span><span class="p">]})</span>
</pre></div>
</div>
<p>In case you have your own interaction data, and you wish to merge this into the network, you first need to define its format, and then call the load_resource() function. In this simple example, the proteins are noted by Gene Symbols, their names are in columns #0 and #1, while from column #2 an edge attribute named &#8216;score&#8217; will be read. The file is tab separated, and the file name is &#8216;mylist.sif&#8217;. (This won&#8217;t work unless you really have a file like this.)</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">mylist</span> <span class="o">=</span> <span class="n">pypath</span><span class="o">.</span><span class="n">input_formats</span><span class="o">.</span><span class="n">ReadSettings</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;mylist&quot;</span><span class="p">,</span>
    <span class="n">separator</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">nameColA</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">nameColB</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
    <span class="n">nameTypeA</span> <span class="o">=</span> <span class="s2">&quot;genesymbol&quot;</span><span class="p">,</span> <span class="n">nameTypeB</span> <span class="o">=</span> <span class="s2">&quot;genesymbol&quot;</span><span class="p">,</span>
    <span class="n">isDirected</span> <span class="o">=</span> <span class="bp">False</span><span class="p">,</span> <span class="n">inFile</span> <span class="o">=</span> <span class="s1">&#39;mylist.sif&#39;</span><span class="p">,</span>
    <span class="n">extraEdgeAttrs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;score&#39;</span><span class="p">:</span> <span class="mi">2</span>
    <span class="p">})</span>
<span class="n">pa</span><span class="o">.</span><span class="n">load_resource</span><span class="p">(</span><span class="n">mylist</span><span class="p">)</span>
</pre></div>
</div>
<p>Although proteins are identified by UniProt IDs, it is useful to have standard gene symbols as labels, to make things more human understandable.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">pa</span><span class="o">.</span><span class="n">genesymbol_labels</span><span class="p">()</span>
</pre></div>
</div>
<p>Now the igraph object is in pa.graph, vertex and edge attributes are accessible the usual way:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">vs</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;name&#39;</span><span class="p">]</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">vs</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;label&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>Let&#8217;s see the default attributes of edges:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;sources&#39;</span><span class="p">]</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;references&#39;</span><span class="p">]</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;dirs&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>The graph is undirected by default, and &#8216;Direction&#8217; objects can be found in &lt;edge&gt;[&#8216;dirs&#8217;] to describe the directions and signs of the edge:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="k">print</span> <span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">111</span><span class="p">][</span><span class="s1">&#39;dirs&#39;</span><span class="p">]</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">111</span><span class="p">][</span><span class="s1">&#39;dirs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">is_directed</span><span class="p">()</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">111</span><span class="p">][</span><span class="s1">&#39;dirs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">is_inhibition</span><span class="p">()</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">111</span><span class="p">][</span><span class="s1">&#39;dirs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">which_dirs</span><span class="p">()</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">111</span><span class="p">][</span><span class="s1">&#39;dirs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">get_dir</span><span class="p">(</span><span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">111</span><span class="p">][</span><span class="s1">&#39;dirs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">which_dirs</span><span class="p">()[</span><span class="mi">0</span><span class="p">])</span>
</pre></div>
</div>
<p>To convert the igraph object to a directed graph, use this function:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">pa</span><span class="o">.</span><span class="n">get_directed</span><span class="p">()</span>
<span class="n">pa</span><span class="o">.</span><span class="n">dgraph</span><span class="o">.</span><span class="n">ecount</span><span class="p">()</span>
<span class="n">pa</span><span class="o">.</span><span class="n">dgraph</span><span class="o">.</span><span class="n">vcount</span><span class="p">()</span>
</pre></div>
</div>
<p>To replace the graph with the giant component of the directed graph:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">pa</span><span class="o">.</span><span class="n">graph</span> <span class="o">=</span> <span class="n">pa</span><span class="o">.</span><span class="n">get_giant</span><span class="p">(</span><span class="n">graph</span> <span class="o">=</span> <span class="n">pa</span><span class="o">.</span><span class="n">dgraph</span><span class="p">)</span>
</pre></div>
</div>
<p>Get compounds for all the proteins in the network:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">pa</span><span class="o">.</span><span class="n">compounds_from_chembl</span><span class="p">((</span><span class="bp">None</span><span class="p">,</span> <span class="s1">&#39;chembl_ebi&#39;</span><span class="p">),</span> <span class="n">pchembl</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<p>This works only if you have a MySQL server hosting an instance of ChEMBL:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">pa</span><span class="o">.</span><span class="n">compounds_from_chembl</span><span class="p">((</span><span class="bp">None</span><span class="p">,</span> <span class="s1">&#39;chembl&#39;</span><span class="p">),</span> <span class="n">pchembl</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<p>You can find the list of compounds and the requested data (now only the pchembl values) in vertex attributes:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">vs</span><span class="p">[</span><span class="mi">2</span><span class="p">][</span><span class="s1">&#39;compounds_chembl&#39;</span><span class="p">]</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">vs</span><span class="p">[</span><span class="mi">2</span><span class="p">][</span><span class="s1">&#39;compounds_data&#39;</span><span class="p">]</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">vs</span><span class="p">[</span><span class="mi">2</span><span class="p">][</span><span class="s1">&#39;compounds_data&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;pchembl&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>Loading PTMs. This function loads PTMs from all the available resources. It takes long time first, but once it saves the downloaded data under <cite>./cache</cite> directory, it is much faster to run again.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">pa</span><span class="o">.</span><span class="n">load_ptms</span><span class="p">()</span>
</pre></div>
</div>
<p>Individual PTM resources can be loaded the following way:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">pa</span><span class="o">.</span><span class="n">load_phospho_dmi</span><span class="p">(</span><span class="n">source</span> <span class="o">=</span> <span class="s1">&#39;phosphoELM&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>In the latter case, this function is needed to merge the identical PTMs loaded from multiple resources:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">pa</span><span class="o">.</span><span class="n">uniq_ptms</span><span class="p">()</span>
</pre></div>
</div>
<p>PTMs are stored in objects. This example shows how to access the type of modification, and the number and name of the modified residue.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">70</span><span class="p">][</span><span class="s1">&#39;ptm&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">ptm</span><span class="o">.</span><span class="n">typ</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">70</span><span class="p">][</span><span class="s1">&#39;ptm&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">ptm</span><span class="o">.</span><span class="n">protein</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">70</span><span class="p">][</span><span class="s1">&#39;ptm&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">ptm</span><span class="o">.</span><span class="n">residue</span><span class="o">.</span><span class="n">name</span>
<span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">es</span><span class="p">[</span><span class="mi">70</span><span class="p">][</span><span class="s1">&#39;ptm&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">ptm</span><span class="o">.</span><span class="n">residue</span><span class="o">.</span><span class="n">number</span>
</pre></div>
</div>
</div>
<div class="section" id="example-2-using-the-mapper-class-for-translating-ids">
<h2>Example 2: using the Mapper class for translating IDs<a class="headerlink" href="#example-2-using-the-mapper-class-for-translating-ids" title="Permalink to this headline">¶</a></h2>
<p>The mapping submodule of pypath can be used for ID conversion. Here is a basic example:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pypath</span> <span class="kn">import</span> <span class="n">mapping</span>
<span class="n">m</span> <span class="o">=</span> <span class="n">mapping</span><span class="o">.</span><span class="n">Mapper</span><span class="p">(</span><span class="mi">9606</span><span class="p">)</span>
<span class="n">result</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">gene_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;EGFR&#39;</span><span class="p">,</span> <span class="s1">&#39;AKT1&#39;</span><span class="p">,</span> <span class="s1">&#39;GABARAPL1&#39;</span><span class="p">,</span> <span class="s1">&#39;TP53&#39;</span><span class="p">]</span>
  <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">gene_list</span><span class="p">:</span>
    <span class="n">result</span><span class="p">[</span><span class="n">g</span><span class="p">]</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">map_name</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="s1">&#39;genesymbol&#39;</span><span class="p">,</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="example-3-pathways-annotations">
<h2>Example 3: pathways annotations<a class="headerlink" href="#example-3-pathways-annotations" title="Permalink to this headline">¶</a></h2>
<p>Pathways are functional annotations of molecules in molecular networks. Currently pathway annotations from 4 sources are available in pypath: from KEGG, NetPath, SignaLink and Signor. NetPath and SignaLink will be loaded automatically with the network data, the other two need to be loaded separately:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">pa</span><span class="o">.</span><span class="n">kegg_pathways</span><span class="p">()</span>
<span class="n">pa</span><span class="o">.</span><span class="n">signor_pathways</span><span class="p">()</span>
</pre></div>
</div>
<p>SignaLink assigns pathway annotations to proteins, while the other resources assignes this to interactions (at least the data read this way). In addition, proteins classified as autophagy proteins in the Autophagy Regulatory Network will appear as a SignaLink pathway named <cite>Autophagy</cite>. To have uniform annotations (from all sources, for both proteins and interactions, with <code class="docutils literal"><span class="pre">&lt;source&gt;_pathways</span></code> attribute names), use this method to do the necessary conversion:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">pa</span><span class="o">.</span><span class="n">pathway_attributes</span><span class="p">()</span>
</pre></div>
</div>
<p>After this, you can access the pathway annotations in <cite>kegg_pathways</cite>, <cite>netpath_pathways</cite>, <cite>signalink_pathways</cite> and <cite>signor_pathways</cite> edge and vertex attributes:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="k">print</span> <span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">vs</span><span class="p">[</span><span class="mi">333</span><span class="p">][</span><span class="s1">&#39;signor_pathways&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>You can simply do all the steps above by calling one method:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">pa</span><span class="o">.</span><span class="n">load_all_pathways</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="section" id="example-4-other-functional-annotations">
<h2>Example 4: other functional annotations<a class="headerlink" href="#example-4-other-functional-annotations" title="Permalink to this headline">¶</a></h2>
<p>Pathways as defined above usually follow the well known ways of information flow as it is presented in textbooks and reviews. But they are incomplete and biased. More complete and less biased functional annotations are Gene Ontology and GeneSets from the Molecular Signature Database. Methods are available in pypath, so you can load these annotations into network attributes.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1"># load only the biological process aspect:</span>
<span class="n">pa</span><span class="o">.</span><span class="n">load_go</span><span class="p">([</span><span class="s1">&#39;P&#39;</span><span class="p">])</span>
<span class="c1"># get the GO BP terms for AKT1:</span>
<span class="n">pa</span><span class="o">.</span><span class="n">gs</span><span class="p">(</span><span class="s1">&#39;AKT1&#39;</span><span class="p">)[</span><span class="s1">&#39;go&#39;</span><span class="p">][</span><span class="s1">&#39;P&#39;</span><span class="p">]</span>
<span class="c1"># get the GO annotation:</span>
<span class="n">pa</span><span class="o">.</span><span class="n">go_dict</span><span class="p">()</span>
<span class="c1"># list names instead of IDs:</span>
<span class="c1"># (9606 is an NCBI taxonomy ID)</span>
<span class="nb">map</span><span class="p">(</span><span class="n">pa</span><span class="o">.</span><span class="n">go</span><span class="p">[</span><span class="mi">9606</span><span class="p">]</span><span class="o">.</span><span class="n">get_name</span><span class="p">,</span> <span class="n">pa</span><span class="o">.</span><span class="n">gs</span><span class="p">(</span><span class="s1">&#39;AKT1&#39;</span><span class="p">)[</span><span class="s1">&#39;go&#39;</span><span class="p">][</span><span class="s1">&#39;P&#39;</span><span class="p">])</span>
<span class="c1"># calculate enrichment:</span>
<span class="c1"># this is a simple Fisher test</span>
<span class="c1"># with multiple p-values correction</span>
<span class="c1"># for example, get the enriched terms for the Notch pathway:</span>
<span class="n">pa</span><span class="o">.</span><span class="n">load_all_pathways</span><span class="p">()</span>
<span class="n">notch</span> <span class="o">=</span> <span class="n">pa</span><span class="o">.</span><span class="n">pathway_members</span><span class="p">(</span><span class="s1">&#39;Notch&#39;</span><span class="p">,</span> <span class="s1">&#39;signalink&#39;</span><span class="p">)</span>
<span class="n">pa</span><span class="o">.</span><span class="n">go_dict</span><span class="p">()</span>
<span class="n">enr</span> <span class="o">=</span> <span class="n">pa</span><span class="o">.</span><span class="n">go_enrichment</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">notch</span><span class="o">.</span><span class="n">up</span><span class="p">()))</span>
<span class="k">print</span> <span class="n">enr</span>
<span class="n">enr</span><span class="o">.</span><span class="n">toplist</span><span class="p">()</span>
<span class="n">enr</span><span class="o">.</span><span class="n">top_terms</span><span class="p">()</span>
<span class="n">enr</span><span class="o">.</span><span class="n">top_ids</span><span class="p">()</span>
<span class="n">enr</span><span class="o">.</span><span class="n">enrichments</span><span class="p">[</span><span class="n">enr</span><span class="o">.</span><span class="n">top_ids</span><span class="p">()[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">pval_adj</span>
<span class="n">enr</span><span class="o">.</span><span class="n">enrichments</span><span class="p">[</span><span class="n">enr</span><span class="o">.</span><span class="n">top_ids</span><span class="p">()[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">significant</span><span class="p">()</span>
</pre></div>
</div>
<p>Using GeneSets from MSigDB:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pypath</span> <span class="kn">import</span> <span class="n">gsea</span>
<span class="c1"># login with an MSigDB user:</span>
<span class="n">g</span> <span class="o">=</span> <span class="n">gsea</span><span class="o">.</span><span class="n">GSEA</span><span class="p">(</span><span class="s1">&#39;user@email.org&#39;</span><span class="p">,</span> <span class="n">mapper</span> <span class="o">=</span> <span class="n">pa</span><span class="o">.</span><span class="n">mapper</span><span class="p">)</span>
<span class="n">g</span><span class="o">.</span><span class="n">show_collections</span><span class="p">()</span>
<span class="n">g</span><span class="o">.</span><span class="n">load_collection</span><span class="p">(</span><span class="s1">&#39;CGP&#39;</span><span class="p">)</span>
<span class="c1"># genesets are loaded into g.sets</span>
<span class="n">enr</span> <span class="o">=</span> <span class="n">gsea</span><span class="o">.</span><span class="n">GSEABinaryEnrichmentSet</span><span class="p">(</span><span class="n">basic_set</span> <span class="o">=</span> <span class="n">pa</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">vs</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">],</span> <span class="n">gsea</span> <span class="o">=</span> <span class="n">g</span><span class="p">)</span>
<span class="c1"># or ``basic_set`` could be ``dataio.all_uniprots()``</span>
<span class="n">enr</span><span class="o">.</span><span class="n">new_set</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">notch</span><span class="o">.</span><span class="n">up</span><span class="p">()))</span>
<span class="c1"># if you see nothing, it means none of the loaded sets</span>
<span class="c1"># are enriched in your list</span>
<span class="k">print</span> <span class="n">enr</span>
<span class="n">enr</span><span class="o">.</span><span class="n">top_genesets</span><span class="p">()</span>
<span class="n">enr</span><span class="o">.</span><span class="n">top_geneset_ids</span><span class="p">()</span>
<span class="c1"># to do the same with methods of the PyPath() object:</span>
<span class="n">pa</span><span class="o">.</span><span class="n">init_gsea</span><span class="p">(</span><span class="s1">&#39;user@email.org&#39;</span><span class="p">)</span>
<span class="n">pa</span><span class="o">.</span><span class="n">add_genesets</span><span class="p">([</span><span class="s1">&#39;CP:KEGG&#39;</span><span class="p">])</span>
<span class="n">enr</span> <span class="o">=</span> <span class="n">pa</span><span class="o">.</span><span class="n">geneset_enrichment</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">notch</span><span class="o">.</span><span class="n">up</span><span class="p">()),</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="example-5-bypass-cache">
<h2>Example 5: bypass cache<a class="headerlink" href="#example-5-bypass-cache" title="Permalink to this headline">¶</a></h2>
<p>Cache makes pypath run much faster. A typical session downloads hundreds MBs of data from dozens of sources, and it takes minutes to do this. In addition it makes pypath sensitive to network connectivity and speed. After the first download, files are saved in <code class="docutils literal"><span class="pre">./cache</span></code> directory, and files with the same URL will be automatically read from there. However, sometimes it is necessary to bypass the cache and download the files again. For example if we suspect there are erroneous or old files there. There is an easy way to disable the cache temporarily while executing data input methods:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1"># here we use the old cache:</span>
<span class="n">pa</span><span class="o">.</span><span class="n">load_signor_ptms</span><span class="p">()</span>

<span class="k">with</span> <span class="n">pypath</span><span class="o">.</span><span class="n">dataio</span><span class="o">.</span><span class="n">cache_off</span><span class="p">():</span>
    <span class="c1"># here we don not read from the cache</span>
    <span class="c1"># but download again and write the</span>
    <span class="c1"># new files into the cache:</span>
    <span class="n">pa</span><span class="o">.</span><span class="n">load_signor_ptms</span><span class="p">()</span>

<span class="c1"># here already the new files are used from the cache:</span>
<span class="n">pa</span><span class="o">.</span><span class="n">load_signor_ptms</span><span class="p">()</span>

<span class="c1"># similarly, if the cache is turned off by default,</span>
<span class="c1"># we can temporarily enable:</span>

<span class="c1"># this way we permanently disable the cache:</span>
<span class="n">pypath</span><span class="o">.</span><span class="n">dataio</span><span class="o">.</span><span class="n">CACHE</span> <span class="o">=</span> <span class="bp">False</span>

<span class="c1"># and here temporarily enable:</span>
<span class="k">with</span> <span class="n">pypath</span><span class="o">.</span><span class="n">dataio</span><span class="o">.</span><span class="n">cache_on</span><span class="p">():</span>
    <span class="n">human_proteome</span> <span class="o">=</span> <span class="n">pypath</span><span class="o">.</span><span class="n">dataio</span><span class="o">.</span><span class="n">all_uniprots</span><span class="p">()</span>

<span class="c1"># the cache is permanently enabled if this variable is ``None`` or ``True``:</span>
<span class="n">pypath</span><span class="o">.</span><span class="n">dataio</span><span class="o">.</span><span class="n">CACHE</span> <span class="o">=</span> <span class="bp">None</span>
</pre></div>
</div>
<p>I plan to introduce more methods to give a more flexible control over the files in cache.</p>
</div>
<div class="section" id="example-6-saving-and-loading-a-session">
<h2>Example 6: saving and loading a session:<a class="headerlink" href="#example-6-saving-and-loading-a-session" title="Permalink to this headline">¶</a></h2>
<p>The network object with its attributes can be saved into a pickle dump, and loaded from there in subsequent sessions.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1"># initialize a PyPath() object:</span>
<span class="n">pa</span> <span class="o">=</span> <span class="n">pypath</span><span class="o">.</span><span class="n">PyPath</span><span class="p">()</span>
<span class="n">pa</span><span class="o">.</span><span class="n">init_network</span><span class="p">()</span>
<span class="n">pa</span><span class="o">.</span><span class="n">load_all_pathways</span><span class="p">()</span>

<span class="c1"># here we save the loaded network</span>
<span class="c1"># with the pathway annotations:</span>
<span class="n">pa</span><span class="o">.</span><span class="n">save_network</span><span class="p">()</span>

<span class="c1"># in another session we load the saved network:</span>
<span class="n">pa</span> <span class="o">=</span> <span class="n">pypath</span><span class="o">.</span><span class="n">PyPath</span><span class="p">()</span>
<span class="n">pa</span><span class="o">.</span><span class="n">init_network</span><span class="p">(</span><span class="n">pfile</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>

<span class="c1"># above the network has been saved into</span>
<span class="c1"># `cache/default_network.pickle`</span>
<span class="c1"># to save/load to/from different file:</span>

<span class="n">pa</span><span class="o">.</span><span class="n">save_network</span><span class="p">(</span><span class="s1">&#39;cache/other_network.pickle&#39;</span><span class="p">)</span>
<span class="n">pa</span> <span class="o">=</span> <span class="n">pypath</span><span class="o">.</span><span class="n">PyPath</span><span class="p">()</span>
<span class="n">pa</span><span class="o">.</span><span class="n">init_network</span><span class="p">(</span><span class="n">pfile</span> <span class="o">=</span> <span class="s1">&#39;cache/other_network.pickle&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="how-to-set-up-a-chembl-mysql-instance">
<h1>How to set up a ChEMBL MySQL instance?<a class="headerlink" href="#how-to-set-up-a-chembl-mysql-instance" title="Permalink to this headline">¶</a></h1>
<p>Currently <code class="docutils literal"><span class="pre">pypath.chembl</span></code> gives some powerful and flexible methods to query compound-target relationships from ChEMBL.
This is implemented using MySQL, and so far I could not find a way to provide the same features with using the webservice.
Using the webservice would be much more convenient for most of the users, so it is only matter of time and I will implement a webservice based ChEMBL module.
Until then, here is a short guide to load ChEMBL on your own MySQL server. To do this, you will need 25GB of free disk space. ChEMBL is huge. 1GB is the downloaded compressed database dump, 8GB is the same uncompressed, and 16GB is the database loaded into MySQL. You can delete the 2 former, so at the end you will sacrifice only 16GB to have your own ChEMBL.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span><span class="c1"># login to the MySQL shell as administrator:</span>

mysql --user<span class="o">=</span>root --password<span class="o">=</span>foobar <span class="o">[</span>--host<span class="o">=</span>217.0.0.1 --port<span class="o">=</span>3306<span class="o">]</span>
</pre></div>
</div>
<div class="highlight-mysql"><div class="highlight"><pre><span></span><span class="cm">/* create a database and user for ChEMBL: */</span>
<span class="k">CREATE</span> <span class="k">DATABASE</span> <span class="n">chembl</span><span class="p">;</span>
<span class="k">CREATE</span> <span class="n">USER</span> <span class="n">chembl</span><span class="p">;</span>
<span class="k">GRANT</span> <span class="k">ALL</span> <span class="k">ON</span> <span class="n">chembl</span><span class="p">.</span><span class="o">*</span> <span class="k">TO</span> <span class="ss">`chembl`</span><span class="o">@</span><span class="ss">`%`</span> <span class="n">IDENTIFIED</span> <span class="k">BY</span> <span class="s1">&#39;a-new-password&#39;</span><span class="p">;</span>
<span class="k">FLUSH</span> <span class="n">PRIVILEGES</span><span class="p">;</span>

<span class="cm">/* optionally create a user which can only read, e.g. if you want</span>
<span class="cm">to share this database with the whole institute: */</span>
<span class="k">CREATE</span> <span class="n">USER</span> <span class="n">chembl_ro</span><span class="p">;</span>
<span class="k">GRANT</span> <span class="k">SELECT</span> <span class="k">ON</span> <span class="n">chembl</span><span class="p">.</span><span class="o">*</span> <span class="k">TO</span> <span class="ss">`chembl_ro`</span><span class="o">@</span><span class="ss">`%`</span> <span class="n">IDENTIFIED</span> <span class="k">BY</span> <span class="s1">&#39;a-new-password&#39;</span><span class="p">;</span>
<span class="k">FLUSH</span> <span class="n">PRIVILEGES</span><span class="p">;</span>

<span class="k">EXIT</span><span class="p">;</span>
</pre></div>
</div>
<div class="highlight-bash"><div class="highlight"><pre><span></span><span class="c1"># let&#39;s download the ChEMBL MySQL dump (warning, it is 1GB!):</span>
curl -O ftp://ftp.ebi.ac.uk/pub/databases/chembl/ChEMBLdb/latest/chembl_21_mysql.tar.gz

<span class="c1"># uncompressing. start this only if you have 8GB free space:</span>
tar -xzvf chembl_21_mysql.tar.gz

<span class="c1"># loading into mysql. this will take up 16GB more space:</span>
mysql --user<span class="o">=</span>chembl --password<span class="o">=</span><span class="s1">&#39;password-of-chembl&#39;</span> <span class="o">[</span>--host<span class="o">=</span>217.0.0.1 --port<span class="o">=</span>3306<span class="o">]</span> &lt; <span class="se">\</span>
    chembl_21_mysql/chembl_21.mysqldump.sql
</pre></div>
</div>
<p>At this point you can query ChEMBL on your own laptop/server. Congratulations! Enjoy!
You might find it necessary to increase the cache and buffer sizes in <code class="docutils literal"><span class="pre">/etc/mysql/my.cnf</span></code>.
Loading the mysqldump takes time, if it fails try it again, but before check if you have enough free space and see the error log of MySQL.</p>
<p>Alternatively, you can download and run the <a class="reference external" href="http://chembl.blogspot.co.uk/2014/06/mychembl-launchpadlaunched.html">myChEMBL virtual machine</a>, and connect to its pre-installed MySQL server via the VirtualBox virtual network.</p>
</div>
<div class="section" id="how-to-access-omnipath-data-by-bioservices">
<h1>How to access OmniPath data by <a class="reference external" href="http://pythonhosted.org/bioservices/">bioservices</a><a class="headerlink" href="#how-to-access-omnipath-data-by-bioservices" title="Permalink to this headline">¶</a></h1>
<p>Bioservices is a Python module to access various webserices. If you need only the network data and kinase-substrate interactions, and you would like to proccess the data further your own way, bioservices offers a convenient way to fetch the data directly from the OmniPath webservice into Python objects.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">bioservices</span> <span class="kn">import</span> <span class="n">omnipath</span>
<span class="n">op</span> <span class="o">=</span> <span class="n">omnipath</span><span class="o">.</span><span class="n">OmniPath</span><span class="p">()</span>

<span class="c1"># all interactions:</span>
<span class="n">i</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">get_interactions</span><span class="p">()</span>

<span class="c1"># or only those of EGFR1 but include the sources and references fields:</span>
<span class="n">i</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">get_interactions</span><span class="p">(</span><span class="s1">&#39;P00533&#39;</span><span class="p">,</span> <span class="n">fields</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sources&#39;</span><span class="p">,</span> <span class="s1">&#39;references&#39;</span><span class="p">])</span>

<span class="c1"># get kinase-substrate interactions of EGFR1 and PRKCA:</span>
<span class="n">ks</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">get_ptms</span><span class="p">(</span><span class="s1">&#39;P00533,Q02156&#39;</span><span class="p">,</span> <span class="n">fields</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sources&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="section" id="reference">
<h1>Reference<a class="headerlink" href="#reference" title="Permalink to this headline">¶</a></h1>
<span class="target" id="module-pypath.main"></span><dl class="class">
<dt id="pypath.main.PyPath">
<em class="property">class </em><code class="descclassname">pypath.main.</code><code class="descname">PyPath</code><span class="sig-paren">(</span><em>ncbi_tax_id=9606</em>, <em>default_name_type={'protein': 'uniprot'</em>, <em>'drug': 'chembl'</em>, <em>'mirna': 'mirbase'}</em>, <em>copy=None</em>, <em>mysql=(None</em>, <em>'mapping')</em>, <em>chembl_mysql=(None</em>, <em>'chembl')</em>, <em>name='unnamed'</em>, <em>outdir='results'</em>, <em>loglevel='INFO'</em>, <em>loops=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.main.PyPath.add_update_vertex">
<code class="descname">add_update_vertex</code><span class="sig-paren">(</span><em>defAttrs</em>, <em>originalName</em>, <em>originalNameType</em>, <em>extraAttrs={}</em>, <em>add=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.add_update_vertex"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.add_update_vertex" title="Permalink to this definition">¶</a></dt>
<dd><p>Updates the attributes of one node in the network.
Optionally it creates a new node and sets the attributes,
but it is not efficient as igraph needs to reindex vertices 
after this operation, so better to create new nodes and 
edges in batch.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.all_between">
<code class="descname">all_between</code><span class="sig-paren">(</span><em>nameA</em>, <em>nameB</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.all_between"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.all_between" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all edges between two given vertex names. Similar to 
straight_between(), but checks both directions, and returns 
list of edge ids in [undirected, straight, reveresed] format,
for both nameA -&gt; nameB and nameB -&gt; nameA edges.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.basic_stats">
<code class="descname">basic_stats</code><span class="sig-paren">(</span><em>latex=False, caption='', latex_hdr=True, fontsize=8, font='HelveticaNeueLTStd-LtCn', fname=None, header_format='%s', row_order=None, by_category=True, use_cats=['p', 'm', 'i', 'r'], urls=True, annots=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.basic_stats"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.basic_stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns basic numbers about the network resources, e.g. edge and 
node counts.</p>
<dl class="docutils">
<dt>latex</dt>
<dd>Return table in a LaTeX document. This can be compiled by
PDFLaTeX: 
latex stats.tex</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.count_sol">
<code class="descname">count_sol</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.count_sol"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.count_sol" title="Permalink to this definition">¶</a></dt>
<dd><p>Counts nodes with zero degree.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.curation_effort">
<code class="descname">curation_effort</code><span class="sig-paren">(</span><em>sum_by_source=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.curation_effort"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.curation_effort" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the total number of reference-interactions pairs.</p>
<dl class="docutils">
<dt>&#64;sum_by_source</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">bool</span><dd>If True, counts the refrence-interaction pairs by
sources, and returns the sum of these values.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.delete_by_taxon">
<code class="descname">delete_by_taxon</code><span class="sig-paren">(</span><em>tax</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.delete_by_taxon"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.delete_by_taxon" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes the proteins of all organisms which are not listed.</p>
<dl class="docutils">
<dt>&#64;tax</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">list</span><dd>List of NCBI Taxonomy IDs of the organisms of interest.
E.g. [7227, 9606]</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.delete_unknown">
<code class="descname">delete_unknown</code><span class="sig-paren">(</span><em>tax</em>, <em>typ='protein'</em>, <em>defaultNameType=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.delete_unknown"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.delete_unknown" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes those proteins which are not in the list of all default
IDs of the organisms. By default, it means to remove all protein 
nodes not having a human SwissProt ID.</p>
<dl class="docutils">
<dt>&#64;tax</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">list</span><dd>List of NCBI Taxonomy IDs of the organisms of interest.
E.g. [7227, 9606]</dd>
<dt>&#64;typ</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>Molecule type. E.g. &#8216;protein&#8217; or &#8216;mirna&#8217;</dd>
<dt>&#64;defaultNameType</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>The default name type of the given molecular species.
For proteins it&#8217;s &#8216;uniprot&#8217; by default.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.dgenesymbol">
<code class="descname">dgenesymbol</code><span class="sig-paren">(</span><em>genesymbol</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.dgenesymbol"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.dgenesymbol" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <code class="docutils literal"><span class="pre">igraph.Vertex()</span></code> object if the GeneSymbol
can be found in the default directed network, 
otherwise <code class="docutils literal"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;genesymbol</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>GeneSymbol.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.dgs">
<code class="descname">dgs</code><span class="sig-paren">(</span><em>genesymbol</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.dgs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <code class="docutils literal"><span class="pre">igraph.Vertex()</span></code> object if the GeneSymbol
can be found in the default directed network, 
otherwise <code class="docutils literal"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;genesymbol</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>GeneSymbol.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.dp">
<code class="descname">dp</code><span class="sig-paren">(</span><em>identifier</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.dp" title="Permalink to this definition">¶</a></dt>
<dd><p>Same as <code class="docutils literal"><span class="pre">PyPath.protein</span></code>, just for the directed graph.
Returns <code class="docutils literal"><span class="pre">igraph.Vertex()</span></code> object if the identifier
is a valid vertex index in the default directed graph,
or a UniProt ID or GeneSymbol which can be found in the
default directed network, otherwise <code class="docutils literal"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;identifier</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">int, str</span><dd>Vertex index (int) or GeneSymbol (str) or UniProt ID (str).</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.dprotein">
<code class="descname">dprotein</code><span class="sig-paren">(</span><em>identifier</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.dprotein"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.dprotein" title="Permalink to this definition">¶</a></dt>
<dd><p>Same as <code class="docutils literal"><span class="pre">PyPath.protein</span></code>, just for the directed graph.
Returns <code class="docutils literal"><span class="pre">igraph.Vertex()</span></code> object if the identifier
is a valid vertex index in the default directed graph,
or a UniProt ID or GeneSymbol which can be found in the
default directed network, otherwise <code class="docutils literal"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;identifier</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">int, str</span><dd>Vertex index (int) or GeneSymbol (str) or UniProt ID (str).</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.duniprot">
<code class="descname">duniprot</code><span class="sig-paren">(</span><em>uniprot</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.duniprot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.duniprot" title="Permalink to this definition">¶</a></dt>
<dd><p>Same as <code class="docutils literal"><span class="pre">PyPath.uniprot(),</span> <span class="pre">just</span> <span class="pre">for</span> <span class="pre">directed</span> <span class="pre">graph.</span>
<span class="pre">Returns</span> <span class="pre">``igraph.Vertex()</span></code> object if the UniProt
can be found in the default directed network, 
otherwise <code class="docutils literal"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;uniprot</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>UniProt ID.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.duniprots">
<code class="descname">duniprots</code><span class="sig-paren">(</span><em>uniprots</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.duniprots"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.duniprots" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns list of <code class="docutils literal"><span class="pre">igraph.Vertex()</span></code> object
for a list of UniProt IDs omitting those
could not be found in the default
directed graph.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.dup">
<code class="descname">dup</code><span class="sig-paren">(</span><em>uniprot</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.dup" title="Permalink to this definition">¶</a></dt>
<dd><p>Same as <code class="docutils literal"><span class="pre">PyPath.uniprot(),</span> <span class="pre">just</span> <span class="pre">for</span> <span class="pre">directed</span> <span class="pre">graph.</span>
<span class="pre">Returns</span> <span class="pre">``igraph.Vertex()</span></code> object if the UniProt
can be found in the default directed network, 
otherwise <code class="docutils literal"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;uniprot</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>UniProt ID.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.dups">
<code class="descname">dups</code><span class="sig-paren">(</span><em>uniprots</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.dups" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns list of <code class="docutils literal"><span class="pre">igraph.Vertex()</span></code> object
for a list of UniProt IDs omitting those
could not be found in the default
directed graph.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.edge_exists">
<code class="descname">edge_exists</code><span class="sig-paren">(</span><em>nameA</em>, <em>nameB</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.edge_exists"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.edge_exists" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a tuple of vertice indices if edge doesn&#8217;t exists, 
otherwise edge id. Not sensitive to direction.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.edges_expression">
<code class="descname">edges_expression</code><span class="sig-paren">(</span><em>func=&lt;function &lt;lambda&gt;&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.edges_expression"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.edges_expression" title="Permalink to this definition">¶</a></dt>
<dd><p>Executes function <cite>func</cite> for each pairs of connected proteins in the 
network, for every expression dataset. By default, <cite>func</cite> simply
gives the product the (normalized) expression values.</p>
<dl class="docutils">
<dt>func</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">callable</span><dd>Function to handle 2 vectors (pandas.Series() objects), should
return one vector of the same length.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.edges_in_comlexes">
<code class="descname">edges_in_comlexes</code><span class="sig-paren">(</span><em>csources=['corum'], graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.edges_in_comlexes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.edges_in_comlexes" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates edge attributes <code class="docutils literal"><span class="pre">complexes</span></code> and <code class="docutils literal"><span class="pre">in_complex</span></code>.
These are both dicts where the keys are complex resources.
The values in <code class="docutils literal"><span class="pre">complexes</span></code> are the list of complex names
both the source and the target vertices belong to.
The values <code class="docutils literal"><span class="pre">in_complex</span></code> are boolean values whether there
is at least one complex in the given resources both the
source and the target vertex of the edge belong to.</p>
<dl class="docutils">
<dt>&#64;csources</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">list</span><dd>List of complex resources. Should be already loaded.</dd>
<dt>&#64;graph</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">igraph.Graph()</span><dd>The graph object to do the calculations on.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.export_dot">
<code class="descname">export_dot</code><span class="sig-paren">(</span><em>nodes=None</em>, <em>edges=None</em>, <em>directed=True</em>, <em>labels='genesymbol'</em>, <em>edges_filter=&lt;function &lt;lambda&gt;&gt;</em>, <em>nodes_filter=&lt;function &lt;lambda&gt;&gt;</em>, <em>edge_sources=None</em>, <em>dir_sources=None</em>, <em>graph=None</em>, <em>return_object=False</em>, <em>save_dot=None</em>, <em>save_graphics=None</em>, <em>prog='neato'</em>, <em>format=None</em>, <em>hide=False</em>, <em>font=None</em>, <em>auto_edges=False</em>, <em>hide_nodes=[]</em>, <em>defaults={}</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.export_dot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.export_dot" title="Permalink to this definition">¶</a></dt>
<dd><p>Builds a pygraphviz.AGraph() object with filtering the edges 
and vertices along arbitrary criteria.
Returns the Agraph object if requesred, or exports the dot
file, or saves the graphics.</p>
<p>&#64;nodes : list
List of vertex ids to be included.
&#64;edges : list
List of edge ids to be included.
&#64;directed : bool
Create a directed or undirected graph.
&#64;labels : str
Name type to be used as id/label in the dot format.
&#64;edges_filter : function
Function to filter edges, accepting igraph.Edge as argument.
&#64;nodes_filter : function
Function to filter vertices, accepting igraph.Vertex as argument.
&#64;edge_sources : list
Sources to be included.
&#64;dir_sources : list
Direction and effect sources to be included.
&#64;graph : igraph.Graph
The graph object to export.
&#64;return_object : bool
Whether to return the pygraphviz.AGraph object.
&#64;save_dot : str
Filename to export the dot file to.
&#64;save_graphics : str
Filename to export the graphics, the extension defines the format.
&#64;prog : str
The graphviz layout algorithm to use.
&#64;format : str
The graphics format passed to pygraphviz.AGrapg().draw().
&#64;hide : bool
Hide filtered edges instead of omit them.
&#64;hide nodes : list
Nodes to hide. List of vertex ids.
&#64;auto_edges : str
Automatic, built-in style for edges. 
&#8216;DIRECTIONS&#8217; or &#8216;RESOURCE_CATEGORIES&#8217; are supported.
&#64;font : str
Font to use for labels. 
For using more than one fonts refer to graphviz attributes with constant values 
or define callbacks or mapping dictionaries.
&#64;defaults : dict
Default values for graphviz attributes, labeled with the entity, e.g.
<cite>{&#8216;edge_penwidth&#8217;: 0.2}</cite>.
&#64;**kwargs : constant, callable or dict
Graphviz attributes, labeled by the target entity. E.g. <cite>edge_penwidth</cite>,
&#8216;vertex_shape` or <cite>graph_label</cite>. 
If the value is constant, this value will be used.
If the value is dict, and has <cite>_name</cite> as key, for every instance of the
given entity, the value of the attribute defined by <cite>_name</cite> will be looked
up in the dict, and the corresponding value will be given to the graphviz
attribute. If the key <cite>_name</cite> is missing from the dict, igraph vertex and
edge indices will be looked up among the keys.
If the value is callable, it will be called with the current instance of
the entity and the returned value will be used for the graphviz attribute.
E.g. <cite>edge_arrowhead(edge)</cite> or <cite>vertex_fillcolor(vertex)</cite>
Example:</p>
<blockquote>
<div><p>import pypath
from pypath import data_formats
net = pypath.PyPath()
net.init_network(pfile = &#8216;cache/default.pickle&#8217;)
#net.init_network({&#8216;arn&#8217;: data_formats.omnipath[&#8216;arn&#8217;]})
tgf = [v.index for v in net.graph.vs if &#8216;TGF&#8217; in v[&#8216;slk_pathways&#8217;]]
dot = net.export_dot(nodes = tgf, save_graphics = &#8216;tgf_slk.pdf&#8217;, prog = &#8216;dot&#8217;,</p>
<blockquote>
<div>main_title = &#8216;TGF-beta pathway&#8217;, return_object = True,
label_font = &#8216;HelveticaNeueLTStd Med Cn&#8217;, 
edge_sources = [&#8216;SignaLink3&#8217;], 
dir_sources = [&#8216;SignaLink3&#8217;], hide = True)</div></blockquote>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.export_edgelist">
<code class="descname">export_edgelist</code><span class="sig-paren">(</span><em>fname, graph=None, names=['name'], edge_attributes=[], sep='\t'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.export_edgelist"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.export_edgelist" title="Permalink to this definition">¶</a></dt>
<dd><p>Write edge list to text file with attributes</p>
<p>&#64;param fname: the name of the file or a stream to read from.
&#64;param graph: the igraph object containing the network
&#64;param names: list with the vertex attribute names to be printed</p>
<blockquote>
<div>for source and target vertices</div></blockquote>
<dl class="docutils">
<dt>&#64;param edge_attributes: list with the edge attribute names</dt>
<dd>to be printed</dd>
</dl>
<p>&#64;param sep: string used to separate columns</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.find_all_paths">
<code class="descname">find_all_paths</code><span class="sig-paren">(</span><em>start</em>, <em>end</em>, <em>mode='OUT'</em>, <em>maxlen=2</em>, <em>graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.find_all_paths"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.find_all_paths" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds all paths up to length <cite>maxlen</cite> between groups of
vertices. This function is needed only becaues igraph`s
get_all_shortest_paths() finds only the shortest, not any
path up to a defined length.</p>
<dl class="docutils">
<dt>&#64;start</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">int or list</span><dd>Indices of the starting node(s) of the paths.</dd>
<dt>&#64;end</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">int or list</span><dd>Indices of the target node(s) of the paths.</dd>
<dt>&#64;mode</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">&#8216;IN&#8217;, &#8216;OUT&#8217;, &#8216;ALL&#8217;</span><dd>Passed to igraph.Graph.neighbors()</dd>
<dt>&#64;maxlen</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">int</span><dd>Maximum length of paths in steps, i.e. if maxlen = 3, then 
the longest path may consist of 3 edges and 4 nodes.</dd>
<dt>&#64;graph</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">igraph.Graph object</span><dd>The graph you want to find paths in. self.graph by default.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.genesymbol">
<code class="descname">genesymbol</code><span class="sig-paren">(</span><em>genesymbol</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.genesymbol"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.genesymbol" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <code class="docutils literal"><span class="pre">igraph.Vertex()</span></code> object if the GeneSymbol
can be found in the default undirected network, 
otherwise <code class="docutils literal"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;genesymbol</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>GeneSymbol.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.genesymbol_labels">
<code class="descname">genesymbol_labels</code><span class="sig-paren">(</span><em>graph=None</em>, <em>remap_all=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.genesymbol_labels"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.genesymbol_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Creats vertex attribute <code class="docutils literal"><span class="pre">label</span></code> and fills up with Gene Symbols
of all proteins where the Gene Symbol can be looked up based on
the default name of the protein vertex.
If the attribute <code class="docutils literal"><span class="pre">label</span></code> had been already initialized, 
updates this attribute or recreates if <code class="docutils literal"><span class="pre">remap_all</span></code> 
is <code class="docutils literal"><span class="pre">True</span></code>.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.get_directed">
<code class="descname">get_directed</code><span class="sig-paren">(</span><em>graph=False</em>, <em>conv_edges=False</em>, <em>mutual=False</em>, <em>ret=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.get_directed"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.get_directed" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts <code class="docutils literal"><span class="pre">graph</span></code> undirected <code class="docutils literal"><span class="pre">igraph.Graph</span></code> object to a directed one.
By default it converts the graph in <code class="docutils literal"><span class="pre">PyPath.graph</span></code> and places the directed
instance in <code class="docutils literal"><span class="pre">PyPath.dgraph</span></code>.</p>
<dl class="docutils">
<dt>&#64;graph</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">igraph.Graph</span><dd>Undirected graph object.</dd>
<dt>&#64;conv_edges</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">bool</span><dd>Whether to convert undirected edges (those without explicit
direction information) to an arbitrary direction edge or
a pair of opposite edges.
Otherwise those will be deleted. Default is <code class="docutils literal"><span class="pre">False</span></code>.</dd>
<dt>&#64;mutual</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">bool</span><dd>If <code class="docutils literal"><span class="pre">conv_edges</span></code> is <code class="docutils literal"><span class="pre">True</span></code>, whether to convert the
undirected edges to a single, arbitrary directed edge,
or a pair of opposite directed edges. Default is <code class="docutils literal"><span class="pre">False</span></code>.</dd>
<dt>&#64;ret</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">bool</span><dd>Return the directed graph instance, or return <code class="docutils literal"><span class="pre">None</span></code>.
Default is <code class="docutils literal"><span class="pre">False</span></code> (returns <code class="docutils literal"><span class="pre">None</span></code>).</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.get_edge">
<code class="descname">get_edge</code><span class="sig-paren">(</span><em>nodes</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.get_edge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.get_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the edge id only if there is an edge from nodes[0] to nodes[1],
returns False if edge exists in opposite direction, or no edge exists 
between the two vertices, or any of the vertice ids doesn&#8217;t exist.
To find edges without regarding their direction, see edge_exists().</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.get_giant">
<code class="descname">get_giant</code><span class="sig-paren">(</span><em>replace=False</em>, <em>graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.get_giant"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.get_giant" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the giant component of the graph, or 
replaces the igraph object with only the giant 
component.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.gs">
<code class="descname">gs</code><span class="sig-paren">(</span><em>genesymbol</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.gs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <code class="docutils literal"><span class="pre">igraph.Vertex()</span></code> object if the GeneSymbol
can be found in the default undirected network, 
otherwise <code class="docutils literal"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;genesymbol</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>GeneSymbol.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.gs_edge">
<code class="descname">gs_edge</code><span class="sig-paren">(</span><em>source</em>, <em>target</em>, <em>directed=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.gs_edge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.gs_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <code class="docutils literal"><span class="pre">igraph.Edge</span></code> object if an edge exist between
the 2 proteins, otherwise <code class="docutils literal"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;source</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>GeneSymbol</dd>
<dt>&#64;target</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>GeneSymbol</dd>
<dt>&#64;directed</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">bool</span><dd>To be passed to igraph.Graph.get_eid()</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.init_network">
<code class="descname">init_network</code><span class="sig-paren">(</span><em>lst={'hprd_p': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'dbptm': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'signalink3': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'domino': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'spike': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'matrixdb': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'innatedb': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'mppi': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'intact': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'nrf2': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'trip': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'arn': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'macrophage': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'death': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'pdz': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'elm': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'dip': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'ca1': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'depod': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'biogrid': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'ccmap': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'hprd': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'guide2pharma': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'phelm': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'signor': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'lmpid': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'psite': &lt;pypath.input_formats.ReadSettings instance&gt;}</em>, <em>exclude=[]</em>, <em>cache_files={}</em>, <em>pfile=False</em>, <em>save=False</em>, <em>reread=False</em>, <em>redownload=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.init_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.init_network" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a lazy way to start the module, load data 
and build the high confidence, literature curated
part of the signaling network.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_ddi">
<code class="descname">load_ddi</code><span class="sig-paren">(</span><em>ddi</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_ddi"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_ddi" title="Permalink to this definition">¶</a></dt>
<dd><p>ddi is either a list of intera.DomainDomain objects, 
or a function resulting this list</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_dmi">
<code class="descname">load_dmi</code><span class="sig-paren">(</span><em>dmi</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_dmi"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_dmi" title="Permalink to this definition">¶</a></dt>
<dd><p>dmi is either a list of intera.DomainMotif objects, 
or a function resulting this list</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_expression">
<code class="descname">load_expression</code><span class="sig-paren">(</span><em>array=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_expression"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_expression" title="Permalink to this definition">¶</a></dt>
<dd><p>Expression data can be loaded into vertex attributes, 
or into a pandas DataFrame – the latter offers faster 
ways to process and use these huge matrices.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_mutations">
<code class="descname">load_mutations</code><span class="sig-paren">(</span><em>attributes=None</em>, <em>gdsc_datadir=None</em>, <em>mutation_file=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_mutations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_mutations" title="Permalink to this definition">¶</a></dt>
<dd><p>Mutations are listed in vertex attributes. Mutation() objects 
offers methods to identify residues and look up in Ptm(), Motif() 
and Domain() objects, to check if those residues are 
modified, or are in some short motif or domain.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.load_resources">
<code class="descname">load_resources</code><span class="sig-paren">(</span><em>lst={'hprd_p': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'dbptm': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'signalink3': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'domino': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'spike': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'matrixdb': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'innatedb': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'mppi': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'intact': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'nrf2': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'trip': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'arn': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'macrophage': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'death': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'pdz': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'elm': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'dip': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'ca1': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'depod': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'biogrid': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'ccmap': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'hprd': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'guide2pharma': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'phelm': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'signor': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'lmpid': &lt;pypath.input_formats.ReadSettings instance&gt;</em>, <em>'psite': &lt;pypath.input_formats.ReadSettings instance&gt;}</em>, <em>exclude=[]</em>, <em>cache_files={}</em>, <em>reread=False</em>, <em>redownload=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.load_resources"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.load_resources" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads multiple resources, and cleans up after. 
Looks up ID types, and loads all ID conversion 
tables from UniProt if necessary. This is much
faster than loading the ID conversion and the 
resources one by one.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.map_list">
<code class="descname">map_list</code><span class="sig-paren">(</span><em>lst</em>, <em>singleList=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.map_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.map_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Only a wrapper for map_edge()</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.map_network">
<code class="descname">map_network</code><span class="sig-paren">(</span><em>mapping=('trembl'</em>, <em>'swissprot')</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.map_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.map_network" title="Permalink to this definition">¶</a></dt>
<dd><p>This function is not needed any more, will be removed soon.</p>
<p>In the whole network changes every uniprot id to primary.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.mutated_edges">
<code class="descname">mutated_edges</code><span class="sig-paren">(</span><em>sample</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.mutated_edges"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.mutated_edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Compares the mutated residues and the modified residues in PTMs.
Interactions are marked as mutated if the target residue in the 
underlying PTM is mutated.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.network_filter">
<code class="descname">network_filter</code><span class="sig-paren">(</span><em>p=2.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.network_filter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.network_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>This function aims to cut the number of edges in the network, 
without loosing nodes, to make the network less connected, 
less hairball-like, more usable for analysis.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.network_stats">
<code class="descname">network_stats</code><span class="sig-paren">(</span><em>outfile=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.network_stats"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.network_stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates basic statistics for the whole network 
and each of sources. Writes the results in a tab file.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.p">
<code class="descname">p</code><span class="sig-paren">(</span><em>identifier</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.p" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <code class="docutils literal"><span class="pre">igraph.Vertex()</span></code> object if the identifier
is a valid vertex index in the default undirected graph,
or a UniProt ID or GeneSymbol which can be found in the
default undirected network, otherwise <code class="docutils literal"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;identifier</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">int, str</span><dd>Vertex index (int) or GeneSymbol (str) or UniProt ID (str).</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.process_dmi">
<code class="descname">process_dmi</code><span class="sig-paren">(</span><em>source</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.process_dmi"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.process_dmi" title="Permalink to this definition">¶</a></dt>
<dd><p>This is an universal function 
for loading domain-motif objects
like load_phospho_dmi() for PTMs.
TODO this will replace load_elm, load_ielm, etc</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.protein">
<code class="descname">protein</code><span class="sig-paren">(</span><em>identifier</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.protein"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.protein" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <code class="docutils literal"><span class="pre">igraph.Vertex()</span></code> object if the identifier
is a valid vertex index in the default undirected graph,
or a UniProt ID or GeneSymbol which can be found in the
default undirected network, otherwise <code class="docutils literal"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;identifier</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">int, str</span><dd>Vertex index (int) or GeneSymbol (str) or UniProt ID (str).</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.protein_edge">
<code class="descname">protein_edge</code><span class="sig-paren">(</span><em>source</em>, <em>target</em>, <em>directed=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.protein_edge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.protein_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <code class="docutils literal"><span class="pre">igraph.Edge</span></code> object if an edge exist between
the 2 proteins, otherwise <code class="docutils literal"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;source</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">int, str</span><dd>Vertex index or UniProt ID or GeneSymbol</dd>
<dt>&#64;target</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">int, str</span><dd>Vertex index or UniProt ID or GeneSymbol</dd>
<dt>&#64;directed</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">bool</span><dd>To be passed to igraph.Graph.get_eid()</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.read_data_file">
<code class="descname">read_data_file</code><span class="sig-paren">(</span><em>settings</em>, <em>keep_raw=False</em>, <em>cache_files={}</em>, <em>reread=False</em>, <em>redownload=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.read_data_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.read_data_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Interaction data with node and edge attributes can be read 
from simple text based files. This function works not only
with files, but with lists as well. Any other function can 
be written to download a preprocess data, and then give it
to this function to finally attach to the network.</p>
<dl class="docutils">
<dt>&#64;settings</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">ReadSettings instance</span><dd>The detailed definition of the input format. Instead of 
the file name you can give a function name, which will 
be executed, and the returned data will be used.</dd>
<dt>&#64;keep_raw</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">boolean</span><dd>To keep the raw data read by this function, in order for
debugging purposes, or further use.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.set_chembl_mysql">
<code class="descname">set_chembl_mysql</code><span class="sig-paren">(</span><em>title</em>, <em>config_file=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.set_chembl_mysql"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.set_chembl_mysql" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the ChEMBL MySQL config according to
<cite>title</cite> section in <cite>config_file</cite> ini style config.</p>
<blockquote>
<div><p>title (str): section title in ini file
config_file (str, NoneType): config file name;</p>
<blockquote>
<div>if None, the <cite>mysql_config/defaults.mysql</cite>
will be used</div></blockquote>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.set_druggability">
<code class="descname">set_druggability</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.set_druggability"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.set_druggability" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a vertex attribute <cite>dgb</cite> with value <em>True</em> if
the protein is druggable, otherwise <em>False</em>.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.set_drugtargets">
<code class="descname">set_drugtargets</code><span class="sig-paren">(</span><em>pchembl=5.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.set_drugtargets"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.set_drugtargets" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a vertex attribute <cite>dtg</cite> with value <em>True</em> if
the protein has at least one compound binding with 
affinity higher than <cite>pchembl</cite>, otherwise <em>False</em>.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.set_kinases">
<code class="descname">set_kinases</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.set_kinases"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.set_kinases" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a vertex attribute <cite>kin</cite> with value <em>True</em> if
the protein is a kinase, otherwise <em>False</em>.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.set_receptors">
<code class="descname">set_receptors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.set_receptors"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.set_receptors" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a vertex attribute <cite>rec</cite> with value <em>True</em> if
the protein is a receptor, otherwise <em>False</em>.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.set_signaling_proteins">
<code class="descname">set_signaling_proteins</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.set_signaling_proteins"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.set_signaling_proteins" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a vertex attribute <cite>kin</cite> with value <em>True</em> if
the protein is a kinase, otherwise <em>False</em>.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.set_transcription_factors">
<code class="descname">set_transcription_factors</code><span class="sig-paren">(</span><em>classes=['a', 'b', 'other']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.set_transcription_factors"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.set_transcription_factors" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a vertex attribute <cite>tf</cite> with value <em>True</em> if
the protein is a transcription factor, otherwise <em>False</em>.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.shortest_path_dist">
<code class="descname">shortest_path_dist</code><span class="sig-paren">(</span><em>graph=None</em>, <em>subset=None</em>, <em>outfile=None</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.shortest_path_dist"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.shortest_path_dist" title="Permalink to this definition">¶</a></dt>
<dd><p>subset is a tuple of two lists if you wish to look for
paths between elements of two groups, or a list if you 
wish to look for shortest paths within this group</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.small_plot">
<code class="descname">small_plot</code><span class="sig-paren">(</span><em>graph</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.small_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.small_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>This method is deprecated, do not use it.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.source_network">
<code class="descname">source_network</code><span class="sig-paren">(</span><em>font='HelveticaNeueLTStd'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.source_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.source_network" title="Permalink to this definition">¶</a></dt>
<dd><p>For EMBL branding, use Helvetica Neue Linotype Standard light</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.straight_between">
<code class="descname">straight_between</code><span class="sig-paren">(</span><em>nameA</em>, <em>nameB</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.straight_between"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.straight_between" title="Permalink to this definition">¶</a></dt>
<dd><p>This does actually the same as get_edge(), but by names
instead of vertex ids.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.sum_in_complex">
<code class="descname">sum_in_complex</code><span class="sig-paren">(</span><em>csources=['corum'], graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.sum_in_complex"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.sum_in_complex" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the total number of edges in the network falling
between two members of the same complex.
Returns as a dict by complex resources.
Calls :py:func:pypath.pypath.Pypath.edges_in_comlexes()
to do the calculations.</p>
<dl class="docutils">
<dt>&#64;csources</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">list</span><dd>List of complex resources. Should be already loaded.</dd>
<dt>&#64;graph</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">igraph.Graph()</span><dd>The graph object to do the calculations on.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.third_source_directions">
<code class="descname">third_source_directions</code><span class="sig-paren">(</span><em>graph=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.third_source_directions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.third_source_directions" title="Permalink to this definition">¶</a></dt>
<dd><p>This method calls a series of methods to get
additional direction &amp; effect information
from sources having no literature curated references,
but giving sufficient evidence about the directionality
for interactions already supported by literature 
evidences from other sources.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.uniprot">
<code class="descname">uniprot</code><span class="sig-paren">(</span><em>uniprot</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.uniprot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.uniprot" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <code class="docutils literal"><span class="pre">igraph.Vertex()</span></code> object if the UniProt
can be found in the default undirected network, 
otherwise <code class="docutils literal"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;uniprot</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>UniProt ID.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.uniprots">
<code class="descname">uniprots</code><span class="sig-paren">(</span><em>uniprots</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.uniprots"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.uniprots" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns list of <code class="docutils literal"><span class="pre">igraph.Vertex()</span></code> object
for a list of UniProt IDs omitting those
could not be found in the default
undirected graph.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.up">
<code class="descname">up</code><span class="sig-paren">(</span><em>uniprot</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.up" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <code class="docutils literal"><span class="pre">igraph.Vertex()</span></code> object if the UniProt
can be found in the default undirected network, 
otherwise <code class="docutils literal"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;uniprot</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>UniProt ID.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.up_edge">
<code class="descname">up_edge</code><span class="sig-paren">(</span><em>source</em>, <em>target</em>, <em>directed=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.up_edge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.up_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <code class="docutils literal"><span class="pre">igraph.Edge</span></code> object if an edge exist between
the 2 proteins, otherwise <code class="docutils literal"><span class="pre">None</span></code>.</p>
<dl class="docutils">
<dt>&#64;source</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>UniProt ID</dd>
<dt>&#64;target</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>UniProt ID</dd>
<dt>&#64;directed</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">bool</span><dd>To be passed to igraph.Graph.get_eid()</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.update_vindex">
<code class="descname">update_vindex</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.update_vindex"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.update_vindex" title="Permalink to this definition">¶</a></dt>
<dd><p>This is deprecated.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.update_vname">
<code class="descname">update_vname</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.update_vname"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.update_vname" title="Permalink to this definition">¶</a></dt>
<dd><p>For fast lookup of node names and indexes, these are 
hold in a list and a dict as well. However, every time
new nodes are added, these should be updated. This 
function is automatically called after all operations
affecting node indices.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.ups">
<code class="descname">ups</code><span class="sig-paren">(</span><em>uniprots</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.main.PyPath.ups" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns list of <code class="docutils literal"><span class="pre">igraph.Vertex()</span></code> object
for a list of UniProt IDs omitting those
could not be found in the default
undirected graph.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.PyPath.vertex_pathways">
<code class="descname">vertex_pathways</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#PyPath.vertex_pathways"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.PyPath.vertex_pathways" title="Permalink to this definition">¶</a></dt>
<dd><p>Some resources assignes interactions some others 
proteins to pathways.
This function converts pathway annotations from 
edge attributes to vertex attributes.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.main.Direction">
<em class="property">class </em><code class="descclassname">pypath.main.</code><code class="descname">Direction</code><span class="sig-paren">(</span><em>nameA</em>, <em>nameB</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.main.Direction.get_dir">
<code class="descname">get_dir</code><span class="sig-paren">(</span><em>direction</em>, <em>sources=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.get_dir"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.get_dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns boolean or list of sources</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.get_dirs">
<code class="descname">get_dirs</code><span class="sig-paren">(</span><em>src</em>, <em>tgt</em>, <em>sources=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.get_dirs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.get_dirs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all directions with boolean values
or list of sources.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.set_dir">
<code class="descname">set_dir</code><span class="sig-paren">(</span><em>direction</em>, <em>source</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.set_dir"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.set_dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds directionality information with 
the corresponding data source named.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.src">
<code class="descname">src</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.src"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.src" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the IDs of effector molecules in this directed 
interaction. If the interaction is bidirectional, the 
list will contain 2 IDs. If the interaction is undirec-
ted, an empty list will be returned.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.tgt">
<code class="descname">tgt</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.tgt"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.tgt" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the IDs of the target moleculess in the inter-
action. Same behaviour as <cite>Direction.src()</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.main.Direction.unset_dir">
<code class="descname">unset_dir</code><span class="sig-paren">(</span><em>direction</em>, <em>source=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/main.html#Direction.unset_dir"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.main.Direction.unset_dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes directionality information,
or single source.</p>
</dd></dl>

</dd></dl>

<span class="target" id="module-pypath.input_formats"></span><dl class="class">
<dt id="pypath.input_formats.ReadSettings">
<em class="property">class </em><code class="descclassname">pypath.input_formats.</code><code class="descname">ReadSettings</code><span class="sig-paren">(</span><em>name='unknown'</em>, <em>separator=None</em>, <em>nameColA=0</em>, <em>nameColB=1</em>, <em>nameTypeA='uniprot'</em>, <em>nameTypeB='uniprot'</em>, <em>typeA='protein'</em>, <em>typeB='protein'</em>, <em>isDirected=False</em>, <em>sign=False</em>, <em>inFile=None</em>, <em>references=False</em>, <em>extraEdgeAttrs={}</em>, <em>extraNodeAttrsA={}</em>, <em>extraNodeAttrsB={}</em>, <em>header=False</em>, <em>taxonA=9606</em>, <em>taxonB=9606</em>, <em>ncbiTaxId=False</em>, <em>interactionType='PPI'</em>, <em>positiveFilters=[]</em>, <em>negativeFilters=[]</em>, <em>inputArgs={}</em>, <em>must_have_references=True</em>, <em>huge=False</em>, <em>resource=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/input_formats.html#ReadSettings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.input_formats.ReadSettings" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.input_formats.ReadList">
<em class="property">class </em><code class="descclassname">pypath.input_formats.</code><code class="descname">ReadList</code><span class="sig-paren">(</span><em>name='unknown'</em>, <em>separator=None</em>, <em>nameCol=0</em>, <em>nameType='uniprot'</em>, <em>typ='protein'</em>, <em>inFile=None</em>, <em>extraAttrs={}</em>, <em>header=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/input_formats.html#ReadList"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.input_formats.ReadList" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.input_formats.UniprotMapping">
<em class="property">class </em><code class="descclassname">pypath.input_formats.</code><code class="descname">UniprotMapping</code><span class="sig-paren">(</span><em>nameType</em>, <em>bi=False</em>, <em>tax=9606</em>, <em>swissprot='yes'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/input_formats.html#UniprotMapping"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.input_formats.UniprotMapping" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.input_formats.PickleMapping">
<em class="property">class </em><code class="descclassname">pypath.input_formats.</code><code class="descname">PickleMapping</code><span class="sig-paren">(</span><em>pickleFile</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/input_formats.html#PickleMapping"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.input_formats.PickleMapping" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.input_formats.FileMapping">
<em class="property">class </em><code class="descclassname">pypath.input_formats.</code><code class="descname">FileMapping</code><span class="sig-paren">(</span><em>input</em>, <em>oneCol</em>, <em>twoCol</em>, <em>separator=None</em>, <em>header=0</em>, <em>bi=False</em>, <em>tax=9606</em>, <em>typ='protein'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/input_formats.html#FileMapping"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.input_formats.FileMapping" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.input_formats.MysqlMapping">
<em class="property">class </em><code class="descclassname">pypath.input_formats.</code><code class="descname">MysqlMapping</code><span class="sig-paren">(</span><em>tableName</em>, <em>fieldOne</em>, <em>fieldTwo</em>, <em>db=None</em>, <em>tax=None</em>, <em>bi=False</em>, <em>mysql=None</em>, <em>typ='protein'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/input_formats.html#MysqlMapping"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.input_formats.MysqlMapping" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<span class="target" id="module-pypath.pyreact"></span><dl class="class">
<dt id="pypath.pyreact.BioPaxReader">
<em class="property">class </em><code class="descclassname">pypath.pyreact.</code><code class="descname">BioPaxReader</code><span class="sig-paren">(</span><em>biopax</em>, <em>source</em>, <em>cleanup_period=800</em>, <em>file_from_archive=None</em>, <em>silent=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#BioPaxReader"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.BioPaxReader" title="Permalink to this definition">¶</a></dt>
<dd><p>This class parses a BioPAX file and exposes its content easily accessible
for further processing. First it opens the file, if necessary it extracts
from the archive. Then an <cite>lxml.etree.iterparse</cite> object is created, so the
iteration is efficient and memory requirements are minimal. The iterparse
object is iterated then, and for each tag included in the
<cite>BioPaxReader.methods</cite> dict, the appropriate method is called. These me-
thods extract information from the BioPAX entity, and store it in arbit-
rary data structures: strings, lists or dicts. These are stored in dicts
where keys are the original IDs of the tags, prefixed with the unique ID
of the parser object. This is necessary to give a way to merge later the
result of parsing more BioPAX files. For example, <cite>id42</cite> may identify
EGFR in one file, but AKT1 in the other. Then, the parser of the first
file has a unique ID of a 5 letter random string, the second parser a
different one, and the molecules with the same ID can be distinguished
at merging, e.g. EGFR will be <cite>ffjh2&#64;id42</cite> and AKT1 will be <cite>tr9gy&#64;id42</cite>.
The methods and the resulted dicts are named after the BioPAX elements,
sometimes abbreviated. For example, <cite>BioPaxReader.protein()</cite> processes
the <cite>&lt;bp:Protein&gt;</cite> elements, and stores the results in
<cite>BioPaxReader.proteins</cite>.</p>
<p>In its current state, this class does not parse every information and
all BioPax entities. For example, nucleic acid related entities and
interactions are omitted. But these easily can be added with minor mo-
difications.</p>
<dl class="method">
<dt id="pypath.pyreact.BioPaxReader.biopax_size">
<code class="descname">biopax_size</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#BioPaxReader.biopax_size"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.BioPaxReader.biopax_size" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the uncompressed size of the BioPax XML. This is needed in
order to have a progress bar. This method should not be called
directly, <code class="docutils literal"><span class="pre">BioPaxReader.process()</span></code> calls it.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pyreact.BioPaxReader.cleanup_hook">
<code class="descname">cleanup_hook</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#BioPaxReader.cleanup_hook"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.BioPaxReader.cleanup_hook" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes the used elements to free up memory.
This method should not be called directly,
<code class="docutils literal"><span class="pre">BioPaxReader.iterate()</span></code> calls it.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pyreact.BioPaxReader.close_biopax">
<code class="descname">close_biopax</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#BioPaxReader.close_biopax"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.BioPaxReader.close_biopax" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes the iterator and closes the file object.
This method should not be called directly,
<code class="docutils literal"><span class="pre">BioPaxReader.process()</span></code> calls it.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pyreact.BioPaxReader.extract">
<code class="descname">extract</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#BioPaxReader.extract"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.BioPaxReader.extract" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts the BioPax file from compressed archive. Creates a
temporary file. This is needed to trace the progress of
processing, which is useful in case of large files.
This method should not be called directly,
<code class="docutils literal"><span class="pre">BioPaxReader.process()</span></code> calls it.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pyreact.BioPaxReader.init_etree">
<code class="descname">init_etree</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#BioPaxReader.init_etree"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.BioPaxReader.init_etree" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates the <code class="docutils literal"><span class="pre">lxml.etree.iterparse</span></code> object.
This method should not be called directly,
<code class="docutils literal"><span class="pre">BioPaxReader.process()</span></code> calls it.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pyreact.BioPaxReader.iterate">
<code class="descname">iterate</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#BioPaxReader.iterate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.BioPaxReader.iterate" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterates the BioPax XML and calls the appropriate methods
for each element.
This method should not be called directly,
<code class="docutils literal"><span class="pre">BioPaxReader.process()</span></code> calls it.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pyreact.BioPaxReader.open_biopax">
<code class="descname">open_biopax</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#BioPaxReader.open_biopax"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.BioPaxReader.open_biopax" title="Permalink to this definition">¶</a></dt>
<dd><p>Opens the BioPax file. This method should not be called directly,
<code class="docutils literal"><span class="pre">BioPaxReader.process()</span></code> calls it.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pyreact.BioPaxReader.process">
<code class="descname">process</code><span class="sig-paren">(</span><em>silent=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#BioPaxReader.process"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.BioPaxReader.process" title="Permalink to this definition">¶</a></dt>
<dd><p>This method executes the total workflow of BioPax processing.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>silent</strong> (<em>bool</em>) &#8211; whether to print status messages and progress bars.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pypath.pyreact.BioPaxReader.set_progress">
<code class="descname">set_progress</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#BioPaxReader.set_progress"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.BioPaxReader.set_progress" title="Permalink to this definition">¶</a></dt>
<dd><p>Initializes a progress bar.
This method should not be called directly,
<code class="docutils literal"><span class="pre">BioPaxReader.process()</span></code> calls it.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt>
<em class="property">class </em><code class="descclassname">pypath.pyreact.</code><code class="descname">BioPaxReader</code><span class="sig-paren">(</span><em>biopax</em>, <em>source</em>, <em>cleanup_period=800</em>, <em>file_from_archive=None</em>, <em>silent=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#BioPaxReader"><span class="viewcode-link">[source]</span></a></dt>
<dd><p>This class parses a BioPAX file and exposes its content easily accessible
for further processing. First it opens the file, if necessary it extracts
from the archive. Then an <cite>lxml.etree.iterparse</cite> object is created, so the
iteration is efficient and memory requirements are minimal. The iterparse
object is iterated then, and for each tag included in the
<cite>BioPaxReader.methods</cite> dict, the appropriate method is called. These me-
thods extract information from the BioPAX entity, and store it in arbit-
rary data structures: strings, lists or dicts. These are stored in dicts
where keys are the original IDs of the tags, prefixed with the unique ID
of the parser object. This is necessary to give a way to merge later the
result of parsing more BioPAX files. For example, <cite>id42</cite> may identify
EGFR in one file, but AKT1 in the other. Then, the parser of the first
file has a unique ID of a 5 letter random string, the second parser a
different one, and the molecules with the same ID can be distinguished
at merging, e.g. EGFR will be <cite>ffjh2&#64;id42</cite> and AKT1 will be <cite>tr9gy&#64;id42</cite>.
The methods and the resulted dicts are named after the BioPAX elements,
sometimes abbreviated. For example, <cite>BioPaxReader.protein()</cite> processes
the <cite>&lt;bp:Protein&gt;</cite> elements, and stores the results in
<cite>BioPaxReader.proteins</cite>.</p>
<p>In its current state, this class does not parse every information and
all BioPax entities. For example, nucleic acid related entities and
interactions are omitted. But these easily can be added with minor mo-
difications.</p>
<dl class="method">
<dt>
<code class="descname">biopax_size</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#BioPaxReader.biopax_size"><span class="viewcode-link">[source]</span></a></dt>
<dd><p>Gets the uncompressed size of the BioPax XML. This is needed in
order to have a progress bar. This method should not be called
directly, <code class="docutils literal"><span class="pre">BioPaxReader.process()</span></code> calls it.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="descname">cleanup_hook</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#BioPaxReader.cleanup_hook"><span class="viewcode-link">[source]</span></a></dt>
<dd><p>Removes the used elements to free up memory.
This method should not be called directly,
<code class="docutils literal"><span class="pre">BioPaxReader.iterate()</span></code> calls it.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="descname">close_biopax</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#BioPaxReader.close_biopax"><span class="viewcode-link">[source]</span></a></dt>
<dd><p>Deletes the iterator and closes the file object.
This method should not be called directly,
<code class="docutils literal"><span class="pre">BioPaxReader.process()</span></code> calls it.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="descname">extract</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#BioPaxReader.extract"><span class="viewcode-link">[source]</span></a></dt>
<dd><p>Extracts the BioPax file from compressed archive. Creates a
temporary file. This is needed to trace the progress of
processing, which is useful in case of large files.
This method should not be called directly,
<code class="docutils literal"><span class="pre">BioPaxReader.process()</span></code> calls it.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="descname">init_etree</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#BioPaxReader.init_etree"><span class="viewcode-link">[source]</span></a></dt>
<dd><p>Creates the <code class="docutils literal"><span class="pre">lxml.etree.iterparse</span></code> object.
This method should not be called directly,
<code class="docutils literal"><span class="pre">BioPaxReader.process()</span></code> calls it.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="descname">iterate</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#BioPaxReader.iterate"><span class="viewcode-link">[source]</span></a></dt>
<dd><p>Iterates the BioPax XML and calls the appropriate methods
for each element.
This method should not be called directly,
<code class="docutils literal"><span class="pre">BioPaxReader.process()</span></code> calls it.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="descname">open_biopax</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#BioPaxReader.open_biopax"><span class="viewcode-link">[source]</span></a></dt>
<dd><p>Opens the BioPax file. This method should not be called directly,
<code class="docutils literal"><span class="pre">BioPaxReader.process()</span></code> calls it.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="descname">process</code><span class="sig-paren">(</span><em>silent=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#BioPaxReader.process"><span class="viewcode-link">[source]</span></a></dt>
<dd><p>This method executes the total workflow of BioPax processing.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>silent</strong> (<em>bool</em>) &#8211; whether to print status messages and progress bars.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt>
<code class="descname">set_progress</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#BioPaxReader.set_progress"><span class="viewcode-link">[source]</span></a></dt>
<dd><p>Initializes a progress bar.
This method should not be called directly,
<code class="docutils literal"><span class="pre">BioPaxReader.process()</span></code> calls it.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.pyreact.PyReact">
<em class="property">class </em><code class="descclassname">pypath.pyreact.</code><code class="descname">PyReact</code><span class="sig-paren">(</span><em>mapper=None</em>, <em>ncbi_tax_id=9606</em>, <em>default_id_types={}</em>, <em>modifications=True</em>, <em>seq=None</em>, <em>silent=False</em>, <em>max_complex_combinations=100</em>, <em>max_reaction_combinations=100</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#PyReact"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.PyReact" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.pyreact.PyReact.gen_cvariations">
<code class="descname">gen_cvariations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#PyReact.gen_cvariations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.PyReact.gen_cvariations" title="Permalink to this definition">¶</a></dt>
<dd><p>Because one key from the BioPax file might represent more complexes,
<em>complexvariations</em> are created to give a way to represent sets of
combinations. These are created for all complexes, even with only
one unambiguous constitution. The keys are the constitutions of
all the combinations listed in alphabetic order, separated by <code class="docutils literal"><span class="pre">|</span></code>.
For example, <code class="docutils literal"><span class="pre">A,B,C|A,B,D|A,B,E</span></code>.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pyreact.PyReact.in_same_component">
<code class="descname">in_same_component</code><span class="sig-paren">(</span><em>by_source=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#PyReact.in_same_component"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.PyReact.in_same_component" title="Permalink to this definition">¶</a></dt>
<dd><p>For all complexes connects all members of the complex with each other.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pyreact.PyReact.merge_complexes">
<code class="descname">merge_complexes</code><span class="sig-paren">(</span><em>this_round=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#PyReact.merge_complexes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.PyReact.merge_complexes" title="Permalink to this definition">¶</a></dt>
<dd><p>Merges complexes from the active <code class="docutils literal"><span class="pre">BioPaxReader</span></code> object.
Protein families and subcomplexes are expanded, and all combinations
are created as separate complexes. The complexes from the same ID
are added to sets in the <code class="docutils literal"><span class="pre">rcomplexes</span></code> dict.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.pyreact.PyReact.merge_cvariations">
<code class="descname">merge_cvariations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#PyReact.merge_cvariations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.PyReact.merge_cvariations" title="Permalink to this definition">¶</a></dt>
<dd><p>This processes those complexes which are in fact a set of complex variations.
As simple complexes also are always extended to complex variations because
they might have not only simple proteins but protein families as members,
here we only add new records to the attributes of already existing complexes.
After <code class="docutils literal"><span class="pre">merge_complexes</span></code> will be called again, to process those simple
complexes which have any of the complex variations processed here among their
subcomplexes.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.pyreact.Entity">
<em class="property">class </em><code class="descclassname">pypath.pyreact.</code><code class="descname">Entity</code><span class="sig-paren">(</span><em>identifier</em>, <em>id_type</em>, <em>sources=[]</em>, <em>attrs=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#Entity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.Entity" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.pyreact.Entity.expand">
<code class="descname">expand</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#Entity.expand"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.Entity.expand" title="Permalink to this definition">¶</a></dt>
<dd><p>With this method it is possible to iterate <code class="docutils literal"><span class="pre">Entity</span></code> objects
just like <code class="docutils literal"><span class="pre">EntitySet</span></code> objects.</p>
<p>Yields string.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.pyreact.Protein">
<em class="property">class </em><code class="descclassname">pypath.pyreact.</code><code class="descname">Protein</code><span class="sig-paren">(</span><em>protein_id</em>, <em>id_type='uniprot'</em>, <em>sources=[]</em>, <em>attrs=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#Protein"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.Protein" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.pyreact.EntitySet">
<em class="property">class </em><code class="descclassname">pypath.pyreact.</code><code class="descname">EntitySet</code><span class="sig-paren">(</span><em>members</em>, <em>sources=[]</em>, <em>sep=';'</em>, <em>parent=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#EntitySet"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.EntitySet" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.pyreact.AttributeHandler">
<em class="property">class </em><code class="descclassname">pypath.pyreact.</code><code class="descname">AttributeHandler</code><a class="reference internal" href="_modules/pypath/pyreact.html#AttributeHandler"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.AttributeHandler" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.pyreact.ProteinFamily">
<em class="property">class </em><code class="descclassname">pypath.pyreact.</code><code class="descname">ProteinFamily</code><span class="sig-paren">(</span><em>members</em>, <em>source</em>, <em>parent=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#ProteinFamily"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.ProteinFamily" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.pyreact.Complex">
<em class="property">class </em><code class="descclassname">pypath.pyreact.</code><code class="descname">Complex</code><span class="sig-paren">(</span><em>members</em>, <em>source</em>, <em>parent=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#Complex"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.Complex" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.pyreact.ComplexVariations">
<em class="property">class </em><code class="descclassname">pypath.pyreact.</code><code class="descname">ComplexVariations</code><span class="sig-paren">(</span><em>members</em>, <em>source</em>, <em>parent=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#ComplexVariations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.ComplexVariations" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.pyreact.ComplexVariations.itermembers">
<code class="descname">itermembers</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#ComplexVariations.itermembers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.ComplexVariations.itermembers" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a convenient iterator for the expand methods
of higher classes like <code class="docutils literal"><span class="pre">ReactionSide</span></code> or <code class="docutils literal"><span class="pre">Control</span></code>.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.pyreact.ReactionSide">
<em class="property">class </em><code class="descclassname">pypath.pyreact.</code><code class="descname">ReactionSide</code><span class="sig-paren">(</span><em>members</em>, <em>source=[]</em>, <em>parent=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#ReactionSide"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.ReactionSide" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.pyreact.ReactionSide.expand">
<code class="descname">expand</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#ReactionSide.expand"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.ReactionSide.expand" title="Permalink to this definition">¶</a></dt>
<dd><p>Expands the <code class="docutils literal"><span class="pre">ReactionSide</span></code> by iterating over all combinations
of all <code class="docutils literal"><span class="pre">ComplexVariation</span></code> and <code class="docutils literal"><span class="pre">ProteinFamily</span></code> members, so
yields <code class="docutils literal"><span class="pre">ReactionSide</span></code> objects with only <code class="docutils literal"><span class="pre">Protein</span></code> and <code class="docutils literal"><span class="pre">Complex</span></code>
members. Yields tuple, because <code class="docutils literal"><span class="pre">ReactionSide</span></code> is initialized in
<code class="docutils literal"><span class="pre">Reaction</span></code>, the tuple is suitable to serve as <code class="docutils literal"><span class="pre">members</span></code> and
<code class="docutils literal"><span class="pre">attrs</span></code>.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.pyreact.Reaction">
<em class="property">class </em><code class="descclassname">pypath.pyreact.</code><code class="descname">Reaction</code><span class="sig-paren">(</span><em>left</em>, <em>right</em>, <em>left_attrs</em>, <em>right_attrs</em>, <em>source=[]</em>, <em>parent=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#Reaction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.Reaction" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.pyreact.Control">
<em class="property">class </em><code class="descclassname">pypath.pyreact.</code><code class="descname">Control</code><span class="sig-paren">(</span><em>er</em>, <em>ed</em>, <em>source=[]</em>, <em>parent=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pyreact.html#Control"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pyreact.Control" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<span class="target" id="module-pypath.plot"></span><dl class="function">
<dt id="pypath.plot.is_opentype_cff_font">
<code class="descclassname">pypath.plot.</code><code class="descname">is_opentype_cff_font</code><span class="sig-paren">(</span><em>filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#is_opentype_cff_font"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.is_opentype_cff_font" title="Permalink to this definition">¶</a></dt>
<dd><p>This is necessary to fix a bug in matplotlib:
<a class="reference external" href="https://github.com/matplotlib/matplotlib/pull/6714">https://github.com/matplotlib/matplotlib/pull/6714</a>
Returns True if the given font is a Postscript Compact Font Format
Font embedded in an OpenType wrapper.  Used by the PostScript and
PDF backends that can not subset these fonts.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.plot.randn">
<code class="descclassname">pypath.plot.</code><code class="descname">randn</code><span class="sig-paren">(</span><em>d0</em>, <em>d1</em>, <em>...</em>, <em>dn</em><span class="sig-paren">)</span><a class="headerlink" href="#pypath.plot.randn" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a sample (or samples) from the &#8220;standard normal&#8221; distribution.</p>
<p>If positive, int_like or int-convertible arguments are provided,
<cite>randn</cite> generates an array of shape <code class="docutils literal"><span class="pre">(d0,</span> <span class="pre">d1,</span> <span class="pre">...,</span> <span class="pre">dn)</span></code>, filled
with random floats sampled from a univariate &#8220;normal&#8221; (Gaussian)
distribution of mean 0 and variance 1 (if any of the  are
floats, they are first converted to integers by truncation). A single
float randomly sampled from the distribution is returned if no
argument is provided.</p>
<p>This is a convenience function.  If you want an interface that takes a
tuple as the first argument, use <cite>numpy.random.standard_normal</cite> instead.</p>
<dl class="docutils">
<dt>d0, d1, ..., dn</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">int, optional</span><dd>The dimensions of the returned array, should be all positive.
If no argument is given a single Python float is returned.</dd>
</dl>
<dl class="docutils">
<dt>Z</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">ndarray or float</span><dd>A <code class="docutils literal"><span class="pre">(d0,</span> <span class="pre">d1,</span> <span class="pre">...,</span> <span class="pre">dn)</span></code>-shaped array of floating-point samples from
the standard normal distribution, or a single such float if
no parameters were supplied.</dd>
</dl>
<p>random.standard_normal : Similar, but takes a tuple as its argument.</p>
<p>For random samples from , use:</p>
<p><code class="docutils literal"><span class="pre">sigma</span> <span class="pre">*</span> <span class="pre">np.random.randn(...)</span> <span class="pre">+</span> <span class="pre">mu</span></code></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randn</span><span class="p">()</span>
<span class="go">2.1923875335537315 #random</span>
</pre></div>
</div>
<p>Two-by-four array of samples from N(3, 6.25):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="mf">2.5</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randn</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span> <span class="o">+</span> <span class="mi">3</span>
<span class="go">array([[-4.49401501,  4.00950034, -1.81814867,  7.29718677],  #random</span>
<span class="go">       [ 0.39924804,  4.68456316,  4.99394529,  4.84057254]]) #random</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="pypath.plot.Plot">
<em class="property">class </em><code class="descclassname">pypath.plot.</code><code class="descname">Plot</code><span class="sig-paren">(</span><em>fname=None</em>, <em>font_family='Helvetica Neue LT Std'</em>, <em>font_style='normal'</em>, <em>font_weight='normal'</em>, <em>font_variant='normal'</em>, <em>font_stretch='normal'</em>, <em>palette=None</em>, <em>context='poster'</em>, <em>lab_size=(9</em>, <em>9)</em>, <em>axis_lab_size=10.0</em>, <em>rc={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#Plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.Plot" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.plot.Plot.finish">
<code class="descname">finish</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#Plot.finish"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.Plot.finish" title="Permalink to this definition">¶</a></dt>
<dd><p>Saves and closes a figure.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.plot.MultiBarplot">
<em class="property">class </em><code class="descclassname">pypath.plot.</code><code class="descname">MultiBarplot</code><span class="sig-paren">(</span><em>x</em>, <em>y</em>, <em>categories=None</em>, <em>cat_names=None</em>, <em>cat_ordr=None</em>, <em>fname=None</em>, <em>figsize=(12</em>, <em>4)</em>, <em>xlab=''</em>, <em>ylab=''</em>, <em>title=''</em>, <em>lab_angle=90</em>, <em>color='#007b7f'</em>, <em>order=False</em>, <em>desc=True</em>, <em>ylog=False</em>, <em>legend=None</em>, <em>fin=True</em>, <em>rc={}</em>, <em>axis_lab_font={}</em>, <em>bar_args={}</em>, <em>ticklabel_font={}</em>, <em>legend_font={}</em>, <em>title_font={}</em>, <em>title_halign='center'</em>, <em>title_valign='top'</em>, <em>y2=None</em>, <em>color2=None</em>, <em>ylim=None</em>, <em>grouped=False</em>, <em>group_labels=[]</em>, <em>summary=False</em>, <em>summary_name=''</em>, <em>uniform_ylim=False</em>, <em>do=True</em>, <em>legloc=0</em>, <em>maketitle=True</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#MultiBarplot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.MultiBarplot" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.plot.MultiBarplot.by_plot">
<code class="descname">by_plot</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#MultiBarplot.by_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.MultiBarplot.by_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets list of lists with x and y values and colors by category.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.MultiBarplot.do_plot">
<code class="descname">do_plot</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#MultiBarplot.do_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.MultiBarplot.do_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Calls the plotting methods in the correct order.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.MultiBarplot.finish">
<code class="descname">finish</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#MultiBarplot.finish"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.MultiBarplot.finish" title="Permalink to this definition">¶</a></dt>
<dd><p>Applies tight layout, draws the figure, writes the file and closes.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.MultiBarplot.init_fig">
<code class="descname">init_fig</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#MultiBarplot.init_fig"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.MultiBarplot.init_fig" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a figure using the object oriented matplotlib interface.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.MultiBarplot.labels">
<code class="descname">labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#MultiBarplot.labels"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.MultiBarplot.labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets properties of axis labels and ticklabels.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.MultiBarplot.make_plots">
<code class="descname">make_plots</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#MultiBarplot.make_plots"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.MultiBarplot.make_plots" title="Permalink to this definition">¶</a></dt>
<dd><p>Does the actual plotting.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.MultiBarplot.plot">
<code class="descname">plot</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#MultiBarplot.plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.MultiBarplot.plot" title="Permalink to this definition">¶</a></dt>
<dd><p>The total workflow of this class.
Calls all methods in the correct order.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.MultiBarplot.plots_order">
<code class="descname">plots_order</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#MultiBarplot.plots_order"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.MultiBarplot.plots_order" title="Permalink to this definition">¶</a></dt>
<dd><p>Defines the order of the subplots.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.MultiBarplot.post_plot">
<code class="descname">post_plot</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#MultiBarplot.post_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.MultiBarplot.post_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Saves the plot into file, and closes the figure.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.MultiBarplot.pre_plot">
<code class="descname">pre_plot</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#MultiBarplot.pre_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.MultiBarplot.pre_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Executes all necessary tasks before plotting in the correct order.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.MultiBarplot.reload">
<code class="descname">reload</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#MultiBarplot.reload"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.MultiBarplot.reload" title="Permalink to this definition">¶</a></dt>
<dd><p>Reloads the module and updates the class instance.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.MultiBarplot.set_categories">
<code class="descname">set_categories</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#MultiBarplot.set_categories"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.MultiBarplot.set_categories" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets a list with category indices (integers) of length equal of x,
and sets dicts to translate between category names and indices.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.MultiBarplot.set_colors">
<code class="descname">set_colors</code><span class="sig-paren">(</span><em>colseries=''</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#MultiBarplot.set_colors"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.MultiBarplot.set_colors" title="Permalink to this definition">¶</a></dt>
<dd><p>Compiles an array of colors equal length of x.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.MultiBarplot.set_figsize">
<code class="descname">set_figsize</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#MultiBarplot.set_figsize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.MultiBarplot.set_figsize" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts width and height to a tuple so can be used for figsize.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.MultiBarplot.set_grid">
<code class="descname">set_grid</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#MultiBarplot.set_grid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.MultiBarplot.set_grid" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets up a grid according to the number of subplots,
with proportions according to the number of elements
in each subplot.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.MultiBarplot.set_title">
<code class="descname">set_title</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#MultiBarplot.set_title"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.MultiBarplot.set_title" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the main title.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.MultiBarplot.sort">
<code class="descname">sort</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#MultiBarplot.sort"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.MultiBarplot.sort" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds the defined or default order, and
sorts the arrays x, y and col accordingly.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.plot.ScatterPlus">
<em class="property">class </em><code class="descclassname">pypath.plot.</code><code class="descname">ScatterPlus</code><span class="sig-paren">(</span><em>x</em>, <em>y</em>, <em>size=None</em>, <em>color='#114477'</em>, <em>labels=None</em>, <em>xlog=False</em>, <em>ylog=False</em>, <em>xlim=None</em>, <em>ylim=None</em>, <em>xtickscale=None</em>, <em>ytickscale=None</em>, <em>legscale=None</em>, <em>fname=None</em>, <em>confi=True</em>, <em>title_font={}</em>, <em>ticklabel_font={}</em>, <em>legend_font={}</em>, <em>axis_lab_font={}</em>, <em>annot_font={}</em>, <em>xlab=''</em>, <em>ylab=''</em>, <em>axis_lab_size=10.0</em>, <em>min_size=5.0</em>, <em>max_size=30.0</em>, <em>log_size=False</em>, <em>alpha=0.5</em>, <em>size_scaling=0.8</em>, <em>lab_angle=90</em>, <em>order=False</em>, <em>desc=True</em>, <em>legend=True</em>, <em>legtitle=''</em>, <em>legstrip=(None</em>, <em>None)</em>, <em>color_labels=[]</em>, <em>legloc=4</em>, <em>size_to_value=&lt;function &lt;lambda&gt;&gt;</em>, <em>value_to_size=&lt;function &lt;lambda&gt;&gt;</em>, <em>figsize=(10.0</em>, <em>7.5)</em>, <em>title=''</em>, <em>title_halign='center'</em>, <em>title_valign='top'</em>, <em>fin=True</em>, <em>rc={}</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#ScatterPlus"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.ScatterPlus" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.plot.ScatterPlus.finish">
<code class="descname">finish</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#ScatterPlus.finish"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.ScatterPlus.finish" title="Permalink to this definition">¶</a></dt>
<dd><p>Applies tight layout, draws the figure, writes the file and closes.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.ScatterPlus.init_fig">
<code class="descname">init_fig</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#ScatterPlus.init_fig"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.ScatterPlus.init_fig" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a figure using the object oriented matplotlib interface.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.ScatterPlus.reload">
<code class="descname">reload</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#ScatterPlus.reload"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.ScatterPlus.reload" title="Permalink to this definition">¶</a></dt>
<dd><p>Reloads the module and updates the class instance.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.ScatterPlus.set_figsize">
<code class="descname">set_figsize</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#ScatterPlus.set_figsize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.ScatterPlus.set_figsize" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts width and height to a tuple so can be used for figsize.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.ScatterPlus.set_title">
<code class="descname">set_title</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#ScatterPlus.set_title"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.ScatterPlus.set_title" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the main title.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.ScatterPlus.values_to_sizes">
<code class="descname">values_to_sizes</code><span class="sig-paren">(</span><em>values</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#ScatterPlus.values_to_sizes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.ScatterPlus.values_to_sizes" title="Permalink to this definition">¶</a></dt>
<dd><p>Transformation converts from size values in data dimension
to dimension of the size graphical parameter.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.plot.StackedBarplot">
<em class="property">class </em><code class="descclassname">pypath.plot.</code><code class="descname">StackedBarplot</code><span class="sig-paren">(</span><em>x</em>, <em>y</em>, <em>fname</em>, <em>names</em>, <em>colors</em>, <em>xlab=''</em>, <em>ylab=''</em>, <em>title=''</em>, <em>title_halign='center'</em>, <em>title_valign='top'</em>, <em>bar_args={}</em>, <em>axis_lab_font={}</em>, <em>ticklabel_font={}</em>, <em>title_font={}</em>, <em>legend_font={}</em>, <em>lab_angle=90</em>, <em>figsize=(9</em>, <em>6)</em>, <em>legend=True</em>, <em>order=False</em>, <em>desc=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#StackedBarplot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.StackedBarplot" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.plot.StackedBarplot.finish">
<code class="descname">finish</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#StackedBarplot.finish"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.StackedBarplot.finish" title="Permalink to this definition">¶</a></dt>
<dd><p>Applies tight layout, draws the figure, writes the file and closes.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.StackedBarplot.init_fig">
<code class="descname">init_fig</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#StackedBarplot.init_fig"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.StackedBarplot.init_fig" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a figure using the object oriented matplotlib interface.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.StackedBarplot.plot">
<code class="descname">plot</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#StackedBarplot.plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.StackedBarplot.plot" title="Permalink to this definition">¶</a></dt>
<dd><p>The total workflow of this class.
Calls all methods in the correct order.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.StackedBarplot.reload">
<code class="descname">reload</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#StackedBarplot.reload"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.StackedBarplot.reload" title="Permalink to this definition">¶</a></dt>
<dd><p>Reloads the module and updates the class instance.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.StackedBarplot.set_figsize">
<code class="descname">set_figsize</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#StackedBarplot.set_figsize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.StackedBarplot.set_figsize" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts width and height to a tuple so can be used for figsize.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.StackedBarplot.set_title">
<code class="descname">set_title</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#StackedBarplot.set_title"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.StackedBarplot.set_title" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the main title.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.plot.SimilarityGraph">
<em class="property">class </em><code class="descclassname">pypath.plot.</code><code class="descname">SimilarityGraph</code><span class="sig-paren">(</span><em>pp</em>, <em>fname</em>, <em>similarity</em>, <em>size</em>, <em>layout_method='fruchterman_reingold'</em>, <em>layout_param={}</em>, <em>width=1024</em>, <em>height=1024</em>, <em>margin=124</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#SimilarityGraph"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.SimilarityGraph" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.plot.SimilarityGraph.sizes_edge">
<code class="descname">sizes_edge</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#SimilarityGraph.sizes_edge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.SimilarityGraph.sizes_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the size according to number of edges for each resource.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.plot.Histogram">
<em class="property">class </em><code class="descclassname">pypath.plot.</code><code class="descname">Histogram</code><span class="sig-paren">(</span><em>data</em>, <em>labels</em>, <em>fname</em>, <em>font_family='Helvetica Neue LT Std'</em>, <em>font_style='normal'</em>, <em>font_weight='normal'</em>, <em>font_variant='normal'</em>, <em>font_stretch='normal'</em>, <em>xlab=''</em>, <em>ylab=''</em>, <em>title=''</em>, <em>axis_lab_size=10.0</em>, <em>lab_angle=90</em>, <em>lab_size=(9</em>, <em>9)</em>, <em>color=None</em>, <em>palette=None</em>, <em>rc={}</em>, <em>context='poster'</em>, <em>figsize=(5.0</em>, <em>3.0)</em>, <em>bins=None</em>, <em>nbins=None</em>, <em>x_log=False</em>, <em>y_log=False</em>, <em>tone=2</em>, <em>alpha=0.5</em>, <em>legend_size=6</em>, <em>xlim=None</em>, <em>kde_base=0.2</em>, <em>kde_perc=12.0</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#Histogram"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.Histogram" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.plot.HistoryTree">
<em class="property">class </em><code class="descclassname">pypath.plot.</code><code class="descname">HistoryTree</code><span class="sig-paren">(</span><em>fname</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#HistoryTree"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.HistoryTree" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.plot.HistoryTree.run_latex">
<code class="descname">run_latex</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#HistoryTree.run_latex"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.HistoryTree.run_latex" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs LaTeX to compile the TeX file.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.HistoryTree.write_tex">
<code class="descname">write_tex</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#HistoryTree.write_tex"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.HistoryTree.write_tex" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes the TeX markup to file.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.plot.HtpCharacteristics">
<em class="property">class </em><code class="descclassname">pypath.plot.</code><code class="descname">HtpCharacteristics</code><span class="sig-paren">(</span><em>pp</em>, <em>fname</em>, <em>upper=200</em>, <em>lower=5</em>, <em>axis_lab_font={}</em>, <em>ticklabel_font={}</em>, <em>title_font={}</em>, <em>title=''</em>, <em>htdata={}</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#HtpCharacteristics"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.HtpCharacteristics" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.plot.HtpCharacteristics.finish">
<code class="descname">finish</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#HtpCharacteristics.finish"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.HtpCharacteristics.finish" title="Permalink to this definition">¶</a></dt>
<dd><p>Applies tight layout, draws the figure, writes the file and closes.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.HtpCharacteristics.init_fig">
<code class="descname">init_fig</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#HtpCharacteristics.init_fig"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.HtpCharacteristics.init_fig" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a figure using the object oriented matplotlib interface.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.HtpCharacteristics.set_figsize">
<code class="descname">set_figsize</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#HtpCharacteristics.set_figsize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.HtpCharacteristics.set_figsize" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts width and height to a tuple so can be used for figsize.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.HtpCharacteristics.set_grid">
<code class="descname">set_grid</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#HtpCharacteristics.set_grid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.HtpCharacteristics.set_grid" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets up a grid according to the number of subplots,
with one additional column of zero width on the left
to have aligned y axis labels.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.HtpCharacteristics.set_title">
<code class="descname">set_title</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#HtpCharacteristics.set_title"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.HtpCharacteristics.set_title" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the main title.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.plot.RefsComposite">
<em class="property">class </em><code class="descclassname">pypath.plot.</code><code class="descname">RefsComposite</code><span class="sig-paren">(</span><em>pp</em>, <em>fname</em>, <em>axis_lab_font={}</em>, <em>ticklabel_font={}</em>, <em>title_font={}</em>, <em>legend_font={}</em>, <em>bar_args={}</em>, <em>title=''</em>, <em>color='#88CCEE'</em>, <em>hcolor='#88CCEE'</em>, <em>all_color='#44AA99'</em>, <em>htp_threshold=20</em>, <em>figsize=(12.8</em>, <em>8.8)</em>, <em>all_name='All'</em>, <em>curation_plot=False</em>, <em>pubmeds=None</em>, <em>earliest=None</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#RefsComposite"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.RefsComposite" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.plot.RefsComposite.finish">
<code class="descname">finish</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#RefsComposite.finish"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.RefsComposite.finish" title="Permalink to this definition">¶</a></dt>
<dd><p>Applies tight layout, draws the figure, writes the file and closes.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.RefsComposite.init_fig">
<code class="descname">init_fig</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#RefsComposite.init_fig"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.RefsComposite.init_fig" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a figure using the object oriented matplotlib interface.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.RefsComposite.set_figsize">
<code class="descname">set_figsize</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#RefsComposite.set_figsize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.RefsComposite.set_figsize" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts width and height to a tuple so can be used for figsize.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.RefsComposite.set_grid">
<code class="descname">set_grid</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#RefsComposite.set_grid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.RefsComposite.set_grid" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets up a grid according to the number of subplots,
with one additional column of zero width on the left
to have aligned y axis labels.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.RefsComposite.set_title">
<code class="descname">set_title</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#RefsComposite.set_title"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.RefsComposite.set_title" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the main title.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.plot.CurationPlot">
<em class="property">class </em><code class="descclassname">pypath.plot.</code><code class="descname">CurationPlot</code><span class="sig-paren">(</span><em>pp</em>, <em>fname</em>, <em>colors</em>, <em>pubmeds=None</em>, <em>earliest=None</em>, <em>axis_lab_font={}</em>, <em>ticklabel_font={}</em>, <em>legend_font={}</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#CurationPlot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.CurationPlot" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.plot.CurationPlot.finish">
<code class="descname">finish</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#CurationPlot.finish"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.CurationPlot.finish" title="Permalink to this definition">¶</a></dt>
<dd><p>Applies tight layout, draws the figure, writes the file and closes.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.CurationPlot.init_fig">
<code class="descname">init_fig</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#CurationPlot.init_fig"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.CurationPlot.init_fig" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a figure using the object oriented matplotlib interface.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.CurationPlot.set_figsize">
<code class="descname">set_figsize</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#CurationPlot.set_figsize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.CurationPlot.set_figsize" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts width and height to a tuple so can be used for figsize.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.plot.BarplotsGrid">
<em class="property">class </em><code class="descclassname">pypath.plot.</code><code class="descname">BarplotsGrid</code><span class="sig-paren">(</span><em>pp</em>, <em>x</em>, <em>by</em>, <em>fname</em>, <em>ylab</em>, <em>data=None</em>, <em>color='#77AADD'</em>, <em>xlim=None</em>, <em>uniform_xlim=True</em>, <em>uniform_ylim=False</em>, <em>full_range_x=True</em>, <em>sort=False</em>, <em>desc=False</em>, <em>ylog=False</em>, <em>hoffset=0.0</em>, <em>woffset=0.0</em>, <em>axis_lab_font={}</em>, <em>ticklabel_font={}</em>, <em>small_ticklabel_font={}</em>, <em>title_font={}</em>, <em>bar_args={}</em>, <em>htp_threshold=20</em>, <em>xmin=None</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#BarplotsGrid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.BarplotsGrid" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.plot.BarplotsGrid.finish">
<code class="descname">finish</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#BarplotsGrid.finish"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.BarplotsGrid.finish" title="Permalink to this definition">¶</a></dt>
<dd><p>Applies tight layout, draws the figure, writes the file and closes.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.BarplotsGrid.init_fig">
<code class="descname">init_fig</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#BarplotsGrid.init_fig"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.BarplotsGrid.init_fig" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a figure using the object oriented matplotlib interface.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.BarplotsGrid.set_figsize">
<code class="descname">set_figsize</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#BarplotsGrid.set_figsize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.BarplotsGrid.set_figsize" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts width and height to a tuple so can be used for figsize.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.BarplotsGrid.set_grid">
<code class="descname">set_grid</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#BarplotsGrid.set_grid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.BarplotsGrid.set_grid" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets up a grid according to the number of subplots,
with one additional column of zero width on the left
to have aligned y axis labels.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.BarplotsGrid.set_title">
<code class="descname">set_title</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#BarplotsGrid.set_title"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.BarplotsGrid.set_title" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the main title.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.plot.Dendrogram">
<em class="property">class </em><code class="descclassname">pypath.plot.</code><code class="descname">Dendrogram</code><span class="sig-paren">(</span><em>fname</em>, <em>data</em>, <em>color='#4477AA'</em>, <em>axis_lab_font={}</em>, <em>ticklabel_font={}</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#Dendrogram"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.Dendrogram" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.plot.Dendrogram.finish">
<code class="descname">finish</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#Dendrogram.finish"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.Dendrogram.finish" title="Permalink to this definition">¶</a></dt>
<dd><p>Applies tight layout, draws the figure, writes the file and closes.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.Dendrogram.init_fig">
<code class="descname">init_fig</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#Dendrogram.init_fig"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.Dendrogram.init_fig" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a figure using the object oriented matplotlib interface.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.plot.Dendrogram.set_figsize">
<code class="descname">set_figsize</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/plot.html#Dendrogram.set_figsize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.plot.Dendrogram.set_figsize" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts width and height to a tuple so can be used for figsize.</p>
</dd></dl>

</dd></dl>

<span class="target" id="module-pypath.analysis"></span><dl class="class">
<dt id="pypath.analysis.Workflow">
<em class="property">class </em><code class="descclassname">pypath.analysis.</code><code class="descname">Workflow</code><span class="sig-paren">(</span><em>name</em>, <em>network_datasets=[]</em>, <em>do_main_table=True</em>, <em>do_compile_main_table=True</em>, <em>do_curation_table=True</em>, <em>do_compile_curation_table=True</em>, <em>do_simgraphs=True</em>, <em>do_multi_barplots=True</em>, <em>do_coverage_groups=True</em>, <em>do_htp_char=True</em>, <em>do_ptms_barplot=True</em>, <em>do_scatterplots=True</em>, <em>do_history_tree=True</em>, <em>do_compile_history_tree=True</em>, <em>do_refs_journals_grid=True</em>, <em>do_refs_years_grid=True</em>, <em>do_dirs_stacked=True</em>, <em>do_refs_composite=True</em>, <em>do_curation_plot=True</em>, <em>do_refs_by_j=True</em>, <em>do_refs_by_db=True</em>, <em>do_refs_by_year=True</em>, <em>do_resource_list=True</em>, <em>do_compile_resource_list=True</em>, <em>do_consistency_dedrogram=True</em>, <em>do_consistency_table=True</em>, <em>title=None</em>, <em>outdir=None</em>, <em>htdata={}</em>, <em>inc_raw=None</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/analysis.html#Workflow"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.analysis.Workflow" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<span class="target" id="module-pypath.intera"></span><dl class="class">
<dt id="pypath.intera.Residue">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">Residue</code><span class="sig-paren">(</span><em>number</em>, <em>name</em>, <em>identifier</em>, <em>id_type='uniprot'</em>, <em>isoform=1</em>, <em>mutated=False</em>, <em>seq=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Residue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Residue" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.intera.Mutation">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">Mutation</code><span class="sig-paren">(</span><em>original</em>, <em>mutated</em>, <em>sample</em>, <em>properties={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Mutation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Mutation" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.intera.Motif">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">Motif</code><span class="sig-paren">(</span><em>protein</em>, <em>start</em>, <em>end</em>, <em>id_type='uniprot'</em>, <em>regex=None</em>, <em>instance=None</em>, <em>isoform=1</em>, <em>motif_name=None</em>, <em>prob=None</em>, <em>elm=None</em>, <em>description=None</em>, <em>seq=None</em>, <em>source=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Motif"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Motif" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.intera.Ptm">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">Ptm</code><span class="sig-paren">(</span><em>protein</em>, <em>id_type='uniprot'</em>, <em>typ='unknown'</em>, <em>motif=None</em>, <em>residue=None</em>, <em>source=None</em>, <em>isoform=1</em>, <em>seq=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Ptm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Ptm" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.intera.Domain">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">Domain</code><span class="sig-paren">(</span><em>protein</em>, <em>id_type='uniprot'</em>, <em>domain=None</em>, <em>domain_id_type='pfam'</em>, <em>start=None</em>, <em>end=None</em>, <em>isoform=1</em>, <em>chains={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Domain"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Domain" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.intera.DomainMotif">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">DomainMotif</code><span class="sig-paren">(</span><em>domain</em>, <em>ptm</em>, <em>sources=None</em>, <em>refs=None</em>, <em>pdbs=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#DomainMotif"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.DomainMotif" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.intera.DomainDomain">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">DomainDomain</code><span class="sig-paren">(</span><em>domain_a</em>, <em>domain_b</em>, <em>pdbs=None</em>, <em>sources=None</em>, <em>refs=None</em>, <em>contact_residues=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#DomainDomain"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.DomainDomain" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.intera.Interface">
<em class="property">class </em><code class="descclassname">pypath.intera.</code><code class="descname">Interface</code><span class="sig-paren">(</span><em>id_a</em>, <em>id_b</em>, <em>source</em>, <em>id_type='uniprot'</em>, <em>pdb=None</em>, <em>css=None</em>, <em>stab_en=None</em>, <em>solv_en=None</em>, <em>area=None</em>, <em>isoform_a=1</em>, <em>isoform_b=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Interface"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Interface" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.intera.Interface.add_residues">
<code class="descname">add_residues</code><span class="sig-paren">(</span><em>res_a</em>, <em>res_b</em>, <em>typ='undefined'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Interface.add_residues"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Interface.add_residues" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds one pair of residues of type <cite>typ</cite>, 
where <cite>res_a</cite> and <cite>res_b</cite> are tuples of 
residue number in sequence and residue type,
e.g. (124, &#8216;S&#8217;) &#8211; (means Serine #124)
<cite>typ</cite> can be undefined, hbonds, sbridges, ssbonds or covbonds</p>
</dd></dl>

<dl class="method">
<dt id="pypath.intera.Interface.bond_types">
<code class="descname">bond_types</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Interface.bond_types"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Interface.bond_types" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the bond types present in this interface</p>
</dd></dl>

<dl class="method">
<dt id="pypath.intera.Interface.get_bonds">
<code class="descname">get_bonds</code><span class="sig-paren">(</span><em>typ=None</em>, <em>mode=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Interface.get_bonds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Interface.get_bonds" title="Permalink to this definition">¶</a></dt>
<dd><p>Gives a generator to iterate throught bonds in 
this interface. If no type given, bonds of all types 
returned.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.intera.Interface.numof_residues">
<code class="descname">numof_residues</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/intera.html#Interface.numof_residues"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.intera.Interface.numof_residues" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of residue pairs by bound type</p>
</dd></dl>

</dd></dl>

<span class="target" id="module-pypath.pdb"></span><dl class="class">
<dt id="pypath.pdb.ResidueMapper">
<em class="property">class </em><code class="descclassname">pypath.pdb.</code><code class="descname">ResidueMapper</code><a class="reference internal" href="_modules/pypath/pdb.html#ResidueMapper"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pdb.ResidueMapper" title="Permalink to this definition">¶</a></dt>
<dd><p>This class stores and serves the PDB &#8211;&gt; UniProt 
residue level mapping. Attempts to download the 
mapping, and stores it for further use. Converts 
PDB residue numbers to the corresponding UniProt ones.</p>
<dl class="method">
<dt id="pypath.pdb.ResidueMapper.clean">
<code class="descname">clean</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pdb.html#ResidueMapper.clean"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.pdb.ResidueMapper.clean" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes cached mappings, freeing up memory.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt>
<em class="property">class </em><code class="descclassname">pypath.pdb.</code><code class="descname">ResidueMapper</code><a class="reference internal" href="_modules/pypath/pdb.html#ResidueMapper"><span class="viewcode-link">[source]</span></a></dt>
<dd><p>This class stores and serves the PDB &#8211;&gt; UniProt 
residue level mapping. Attempts to download the 
mapping, and stores it for further use. Converts 
PDB residue numbers to the corresponding UniProt ones.</p>
<dl class="method">
<dt>
<code class="descname">clean</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/pdb.html#ResidueMapper.clean"><span class="viewcode-link">[source]</span></a></dt>
<dd><p>Removes cached mappings, freeing up memory.</p>
</dd></dl>

</dd></dl>

<span class="target" id="module-pypath.proteomicsdb"></span><dl class="class">
<dt id="pypath.proteomicsdb.ProteomicsDB">
<em class="property">class </em><code class="descclassname">pypath.proteomicsdb.</code><code class="descname">ProteomicsDB</code><span class="sig-paren">(</span><em>username</em>, <em>password</em>, <em>output_format='json'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/proteomicsdb.html#ProteomicsDB"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.proteomicsdb.ProteomicsDB" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<span class="target" id="module-pypath.mapping"></span><dl class="class">
<dt id="pypath.mapping.MappingTable">
<em class="property">class </em><code class="descclassname">pypath.mapping.</code><code class="descname">MappingTable</code><span class="sig-paren">(</span><em>one</em>, <em>two</em>, <em>typ</em>, <em>source</em>, <em>param</em>, <em>ncbi_tax_id</em>, <em>mysql=None</em>, <em>log=None</em>, <em>cache=True</em>, <em>cachedir='cache'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#MappingTable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.MappingTable" title="Permalink to this definition">¶</a></dt>
<dd><p>To initialize ID conversion tables for the first time
data is downloaded from UniProt and read to dictionaries.
It takes a couple of seconds. Data is saved to pickle 
dumps, this way after tables load much faster.</p>
<dl class="method">
<dt id="pypath.mapping.MappingTable.read_mapping_uniprot">
<code class="descname">read_mapping_uniprot</code><span class="sig-paren">(</span><em>param</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#MappingTable.read_mapping_uniprot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.MappingTable.read_mapping_uniprot" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads ID mappings directly from UniProt.
See the names of possible identifiers here:
<a class="reference external" href="http://www.uniprot.org/help/programmatic_access">http://www.uniprot.org/help/programmatic_access</a></p>
<p>&#64;param : UniprotMapping instance</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.mapping.Mapper">
<em class="property">class </em><code class="descclassname">pypath.mapping.</code><code class="descname">Mapper</code><span class="sig-paren">(</span><em>ncbi_tax_id=9606</em>, <em>mysql_conf=(None</em>, <em>'mapping')</em>, <em>log=None</em>, <em>cache=True</em>, <em>cachedir='cache'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#Mapper"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.Mapper" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.mapping.Mapper.load_mappings">
<code class="descname">load_mappings</code><span class="sig-paren">(</span><em>maplst=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#Mapper.load_mappings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.Mapper.load_mappings" title="Permalink to this definition">¶</a></dt>
<dd><p>mapList is a list of mappings to load;
elements of mapList are dicts containing the 
id names, molecule type, and preferred source
e.g. (&#8220;one&#8221;: &#8220;uniprot&#8221;, &#8220;two&#8221;: &#8220;refseq&#8221;, &#8220;typ&#8221;: &#8220;protein&#8221;, 
&#8220;src&#8221;: &#8220;mysql&#8221;, &#8220;par&#8221;: &#8220;mysql_param/file_param&#8221;)
by default those are loaded from pickle files</p>
</dd></dl>

<dl class="method">
<dt id="pypath.mapping.Mapper.load_uniprot_mapping">
<code class="descname">load_uniprot_mapping</code><span class="sig-paren">(</span><em>filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#Mapper.load_uniprot_mapping"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.Mapper.load_uniprot_mapping" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a wrapper to load a ... mapping table.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.mapping.Mapper.map_name">
<code class="descname">map_name</code><span class="sig-paren">(</span><em>name</em>, <em>nameType</em>, <em>targetNameType</em>, <em>strict=False</em>, <em>silent=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#Mapper.map_name"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.Mapper.map_name" title="Permalink to this definition">¶</a></dt>
<dd><p>This function should be used to convert individual IDs.
It takes care about everything, you don&#8217;t need to think
on the details. How does it work: looks up dictionaries 
between the original and target ID type, if doesn&#8217;t 
find, attempts to load from the predefined inputs.
If the original name is genesymbol, first it looks up
among the preferred gene names from UniProt, if not 
found, it takes an attempt with the alternative gene
names. If the gene symbol still couldn&#8217;t be found, and 
strict = False, the last attempt only the first 5 chara-
cters of the gene symbol matched. If the target name 
type is uniprot, then it converts all the ACs to primary. 
Then, for the Trembl IDs it looks up the preferred gene 
names, and find Swissprot IDs with the same preferred 
gene name.</p>
<dl class="docutils">
<dt>&#64;name</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>The original name which shall be converted.</dd>
<dt>&#64;nameType</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>The type of the name.
Available by default:
- genesymbol (gene name)
- entrez (Entrez Gene ID [#])
- refseqp (NCBI RefSeq Protein ID [NP_*|XP_*])
- ensp (Ensembl protein ID [ENSP*])
- enst (Ensembl transcript ID [ENST*])
- ensg (Ensembl genomic DNA ID [ENSG*])
- hgnc (HGNC ID [HGNC:#])
- gi (GI number [#])
- embl (DDBJ/EMBL/GeneBank CDS accession)
- embl_id (DDBJ/EMBL/GeneBank accession)
To use other IDs, you need to define the input method
and load the table before calling :py:func:Mapper.map_name().</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pypath.mapping.Mapper.primary_uniprot">
<code class="descname">primary_uniprot</code><span class="sig-paren">(</span><em>lst</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#Mapper.primary_uniprot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.Mapper.primary_uniprot" title="Permalink to this definition">¶</a></dt>
<dd><p>For a list of UniProt IDs returns the list of primary ids.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.mapping.Mapper.trembl_swissprot">
<code class="descname">trembl_swissprot</code><span class="sig-paren">(</span><em>lst</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#Mapper.trembl_swissprot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.Mapper.trembl_swissprot" title="Permalink to this definition">¶</a></dt>
<dd><p>For a list of Trembl and Swissprot IDs, returns possibly
only Swissprot, mapping from Trembl to gene names, and 
then back to Swissprot.</p>
</dd></dl>

<dl class="method">
<dt id="pypath.mapping.Mapper.which_table">
<code class="descname">which_table</code><span class="sig-paren">(</span><em>nameType</em>, <em>targetNameType</em>, <em>load=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#Mapper.which_table"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mapping.Mapper.which_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the table which is suitable to convert an ID of 
nameType to targetNameType. If no such table have been loaded
yet, it attempts to load from UniProt.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt>
<em class="property">class </em><code class="descclassname">pypath.mapping.</code><code class="descname">MappingTable</code><span class="sig-paren">(</span><em>one</em>, <em>two</em>, <em>typ</em>, <em>source</em>, <em>param</em>, <em>ncbi_tax_id</em>, <em>mysql=None</em>, <em>log=None</em>, <em>cache=True</em>, <em>cachedir='cache'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mapping.html#MappingTable"><span class="viewcode-link">[source]</span></a></dt>
<dd><p>To initialize ID conversion tables for the first time
data is downloaded from UniProt and read to dictionaries.
It takes a couple of seconds. Data is saved to pickle 
dumps, this way after tables load much faster.</p>
</dd></dl>

<span class="target" id="module-pypath.maps"></span><span class="target" id="module-pypath.reflists"></span><dl class="class">
<dt id="pypath.reflists.ReferenceList">
<em class="property">class </em><code class="descclassname">pypath.reflists.</code><code class="descname">ReferenceList</code><span class="sig-paren">(</span><em>nameType</em>, <em>typ</em>, <em>tax</em>, <em>inFile</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/reflists.html#ReferenceList"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.reflists.ReferenceList" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="pypath.refs">
<code class="descclassname">pypath.</code><code class="descname">refs</code><a class="headerlink" href="#pypath.refs" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <a class="reference internal" href="#pypath.refs" title="pypath.refs"><code class="xref py py-class docutils literal"><span class="pre">pypath.refs</span></code></a></p>
</dd></dl>

<dl class="class">
<dt id="pypath.refs.Reference">
<em class="property">class </em><code class="descclassname">pypath.refs.</code><code class="descname">Reference</code><span class="sig-paren">(</span><em>pmid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/refs.html#Reference"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.refs.Reference" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<span class="target" id="module-pypath.moduland"></span><span class="target" id="module-pypath.dataio"></span><dl class="class">
<dt id="pypath.dataio.ResidueMapper">
<em class="property">class </em><code class="descclassname">pypath.dataio.</code><code class="descname">ResidueMapper</code><a class="reference internal" href="_modules/pypath/dataio.html#ResidueMapper"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.ResidueMapper" title="Permalink to this definition">¶</a></dt>
<dd><p>This class stores and serves the PDB &#8211;&gt; UniProt 
residue level mapping. Attempts to download the 
mapping, and stores it for further use. Converts 
PDB residue numbers to the corresponding UniProt ones.</p>
<dl class="method">
<dt id="pypath.dataio.ResidueMapper.clean">
<code class="descname">clean</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#ResidueMapper.clean"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.ResidueMapper.clean" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes cached mappings, freeing up memory.</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="pypath.dataio.acsn_ppi">
<code class="descclassname">pypath.dataio.</code><code class="descname">acsn_ppi</code><span class="sig-paren">(</span><em>keep_in_complex_interactions=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#acsn_ppi"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.acsn_ppi" title="Permalink to this definition">¶</a></dt>
<dd><p>Processes ACSN data from local file.
Returns list of interactions.</p>
<dl class="docutils">
<dt>&#64;keep_in_complex_interactions</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">bool</span><dd>Whether to include interactions from complex expansion.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.biogrid_interactions">
<code class="descclassname">pypath.dataio.</code><code class="descname">biogrid_interactions</code><span class="sig-paren">(</span><em>organism=9606</em>, <em>htp_limit=1</em>, <em>ltp=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#biogrid_interactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.biogrid_interactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes BioGRID interactions.
Keeps only the &#8220;low throughput&#8221; interactions.
Returns list of interactions.</p>
<dl class="docutils">
<dt>&#64;organism</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">int</span><dd>NCBI Taxonomy ID of organism.</dd>
<dt>&#64;htp_limit</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">int</span><dd>Exclude interactions only from references
cited at more than this number of interactions.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_acsn_effects">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_acsn_effects</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_acsn_effects"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_acsn_effects" title="Permalink to this definition">¶</a></dt>
<dd><p>Processes ACSN data, returns list of effects.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_ca1">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_ca1</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_ca1"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_ca1" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes the CA1 signaling network (Ma&#8217;ayan 2005).
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_ccmap">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_ccmap</code><span class="sig-paren">(</span><em>organism=9606</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_ccmap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_ccmap" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes CancerCellMap.
Returns list of interactions.</p>
<dl class="docutils">
<dt>&#64;organism</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">int</span><dd>NCBI Taxonomy ID to match column #7 in nodes file.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_complexportal">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_complexportal</code><span class="sig-paren">(</span><em>species=9606</em>, <em>zipped=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_complexportal"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_complexportal" title="Permalink to this definition">¶</a></dt>
<dd><p>Complex dataset from IntAct.
See more:
<a class="reference external" href="http://www.ebi.ac.uk/intact/complex/">http://www.ebi.ac.uk/intact/complex/</a>
<a class="reference external" href="http://nar.oxfordjournals.org/content/early/2014/10/13/nar.gku975.full.pdf">http://nar.oxfordjournals.org/content/early/2014/10/13/nar.gku975.full.pdf</a></p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_dgidb">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_dgidb</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_dgidb"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_dgidb" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes the list of all human druggable proteins.
Returns a list of GeneSymbols.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_disgenet">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_disgenet</code><span class="sig-paren">(</span><em>dataset='curated'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_disgenet"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_disgenet" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes the list of all human disease related proteins
from DisGeNet.
Returns dict of dicts.</p>
<dl class="docutils">
<dt>&#64;dataset</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>Name of DisGeNet dataset to be obtained:
<cite>curated</cite>, <cite>literature</cite>, <cite>befree</cite> or <cite>all</cite>.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_domino_ptms">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_domino_ptms</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_domino_ptms"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_domino_ptms" title="Permalink to this definition">¶</a></dt>
<dd><p>The table comes from dataio.get_domino(), having the following fields:
header = [&#8216;uniprot-A&#8217;, &#8216;uniprot-B&#8217;, &#8216;isoform-A&#8217;, &#8216;isoform-B&#8217;, #3
&#8216;exp. method&#8217;, &#8216;references&#8217;, &#8216;taxon-A&#8217;, &#8216;taxon-B&#8217;, #7
&#8216;role-A&#8217;, &#8216;role-B&#8217;, &#8216;binding-site-range-A&#8217;, &#8216;binding-site-range-B&#8217;, #11
&#8216;domains-A&#8217;, &#8216;domains-B&#8217;, &#8216;ptm-residue-A&#8217;, &#8216;ptm-residue-B&#8217;, #15
&#8216;ptm-type-mi-A&#8217;, &#8216;ptm-type-mi-B&#8217;, &#8216;ptm-type-A&#8217;, &#8216;ptm-type-B&#8217;, #19
&#8216;ptm-res-name-A&#8217;, &#8216;ptm-res-name-B&#8217;, &#8216;mutations-A&#8217;, &#8216;mutations-B&#8217;, #23
&#8216;mutation-effects-A&#8217;, &#8216;mutation-effects-B&#8217;, &#8216;domains-interpro-A&#8217;, #26 
&#8216;domains-interpro-B&#8217;, &#8216;negative&#8217;] #28</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_elm_interactions">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_elm_interactions</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_elm_interactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_elm_interactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Downlods manually curated interactions from ELM.
This is the gold standard set of ELM.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_graphviz_attrs">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_graphviz_attrs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_graphviz_attrs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_graphviz_attrs" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads graphviz attribute list from graphviz.org.
Returns 3 dicts of dicts: graph_attrs, vertex_attrs and edge_attrs.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_guide2pharma">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_guide2pharma</code><span class="sig-paren">(</span><em>organism='human'</em>, <em>endogenous=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_guide2pharma"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_guide2pharma" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes Guide to Pharmacology data.
Returns list of dicts.</p>
<dl class="docutils">
<dt>&#64;organism</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>Name of the organism, e.g. <cite>human</cite>.</dd>
<dt>&#64;endogenous</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">bool</span><dd>Whether to include only endogenous ligands interactions.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_hpmr">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_hpmr</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_hpmr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_hpmr" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes the list of all human receptors from
human receptor census (HPMR &#8211; Human Plasma Membrane Receptome).
Returns list of GeneSymbols.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_hprd">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_hprd</code><span class="sig-paren">(</span><em>in_vivo=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_hprd"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_hprd" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and preprocesses HPRD data.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_hprd_ptms">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_hprd_ptms</code><span class="sig-paren">(</span><em>in_vivo=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_hprd_ptms"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_hprd_ptms" title="Permalink to this definition">¶</a></dt>
<dd><p>Processes HPRD data and extracts PTMs.
Returns list of kinase-substrate interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_hsn">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_hsn</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_hsn"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_hsn" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes HumanSignalingNetwork version 6
(published 2014 Jan by Edwin Wang).
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_i3d">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_i3d</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_i3d"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_i3d" title="Permalink to this definition">¶</a></dt>
<dd><p>Interaction3D contains residue numbers in given chains in 
given PDB stuctures, so we need to add an offset to get the residue 
numbers valid for UniProt sequences. Offsets can be obtained from
Instruct, or from the Pfam PDB-chain-UniProt mapping table.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_instruct">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_instruct</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_instruct"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_instruct" title="Permalink to this definition">¶</a></dt>
<dd><p>Instruct contains residue numbers in UniProt sequences, it means
no further calculations of offsets in chains of PDB structures needed.
Chains are not given, only a set of PDB structures supporting the 
domain-domain // protein-protein interaction.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_instruct_offsets">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_instruct_offsets</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_instruct_offsets"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_instruct_offsets" title="Permalink to this definition">¶</a></dt>
<dd><p>These offsets should be understood as from UniProt to PDB.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_kegg">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_kegg</code><span class="sig-paren">(</span><em>mapper=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_kegg"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_kegg" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes KEGG Pathways.
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_kinases">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_kinases</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_kinases"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_kinases" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes the list of all human kinases.
Returns a list of GeneSymbols.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_laudanna_directions">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_laudanna_directions</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_laudanna_directions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_laudanna_directions" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes the SignalingFlow edge attributes
from Laudanna Lab.
Returns list of directions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_laudanna_effects">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_laudanna_effects</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_laudanna_effects"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_laudanna_effects" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes the SignalingDirection edge attributes
from Laudanna Lab.
Returns list of effects.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_li2012">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_li2012</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_li2012"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_li2012" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads supplementary data of Li 2012 from local file.
Returns table (list of lists).</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_lit_bm_13">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_lit_bm_13</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_lit_bm_13"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_lit_bm_13" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes Lit-BM-13 dataset, the high confidence
literature curated interactions from CCSB.
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_phosphosite">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_phosphosite</code><span class="sig-paren">(</span><em>cache=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_phosphosite"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_phosphosite" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads curated and HTP data from Phosphosite,
from preprocessed cache file if available.
Processes BioPAX format.
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_phosphosite_curated">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_phosphosite_curated</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_phosphosite_curated"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_phosphosite_curated" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads literature curated PhosphoSite data,
from preprocessed cache file if available.
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_phosphosite_noref">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_phosphosite_noref</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_phosphosite_noref"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_phosphosite_noref" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads HTP PhosphoSite data,
from preprocessed cache file if available.
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_pmid">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_pmid</code><span class="sig-paren">(</span><em>idList</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_pmid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_pmid" title="Permalink to this definition">¶</a></dt>
<dd><p>For a list of doi or PMC IDs 
fetches the corresponding PMIDs.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_switches_elm">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_switches_elm</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_switches_elm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_switches_elm" title="Permalink to this definition">¶</a></dt>
<dd><p>switches.elm is a resource containing functional switches in molecular regulation,
in domain-motif level resolution, classified into categories according to their
mechanism.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_tfcensus">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_tfcensus</code><span class="sig-paren">(</span><em>classes=['a', 'b', 'other']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_tfcensus"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_tfcensus" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes list of all human transcripton factors.
Returns dict with lists of ENSGene IDs and HGNC Gene Names.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.get_wang_effects">
<code class="descclassname">pypath.dataio.</code><code class="descname">get_wang_effects</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#get_wang_effects"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.get_wang_effects" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes Wang Lab HumanSignalingNetwork.
Returns list of effects.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.hprd_interactions">
<code class="descclassname">pypath.dataio.</code><code class="descname">hprd_interactions</code><span class="sig-paren">(</span><em>in_vivo=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#hprd_interactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.hprd_interactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Processes HPRD data and extracts interactions.
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.li2012_dmi">
<code class="descclassname">pypath.dataio.</code><code class="descname">li2012_dmi</code><span class="sig-paren">(</span><em>mapper=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#li2012_dmi"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.li2012_dmi" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts table read by <cite>pypath.dataio.get_li2012()</cite> to
list of <cite>pypath.intera.DomainMotif()</cite> objects.
Translates GeneSymbols to UniProt IDs.</p>
<dl class="docutils">
<dt>&#64;mapper</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">pypath.mapping.Mapper()</span><dd>If not provided, a new <cite>Mapper()</cite> instance will be 
initialized, reserving more memory.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.li2012_interactions">
<code class="descclassname">pypath.dataio.</code><code class="descname">li2012_interactions</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#li2012_interactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.li2012_interactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts table read by <cite>pypath.dataio.get_li2012()</cite> to
list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.li2012_phospho">
<code class="descclassname">pypath.dataio.</code><code class="descname">li2012_phospho</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#li2012_phospho"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.li2012_phospho" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts table read by <cite>pypath.dataio.get_li2012()</cite> to
list of dicts of kinase-substrate interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.lmpid_dmi">
<code class="descclassname">pypath.dataio.</code><code class="descname">lmpid_dmi</code><span class="sig-paren">(</span><em>fname='LMPID_DATA_pubmed_ref.xml'</em>, <em>organism=9606</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#lmpid_dmi"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.lmpid_dmi" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts list of domain-motif interactions supplied by
<cite>pypath.dataio.load_lmpid()</cite> to list of 
<a href="#id2"><span class="problematic" id="id3">`</span></a>pypath.intera.DomainMotif() objects.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.lmpid_interactions">
<code class="descclassname">pypath.dataio.</code><code class="descname">lmpid_interactions</code><span class="sig-paren">(</span><em>fname='LMPID_DATA_pubmed_ref.xml'</em>, <em>organism=9606</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#lmpid_interactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.lmpid_interactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts list of domain-motif interactions supplied by
<cite>pypath.dataio.load_lmpid()</cite> to list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.load_lmpid">
<code class="descclassname">pypath.dataio.</code><code class="descname">load_lmpid</code><span class="sig-paren">(</span><em>fname='LMPID_DATA_pubmed_ref.xml'</em>, <em>organism=9606</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#load_lmpid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.load_lmpid" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads and processes LMPID data from local file
<cite>pypath.data/LMPID_DATA_pubmed_ref.xml</cite>.
The file was provided by LMPID authors and is now
redistributed with the module.
Returns list of domain-motif interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.load_macrophage">
<code class="descclassname">pypath.dataio.</code><code class="descname">load_macrophage</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#load_macrophage"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.load_macrophage" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads Macrophage from local file.
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.load_signor_ptms">
<code class="descclassname">pypath.dataio.</code><code class="descname">load_signor_ptms</code><span class="sig-paren">(</span><em>fname='signor_22052015.tab'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#load_signor_ptms"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.load_signor_ptms" title="Permalink to this definition">¶</a></dt>
<dd><p>This function is deprecated, you should not use it.
Loads and processes Signor PTMs from local file.
Returns dict of dicts.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.only_pmids">
<code class="descclassname">pypath.dataio.</code><code class="descname">only_pmids</code><span class="sig-paren">(</span><em>idList</em>, <em>strict=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#only_pmids"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.only_pmids" title="Permalink to this definition">¶</a></dt>
<dd><p>Return elements unchanged which compy to PubMed ID format,
and attempts to translate the DOIs and PMC IDs using NCBI
E-utils.
Returns list containing only PMIDs.</p>
<dl class="docutils">
<dt>&#64;idList</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">list, str</span><dd>List of IDs or one single ID.</dd>
<dt>&#64;strict</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">bool</span><dd>Whether keep in the list those IDs which are not PMIDs,
neither DOIs or PMC IDs or NIH manuscript IDs.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.open_pubmed">
<code class="descclassname">pypath.dataio.</code><code class="descname">open_pubmed</code><span class="sig-paren">(</span><em>pmid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#open_pubmed"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.open_pubmed" title="Permalink to this definition">¶</a></dt>
<dd><p>Opens PubMed record in web browser.</p>
<dl class="docutils">
<dt>&#64;pmid</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str or int</span><dd>PubMed ID</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.phosphosite_directions">
<code class="descclassname">pypath.dataio.</code><code class="descname">phosphosite_directions</code><span class="sig-paren">(</span><em>organism='human'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#phosphosite_directions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.phosphosite_directions" title="Permalink to this definition">¶</a></dt>
<dd><p>From curated and HTP PhosphoSite data generates a
list of directions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.reactions_biopax">
<code class="descclassname">pypath.dataio.</code><code class="descname">reactions_biopax</code><span class="sig-paren">(</span><em>biopax_file</em>, <em>organism=9606</em>, <em>protein_name_type='UniProt'</em>, <em>clean=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#reactions_biopax"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.reactions_biopax" title="Permalink to this definition">¶</a></dt>
<dd><p>Processes a BioPAX file and extracts binary interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.reactome_biopax">
<code class="descclassname">pypath.dataio.</code><code class="descname">reactome_biopax</code><span class="sig-paren">(</span><em>organism=9606</em>, <em>cache=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#reactome_biopax"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.reactome_biopax" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads Reactome human reactions in SBML format.
Returns File object.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.reactome_interactions">
<code class="descclassname">pypath.dataio.</code><code class="descname">reactome_interactions</code><span class="sig-paren">(</span><em>cacheFile=None</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#reactome_interactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.reactome_interactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and processes Reactome BioPAX.
Extracts binary interactions.
The applied criteria are very stringent, yields very few interactions.
Requires large free memory, approx. 2G.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.reactome_sbml">
<code class="descclassname">pypath.dataio.</code><code class="descname">reactome_sbml</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#reactome_sbml"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.reactome_sbml" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads Reactome human reactions in SBML format.
Returns gzip.GzipFile object.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.read_complexes_havugimana">
<code class="descclassname">pypath.dataio.</code><code class="descname">read_complexes_havugimana</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#read_complexes_havugimana"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.read_complexes_havugimana" title="Permalink to this definition">¶</a></dt>
<dd><p>Supplement Table S3/1 from Havugimana 2012
Cell. 150(5): 1068–1081.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.read_table">
<code class="descclassname">pypath.dataio.</code><code class="descname">read_table</code><span class="sig-paren">(</span><em>cols</em>, <em>fileObject=None</em>, <em>data=None</em>, <em>sep='\t'</em>, <em>sep2=None</em>, <em>rem=[]</em>, <em>hdr=None</em>, <em>encoding='ascii'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#read_table"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.read_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Generic function to read data tables.</p>
<dl class="docutils">
<dt>fileObject</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">file-like</span><dd>Any file like object: file opened for read, or StringIO buffer</dd>
<dt>cols</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">dict</span><dd>Dictionary of columns to read. Keys identifying fields are returned 
in the result. Values are column numbers.</dd>
<dt>sepLevel1</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>Field separator of the file.</dd>
<dt>sepLevel2</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">dict</span><dd>Subfield separators and prefixes. 
E.g. {2: &#8216;,&#8217;, 3: &#8216;|&#8217;}</dd>
<dt>hdr</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">int</span><dd>Number of header lines. If None, no headers assumed.</dd>
<dt>rem</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">list</span><dd>Strings to remove. For each line these elements will be replaced with &#8216;&#8217;.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.read_xls">
<code class="descclassname">pypath.dataio.</code><code class="descname">read_xls</code><span class="sig-paren">(</span><em>xls_file</em>, <em>sheet=''</em>, <em>csv_file=None</em>, <em>return_table=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#read_xls"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.read_xls" title="Permalink to this definition">¶</a></dt>
<dd><p>Generic function to read MS Excel XLS file, and convert one sheet
to CSV, or return as a list of lists</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.rolland_hi_ii_14">
<code class="descclassname">pypath.dataio.</code><code class="descname">rolland_hi_ii_14</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#rolland_hi_ii_14"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.rolland_hi_ii_14" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads the HI-II-14 unbiased interactome from the large scale screening
of from Rolland 2014.
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.signalink_interactions">
<code class="descclassname">pypath.dataio.</code><code class="descname">signalink_interactions</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#signalink_interactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.signalink_interactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads and processes SignaLink3 interactions from local file.
Returns list of interactions.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.signor_interactions">
<code class="descclassname">pypath.dataio.</code><code class="descname">signor_interactions</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#signor_interactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.signor_interactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads the full dataset from Signor.
Return the file contents.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.signor_urls">
<code class="descclassname">pypath.dataio.</code><code class="descname">signor_urls</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#signor_urls"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.signor_urls" title="Permalink to this definition">¶</a></dt>
<dd><p>This function is deprecated.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.take_a_trip">
<code class="descclassname">pypath.dataio.</code><code class="descname">take_a_trip</code><span class="sig-paren">(</span><em>cachefile='trip.pickle'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#take_a_trip"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.take_a_trip" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads TRIP data from webpage and preprocesses it.
Saves preprocessed data into <cite>cachefile</cite> and next
time loads from this file.</p>
<dl class="docutils">
<dt>&#64;cachefile</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>Filename, located in <cite>./cache</cite>.
To disable cache, pass <cite>None</cite>.
To download again, remove file from <cite>./cache</cite>.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.trip_find_uniprot">
<code class="descclassname">pypath.dataio.</code><code class="descname">trip_find_uniprot</code><span class="sig-paren">(</span><em>soup</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#trip_find_uniprot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.trip_find_uniprot" title="Permalink to this definition">¶</a></dt>
<dd><p>Looks up a UniProt name in table downloaded from TRIP
webpage.</p>
<dl class="docutils">
<dt>&#64;soup</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">bs4.BeautifulSoup</span><dd>The <cite>BeautifulSoup</cite> instance returned by <cite>pypath.dataio.trip_get_uniprot()</cite>.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.trip_get_uniprot">
<code class="descclassname">pypath.dataio.</code><code class="descname">trip_get_uniprot</code><span class="sig-paren">(</span><em>syn</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#trip_get_uniprot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.trip_get_uniprot" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads table from TRIP webpage and UniProt attempts to
look up the UniProt ID for one synonym.</p>
<dl class="docutils">
<dt>&#64;syn</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>The synonym as shown on TRIP webpage.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.trip_interactions">
<code class="descclassname">pypath.dataio.</code><code class="descname">trip_interactions</code><span class="sig-paren">(</span><em>exclude_methods=['Inference', 'Speculation'], predictions=False, species='Human', strict=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#trip_interactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.trip_interactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Obtains processed TRIP interactions by calling <cite>pypath.dataio.trip_process()</cite>
and returns list of interactions. All arguments are passed to 
<cite>trip_process()</cite>, see their definition there.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.trip_process">
<code class="descclassname">pypath.dataio.</code><code class="descname">trip_process</code><span class="sig-paren">(</span><em>exclude_methods=['Inference', 'Speculation'], predictions=False, species='Human', strict=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#trip_process"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.trip_process" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads TRIP data by calling <cite>pypath.dadio.take_a_trip()</cite> and
further provcesses it.
Returns dict of dict with TRIP data.</p>
<dl class="docutils">
<dt>&#64;exclude_methods</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">list</span><dd>Interaction detection methods to be discarded.</dd>
<dt>&#64;predictions</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">bool</span><dd>Whether to include predicted interactions.</dd>
<dt>&#64;species</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>Organism name, e.g. <cite>Human</cite>.</dd>
<dt>&#64;strict</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">bool</span><dd>Whether include interactions with species not 
used as a bait or not specified.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.trip_process_table">
<code class="descclassname">pypath.dataio.</code><code class="descname">trip_process_table</code><span class="sig-paren">(</span><em>tab</em>, <em>result</em>, <em>intrs</em>, <em>trp_uniprot</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#trip_process_table"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.trip_process_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Processes one HTML table downloaded from TRIP webpage.</p>
<dl class="docutils">
<dt>&#64;tab</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">bs4.element.Tag()</span><dd>One table of interactions from TRIP webpage.</dd>
<dt>&#64;result</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">dict</span><dd>Dictionary the data should be filled in.</dd>
<dt>&#64;intrs</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">dict</span><dd>Dictionary of already converted interactor IDs.
This serves as a cache so do not need to look up
the same ID twice.</dd>
<dt>&#64;trp_uniprot</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>UniProt ID of TRP domain containing protein.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pypath.dataio.vidal_hi_iii">
<code class="descclassname">pypath.dataio.</code><code class="descname">vidal_hi_iii</code><span class="sig-paren">(</span><em>fname</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/dataio.html#vidal_hi_iii"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.dataio.vidal_hi_iii" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads the HI-III  unbiased interactome from preliminary data of
the next large scale screening of Vidal Lab.</p>
<dl class="docutils">
<dt>The data is accessible here:</dt>
<dd><a class="reference external" href="http://interactome.dfci.harvard.edu/H_sapiens/dload_trk.php">http://interactome.dfci.harvard.edu/H_sapiens/dload_trk.php</a></dd>
</dl>
<p>You need to register and accept the license terms.</p>
<p>Returns list of interactions.</p>
</dd></dl>

<span class="target" id="module-pypath.uniprot_input"></span><span class="target" id="module-pypath.mapping_input"></span><span class="target" id="module-pypath.server"></span><dl class="class">
<dt id="pypath.server.RestResource">
<em class="property">class </em><code class="descclassname">pypath.server.</code><code class="descname">RestResource</code><span class="sig-paren">(</span><em>pypath</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/server.html#RestResource"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.server.RestResource" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.server.Rest">
<em class="property">class </em><code class="descclassname">pypath.server.</code><code class="descname">Rest</code><span class="sig-paren">(</span><em>pypath</em>, <em>port</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/server.html#Rest"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.server.Rest" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<span class="target" id="module-pypath.curl"></span><dl class="class">
<dt id="pypath.curl.Curl">
<em class="property">class </em><code class="descclassname">pypath.curl.</code><code class="descname">Curl</code><span class="sig-paren">(</span><em>url</em>, <em>silent=True</em>, <em>get=None</em>, <em>post=None</em>, <em>req_headers=None</em>, <em>cache=True</em>, <em>debug=False</em>, <em>outf=None</em>, <em>compr=None</em>, <em>encoding=None</em>, <em>files_needed=None</em>, <em>timeout=300</em>, <em>init_url=None</em>, <em>init_fun='get_jsessionid'</em>, <em>follow=True</em>, <em>large=False</em>, <em>override_post=False</em>, <em>init_headers=False</em>, <em>return_headers=False</em>, <em>binary_data=None</em>, <em>write_cache=True</em>, <em>force_quote=False</em>, <em>sftp_user=None</em>, <em>sftp_passwd=None</em>, <em>sftp_passwd_file='.secrets'</em>, <em>sftp_port=22</em>, <em>sftp_host=None</em>, <em>sftp_ask=None</em>, <em>setup=True</em>, <em>call=True</em>, <em>process=True</em>, <em>retries=3</em>, <em>cache_dir='cache'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/curl.html#Curl"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.curl.Curl" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.curl.Curl.is_quoted">
<code class="descname">is_quoted</code><span class="sig-paren">(</span><em>string</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/curl.html#Curl.is_quoted"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.curl.Curl.is_quoted" title="Permalink to this definition">¶</a></dt>
<dd><p>From <a class="reference external" href="http://stackoverflow.com/questions/1637762/test-if-string-is-url-encoded-in-php">http://stackoverflow.com/questions/1637762/test-if-string-is-url-encoded-in-php</a></p>
</dd></dl>

<dl class="method">
<dt id="pypath.curl.Curl.url_fix">
<code class="descname">url_fix</code><span class="sig-paren">(</span><em>charset='utf-8'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/curl.html#Curl.url_fix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.curl.Curl.url_fix" title="Permalink to this definition">¶</a></dt>
<dd><p>From <a class="reference external" href="http://stackoverflow.com/a/121017/854988">http://stackoverflow.com/a/121017/854988</a></p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pypath.curl.FileOpener">
<em class="property">class </em><code class="descclassname">pypath.curl.</code><code class="descname">FileOpener</code><span class="sig-paren">(</span><em>file_param</em>, <em>compr=None</em>, <em>extract=True</em>, <em>_open=True</em>, <em>set_fileobj=True</em>, <em>files_needed=None</em>, <em>large=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/curl.html#FileOpener"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.curl.FileOpener" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<span class="target" id="module-pypath.chembl"></span><dl class="class">
<dt id="pypath.chembl.Chembl">
<em class="property">class </em><code class="descclassname">pypath.chembl.</code><code class="descname">Chembl</code><span class="sig-paren">(</span><em>chembl_mysql=(None</em>, <em>'chembl_ebi')</em>, <em>ncbi_tax_id=9606</em>, <em>mapping_mysql=None</em>, <em>mapper=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/chembl.html#Chembl"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.chembl.Chembl" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<span class="target" id="module-pypath.unichem"></span><span class="target" id="module-pypath.seq"></span><dl class="class">
<dt id="pypath.seq.Seq">
<em class="property">class </em><code class="descclassname">pypath.seq.</code><code class="descname">Seq</code><span class="sig-paren">(</span><em>protein</em>, <em>sequence</em>, <em>isoform=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/seq.html#Seq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.seq.Seq" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<span class="target" id="module-pypath.mysql"></span><dl class="class">
<dt id="pypath.mysql.MysqlRunner">
<em class="property">class </em><code class="descclassname">pypath.mysql.</code><code class="descname">MysqlRunner</code><span class="sig-paren">(</span><em>param</em>, <em>cursor='serverside'</em>, <em>concurrent_queries=4</em>, <em>log=None</em>, <em>silent=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mysql.html#MysqlRunner"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mysql.MysqlRunner" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="pypath.mysql.MysqlRunner.get_qid">
<code class="descname">get_qid</code><span class="sig-paren">(</span><em>query</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mysql.html#MysqlRunner.get_qid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mysql.MysqlRunner.get_qid" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the 32 byte md5sum of a string:
this serves as a unique identifier of queries,
referred as <cite>qid</cite> in this module.</p>
<dl class="docutils">
<dt>&#64;query</dt>
 <span class="classifier-delimiter">:</span> <span class="classifier">str</span><dd>MySQL query or any other string.</dd>
</dl>
</dd></dl>

</dd></dl>

<span class="target" id="module-pypath.mysql_connect"></span><dl class="class">
<dt id="pypath.mysql_connect.MysqlConnect">
<em class="property">class </em><code class="descclassname">pypath.mysql_connect.</code><code class="descname">MysqlConnect</code><span class="sig-paren">(</span><em>config=None</em>, <em>log=None</em>, <em>timeout=12</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/mysql_connect.html#MysqlConnect"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.mysql_connect.MysqlConnect" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<span class="target" id="module-pypath.progress"></span><dl class="class">
<dt id="pypath.progress.Progress">
<em class="property">class </em><code class="descclassname">pypath.progress.</code><code class="descname">Progress</code><span class="sig-paren">(</span><em>total=None</em>, <em>name='Progress'</em>, <em>interval=3000</em>, <em>percent=True</em>, <em>status='initializing'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/progress.html#Progress"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.progress.Progress" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<span class="target" id="module-pypath.colorgen"></span><dl class="function">
<dt id="pypath.colorgen.flatten">
<code class="descclassname">pypath.colorgen.</code><code class="descname">flatten</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pypath.colorgen.flatten" title="Permalink to this definition">¶</a></dt>
<dd><p>chain.from_iterable(iterable) &#8211;&gt; chain object</p>
<p>Alternate chain() contructor taking a single iterable argument
that evaluates lazily.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.colorgen.getfracs">
<code class="descclassname">pypath.colorgen.</code><code class="descname">getfracs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/colorgen.html#getfracs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.colorgen.getfracs" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>[Fraction(0, 1), Fraction(1, 2), Fraction(1, 4), Fraction(3, 4),</dt>
<dd>Fraction(1, 8), Fraction(3, 8), Fraction(5, 8), Fraction(7, 8),
Fraction(1, 16), Fraction(3, 16), ...]</dd>
</dl>
<p>[0.0, 0.5, 0.25, 0.75, 0.125, 0.375, 0.625, 0.875, 0.0625, 0.1875, ...]</p>
</dd></dl>

<dl class="function">
<dt id="pypath.colorgen.zenos_dichotomy">
<code class="descclassname">pypath.colorgen.</code><code class="descname">zenos_dichotomy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/colorgen.html#zenos_dichotomy"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.colorgen.zenos_dichotomy" title="Permalink to this definition">¶</a></dt>
<dd><p><a class="reference external" href="http://en.wikipedia.org/wiki/1/2_%2B_1/4_%2B_1/8_%2B_1/16_%2B_%C2%B7_%C2%B7_%C2%B7">http://en.wikipedia.org/wiki/1/2_%2B_1/4_%2B_1/8_%2B_1/16_%2B_%C2%B7_%C2%B7_%C2%B7</a></p>
</dd></dl>

<span class="target" id="module-pypath.common"></span><dl class="function">
<dt id="pypath.common.uniqList">
<code class="descclassname">pypath.common.</code><code class="descname">uniqList</code><span class="sig-paren">(</span><em>seq</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/common.html#uniqList"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.common.uniqList" title="Permalink to this definition">¶</a></dt>
<dd><p>Not order preserving
From <a class="reference external" href="http://www.peterbe.com/plog/uniqifiers-benchmark">http://www.peterbe.com/plog/uniqifiers-benchmark</a></p>
</dd></dl>

<dl class="function">
<dt id="pypath.common.rotate">
<code class="descclassname">pypath.common.</code><code class="descname">rotate</code><span class="sig-paren">(</span><em>point</em>, <em>angle</em>, <em>center=(0.0</em>, <em>0.0)</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/common.html#rotate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.common.rotate" title="Permalink to this definition">¶</a></dt>
<dd><p>from <a class="reference external" href="http://stackoverflow.com/a/20024348/854988">http://stackoverflow.com/a/20024348/854988</a>
Rotates a point around center. Angle is in degrees.
Rotation is counter-clockwise</p>
</dd></dl>

<dl class="function">
<dt id="pypath.common.cleanDict">
<code class="descclassname">pypath.common.</code><code class="descname">cleanDict</code><span class="sig-paren">(</span><em>dct</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/common.html#cleanDict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.common.cleanDict" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes <code class="docutils literal"><span class="pre">None</span></code> values from dict and casts everything else to <code class="docutils literal"><span class="pre">str</span></code>.</p>
</dd></dl>

<dl class="function">
<dt id="pypath.common.md5">
<code class="descclassname">pypath.common.</code><code class="descname">md5</code><span class="sig-paren">(</span><em>value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/common.html#md5"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.common.md5" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the ms5sum of <code class="docutils literal"><span class="pre">value</span></code> as string.</p>
</dd></dl>

<span class="target" id="module-pypath.enrich"></span><dl class="class">
<dt id="pypath.enrich.Enrichment">
<em class="property">class </em><code class="descclassname">pypath.enrich.</code><code class="descname">Enrichment</code><span class="sig-paren">(</span><em>set_count</em>, <em>pop_count</em>, <em>set_size</em>, <em>pop_size</em>, <em>data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/enrich.html#Enrichment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.enrich.Enrichment" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.enrich.EnrichmentSet">
<em class="property">class </em><code class="descclassname">pypath.enrich.</code><code class="descname">EnrichmentSet</code><span class="sig-paren">(</span><em>data</em>, <em>pop_size</em>, <em>correction_method='hommel'</em>, <em>alpha=0.05</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/enrich.html#EnrichmentSet"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.enrich.EnrichmentSet" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<span class="target" id="module-pypath.gsea"></span><dl class="class">
<dt id="pypath.gsea.GSEA">
<em class="property">class </em><code class="descclassname">pypath.gsea.</code><code class="descname">GSEA</code><span class="sig-paren">(</span><em>user=None</em>, <em>mapper=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/gsea.html#GSEA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.gsea.GSEA" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.gsea.GSEABinaryEnrichmentSet">
<em class="property">class </em><code class="descclassname">pypath.gsea.</code><code class="descname">GSEABinaryEnrichmentSet</code><span class="sig-paren">(</span><em>basic_set</em>, <em>gsea=None</em>, <em>geneset_ids=None</em>, <em>alpha=0.05</em>, <em>correction_method='hommel'</em>, <em>user=None</em>, <em>mapper=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/gsea.html#GSEABinaryEnrichmentSet"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.gsea.GSEABinaryEnrichmentSet" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<span class="target" id="module-pypath.go"></span><dl class="function">
<dt id="pypath.go.load_go">
<code class="descclassname">pypath.go.</code><code class="descname">load_go</code><span class="sig-paren">(</span><em>graph, aspect=['C', 'F', 'P']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/go.html#load_go"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.go.load_go" title="Permalink to this definition">¶</a></dt>
<dd><p>&#64;graph : igraph.Graph
Any igraph.Graph object with uniprot IDs in its <cite>name</cite> vertex attribute.</p>
</dd></dl>

<dl class="class">
<dt id="pypath.go.GOAnnotation">
<em class="property">class </em><code class="descclassname">pypath.go.</code><code class="descname">GOAnnotation</code><span class="sig-paren">(</span><em>organism=9606</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/go.html#GOAnnotation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.go.GOAnnotation" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pypath.go.GOEnrichmentSet">
<em class="property">class </em><code class="descclassname">pypath.go.</code><code class="descname">GOEnrichmentSet</code><span class="sig-paren">(</span><em>aspect</em>, <em>organism=9606</em>, <em>annotation=None</em>, <em>basic_set=None</em>, <em>alpha=0.05</em>, <em>correction_method='hommel'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/go.html#GOEnrichmentSet"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.go.GOEnrichmentSet" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<span class="target" id="module-pypath.drawing"></span><dl class="class">
<dt id="pypath.drawing.Plot">
<em class="property">class </em><code class="descclassname">pypath.drawing.</code><code class="descname">Plot</code><span class="sig-paren">(</span><em>graph=None</em>, <em>filename=None</em>, <em>graphix_dir='pdf'</em>, <em>graphix_format='pdf'</em>, <em>name=None</em>, <em>title_text=None</em>, <em>title_font_family=None</em>, <em>title_font_size=None</em>, <em>title_color='#646567'</em>, <em>size=None</em>, <em>layout='fruchterman_reingold'</em>, <em>layout_param=None</em>, <em>vertex_label=None</em>, <em>vertex_size=None</em>, <em>vertex_label_size='degree_label_size'</em>, <em>edge_width=None</em>, <em>vertex_color='#6EA945'</em>, <em>vertex_label_color='#007B7F'</em>, <em>vertex_alpha='AA'</em>, <em>vertex_frame_color='#FFFFFF00'</em>, <em>vertex_frame_width=0</em>, <em>edge_label=None</em>, <em>edge_label_size=None</em>, <em>edge_label_color='#007B7F'</em>, <em>edge_curved=None</em>, <em>edge_color='#818284'</em>, <em>edge_alpha='AA'</em>, <em>autocurve=None</em>, <em>vertex_label_font='sans-serif'</em>, <em>edge_label_font='sans-serif'</em>, <em>edge_arrow_size=1.0</em>, <em>edge_arrow_width=1.0</em>, <em>palettes={}</em>, <em>bbox=None</em>, <em>margin=10</em>, <em>small=None</em>, <em>dimensions=(1280</em>, <em>1280)</em>, <em>grouping=None</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pypath/drawing.html#Plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pypath.drawing.Plot" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<span class="target" id="module-pypath.logn"></span></div>


           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2016, Dénes Türei.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.5.1',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>